<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 0: Introductions and Environment Setup | RNA-seq Analysis 2025</title>
  <meta name="description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 0: Introductions and Environment Setup | RNA-seq Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/RNA_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 0: Introductions and Environment Setup | RNA-seq Analysis 2025" />
  
  <meta name="twitter:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Malachi Griffith and Obi Griffith" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="meet-your-faculty.html"/>
<link rel="next" href="module-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>RNA-Seq Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_RNA_Icon.png" width="70%" alt="WORKSHOP NAME logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html"><i class="fa fa-check"></i>Module 0: Introductions and Environment Setup</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#environment-setup"><i class="fa fa-check"></i>Environment Setup</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tool-installation"><i class="fa fa-check"></i>Tool Installation</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#note"><i class="fa fa-check"></i>Note</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#samtools"><i class="fa fa-check"></i>SAMtools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bam-readcount"><i class="fa fa-check"></i>bam-readcount</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#hisat2"><i class="fa fa-check"></i>HISAT2</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#stringtie"><i class="fa fa-check"></i>StringTie</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gffcompare"><i class="fa fa-check"></i>gffcompare</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#htseq-count"><i class="fa fa-check"></i>htseq-count</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tophat"><i class="fa fa-check"></i>TopHat</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#kallisto"><i class="fa fa-check"></i>kallisto</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastqc"><i class="fa fa-check"></i>FastQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastp"><i class="fa fa-check"></i>Fastp</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#multiqc"><i class="fa fa-check"></i>MultiQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#picard"><i class="fa fa-check"></i>Picard</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#regtools"><i class="fa fa-check"></i>RegTools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#rseqc"><i class="fa fa-check"></i>RSeQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bedops"><i class="fa fa-check"></i>bedops</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gtftogenepred"><i class="fa fa-check"></i>gtfToGenePred</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#genepredtobed"><i class="fa fa-check"></i>genePredToBed</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#how_are_we_stranded_here"><i class="fa fa-check"></i>how_are_we_stranded_here</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-cell-ranger"><i class="fa fa-check"></i>Install Cell Ranger</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-r"><i class="fa fa-check"></i>Install R</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#r-libraries"><i class="fa fa-check"></i>R Libraries</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bioconductor"><i class="fa fa-check"></i>Bioconductor</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#sleuth"><i class="fa fa-check"></i>Sleuth</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#practical-exercise-1---software-installation"><i class="fa fa-check"></i>PRACTICAL EXERCISE 1 - Software Installation</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#add-locally-installed-tools-to-your-path-optional"><i class="fa fa-check"></i>Add locally installed tools to your PATH [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-from-official-ubuntu-packages-optional"><i class="fa fa-check"></i>Installing tools from official ubuntu packages [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image"><i class="fa fa-check"></i>Installing tools by Docker image</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image-using-singularity"><i class="fa fa-check"></i>Installing tools by Docker image (using Singularity)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html"><i class="fa fa-check"></i>Module 1</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#labs"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---key-concepts"><i class="fa fa-check"></i>Module 1 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---learning-objectives"><i class="fa fa-check"></i>Module 1 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-a-reference-genome-from-ensembl-igenomes-ncbi-or-ucsc."><i class="fa fa-check"></i>Obtain a reference genome from Ensembl, iGenomes, NCBI or UCSC.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#note-on-complex-commands-and-scripting-in-unix"><i class="fa fa-check"></i>Note on complex commands and scripting in Unix</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-2-advanced"><i class="fa fa-check"></i>PRACTICAL EXERCISE 2 (ADVANCED)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture-1"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-known-genetranscript-annotations"><i class="fa fa-check"></i>Obtain Known Gene/Transcript Annotations</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#definitions"><i class="fa fa-check"></i>Definitions:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#the-purpose-of-gene-annotations-.gtf-file"><i class="fa fa-check"></i>The Purpose of Gene Annotations (.gtf file)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#sources-for-obtaining-gene-annotation-files-formatted-for-hisat2stringtieballgown"><i class="fa fa-check"></i>Sources for obtaining gene annotation files formatted for HISAT2/StringTie/Ballgown</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#important-notes"><i class="fa fa-check"></i>Important notes:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#indexing-mini-lecture"><i class="fa fa-check"></i>Indexing mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#create-a-hisat2-index"><i class="fa fa-check"></i>Create a HISAT2 index</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-rna-seq-test-data."><i class="fa fa-check"></i>Obtain RNA-seq test data.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#determining-the-strandedness-of-rna-seq-data-optional"><i class="fa fa-check"></i>Determining the strandedness of RNA-seq data (Optional)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-3"><i class="fa fa-check"></i>PRACTICAL EXERCISE 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html"><i class="fa fa-check"></i>Module 2</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#labs-1"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#module-2---key-concepts"><i class="fa fa-check"></i>Module 2 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#module-2---learning-objectives"><i class="fa fa-check"></i>Module 2 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lectures"><i class="fa fa-check"></i>Lectures</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#read-trimming-with-fastp"><i class="fa fa-check"></i>Read trimming with Fastp</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#use-fastqc-and-multiqc-to-compare-the-impact-of-trimming"><i class="fa fa-check"></i>Use FastQC and multiqc to compare the impact of trimming</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#clean-up"><i class="fa fa-check"></i>Clean up</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-5"><i class="fa fa-check"></i>PRACTICAL EXERCISE 5</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#alignment-mini-lecture"><i class="fa fa-check"></i>Alignment mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#hisat2-alignment"><i class="fa fa-check"></i>HISAT2 alignment</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#merge-hisat2-bam-files"><i class="fa fa-check"></i>Merge HISAT2 BAM files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-6"><i class="fa fa-check"></i>PRACTICAL EXERCISE 6</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-1-getting-familiar-with-igv"><i class="fa fa-check"></i>Visualization Part 1: Getting familiar with IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-2-inspecting-snps-snvs-and-svs"><i class="fa fa-check"></i>Visualization Part 2: Inspecting SNPs, SNVs, and SVs</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-3-automating-tasks-in-igv"><i class="fa fa-check"></i>Visualization Part 3: Automating Tasks in IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#indexing-bam-files-with-samtools"><i class="fa fa-check"></i>Indexing BAM files with samtools</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualize-alignments-with-igv"><i class="fa fa-check"></i>Visualize alignments with IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-7"><i class="fa fa-check"></i>PRACTICAL EXERCISE 7</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#bam-read-counting"><i class="fa fa-check"></i>BAM Read Counting</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-7-1"><i class="fa fa-check"></i>PRACTICAL EXERCISE 7</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#alignment-qc-mini-lecture"><i class="fa fa-check"></i>Alignment QC mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#use-samtools-and-fastqc-to-evaluate-the-alignments"><i class="fa fa-check"></i>Use samtools and FastQC to evaluate the alignments</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#create-versions-of-our-bam-files-with-only-the-chromosome-22-alignments"><i class="fa fa-check"></i>Create versions of our BAM files with only the chromosome 22 alignments</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#using-fastqc"><i class="fa fa-check"></i>Using FastQC</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#using-picard"><i class="fa fa-check"></i>Using Picard</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#rseqc-optional"><i class="fa fa-check"></i>RSeQC [optional]</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#multiqc-1"><i class="fa fa-check"></i>MultiQC</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#view-a-pre-generated-multiqc-report-for-full-bam-files"><i class="fa fa-check"></i>View a pre-generated MultiQC report for full bam files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-i---obtaining-the-dataset-reference-files"><i class="fa fa-check"></i>Part I - Obtaining the dataset &amp; reference files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-ii---data-preprocessing-qc-trimming"><i class="fa fa-check"></i>Part II - Data Preprocessing (QC &amp; Trimming)</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-iii---alignment"><i class="fa fa-check"></i>Part III - Alignment</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-iv---post-alignment-qc-igv-visualization"><i class="fa fa-check"></i>Part IV - Post-alignment QC &amp; IGV Visualization</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#presenting-your-results"><i class="fa fa-check"></i>Presenting Your Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html"><i class="fa fa-check"></i>Module 3</a>
<ul>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#labs-2"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#module-3---key-concepts"><i class="fa fa-check"></i>Module 3 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#module-3---learning-objectives"><i class="fa fa-check"></i>Module 3 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#lectures-1"><i class="fa fa-check"></i>Lectures</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#expression-mini-lecture"><i class="fa fa-check"></i>Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#use-stringtie-to-generate-expression-estimates-from-the-sambam-files-generated-by-hisat2-in-the-previous-module"><i class="fa fa-check"></i>Use Stringtie to generate expression estimates from the SAM/BAM files generated by HISAT2 in the previous module</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#practical-exercise-8"><i class="fa fa-check"></i>PRACTICAL EXERCISE 8</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ballgown-de-analysis"><i class="fa fa-check"></i>Ballgown DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#practical-exercise-9"><i class="fa fa-check"></i>PRACTICAL EXERCISE 9</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ercc-de-analysis"><i class="fa fa-check"></i>ERCC DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture-1"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#edger-de-analysis"><i class="fa fa-check"></i>edgeR DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture-2"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#deseq2-de-analysis"><i class="fa fa-check"></i>DESeq2 DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#format-htseq-counts-data-to-work-with-deseq2"><i class="fa fa-check"></i>Format htseq counts data to work with DESeq2</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#filter-raw-counts"><i class="fa fa-check"></i>Filter raw counts</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#specifying-the-experimental-design"><i class="fa fa-check"></i>Specifying the experimental design</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#construct-the-deseq2-object-piecing-all-the-data-together"><i class="fa fa-check"></i>Construct the DESeq2 object piecing all the data together</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#running-deseq2"><i class="fa fa-check"></i>Running DESeq2</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#log-fold-change-shrinkage"><i class="fa fa-check"></i>Log-fold change shrinkage</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#annotate-gene-symbols-onto-the-de-results"><i class="fa fa-check"></i>Annotate gene symbols onto the DE results</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#data-manipulation"><i class="fa fa-check"></i>Data manipulation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#save-results-to-files"><i class="fa fa-check"></i>Save results to files</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#briefly-examine-the-top-over-expressed-genes"><i class="fa fa-check"></i>Briefly examine the top over-expressed genes</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#perform-some-preliminary-exploration-of-de-genes-using-webtools"><i class="fa fa-check"></i>Perform some preliminary exploration of DE genes using webtools</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#visualize-overlap-with-a-venn-diagram.-this-can-be-done-with-simple-web-tools-like"><i class="fa fa-check"></i>Visualize overlap with a venn diagram. This can be done with simple web tools like:</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ballgown-de-visualization"><i class="fa fa-check"></i>Ballgown DE Visualization</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-visualization"><i class="fa fa-check"></i>Differential Expression Visualization</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#viewing-pairwise-sample-clustering"><i class="fa fa-check"></i>Viewing pairwise sample clustering</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-i---expression-estimation"><i class="fa fa-check"></i>Part I - Expression Estimation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-ii---differential-expression"><i class="fa fa-check"></i>Part II - Differential Expression</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#presenting-your-results-1"><i class="fa fa-check"></i>Presenting Your Results</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-0-obtaining-data-and-references"><i class="fa fa-check"></i>PART 0 : Obtaining Data and References</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-1-data-preprocessing"><i class="fa fa-check"></i>Part 1 : Data preprocessing</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-2-data-alignment"><i class="fa fa-check"></i>Part 2: Data alignment</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-3-expression-estimation"><i class="fa fa-check"></i>Part 3: Expression Estimation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-4-differential-expression-analysis"><i class="fa fa-check"></i>Part 4: Differential Expression Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-5-differential-expression-analysis-visualization"><i class="fa fa-check"></i>Part 5: Differential Expression Analysis Visualization</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNA-seq Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-0-introductions-and-environment-setup" class="section level1 hasAnchor">
<h1>Module 0: Introductions and Environment Setup<a href="module-0-introductions-and-environment-setup.html#module-0-introductions-and-environment-setup" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture" class="section level2 hasAnchor">
<h2>Lecture<a href="module-0-introductions-and-environment-setup.html#lecture" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe width="640" height="360" src="https://www.youtube.com/embed/OX6Dhqx3R1s?si=Vf7tNc_SJIHdfpu0" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
</iframe>
<iframe src="https://drive.google.com/file/d/1MiLebdA0sNDn8UR1fr52fc5nXr0r58Pl/preview" width="640" height="480" allow="autoplay">
</iframe>
</div>
<div id="environment-setup" class="section level2 hasAnchor">
<h2>Environment Setup<a href="module-0-introductions-and-environment-setup.html#environment-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This tutorial assumes use of a Linux computer with an ‘x86_64’ architecture. The rest of the tutorial should be conducted in a linux Terminal session. In other words you must already be logged into the Amazon EC2 instance as described in the previous section.</p>
<p>Before proceeding you must define a global working directory by setting the environment variable: ‘RNA_HOME’
Log into a server and SET THIS BEFORE RUNNING EVERYTHING.</p>
<p>Create a working directory and set the ‘RNA_HOME’ environment variable</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="module-0-introductions-and-environment-setup.html#cb1-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/workspace/rnaseq/</span>
<span id="cb1-2"><a href="module-0-introductions-and-environment-setup.html#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="module-0-introductions-and-environment-setup.html#cb1-3" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_HOME</span><span class="op">=</span>~/workspace/rnaseq</span></code></pre></div>
<p>Make sure whatever the working dir is, that it is set and is valid</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="module-0-introductions-and-environment-setup.html#cb2-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_HOME</span></span></code></pre></div>
<p>Since all the environment variables we set up for the RNA-seq workshop start with ‘RNA’ we can easily view them all by combined use of the <code>env</code> and <code>grep</code> commands as shown below. The <code>env</code> command shows all environment variables currently defined and the <code>grep</code> command identifies string matches.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="module-0-introductions-and-environment-setup.html#cb3-1" tabindex="-1"></a><span class="fu">env</span> <span class="kw">|</span> <span class="fu">grep</span> RNA</span></code></pre></div>
<p>You can place the RNA_HOME variable (and other environment variables) in your .bashrc and then logout and login again to avoid having to worry about it. A <code>.bashrc</code> file with these variables has already been created for you.</p>
<p>In order to view the contents of this file, you can type:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="module-0-introductions-and-environment-setup.html#cb4-1" tabindex="-1"></a><span class="fu">less</span> ~/.bashrc</span></code></pre></div>
<p>To exit the file, type <code>q</code>.</p>
<p>Environment variables used throughout this tutorial:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="module-0-introductions-and-environment-setup.html#cb5-1" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_HOME</span><span class="op">=</span>~/workspace/rnaseq</span>
<span id="cb5-2"><a href="module-0-introductions-and-environment-setup.html#cb5-2" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_DATA_DIR</span><span class="op">=</span><span class="va">$RNA_HOME</span>/data</span>
<span id="cb5-3"><a href="module-0-introductions-and-environment-setup.html#cb5-3" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_DATA_TRIM_DIR</span><span class="op">=</span><span class="va">$RNA_DATA_DIR</span>/trimmed</span>
<span id="cb5-4"><a href="module-0-introductions-and-environment-setup.html#cb5-4" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_REFS_DIR</span><span class="op">=</span><span class="va">$RNA_HOME</span>/refs</span>
<span id="cb5-5"><a href="module-0-introductions-and-environment-setup.html#cb5-5" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_REF_INDEX</span><span class="op">=</span><span class="va">$RNA_REFS_DIR</span>/chr22_with_ERCC92</span>
<span id="cb5-6"><a href="module-0-introductions-and-environment-setup.html#cb5-6" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_REF_FASTA</span><span class="op">=</span><span class="va">$RNA_REF_INDEX</span>.fa</span>
<span id="cb5-7"><a href="module-0-introductions-and-environment-setup.html#cb5-7" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_REF_GTF</span><span class="op">=</span><span class="va">$RNA_REF_INDEX</span>.gtf</span>
<span id="cb5-8"><a href="module-0-introductions-and-environment-setup.html#cb5-8" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_ALIGN_DIR</span><span class="op">=</span><span class="va">$RNA_HOME</span>/alignments/hisat2</span></code></pre></div>
<p>We will be using picard tools throughout this workshop. To follow along, you will need to set an environment variable pointing to your picard installation.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="module-0-introductions-and-environment-setup.html#cb6-1" tabindex="-1"></a><span class="bu">export</span> <span class="va">PICARD</span><span class="op">=</span>/home/ubuntu/bin/picard.jar</span></code></pre></div>
<p>For simplicity, we are going to download a preconfigured <code>.bashrc</code> file to use.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="module-0-introductions-and-environment-setup.html#cb7-1" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb7-2"><a href="module-0-introductions-and-environment-setup.html#cb7-2" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/bashrc_copy</span>
<span id="cb7-3"><a href="module-0-introductions-and-environment-setup.html#cb7-3" tabindex="-1"></a><span class="fu">mv</span> bashrc_copy ~/.bashrc</span>
<span id="cb7-4"><a href="module-0-introductions-and-environment-setup.html#cb7-4" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code></pre></div>
<p>Now if you run the following command, you should see the RNA environment variables present.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="module-0-introductions-and-environment-setup.html#cb8-1" tabindex="-1"></a><span class="fu">env</span> <span class="kw">|</span> <span class="fu">grep</span> RNA</span></code></pre></div>
<p>Alternatively, you could have add these enivroment variables manually if they were not part of your .bashrc. First, you can open your .bashrc file with nano by simply typing:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="module-0-introductions-and-environment-setup.html#cb9-1" tabindex="-1"></a><span class="fu">nano</span> ~/.bashrc</span></code></pre></div>
<p>You can now see the contents of this file. Then, you want to add the above environment variables to the bottom of the file. You can do this by copying and pasting. Once you have the variables in the file, you’ll want to type <code>ctrl</code> + <code>o</code> to save the file, then <code>enter</code> to confirm you want the same filename, then <code>ctrl</code> + <code>x</code> to exit nano.</p>
<p>Again, check all the RNA related environment variables to make sure things look right.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="module-0-introductions-and-environment-setup.html#cb10-1" tabindex="-1"></a><span class="fu">env</span> <span class="kw">|</span> <span class="fu">grep</span> RNA</span></code></pre></div>
<p>Note that if you are doing this course on the Google Cloud Platform instead of AWS, you should instead use this <code>.bashrc</code> file: <a href="http://genomedata.org/rnaseq-tutorial/bashrc_copy_gcp.sh" class="uri">http://genomedata.org/rnaseq-tutorial/bashrc_copy_gcp.sh</a></p>
</div>
<div id="tool-installation" class="section level2 hasAnchor">
<h2>Tool Installation<a href="module-0-introductions-and-environment-setup.html#tool-installation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="note" class="section level3 hasAnchor">
<h3>Note<a href="module-0-introductions-and-environment-setup.html#note" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, make sure your environment is set up correctly.</p>
<hr />
<p>Tools needed for this analysis are: samtools, bam-readcount, HISAT2, stringtie, gffcompare, htseq-count, gtf_to_fasta (TopHat), kallisto, FastQC, Fastp, MultiQC, Picard, Regtools, RSeqQC, bedops, gtfToGenePred, genePredToBed, how_are_we_stranded_here, CellRanger, R, BioConductor, ballgown, and other R packages. In the following installation example, the installs are local and will work whether you have root (i.e. admin) access or not. However, if root is available some binaries can/will be copied to system-wide locations (e.g., ~/bin/).</p>
<p>Set up tool installation location:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="module-0-introductions-and-environment-setup.html#cb11-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span></span>
<span id="cb11-2"><a href="module-0-introductions-and-environment-setup.html#cb11-2" tabindex="-1"></a><span class="fu">mkdir</span> student_tools</span>
<span id="cb11-3"><a href="module-0-introductions-and-environment-setup.html#cb11-3" tabindex="-1"></a><span class="bu">cd</span> student_tools</span></code></pre></div>
</div>
<div id="samtools" class="section level3 hasAnchor">
<h3><a href="http://www.htslib.org/">SAMtools</a><a href="module-0-introductions-and-environment-setup.html#samtools" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: build C++ binary from source code using <code>make</code>. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/19505943/">PMID: 19505943</a>.</p>
<p>The following tool is installed by downloading a compressed archive using <code>wget</code>, decompressing it using <code>bunzip2</code>, unpacking the archive using <code>tar</code>, and building the source code using <code>make</code> to run compiler commands in the “Makefile” provided with the tool. When <code>make</code> is run without options, it attempts the “default goal” in the make file which is the first “target” defined. In this case the first “target” is <code>:all</code>. Once the build is complete, we test that it worked by attempting to execute the <code>samtools</code> binary. Remember that the <code>./</code> in <code>./samtools</code> tells the commandline that you want to execute the <code>samtools</code> binary in the current directory. We do this because there may be other <code>samtools</code> binaries in our PATH. Try <code>which samtools</code> to see the samtools binary that appears first in our PATH and therefore will be the one used when we specify <code>samtools</code> without specifying a particular location of the binary.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="module-0-introductions-and-environment-setup.html#cb12-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb12-2"><a href="module-0-introductions-and-environment-setup.html#cb12-2" tabindex="-1"></a><span class="fu">wget</span> https://github.com/samtools/samtools/releases/download/1.18/samtools-1.18.tar.bz2</span>
<span id="cb12-3"><a href="module-0-introductions-and-environment-setup.html#cb12-3" tabindex="-1"></a><span class="fu">bunzip2</span> samtools-1.18.tar.bz2</span>
<span id="cb12-4"><a href="module-0-introductions-and-environment-setup.html#cb12-4" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> samtools-1.18.tar</span>
<span id="cb12-5"><a href="module-0-introductions-and-environment-setup.html#cb12-5" tabindex="-1"></a><span class="bu">cd</span> samtools-1.18</span>
<span id="cb12-6"><a href="module-0-introductions-and-environment-setup.html#cb12-6" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb12-7"><a href="module-0-introductions-and-environment-setup.html#cb12-7" tabindex="-1"></a><span class="ex">./samtools</span></span></code></pre></div>
</div>
<div id="bam-readcount" class="section level3 hasAnchor">
<h3><a href="https://github.com/genome/bam-readcount">bam-readcount</a><a href="module-0-introductions-and-environment-setup.html#bam-readcount" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: build C++ binary from source code using <code>cmake</code> and <code>make</code>. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/34341766/">PMID: 34341766</a>.</p>
<p>Installation of the bam-readcount tool involves “cloning” the source code with a code version control system called <code>git</code>. The code is then compiled using <code>cmake</code> and <code>make</code>. <code>cmake</code> is an application for managing the build process of software using a compiler-independent method. It is used in conjunction with native build environments such as <code>make</code> (<a href="https://en.wikipedia.org/wiki/CMake">cmake ref</a>). Note that bam-readcount relies on another tool, samtools, as a dependency. An environment variable is used to specify the path to the samtools install.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="module-0-introductions-and-environment-setup.html#cb13-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb13-2"><a href="module-0-introductions-and-environment-setup.html#cb13-2" tabindex="-1"></a><span class="bu">export</span> <span class="va">SAMTOOLS_ROOT</span><span class="op">=</span><span class="va">$RNA_HOME</span>/student_tools/samtools-1.18</span>
<span id="cb13-3"><a href="module-0-introductions-and-environment-setup.html#cb13-3" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/genome/bam-readcount </span>
<span id="cb13-4"><a href="module-0-introductions-and-environment-setup.html#cb13-4" tabindex="-1"></a><span class="bu">cd</span> bam-readcount</span>
<span id="cb13-5"><a href="module-0-introductions-and-environment-setup.html#cb13-5" tabindex="-1"></a><span class="fu">mkdir</span> build</span>
<span id="cb13-6"><a href="module-0-introductions-and-environment-setup.html#cb13-6" tabindex="-1"></a><span class="bu">cd</span> build</span>
<span id="cb13-7"><a href="module-0-introductions-and-environment-setup.html#cb13-7" tabindex="-1"></a><span class="fu">cmake</span> ..</span>
<span id="cb13-8"><a href="module-0-introductions-and-environment-setup.html#cb13-8" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb13-9"><a href="module-0-introductions-and-environment-setup.html#cb13-9" tabindex="-1"></a><span class="ex">./bin/bam-readcount</span></span></code></pre></div>
</div>
<div id="hisat2" class="section level3 hasAnchor">
<h3><a href="https://ccb.jhu.edu/software/hisat2/index.shtml">HISAT2</a><a href="module-0-introductions-and-environment-setup.html#hisat2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download a precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/31375807/">PMID: 31375807</a>.</p>
<p>The <code>hisat2</code> aligner is installed below by simply downloading an archive of binaries using <code>wget</code>, unpacking them with <code>unzip</code>, and testing the tool to make sure it executes without error on the current system. This approach relies on understanding the architecture of your system and downloading the correct precompiled binary. The <code>uname -m</code> command lists the current system architecture.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="module-0-introductions-and-environment-setup.html#cb14-1" tabindex="-1"></a><span class="fu">uname</span> <span class="at">-m</span></span>
<span id="cb14-2"><a href="module-0-introductions-and-environment-setup.html#cb14-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb14-3"><a href="module-0-introductions-and-environment-setup.html#cb14-3" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-s</span> https://cloud.biohpc.swmed.edu/index.php/s/oTtGWbWjaxsQ2Ho/download <span class="op">&gt;</span> hisat2-2.2.1-Linux_x86_64.zip</span>
<span id="cb14-4"><a href="module-0-introductions-and-environment-setup.html#cb14-4" tabindex="-1"></a><span class="fu">unzip</span> hisat2-2.2.1-Linux_x86_64.zip</span>
<span id="cb14-5"><a href="module-0-introductions-and-environment-setup.html#cb14-5" tabindex="-1"></a><span class="bu">cd</span> hisat2-2.2.1</span>
<span id="cb14-6"><a href="module-0-introductions-and-environment-setup.html#cb14-6" tabindex="-1"></a><span class="ex">./hisat2</span> <span class="at">-h</span></span></code></pre></div>
</div>
<div id="stringtie" class="section level3 hasAnchor">
<h3><a href="https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual">StringTie</a><a href="module-0-introductions-and-environment-setup.html#stringtie" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download a precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/25690850/">PMID: 25690850</a>.</p>
<p>The <code>stringtie</code> reference guided transcript assembly and abundance estimation tool is installed below by simply downloading an archive with <code>wget</code>, unpacking the archive with <code>tar</code>, and executing <code>stringtie</code> to confirm it runs without error on our system.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="module-0-introductions-and-environment-setup.html#cb15-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb15-2"><a href="module-0-introductions-and-environment-setup.html#cb15-2" tabindex="-1"></a><span class="fu">wget</span> http://ccb.jhu.edu/software/stringtie/dl/stringtie-2.2.1.tar.gz</span>
<span id="cb15-3"><a href="module-0-introductions-and-environment-setup.html#cb15-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzvf</span> stringtie-2.2.1.tar.gz</span>
<span id="cb15-4"><a href="module-0-introductions-and-environment-setup.html#cb15-4" tabindex="-1"></a><span class="bu">cd</span> stringtie-2.2.1</span>
<span id="cb15-5"><a href="module-0-introductions-and-environment-setup.html#cb15-5" tabindex="-1"></a><span class="fu">make</span> release</span>
<span id="cb15-6"><a href="module-0-introductions-and-environment-setup.html#cb15-6" tabindex="-1"></a><span class="ex">./stringtie</span> <span class="at">-h</span></span></code></pre></div>
</div>
<div id="gffcompare" class="section level3 hasAnchor">
<h3><a href="http://ccb.jhu.edu/software/stringtie/gff.shtml#gffcompare">gffcompare</a><a href="module-0-introductions-and-environment-setup.html#gffcompare" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download a precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/25690850/">PMID: 25690850</a>.</p>
<p>The <code>gffcompare</code> tool for comparing transcript annotations is installed below by simply downloading an archive with <code>wget</code>, unpacking it with <code>tar</code>, and executing <code>gffcompare</code> to ensure it runs without error on our system.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="module-0-introductions-and-environment-setup.html#cb16-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb16-2"><a href="module-0-introductions-and-environment-setup.html#cb16-2" tabindex="-1"></a><span class="fu">wget</span> http://ccb.jhu.edu/software/stringtie/dl/gffcompare-0.12.6.Linux_x86_64.tar.gz</span>
<span id="cb16-3"><a href="module-0-introductions-and-environment-setup.html#cb16-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzvf</span> gffcompare-0.12.6.Linux_x86_64.tar.gz</span>
<span id="cb16-4"><a href="module-0-introductions-and-environment-setup.html#cb16-4" tabindex="-1"></a><span class="bu">cd</span> gffcompare-0.12.6.Linux_x86_64/</span>
<span id="cb16-5"><a href="module-0-introductions-and-environment-setup.html#cb16-5" tabindex="-1"></a><span class="ex">./gffcompare</span></span></code></pre></div>
</div>
<div id="htseq-count" class="section level3 hasAnchor">
<h3><a href="https://htseq.readthedocs.io/en/master/install.html">htseq-count</a><a href="module-0-introductions-and-environment-setup.html#htseq-count" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: pip install. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/25260700/">PMID: 25260700</a>.</p>
<p>The htseq-count read counting tools is a python package that can be installed using <code>pip</code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="module-0-introductions-and-environment-setup.html#cb17-1" tabindex="-1"></a><span class="ex">pip3</span> install <span class="at">--user</span> HTSeq</span></code></pre></div>
</div>
<div id="tophat" class="section level3 hasAnchor">
<h3><a href="https://ccb.jhu.edu/software/tophat/index.shtml">TopHat</a><a href="module-0-introductions-and-environment-setup.html#tophat" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: dowload a precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/19289445/">PMID: 19289445</a>.</p>
<p>Note, this tool is currently only installed for the gtf_to_fasta tool used in kallisto section.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="module-0-introductions-and-environment-setup.html#cb18-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb18-2"><a href="module-0-introductions-and-environment-setup.html#cb18-2" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/tophat-2.1.1.Linux_x86_64.tar.gz</span>
<span id="cb18-3"><a href="module-0-introductions-and-environment-setup.html#cb18-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-zxvf</span> tophat-2.1.1.Linux_x86_64.tar.gz</span>
<span id="cb18-4"><a href="module-0-introductions-and-environment-setup.html#cb18-4" tabindex="-1"></a><span class="bu">cd</span> tophat-2.1.1.Linux_x86_64/</span>
<span id="cb18-5"><a href="module-0-introductions-and-environment-setup.html#cb18-5" tabindex="-1"></a><span class="ex">./gtf_to_fasta</span></span></code></pre></div>
</div>
<div id="kallisto" class="section level3 hasAnchor">
<h3><a href="https://pachterlab.github.io/kallisto/">kallisto</a><a href="module-0-introductions-and-environment-setup.html#kallisto" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download a precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/27043002/">PMID: 27043002</a>.</p>
<p>The kallisto alignment free expression estimation tool is installed below simply by downloading an archive with <code>wget</code>, unpacking the archive with <code>tar</code>, and testing the binary to ensure it runs on our system.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="module-0-introductions-and-environment-setup.html#cb19-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb19-2"><a href="module-0-introductions-and-environment-setup.html#cb19-2" tabindex="-1"></a><span class="fu">wget</span> https://github.com/pachterlab/kallisto/releases/download/v0.44.0/kallisto_linux-v0.44.0.tar.gz</span>
<span id="cb19-3"><a href="module-0-introductions-and-environment-setup.html#cb19-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-zxvf</span> kallisto_linux-v0.44.0.tar.gz</span>
<span id="cb19-4"><a href="module-0-introductions-and-environment-setup.html#cb19-4" tabindex="-1"></a><span class="bu">cd</span> kallisto_linux-v0.44.0/</span>
<span id="cb19-5"><a href="module-0-introductions-and-environment-setup.html#cb19-5" tabindex="-1"></a><span class="ex">./kallisto</span></span></code></pre></div>
</div>
<div id="fastqc" class="section level3 hasAnchor">
<h3><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a><a href="module-0-introductions-and-environment-setup.html#fastqc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download precompiled binary. Citation: <a href="https://github.com/s-andrews/FastQC">s-andrews/FastQC</a>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="module-0-introductions-and-environment-setup.html#cb20-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb20-2"><a href="module-0-introductions-and-environment-setup.html#cb20-2" tabindex="-1"></a><span class="fu">wget</span> https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip</span>
<span id="cb20-3"><a href="module-0-introductions-and-environment-setup.html#cb20-3" tabindex="-1"></a><span class="fu">unzip</span> fastqc_v0.12.1.zip</span>
<span id="cb20-4"><a href="module-0-introductions-and-environment-setup.html#cb20-4" tabindex="-1"></a><span class="bu">cd</span> FastQC/</span>
<span id="cb20-5"><a href="module-0-introductions-and-environment-setup.html#cb20-5" tabindex="-1"></a><span class="fu">chmod</span> 755 fastqc</span>
<span id="cb20-6"><a href="module-0-introductions-and-environment-setup.html#cb20-6" tabindex="-1"></a><span class="ex">./fastqc</span> <span class="at">--help</span></span></code></pre></div>
</div>
<div id="fastp" class="section level3 hasAnchor">
<h3><a href="https://github.com/OpenGene/fastp">Fastp</a><a href="module-0-introductions-and-environment-setup.html#fastp" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/30423086/">PMID: 30423086</a></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="module-0-introductions-and-environment-setup.html#cb21-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb21-2"><a href="module-0-introductions-and-environment-setup.html#cb21-2" tabindex="-1"></a><span class="fu">mkdir</span> fastp</span>
<span id="cb21-3"><a href="module-0-introductions-and-environment-setup.html#cb21-3" tabindex="-1"></a><span class="bu">cd</span> fastp</span>
<span id="cb21-4"><a href="module-0-introductions-and-environment-setup.html#cb21-4" tabindex="-1"></a><span class="fu">wget</span> http://opengene.org/fastp/fastp</span>
<span id="cb21-5"><a href="module-0-introductions-and-environment-setup.html#cb21-5" tabindex="-1"></a><span class="fu">chmod</span> a+x ./fastp</span>
<span id="cb21-6"><a href="module-0-introductions-and-environment-setup.html#cb21-6" tabindex="-1"></a><span class="ex">./fastp</span></span></code></pre></div>
</div>
<div id="multiqc" class="section level3 hasAnchor">
<h3><a href="http://multiqc.info/">MultiQC</a><a href="module-0-introductions-and-environment-setup.html#multiqc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: use pip. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/27312411/">PMID: 27312411</a>.</p>
<p>Multiqc, a tool for assembling QC reports is a python package that can be installed using the python package manager <code>pip</code>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="module-0-introductions-and-environment-setup.html#cb22-1" tabindex="-1"></a><span class="ex">pip3</span> install multiqc</span>
<span id="cb22-2"><a href="module-0-introductions-and-environment-setup.html#cb22-2" tabindex="-1"></a><span class="bu">export</span> <span class="va">PATH</span><span class="op">=</span>/home/ubuntu/.local/bin:<span class="va">$PATH</span></span>
<span id="cb22-3"><a href="module-0-introductions-and-environment-setup.html#cb22-3" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--help</span></span></code></pre></div>
</div>
<div id="picard" class="section level3 hasAnchor">
<h3><a href="https://broadinstitute.github.io/picard/">Picard</a><a href="module-0-introductions-and-environment-setup.html#picard" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download java jar file. Citation: <a href="https://github.com/broadinstitute/picard">broadinstitute/picard</a>.</p>
<p>Picard is a rich tool kit for BAM file manipulation that is installed below simply by downloading a jar file. The jar file is tested using Java, a dependency that must also be installed (it should already be present in many systems).</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="module-0-introductions-and-environment-setup.html#cb23-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb23-2"><a href="module-0-introductions-and-environment-setup.html#cb23-2" tabindex="-1"></a><span class="fu">wget</span> https://github.com/broadinstitute/picard/releases/download/2.26.4/picard.jar <span class="at">-O</span> picard.jar</span>
<span id="cb23-3"><a href="module-0-introductions-and-environment-setup.html#cb23-3" tabindex="-1"></a><span class="ex">java</span> <span class="at">-jar</span> <span class="va">$RNA_HOME</span>/student_tools/picard.jar</span></code></pre></div>
</div>
<div id="regtools" class="section level3 hasAnchor">
<h3><a href="https://github.com/griffithlab/regtools#regtools">RegTools</a><a href="module-0-introductions-and-environment-setup.html#regtools" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: compile from source code using <code>cmake</code> and <code>make</code>. Citation: <a href="https://www.biorxiv.org/content/10.1101/436634v2">bioRXiv: 10.1101/436634v2</a>.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="module-0-introductions-and-environment-setup.html#cb24-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb24-2"><a href="module-0-introductions-and-environment-setup.html#cb24-2" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/griffithlab/regtools</span>
<span id="cb24-3"><a href="module-0-introductions-and-environment-setup.html#cb24-3" tabindex="-1"></a><span class="bu">cd</span> regtools/</span>
<span id="cb24-4"><a href="module-0-introductions-and-environment-setup.html#cb24-4" tabindex="-1"></a><span class="fu">mkdir</span> build</span>
<span id="cb24-5"><a href="module-0-introductions-and-environment-setup.html#cb24-5" tabindex="-1"></a><span class="bu">cd</span> build/</span>
<span id="cb24-6"><a href="module-0-introductions-and-environment-setup.html#cb24-6" tabindex="-1"></a><span class="fu">cmake</span> ..</span>
<span id="cb24-7"><a href="module-0-introductions-and-environment-setup.html#cb24-7" tabindex="-1"></a><span class="fu">make</span></span>
<span id="cb24-8"><a href="module-0-introductions-and-environment-setup.html#cb24-8" tabindex="-1"></a><span class="ex">./regtools</span></span></code></pre></div>
</div>
<div id="rseqc" class="section level3 hasAnchor">
<h3><a href="http://rseqc.sourceforge.net/">RSeQC</a><a href="module-0-introductions-and-environment-setup.html#rseqc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: use pip. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/22743226/">PMID: 22743226</a>.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb25-1"><a href="module-0-introductions-and-environment-setup.html#cb25-1" tabindex="-1"></a><span class="ex">pip3</span> install RSeQC</span>
<span id="cb25-2"><a href="module-0-introductions-and-environment-setup.html#cb25-2" tabindex="-1"></a><span class="ex">read_GC.py</span></span></code></pre></div>
</div>
<div id="bedops" class="section level3 hasAnchor">
<h3><a href="https://bedops.readthedocs.io/en/latest/">bedops</a><a href="module-0-introductions-and-environment-setup.html#bedops" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download precompiled binary. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/22576172/">PMID: 22576172</a>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb26-1"><a href="module-0-introductions-and-environment-setup.html#cb26-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb26-2"><a href="module-0-introductions-and-environment-setup.html#cb26-2" tabindex="-1"></a><span class="fu">mkdir</span> bedops_linux_x86_64-v2.4.41</span>
<span id="cb26-3"><a href="module-0-introductions-and-environment-setup.html#cb26-3" tabindex="-1"></a><span class="bu">cd</span> bedops_linux_x86_64-v2.4.41</span>
<span id="cb26-4"><a href="module-0-introductions-and-environment-setup.html#cb26-4" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> https://github.com/bedops/bedops/releases/download/v2.4.41/bedops_linux_x86_64-v2.4.41.tar.bz2</span>
<span id="cb26-5"><a href="module-0-introductions-and-environment-setup.html#cb26-5" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-jxvf</span> bedops_linux_x86_64-v2.4.41.tar.bz2</span>
<span id="cb26-6"><a href="module-0-introductions-and-environment-setup.html#cb26-6" tabindex="-1"></a><span class="ex">./bin/bedops</span></span></code></pre></div>
</div>
<div id="gtftogenepred" class="section level3 hasAnchor">
<h3><a href="https://bioconda.github.io/recipes/ucsc-gtftogenepred/README.html">gtfToGenePred</a><a href="module-0-introductions-and-environment-setup.html#gtftogenepred" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download precompiled binary.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb27-1"><a href="module-0-introductions-and-environment-setup.html#cb27-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb27-2"><a href="module-0-introductions-and-environment-setup.html#cb27-2" tabindex="-1"></a><span class="fu">mkdir</span> gtfToGenePred</span>
<span id="cb27-3"><a href="module-0-introductions-and-environment-setup.html#cb27-3" tabindex="-1"></a><span class="bu">cd</span> gtfToGenePred</span>
<span id="cb27-4"><a href="module-0-introductions-and-environment-setup.html#cb27-4" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/gtfToGenePred</span>
<span id="cb27-5"><a href="module-0-introductions-and-environment-setup.html#cb27-5" tabindex="-1"></a><span class="fu">chmod</span> a+x gtfToGenePred</span>
<span id="cb27-6"><a href="module-0-introductions-and-environment-setup.html#cb27-6" tabindex="-1"></a><span class="ex">./gtfToGenePred</span></span></code></pre></div>
</div>
<div id="genepredtobed" class="section level3 hasAnchor">
<h3><a href="https://bioconda.github.io/recipes/ucsc-genepredtobed/README.html">genePredToBed</a><a href="module-0-introductions-and-environment-setup.html#genepredtobed" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: download precompiled binary.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb28-1"><a href="module-0-introductions-and-environment-setup.html#cb28-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb28-2"><a href="module-0-introductions-and-environment-setup.html#cb28-2" tabindex="-1"></a><span class="fu">mkdir</span> genePredToBed</span>
<span id="cb28-3"><a href="module-0-introductions-and-environment-setup.html#cb28-3" tabindex="-1"></a><span class="bu">cd</span> genePredToBed</span>
<span id="cb28-4"><a href="module-0-introductions-and-environment-setup.html#cb28-4" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-c</span> http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/genePredToBed</span>
<span id="cb28-5"><a href="module-0-introductions-and-environment-setup.html#cb28-5" tabindex="-1"></a><span class="fu">chmod</span> a+x genePredToBed</span>
<span id="cb28-6"><a href="module-0-introductions-and-environment-setup.html#cb28-6" tabindex="-1"></a><span class="ex">./genePredToBed</span></span></code></pre></div>
</div>
<div id="how_are_we_stranded_here" class="section level3 hasAnchor">
<h3><a href="https://github.com/betsig/how_are_we_stranded_here">how_are_we_stranded_here</a><a href="module-0-introductions-and-environment-setup.html#how_are_we_stranded_here" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb29-1"><a href="module-0-introductions-and-environment-setup.html#cb29-1" tabindex="-1"></a><span class="ex">pip3</span> install git+https://github.com/kcotto/how_are_we_stranded_here.git</span>
<span id="cb29-2"><a href="module-0-introductions-and-environment-setup.html#cb29-2" tabindex="-1"></a><span class="ex">check_strandedness</span></span></code></pre></div>
</div>
<div id="install-cell-ranger" class="section level3 hasAnchor">
<h3>Install <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation">Cell Ranger</a><a href="module-0-introductions-and-environment-setup.html#install-cell-ranger" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Must register to get download link, modify command below to match downloaded tar</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb30-1"><a href="module-0-introductions-and-environment-setup.html#cb30-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span>
<span id="cb30-2"><a href="module-0-introductions-and-environment-setup.html#cb30-2" tabindex="-1"></a><span class="fu">wget</span> <span class="kw">`</span><span class="ex">download_link</span><span class="kw">`</span></span>
<span id="cb30-3"><a href="module-0-introductions-and-environment-setup.html#cb30-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzvf</span> cellranger-7.2.0.tar.gz</span></code></pre></div>
</div>
<div id="install-r" class="section level3 hasAnchor">
<h3>Install <a href="http://www.r-project.org/">R</a><a href="module-0-introductions-and-environment-setup.html#install-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb31-1"><a href="module-0-introductions-and-environment-setup.html#cb31-1" tabindex="-1"></a><span class="co">#sudo apt-get remove r-base-core</span></span>
<span id="cb31-2"><a href="module-0-introductions-and-environment-setup.html#cb31-2" tabindex="-1"></a></span>
<span id="cb31-3"><a href="module-0-introductions-and-environment-setup.html#cb31-3" tabindex="-1"></a><span class="co">#wget -qO- https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc | sudo gpg --dearmor -o /usr/share/keyrings/r-project.gpg</span></span>
<span id="cb31-4"><a href="module-0-introductions-and-environment-setup.html#cb31-4" tabindex="-1"></a><span class="co">#echo &quot;deb [signed-by=/usr/share/keyrings/r-project.gpg] https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/&quot; | sudo tee -a /etc/apt/sources.list.d/r-project.list</span></span>
<span id="cb31-5"><a href="module-0-introductions-and-environment-setup.html#cb31-5" tabindex="-1"></a><span class="co">#sudo apt update</span></span>
<span id="cb31-6"><a href="module-0-introductions-and-environment-setup.html#cb31-6" tabindex="-1"></a><span class="co">#sudo apt install --no-install-recommends r-base</span></span></code></pre></div>
<p>Note, if X11 libraries are not available you may need to use <code>--with-x=no</code> during config, on a regular linux system you would not use this option.
Also, linking the R-patched <code>bin</code> directory into your <code>PATH</code> may cause weird things to happen, such as man pages or <code>git log</code> to not display. This can be circumvented by directly linking the <code>R*</code> executables (<code>R</code>, <code>RScript</code>, <code>RCmd</code>, etc.) into a <code>PATH</code> directory.</p>
</div>
<div id="r-libraries" class="section level3 hasAnchor">
<h3>R Libraries<a href="module-0-introductions-and-environment-setup.html#r-libraries" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: add new base R libraries to an R installation.</p>
<p>For this tutorial we require:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/devtools/index.html">devtools</a></li>
<li><a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a></li>
<li><a href="http://cran.r-project.org/web/packages/gplots/index.html">gplots</a></li>
<li><a href="https://ggplot2.tidyverse.org/">ggplot2</a></li>
</ul>
<p>launch R (enter <code>R</code> at linux command prompt) and type the following at an R command prompt. NOTE: This has been pre-installed for you, so these commands can be skipped.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb32-1"><a href="module-0-introductions-and-environment-setup.html#cb32-1" tabindex="-1"></a><span class="co">#R</span></span>
<span id="cb32-2"><a href="module-0-introductions-and-environment-setup.html#cb32-2" tabindex="-1"></a><span class="co">#install.packages(c(&quot;devtools&quot;,&quot;dplyr&quot;,&quot;gplots&quot;,&quot;ggplot2&quot;),repos=&quot;http://cran.us.r-project.org&quot;)</span></span>
<span id="cb32-3"><a href="module-0-introductions-and-environment-setup.html#cb32-3" tabindex="-1"></a><span class="co">#quit(save=&quot;no&quot;)</span></span></code></pre></div>
</div>
<div id="bioconductor" class="section level3 hasAnchor">
<h3><a href="http://www.bioconductor.org/">Bioconductor</a><a href="module-0-introductions-and-environment-setup.html#bioconductor" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: add bioconductor libraries to an R installation. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/15461798/">PMID: 15461798</a>.</p>
<p>For this tutorial we require:</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/html/genefilter.html">genefilter</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/ballgown.html">ballgown</a>. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/25748911/">PMID: 25748911</a>.</li>
<li><a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/19910308/">PMID: 19910308</a>.</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html">rhdf5</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/21930506/">PMID: 21930506</a>.</li>
</ul>
<p>launch R (enter <code>R</code> at linux command prompt) and type the following at an R command prompt. If prompted, type “a” to update all old packages. NOTE: This has been pre-installed for you, so these commands can be skipped.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb33-1"><a href="module-0-introductions-and-environment-setup.html#cb33-1" tabindex="-1"></a><span class="co">#R</span></span>
<span id="cb33-2"><a href="module-0-introductions-and-environment-setup.html#cb33-2" tabindex="-1"></a><span class="co">#source(&quot;http://bioconductor.org/biocLite.R&quot;)</span></span>
<span id="cb33-3"><a href="module-0-introductions-and-environment-setup.html#cb33-3" tabindex="-1"></a><span class="co">#biocLite(c(&quot;genefilter&quot;,&quot;ballgown&quot;,&quot;edgeR&quot;,&quot;GenomicRanges&quot;,&quot;rhdf5&quot;,&quot;biomaRt&quot;))</span></span>
<span id="cb33-4"><a href="module-0-introductions-and-environment-setup.html#cb33-4" tabindex="-1"></a><span class="co">#quit(save=&quot;no&quot;)</span></span></code></pre></div>
</div>
<div id="sleuth" class="section level3 hasAnchor">
<h3><a href="https://pachterlab.github.io/sleuth/download">Sleuth</a><a href="module-0-introductions-and-environment-setup.html#sleuth" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Installation type: R package installation from a git repository. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/28581496/">PMID: 28581496</a>.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb34-1"><a href="module-0-introductions-and-environment-setup.html#cb34-1" tabindex="-1"></a><span class="co">#R</span></span>
<span id="cb34-2"><a href="module-0-introductions-and-environment-setup.html#cb34-2" tabindex="-1"></a><span class="co">#install.packages(&quot;devtools&quot;)</span></span>
<span id="cb34-3"><a href="module-0-introductions-and-environment-setup.html#cb34-3" tabindex="-1"></a><span class="co">#devtools::install_github(&quot;pachterlab/sleuth&quot;)</span></span>
<span id="cb34-4"><a href="module-0-introductions-and-environment-setup.html#cb34-4" tabindex="-1"></a><span class="co">#quit(save=&quot;no&quot;)</span></span></code></pre></div>
<hr />
</div>
<div id="practical-exercise-1---software-installation" class="section level3 hasAnchor">
<h3>PRACTICAL EXERCISE 1 - Software Installation<a href="module-0-introductions-and-environment-setup.html#practical-exercise-1---software-installation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Assignment: Install bedtools on your own. Make sure you install it in your tools folder. Download, unpack, compile, and test the bedtools software. Citation: <a href="https://pubmed.ncbi.nlm.nih.gov/20110278/">PMID: 20110278</a>.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb35-1"><a href="module-0-introductions-and-environment-setup.html#cb35-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/student_tools/</span></code></pre></div>
<ul>
<li>Hint: google “bedtools” to find the source code</li>
<li>Hint: there is a README file that will give you hints on how to install</li>
<li>Hint: If your install has worked you should be able to run bedtools as follows:</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb36-1"><a href="module-0-introductions-and-environment-setup.html#cb36-1" tabindex="-1"></a><span class="va">$RNA_HOME</span><span class="ex">/student_tools/bedtools2/bin/bedtools</span></span></code></pre></div>
<div id="questions" class="section level4 hasAnchor">
<h4>Questions<a href="module-0-introductions-and-environment-setup.html#questions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>What happens when you run bedtools without any options?</li>
<li>Where can you find detailed documentation on how to use bedtools?</li>
<li>How many general categories of analysis can you perform with bedtools? What are they?</li>
</ul>
<p>Solution: When you are ready you can check your approach against the <a href="">Solutions</a></p>
<hr />
</div>
</div>
<div id="add-locally-installed-tools-to-your-path-optional" class="section level3 hasAnchor">
<h3>Add locally installed tools to your PATH [OPTIONAL]<a href="module-0-introductions-and-environment-setup.html#add-locally-installed-tools-to-your-path-optional" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To use the locally installed version of each tool without having to specify complete paths, you could add the install directory of each tool to your ‘$PATH’ variable and set some other environment variables:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb37-1"><a href="module-0-introductions-and-environment-setup.html#cb37-1" tabindex="-1"></a><span class="va">PATH</span><span class="op">=</span><span class="va">$RNA_HOME</span>/student_tools/genePredToBed:<span class="va">$RNA_HOME</span>/student_tools/gtfToGenePred:<span class="va">$RNA_HOME</span>/student_tools/bedops_linux_x86_64-v2.4.41/bin:<span class="va">$RNA_HOME</span>/student_tools/samtools-1.18:<span class="va">$RNA_HOME</span>/student_tools/bam-readcount/bin:<span class="va">$RNA_HOME</span>/student_tools/hisat2-2.2.1:<span class="va">$RNA_HOME</span>/student_tools/stringtie-2.2.1:<span class="va">$RNA_HOME</span>/student_tools/gffcompare-0.12.6.Linux_x86_64:<span class="va">$RNA_HOME</span>/student_tools/tophat-2.1.1.Linux_x86_64:<span class="va">$RNA_HOME</span>/student_tools/kallisto_linux-v0.44.0:<span class="va">$RNA_HOME</span>/student_tools/FastQC:<span class="va">$RNA_HOME</span>/student_tools/fastp:<span class="va">$RNA_HOME</span>/student_tools/regtools/build:/home/ubuntu/bin/bedtools2/bin:/home/ubuntu/.local/bin:<span class="va">$PATH</span></span>
<span id="cb37-2"><a href="module-0-introductions-and-environment-setup.html#cb37-2" tabindex="-1"></a></span>
<span id="cb37-3"><a href="module-0-introductions-and-environment-setup.html#cb37-3" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$PATH</span></span>
<span id="cb37-4"><a href="module-0-introductions-and-environment-setup.html#cb37-4" tabindex="-1"></a></span>
<span id="cb37-5"><a href="module-0-introductions-and-environment-setup.html#cb37-5" tabindex="-1"></a><span class="bu">export</span> <span class="va">PICARD</span><span class="op">=</span><span class="va">$RNA_HOME</span>/student_tools/picard.jar</span></code></pre></div>
<p>You can make these changes permanent by adding the above lines to your .bashrc file
use a text editor to open your bashrc file. For example:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb38-1"><a href="module-0-introductions-and-environment-setup.html#cb38-1" tabindex="-1"></a><span class="fu">vi</span> ~/.bashrc</span></code></pre></div>
<div id="vi-instructions" class="section level4 hasAnchor">
<h4>Vi instructions<a href="module-0-introductions-and-environment-setup.html#vi-instructions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ol style="list-style-type: decimal">
<li>Using your cursor, navigate down to the “export PATH” commands at the end of the file.</li>
<li>Delete the line starting with PATH using the vi command “dd”.</li>
<li>Press the “i” key to enter insert mode. Go to an empty line with you cursor and copy paste the new RNA_HOME and PATH commands into the file</li>
<li>Press the “esc” key to exit insert mode.</li>
<li>Press the “:” key to enter command mode.</li>
<li>Type “wq” to save and quit vi</li>
</ol>
<p>If you would like to learn more about how to use vi, try this tutorial/game: <a href="http://vim-adventures.com/">VIM Adventures</a></p>
<p>NOTE: If you are worried your .bashrc is messed up you can redownload as follows:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb39-1"><a href="module-0-introductions-and-environment-setup.html#cb39-1" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb39-2"><a href="module-0-introductions-and-environment-setup.html#cb39-2" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/bashrc_copy</span>
<span id="cb39-3"><a href="module-0-introductions-and-environment-setup.html#cb39-3" tabindex="-1"></a><span class="fu">mv</span> bashrc_copy ~/.bashrc</span>
<span id="cb39-4"><a href="module-0-introductions-and-environment-setup.html#cb39-4" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code></pre></div>
</div>
</div>
<div id="installing-tools-from-official-ubuntu-packages-optional" class="section level3 hasAnchor">
<h3>Installing tools from official ubuntu packages [OPTIONAL]<a href="module-0-introductions-and-environment-setup.html#installing-tools-from-official-ubuntu-packages-optional" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Some useful tools are available as official ubuntu packages. These can be installed using the linux package management system <code>apt</code>. Most bioinformatic tools (especially the latest versions) are not available as official packages. Nevertheless, here is how you would update your <code>apt</code> library, upgrade existing packages, and install an Ubuntu tool called <code>tree</code>.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb40-1"><a href="module-0-introductions-and-environment-setup.html#cb40-1" tabindex="-1"></a><span class="co">#sudo apt-get update</span></span>
<span id="cb40-2"><a href="module-0-introductions-and-environment-setup.html#cb40-2" tabindex="-1"></a><span class="co">#sudo apt-get upgrade</span></span>
<span id="cb40-3"><a href="module-0-introductions-and-environment-setup.html#cb40-3" tabindex="-1"></a><span class="co">#sudo apt-get install tree</span></span>
<span id="cb40-4"><a href="module-0-introductions-and-environment-setup.html#cb40-4" tabindex="-1"></a><span class="co">#tree</span></span></code></pre></div>
</div>
<div id="installing-tools-by-docker-image" class="section level3 hasAnchor">
<h3>Installing tools by Docker image<a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Some tools have complex dependencies that are difficult to reproduce across systems or make work in the same environment with tools that require different versions of the same dependencies. Container systems such as Docker and Singularity allow you to isolate a tool’s environment giving you almost complete control over dependency issues. For this reason, many tool developers have started to distribute their tools as docker images. Many of these are placed in container image repositories such as <a href="https://hub.docker.com/">DockerHub</a>. Here is an example tool installation using <code>docker</code>.</p>
<p>Install samtools:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb41-1"><a href="module-0-introductions-and-environment-setup.html#cb41-1" tabindex="-1"></a><span class="ex">docker</span> pull biocontainers/samtools:v1.9-4-deb_cv1</span>
<span id="cb41-2"><a href="module-0-introductions-and-environment-setup.html#cb41-2" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-t</span> biocontainers/samtools:v1.9-4-deb_cv1 samtools <span class="at">--help</span></span></code></pre></div>
<p>Install pvactools for personalized cancer vaccine designs:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb42-1"><a href="module-0-introductions-and-environment-setup.html#cb42-1" tabindex="-1"></a><span class="co">#docker pull griffithlab/pvactools:latest</span></span>
<span id="cb42-2"><a href="module-0-introductions-and-environment-setup.html#cb42-2" tabindex="-1"></a><span class="co">#docker run -t griffithlab/pvactools:latest pvacseq --help</span></span></code></pre></div>
</div>
<div id="installing-tools-by-docker-image-using-singularity" class="section level3 hasAnchor">
<h3>Installing tools by Docker image (using Singularity)<a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image-using-singularity" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Some systems do not allow <code>docker</code> to be run for various reasons. Sometimes <code>singularity</code> is used instead. The equivalent to the above but using singularity looks like the following:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb43-1"><a href="module-0-introductions-and-environment-setup.html#cb43-1" tabindex="-1"></a><span class="co">#singularity pull docker://griffithlab/pvactools:latest</span></span>
<span id="cb43-2"><a href="module-0-introductions-and-environment-setup.html#cb43-2" tabindex="-1"></a><span class="co">#singularity run docker://griffithlab/pvactools:latest pvacseq -h</span></span></code></pre></div>
<p>Note that if you encounter errors with /tmp space usage or would like to control where singularity stores its temp files, you can set the environment variables:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb44-1"><a href="module-0-introductions-and-environment-setup.html#cb44-1" tabindex="-1"></a><span class="co">#export SINGULARITY_CACHEDIR=/media/workspace/.singularity</span></span>
<span id="cb44-2"><a href="module-0-introductions-and-environment-setup.html#cb44-2" tabindex="-1"></a><span class="co">#export TMPDIR=/media/workspace/temp</span></span></code></pre></div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="meet-your-faculty.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/RNA_2025/blob/main/010-module-1.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
