<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3 | RNA-seq Analysis 2025</title>
  <meta name="description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3 | RNA-seq Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/RNA_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3 | RNA-seq Analysis 2025" />
  
  <meta name="twitter:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Malachi Griffith and Obi Griffith" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-2.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>RNA-Seq Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_RNA_Icon.png" width="70%" alt="WORKSHOP NAME logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html"><i class="fa fa-check"></i>Module 0: Introductions and Environment Setup</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#environment-setup"><i class="fa fa-check"></i>Environment Setup</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tool-installation"><i class="fa fa-check"></i>Tool Installation</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#note"><i class="fa fa-check"></i>Note</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#samtools"><i class="fa fa-check"></i>SAMtools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bam-readcount"><i class="fa fa-check"></i>bam-readcount</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#hisat2"><i class="fa fa-check"></i>HISAT2</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#stringtie"><i class="fa fa-check"></i>StringTie</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gffcompare"><i class="fa fa-check"></i>gffcompare</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#htseq-count"><i class="fa fa-check"></i>htseq-count</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tophat"><i class="fa fa-check"></i>TopHat</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#kallisto"><i class="fa fa-check"></i>kallisto</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastqc"><i class="fa fa-check"></i>FastQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastp"><i class="fa fa-check"></i>Fastp</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#multiqc"><i class="fa fa-check"></i>MultiQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#picard"><i class="fa fa-check"></i>Picard</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#regtools"><i class="fa fa-check"></i>RegTools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#rseqc"><i class="fa fa-check"></i>RSeQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bedops"><i class="fa fa-check"></i>bedops</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gtftogenepred"><i class="fa fa-check"></i>gtfToGenePred</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#genepredtobed"><i class="fa fa-check"></i>genePredToBed</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#how_are_we_stranded_here"><i class="fa fa-check"></i>how_are_we_stranded_here</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-cell-ranger"><i class="fa fa-check"></i>Install Cell Ranger</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-r"><i class="fa fa-check"></i>Install R</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#r-libraries"><i class="fa fa-check"></i>R Libraries</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bioconductor"><i class="fa fa-check"></i>Bioconductor</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#sleuth"><i class="fa fa-check"></i>Sleuth</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#practical-exercise-1---software-installation"><i class="fa fa-check"></i>PRACTICAL EXERCISE 1 - Software Installation</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#add-locally-installed-tools-to-your-path-optional"><i class="fa fa-check"></i>Add locally installed tools to your PATH [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-from-official-ubuntu-packages-optional"><i class="fa fa-check"></i>Installing tools from official ubuntu packages [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image"><i class="fa fa-check"></i>Installing tools by Docker image</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image-using-singularity"><i class="fa fa-check"></i>Installing tools by Docker image (using Singularity)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html"><i class="fa fa-check"></i>Module 1</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#labs"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---key-concepts"><i class="fa fa-check"></i>Module 1 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---learning-objectives"><i class="fa fa-check"></i>Module 1 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-a-reference-genome-from-ensembl-igenomes-ncbi-or-ucsc."><i class="fa fa-check"></i>Obtain a reference genome from Ensembl, iGenomes, NCBI or UCSC.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#note-on-complex-commands-and-scripting-in-unix"><i class="fa fa-check"></i>Note on complex commands and scripting in Unix</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-2-advanced"><i class="fa fa-check"></i>PRACTICAL EXERCISE 2 (ADVANCED)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture-1"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-known-genetranscript-annotations"><i class="fa fa-check"></i>Obtain Known Gene/Transcript Annotations</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#definitions"><i class="fa fa-check"></i>Definitions:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#the-purpose-of-gene-annotations-.gtf-file"><i class="fa fa-check"></i>The Purpose of Gene Annotations (.gtf file)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#sources-for-obtaining-gene-annotation-files-formatted-for-hisat2stringtieballgown"><i class="fa fa-check"></i>Sources for obtaining gene annotation files formatted for HISAT2/StringTie/Ballgown</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#important-notes"><i class="fa fa-check"></i>Important notes:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#indexing-mini-lecture"><i class="fa fa-check"></i>Indexing mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#create-a-hisat2-index"><i class="fa fa-check"></i>Create a HISAT2 index</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-rna-seq-test-data."><i class="fa fa-check"></i>Obtain RNA-seq test data.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#determining-the-strandedness-of-rna-seq-data-optional"><i class="fa fa-check"></i>Determining the strandedness of RNA-seq data (Optional)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-3"><i class="fa fa-check"></i>PRACTICAL EXERCISE 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html"><i class="fa fa-check"></i>Module 2</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#labs-1"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#module-2---key-concepts"><i class="fa fa-check"></i>Module 2 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#module-2---learning-objectives"><i class="fa fa-check"></i>Module 2 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lectures"><i class="fa fa-check"></i>Lectures</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#read-trimming-with-fastp"><i class="fa fa-check"></i>Read trimming with Fastp</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#use-fastqc-and-multiqc-to-compare-the-impact-of-trimming"><i class="fa fa-check"></i>Use FastQC and multiqc to compare the impact of trimming</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#clean-up"><i class="fa fa-check"></i>Clean up</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-5"><i class="fa fa-check"></i>PRACTICAL EXERCISE 5</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#alignment-mini-lecture"><i class="fa fa-check"></i>Alignment mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#hisat2-alignment"><i class="fa fa-check"></i>HISAT2 alignment</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#merge-hisat2-bam-files"><i class="fa fa-check"></i>Merge HISAT2 BAM files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-6"><i class="fa fa-check"></i>PRACTICAL EXERCISE 6</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-1-getting-familiar-with-igv"><i class="fa fa-check"></i>Visualization Part 1: Getting familiar with IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-2-inspecting-snps-snvs-and-svs"><i class="fa fa-check"></i>Visualization Part 2: Inspecting SNPs, SNVs, and SVs</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualization-part-3-automating-tasks-in-igv"><i class="fa fa-check"></i>Visualization Part 3: Automating Tasks in IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#indexing-bam-files-with-samtools"><i class="fa fa-check"></i>Indexing BAM files with samtools</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#visualize-alignments-with-igv"><i class="fa fa-check"></i>Visualize alignments with IGV</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-7"><i class="fa fa-check"></i>PRACTICAL EXERCISE 7</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#bam-read-counting"><i class="fa fa-check"></i>BAM Read Counting</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#practical-exercise-7-1"><i class="fa fa-check"></i>PRACTICAL EXERCISE 7</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#alignment-qc-mini-lecture"><i class="fa fa-check"></i>Alignment QC mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#use-samtools-and-fastqc-to-evaluate-the-alignments"><i class="fa fa-check"></i>Use samtools and FastQC to evaluate the alignments</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#create-versions-of-our-bam-files-with-only-the-chromosome-22-alignments"><i class="fa fa-check"></i>Create versions of our BAM files with only the chromosome 22 alignments</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#using-fastqc"><i class="fa fa-check"></i>Using FastQC</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#using-picard"><i class="fa fa-check"></i>Using Picard</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#rseqc-optional"><i class="fa fa-check"></i>RSeQC [optional]</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#multiqc-1"><i class="fa fa-check"></i>MultiQC</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#view-a-pre-generated-multiqc-report-for-full-bam-files"><i class="fa fa-check"></i>View a pre-generated MultiQC report for full bam files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-i---obtaining-the-dataset-reference-files"><i class="fa fa-check"></i>Part I - Obtaining the dataset &amp; reference files</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-ii---data-preprocessing-qc-trimming"><i class="fa fa-check"></i>Part II - Data Preprocessing (QC &amp; Trimming)</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-iii---alignment"><i class="fa fa-check"></i>Part III - Alignment</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#part-iv---post-alignment-qc-igv-visualization"><i class="fa fa-check"></i>Part IV - Post-alignment QC &amp; IGV Visualization</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#presenting-your-results"><i class="fa fa-check"></i>Presenting Your Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html"><i class="fa fa-check"></i>Module 3</a>
<ul>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#labs-2"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#module-3---key-concepts"><i class="fa fa-check"></i>Module 3 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#module-3---learning-objectives"><i class="fa fa-check"></i>Module 3 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#lectures-1"><i class="fa fa-check"></i>Lectures</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#expression-mini-lecture"><i class="fa fa-check"></i>Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#use-stringtie-to-generate-expression-estimates-from-the-sambam-files-generated-by-hisat2-in-the-previous-module"><i class="fa fa-check"></i>Use Stringtie to generate expression estimates from the SAM/BAM files generated by HISAT2 in the previous module</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#practical-exercise-8"><i class="fa fa-check"></i>PRACTICAL EXERCISE 8</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ballgown-de-analysis"><i class="fa fa-check"></i>Ballgown DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#practical-exercise-9"><i class="fa fa-check"></i>PRACTICAL EXERCISE 9</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ercc-de-analysis"><i class="fa fa-check"></i>ERCC DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture-1"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#edger-de-analysis"><i class="fa fa-check"></i>edgeR DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-mini-lecture-2"><i class="fa fa-check"></i>Differential Expression mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#deseq2-de-analysis"><i class="fa fa-check"></i>DESeq2 DE Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#format-htseq-counts-data-to-work-with-deseq2"><i class="fa fa-check"></i>Format htseq counts data to work with DESeq2</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#filter-raw-counts"><i class="fa fa-check"></i>Filter raw counts</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#specifying-the-experimental-design"><i class="fa fa-check"></i>Specifying the experimental design</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#construct-the-deseq2-object-piecing-all-the-data-together"><i class="fa fa-check"></i>Construct the DESeq2 object piecing all the data together</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#running-deseq2"><i class="fa fa-check"></i>Running DESeq2</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#log-fold-change-shrinkage"><i class="fa fa-check"></i>Log-fold change shrinkage</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#annotate-gene-symbols-onto-the-de-results"><i class="fa fa-check"></i>Annotate gene symbols onto the DE results</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#data-manipulation"><i class="fa fa-check"></i>Data manipulation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#save-results-to-files"><i class="fa fa-check"></i>Save results to files</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#briefly-examine-the-top-over-expressed-genes"><i class="fa fa-check"></i>Briefly examine the top over-expressed genes</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#perform-some-preliminary-exploration-of-de-genes-using-webtools"><i class="fa fa-check"></i>Perform some preliminary exploration of DE genes using webtools</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#visualize-overlap-with-a-venn-diagram.-this-can-be-done-with-simple-web-tools-like"><i class="fa fa-check"></i>Visualize overlap with a venn diagram. This can be done with simple web tools like:</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#ballgown-de-visualization"><i class="fa fa-check"></i>Ballgown DE Visualization</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#differential-expression-visualization"><i class="fa fa-check"></i>Differential Expression Visualization</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#viewing-pairwise-sample-clustering"><i class="fa fa-check"></i>Viewing pairwise sample clustering</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-i---expression-estimation"><i class="fa fa-check"></i>Part I - Expression Estimation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-ii---differential-expression"><i class="fa fa-check"></i>Part II - Differential Expression</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#presenting-your-results-1"><i class="fa fa-check"></i>Presenting Your Results</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-0-obtaining-data-and-references"><i class="fa fa-check"></i>PART 0 : Obtaining Data and References</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-1-data-preprocessing"><i class="fa fa-check"></i>Part 1 : Data preprocessing</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-2-data-alignment"><i class="fa fa-check"></i>Part 2: Data alignment</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-3-expression-estimation"><i class="fa fa-check"></i>Part 3: Expression Estimation</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-4-differential-expression-analysis"><i class="fa fa-check"></i>Part 4: Differential Expression Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3.html"><a href="module-3.html#part-5-differential-expression-analysis-visualization"><i class="fa fa-check"></i>Part 5: Differential Expression Analysis Visualization</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNA-seq Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3" class="section level1 hasAnchor">
<h1>Module 3<a href="module-3.html#module-3" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-4" class="section level2 hasAnchor">
<h2>Lecture<a href="module-3.html#lecture-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe width="640" height="360" src="https://drive.google.com/file/d/1eQcefbuIWFLhCPQY37srIvLRsNyIzXVF/preview" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
</iframe>
<iframe width="640" height="480" src="https://www.youtube.com/embed/G956a8bTFss?si=LCuy_r5jWFhu-Ao9" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
</iframe>
</div>
<div id="labs-2" class="section level2 hasAnchor">
<h2>Labs<a href="module-3.html#labs-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="module-3---key-concepts" class="section level3 hasAnchor">
<h3>Module 3 - Key concepts<a href="module-3.html#module-3---key-concepts" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Expression estimation, FPKM, TPM, StringTie overview, BallGown overview, multiple testing correction, etc.</li>
</ul>
</div>
<div id="module-3---learning-objectives" class="section level3 hasAnchor">
<h3>Module 3 - Learning objectives<a href="module-3.html#module-3---learning-objectives" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Expression estimation for known genes and transcripts</li>
<li>FPKM/TPM expression estimates vs. raw counts</li>
<li>Differential expression methods</li>
<li>Downstream interpretation of expression and differential estimates</li>
</ul>
</div>
<div id="lectures-1" class="section level3 hasAnchor">
<h3>Lectures<a href="module-3.html#lectures-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_01_AbundanceEstimation.pdf">Abundance Estimations lecture</a></li>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_02_HTSEQ.pdf">HTSEQ lecture</a></li>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_03_DifferentialExpression.pdf">Differential expression lecture</a></li>
</ul>
</div>
<div id="expression-mini-lecture" class="section level3 hasAnchor">
<h3>Expression mini lecture<a href="module-3.html#expression-mini-lecture" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a refresher on expression and abundance estimations, we have made a <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_01_AbundanceEstimation.pdf">mini lecture</a>.</p>
</div>
<div id="use-stringtie-to-generate-expression-estimates-from-the-sambam-files-generated-by-hisat2-in-the-previous-module" class="section level3 hasAnchor">
<h3>Use Stringtie to generate expression estimates from the SAM/BAM files generated by HISAT2 in the previous module<a href="module-3.html#use-stringtie-to-generate-expression-estimates-from-the-sambam-files-generated-by-hisat2-in-the-previous-module" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="note-on-de-novo-transcript-discovery-and-differential-expression-using-stringtie" class="section level4 hasAnchor">
<h4>Note on de novo transcript discovery and differential expression using Stringtie:<a href="module-3.html#note-on-de-novo-transcript-discovery-and-differential-expression-using-stringtie" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In this module, we will run Stringtie in ‘reference only’ mode. For simplicity and to reduce run time, it is sometimes useful to perform expression analysis with only known transcript models. However, Stringtie can predict the transcripts present in each library instead (by dropping the ‘-G’ option in stringtie commands as described in the next module). Stringtie will then assign arbitrary transcript IDs to each transcript assembled from the data and estimate expression for those transcripts. One complication with this method is that in each library a different set of transcripts is likely to be predicted for each library. There may be a lot of similarities but the number of transcripts and their exact structure will differ in the output files for each library. Before you can compare across libraries you therefore need to determine which transcripts correspond to each other across the libraries.</p>
<ul>
<li>Stringtie provides a merge command to combine predicted transcript GTF files from across different libraries</li>
<li>Once you have a merged GTF file you can run Stringtie <strong>again</strong> with this instead of the known transcripts GTF file we used above</li>
<li>Stringtie also provides ‘gffcompare’ to compare predicted transcripts to known transcripts</li>
<li>Refer to the Stringtie manual for a more detailed explanation:</li>
<li><a href="https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual">https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual</a></li>
</ul>
<p>Stringtie basic usage:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb105-1"><a href="module-3.html#cb105-1" tabindex="-1"></a>    <span class="ex">stringtie</span> <span class="op">&lt;</span>aligned_reads.bam<span class="op">&gt;</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]*</span></span></code></pre></div>
<p>Extra options specified below:</p>
<ul>
<li>‘–rf’ tells StringTie that our data is stranded and to use the correct strand specific mode (i.e. assume a stranded library fr-firststrand).</li>
<li>‘-p 4’ tells StringTie to use 4 CPUs</li>
<li>‘-G <known transcripts file>’ reference annotation to use for guiding the assembly process (GTF/GFF3)</li>
<li>‘-e’ only estimate the abundance of given reference transcripts (requires -G)</li>
<li>‘-B’ enable output of Ballgown table files which will be created in the same directory as the output GTF (requires -G, -o recommended)</li>
<li>‘-o’ output path/file name for the assembled transcripts GTF (default: stdout)</li>
<li>‘-A’ output path/file name for gene abundance estimates</li>
</ul>
<div class="sourceCode" id="cb106"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb106-1"><a href="module-3.html#cb106-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/</span>
<span id="cb106-2"><a href="module-3.html#cb106-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> expression/stringtie/ref_only/</span>
<span id="cb106-3"><a href="module-3.html#cb106-3" tabindex="-1"></a><span class="bu">cd</span> expression/stringtie/ref_only/</span>
<span id="cb106-4"><a href="module-3.html#cb106-4" tabindex="-1"></a></span>
<span id="cb106-5"><a href="module-3.html#cb106-5" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> HBR_Rep1/transcripts.gtf <span class="at">-A</span> HBR_Rep1/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep1.bam</span>
<span id="cb106-6"><a href="module-3.html#cb106-6" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> HBR_Rep2/transcripts.gtf <span class="at">-A</span> HBR_Rep2/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep2.bam</span>
<span id="cb106-7"><a href="module-3.html#cb106-7" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> HBR_Rep3/transcripts.gtf <span class="at">-A</span> HBR_Rep3/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep3.bam</span>
<span id="cb106-8"><a href="module-3.html#cb106-8" tabindex="-1"></a></span>
<span id="cb106-9"><a href="module-3.html#cb106-9" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> UHR_Rep1/transcripts.gtf <span class="at">-A</span> UHR_Rep1/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep1.bam</span>
<span id="cb106-10"><a href="module-3.html#cb106-10" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> UHR_Rep2/transcripts.gtf <span class="at">-A</span> UHR_Rep2/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep2.bam</span>
<span id="cb106-11"><a href="module-3.html#cb106-11" tabindex="-1"></a><span class="ex">stringtie</span> <span class="at">--rf</span> <span class="at">-p</span> 4 <span class="at">-G</span> <span class="va">$RNA_REF_GTF</span> <span class="at">-e</span> <span class="at">-B</span> <span class="at">-o</span> UHR_Rep3/transcripts.gtf <span class="at">-A</span> UHR_Rep3/gene_abundances.tsv <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep3.bam</span></code></pre></div>
<p>What does the raw output from Stringtie look like? For details on the Stringtie output files refer to <a href="http://ccb.jhu.edu/software/stringtie/index.shtml?t=manual">Stringtie manual</a> (<a href="http://ccb.jhu.edu/software/stringtie/index.shtml?t=manual#output">outputs section</a>)</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb107-1"><a href="module-3.html#cb107-1" tabindex="-1"></a><span class="fu">less</span> <span class="at">-S</span> UHR_Rep1/transcripts.gtf</span></code></pre></div>
<p>View transcript records only and improve formatting</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb108-1"><a href="module-3.html#cb108-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;^#&quot;</span> UHR_Rep1/transcripts.gtf <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-w</span> <span class="st">&quot;transcript&quot;</span> <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code></pre></div>
<p>Limit the view to transcript records and their expression values (FPKM and TPM values)</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb109-1"><a href="module-3.html#cb109-1" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&#39;{if ($3==&quot;transcript&quot;) print}&#39;</span> UHR_Rep1/transcripts.gtf <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f</span> 1,4,9 <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code></pre></div>
<p>Press ‘q’ to exit the ‘less’ display</p>
<p>Gene and transcript level expression values can also be viewed in these two files:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb110-1"><a href="module-3.html#cb110-1" tabindex="-1"></a><span class="ex">column</span> <span class="at">-t</span> UHR_Rep1/t_data.ctab <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span>
<span id="cb110-2"><a href="module-3.html#cb110-2" tabindex="-1"></a></span>
<span id="cb110-3"><a href="module-3.html#cb110-3" tabindex="-1"></a><span class="fu">less</span> <span class="at">-S</span> <span class="at">-x20</span> UHR_Rep1/gene_abundances.tsv</span></code></pre></div>
<p>Create a tidy expression matrix files for the StringTie results. This will be done at both the gene and transcript level and also will take into account the various expression measures produced: coverage, FPKM, and TPM.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb111-1"><a href="module-3.html#cb111-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/expression/stringtie/ref_only/</span>
<span id="cb111-2"><a href="module-3.html#cb111-2" tabindex="-1"></a><span class="fu">wget</span> https://raw.githubusercontent.com/griffithlab/rnabio.org/master/assets/scripts/stringtie_expression_matrix.pl</span>
<span id="cb111-3"><a href="module-3.html#cb111-3" tabindex="-1"></a><span class="fu">chmod</span> +x stringtie_expression_matrix.pl</span>
<span id="cb111-4"><a href="module-3.html#cb111-4" tabindex="-1"></a></span>
<span id="cb111-5"><a href="module-3.html#cb111-5" tabindex="-1"></a><span class="ex">./stringtie_expression_matrix.pl</span> <span class="at">--expression_metric</span><span class="op">=</span>TPM <span class="at">--result_dirs</span><span class="op">=</span><span class="st">&#39;HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3&#39;</span> <span class="at">--transcript_matrix_file</span><span class="op">=</span>transcript_tpm_all_samples.tsv <span class="at">--gene_matrix_file</span><span class="op">=</span>gene_tpm_all_samples.tsv</span>
<span id="cb111-6"><a href="module-3.html#cb111-6" tabindex="-1"></a></span>
<span id="cb111-7"><a href="module-3.html#cb111-7" tabindex="-1"></a><span class="ex">./stringtie_expression_matrix.pl</span> <span class="at">--expression_metric</span><span class="op">=</span>FPKM <span class="at">--result_dirs</span><span class="op">=</span><span class="st">&#39;HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3&#39;</span> <span class="at">--transcript_matrix_file</span><span class="op">=</span>transcript_fpkm_all_samples.tsv <span class="at">--gene_matrix_file</span><span class="op">=</span>gene_fpkm_all_samples.tsv</span>
<span id="cb111-8"><a href="module-3.html#cb111-8" tabindex="-1"></a></span>
<span id="cb111-9"><a href="module-3.html#cb111-9" tabindex="-1"></a><span class="ex">./stringtie_expression_matrix.pl</span> <span class="at">--expression_metric</span><span class="op">=</span>Coverage <span class="at">--result_dirs</span><span class="op">=</span><span class="st">&#39;HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3&#39;</span> <span class="at">--transcript_matrix_file</span><span class="op">=</span>transcript_coverage_all_samples.tsv <span class="at">--gene_matrix_file</span><span class="op">=</span>gene_coverage_all_samples.tsv</span>
<span id="cb111-10"><a href="module-3.html#cb111-10" tabindex="-1"></a></span>
<span id="cb111-11"><a href="module-3.html#cb111-11" tabindex="-1"></a><span class="ex">column</span> <span class="at">-t</span> transcript_tpm_all_samples.tsv <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span>
<span id="cb111-12"><a href="module-3.html#cb111-12" tabindex="-1"></a><span class="ex">column</span> <span class="at">-t</span> gene_tpm_all_samples.tsv <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code></pre></div>
<p>Later we will use these files to perform various comparisons of expression estimation tools (e.g. stringtie, kallisto, raw counts) and metrics (e.g. FPKM vs TPM).</p>
<hr />
</div>
</div>
<div id="practical-exercise-8" class="section level3 hasAnchor">
<h3>PRACTICAL EXERCISE 8<a href="module-3.html#practical-exercise-8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Assignment: Use StringTie to Calculate transcript-level expression estimates for the alignments (bam files) you created in Practical Exercise 6.</p>
<ul>
<li>Hint: You should have six commands for 3 replicates each of tumor and normal samples.</li>
</ul>
<p>Solution: When you are ready you can check your approach against the <a href="/module-09-appendix/0009/05/01/Practical_Exercise_Solutions/#practical-exercise-8---expression">Solutions</a></p>
<hr />
<div class="float">
<img src="assets/module_3/RNA-seq_Flowchart4-2.png" alt="RNA-seq_Flowchart4" />
<div class="figcaption">RNA-seq_Flowchart4</div>
</div>
<hr />
<div id="mini-lecture" class="section level4 hasAnchor">
<h4>Mini-lecture<a href="module-3.html#mini-lecture" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For more on the differences between abundance estimates like FPKM and count data with HTSeq-count, see this <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_02_HTSEQ.pdf">mini lecture</a>.</p>
<hr />
</div>
<div id="htseq-count-1" class="section level4 hasAnchor">
<h4>HTSEQ-COUNT<a href="module-3.html#htseq-count-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Run htseq-count on alignments instead to produce raw counts instead of FPKM/TPM values for differential expression analysis</p>
<p>Refer to the HTSeq documentation for a more detailed explanation:</p>
<ul>
<li><a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html">https://htseq.readthedocs.io/en/release_0.11.1/count.html</a></li>
</ul>
<p>htseq-count basic usage:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb112-1"><a href="module-3.html#cb112-1" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> <span class="op">&lt;</span>sam_file<span class="op">&gt;</span> <span class="op">&lt;</span>gff_file<span class="op">&gt;</span></span></code></pre></div>
<p>Extra options specified below:</p>
<ul>
<li>‘–format’ specify the input file format one of BAM or SAM. Since we have BAM format files, select ‘bam’ for this option.</li>
<li>‘–order’ provide the expected sort order of the input file. Previously we generated position sorted BAM files so use ‘pos’.</li>
<li>‘–mode’ determines how to deal with reads that overlap more than one feature. We believe the ‘intersection-strict’ mode is best.</li>
<li>‘–stranded’ specifies whether data is stranded or not. The TruSeq strand-specific RNA libraries suggest the ‘reverse’ option for this parameter.</li>
<li>‘–minaqual’ will skip all reads with alignment quality lower than the given minimum value</li>
<li>‘–type’ specifies the feature type (3rd column in GFF file) to be used. (default, suitable for RNA-Seq and Ensembl GTF files: exon)</li>
<li>‘–idattr’ The feature ID used to identify the counts in the output table. The default, suitable for RNA-SEq and Ensembl GTF files, is gene_id.</li>
</ul>
<p>Run htseq-count and calculate gene-level counts:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb113-1"><a href="module-3.html#cb113-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/</span>
<span id="cb113-2"><a href="module-3.html#cb113-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> expression/htseq_counts</span>
<span id="cb113-3"><a href="module-3.html#cb113-3" tabindex="-1"></a><span class="bu">cd</span> expression/htseq_counts</span>
<span id="cb113-4"><a href="module-3.html#cb113-4" tabindex="-1"></a></span>
<span id="cb113-5"><a href="module-3.html#cb113-5" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep1.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> UHR_Rep1_gene.tsv</span>
<span id="cb113-6"><a href="module-3.html#cb113-6" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep2.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> UHR_Rep2_gene.tsv</span>
<span id="cb113-7"><a href="module-3.html#cb113-7" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/UHR_Rep3.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> UHR_Rep3_gene.tsv</span>
<span id="cb113-8"><a href="module-3.html#cb113-8" tabindex="-1"></a></span>
<span id="cb113-9"><a href="module-3.html#cb113-9" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep1.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> HBR_Rep1_gene.tsv</span>
<span id="cb113-10"><a href="module-3.html#cb113-10" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep2.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> HBR_Rep2_gene.tsv</span>
<span id="cb113-11"><a href="module-3.html#cb113-11" tabindex="-1"></a><span class="ex">htseq-count</span> <span class="at">--format</span> bam <span class="at">--order</span> pos <span class="at">--mode</span> intersection-strict <span class="at">--stranded</span> reverse <span class="at">--minaqual</span> 1 <span class="at">--type</span> exon <span class="at">--idattr</span> gene_id <span class="va">$RNA_ALIGN_DIR</span>/HBR_Rep3.bam <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> HBR_Rep3_gene.tsv</span></code></pre></div>
<p>Merge results files into a single matrix for use in edgeR. The following joins the results for each replicate together, adds a header, reformats the result as a tab delimited file, and shows you the first 10 lines of the resulting file :</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb114-1"><a href="module-3.html#cb114-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/expression/htseq_counts/</span>
<span id="cb114-2"><a href="module-3.html#cb114-2" tabindex="-1"></a><span class="fu">join</span> UHR_Rep1_gene.tsv UHR_Rep2_gene.tsv <span class="kw">|</span> <span class="fu">join</span> <span class="at">-</span> UHR_Rep3_gene.tsv <span class="kw">|</span> <span class="fu">join</span> <span class="at">-</span> HBR_Rep1_gene.tsv <span class="kw">|</span> <span class="fu">join</span> <span class="at">-</span> HBR_Rep2_gene.tsv <span class="kw">|</span> <span class="fu">join</span> <span class="at">-</span> HBR_Rep3_gene.tsv <span class="op">&gt;</span> gene_read_counts_table_all.tsv</span>
<span id="cb114-3"><a href="module-3.html#cb114-3" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;GeneID UHR_Rep1 UHR_Rep2 UHR_Rep3 HBR_Rep1 HBR_Rep2 HBR_Rep3&quot;</span> <span class="op">&gt;</span> header.txt</span>
<span id="cb114-4"><a href="module-3.html#cb114-4" tabindex="-1"></a><span class="fu">cat</span> header.txt gene_read_counts_table_all.tsv <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;__&quot;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-v</span> OFS=<span class="st">&quot;\t&quot;</span> <span class="st">&#39;$1=$1&#39;</span> <span class="op">&gt;</span> gene_read_counts_table_all_final.tsv</span>
<span id="cb114-5"><a href="module-3.html#cb114-5" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-f</span> gene_read_counts_table_all.tsv header.txt</span>
<span id="cb114-6"><a href="module-3.html#cb114-6" tabindex="-1"></a><span class="fu">head</span> gene_read_counts_table_all_final.tsv <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span></span></code></pre></div>
<p>-<code>grep -v "__"</code> is being used to filter out the summary lines at the end of the files that <code>ht-seq count</code> gives to summarize reads that had no feature, were ambiguous, did not align at all, did not align due to poor alignment quality, or the alignment was not unique.</p>
<p>-<code>awk -v OFS="\t" '$1=$1'</code> is using <code>awk</code> to replace the single space characters that were in the concatenated version of our <code>header.txt</code> and <code>gene_read_counts_table_all.tsv</code> with a tab character. <code>-v</code> is used to reset the variable <code>OFS</code>, which stands for Output Field Separator. By default, this is a single space. By specifying <code>OFS="\t"</code>, we are telling <code>awk</code> to replace the single space with a tab. The <code>'$1=$1'</code> tells awk to reevaluate the input using the new output variable.</p>
</div>
<div id="prepare-for-de-analysis-using-htseq-count-results" class="section level4 hasAnchor">
<h4>Prepare for DE analysis using htseq-count results<a href="module-3.html#prepare-for-de-analysis-using-htseq-count-results" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Create a directory for the DEseq analysis based on the htseq-count results:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb115-1"><a href="module-3.html#cb115-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/</span>
<span id="cb115-2"><a href="module-3.html#cb115-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> de/htseq_counts</span>
<span id="cb115-3"><a href="module-3.html#cb115-3" tabindex="-1"></a><span class="bu">cd</span> de/htseq_counts</span></code></pre></div>
<p>Note that the htseq-count results provide counts for each gene but uses only the Ensembl Gene ID (e.g. ENSG00000054611). This is not very convenient for biological interpretation. This next step creates a mapping file that will help us translate from ENSG IDs to Symbols. It does this by parsing the GTF transcriptome file we got from Ensembl. That file contains both gene names and IDs. Unfortunately, this file is a bit complex to parse. Furthermore, it contains the ERCC transcripts, and these have their own naming convention which also complicates the parsing.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb116-1"><a href="module-3.html#cb116-1" tabindex="-1"></a></span>
<span id="cb116-2"><a href="module-3.html#cb116-2" tabindex="-1"></a><span class="co"># cut the 9th column with all the gene annotation information</span></span>
<span id="cb116-3"><a href="module-3.html#cb116-3" tabindex="-1"></a><span class="co"># delete all the double quotes from this string</span></span>
<span id="cb116-4"><a href="module-3.html#cb116-4" tabindex="-1"></a><span class="co"># use perl to search for the pattern &quot;gene_id&quot; or &quot;gene_name&quot; followed by space character and then some non-space characters. </span></span>
<span id="cb116-5"><a href="module-3.html#cb116-5" tabindex="-1"></a><span class="co"># the non-space characters are the actual gene ID or Name. Store these in variables $gid and $gname and then print them out</span></span>
<span id="cb116-6"><a href="module-3.html#cb116-6" tabindex="-1"></a><span class="co"># use sort and unique commands to produce a unique list of gene_name, gene_id combinations</span></span>
<span id="cb116-7"><a href="module-3.html#cb116-7" tabindex="-1"></a> </span>
<span id="cb116-8"><a href="module-3.html#cb116-8" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 9 <span class="va">$RNA_REF_GTF</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="st">&#39;&quot;&#39;</span> <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp; if ($_ =~ /gene_id\s+(\S+);/){$gid = $1}; if ($_ =~ /gene_name\s+(\S+);/){$gname = $1}; print &quot;$gid\t$gname\n&quot;&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> ENSG_ID2Name.txt</span>
<span id="cb116-9"><a href="module-3.html#cb116-9" tabindex="-1"></a></span>
<span id="cb116-10"><a href="module-3.html#cb116-10" tabindex="-1"></a><span class="fu">head</span> ENSG_ID2Name.txt</span></code></pre></div>
<p>Determine the number of unique Ensembl Gene IDs and symbols. What does this tell you?</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb117-1"><a href="module-3.html#cb117-1" tabindex="-1"></a></span>
<span id="cb117-2"><a href="module-3.html#cb117-2" tabindex="-1"></a><span class="co">#count unique gene ids</span></span>
<span id="cb117-3"><a href="module-3.html#cb117-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb117-4"><a href="module-3.html#cb117-4" tabindex="-1"></a></span>
<span id="cb117-5"><a href="module-3.html#cb117-5" tabindex="-1"></a><span class="co">#count unique gene names</span></span>
<span id="cb117-6"><a href="module-3.html#cb117-6" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb117-7"><a href="module-3.html#cb117-7" tabindex="-1"></a></span>
<span id="cb117-8"><a href="module-3.html#cb117-8" tabindex="-1"></a><span class="co">#show the most repeated gene names</span></span>
<span id="cb117-9"><a href="module-3.html#cb117-9" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-r</span> <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<hr />
</div>
<div id="ercc-expression-analysis" class="section level4 hasAnchor">
<h4>ERCC expression analysis<a href="module-3.html#ercc-expression-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Based on the above read counts, plot the linearity of the ERCC spike-in read counts observed in our RNA-seq data versus the expected concentration of the ERCC spike-in Mix.</p>
<p>First download a file describing the expected concentrations and fold-change differences for the ERCC spike-in reagent.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb118-1"><a href="module-3.html#cb118-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="va">$RNA_HOME</span>/expression/ercc_spikein_analysis/</span>
<span id="cb118-2"><a href="module-3.html#cb118-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/expression/ercc_spikein_analysis/</span>
<span id="cb118-3"><a href="module-3.html#cb118-3" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/ERCC_Controls_Analysis.txt</span>
<span id="cb118-4"><a href="module-3.html#cb118-4" tabindex="-1"></a><span class="fu">cat</span> ERCC_Controls_Analysis.txt</span></code></pre></div>
<p>We will then merge our experimental RNA-seq read counts, determined for the ERCC transcripts, onto the table of expected concentrations. Finally, we will produce an x-y scatter plot that compares the expected and observed values.</p>
<p>First, start an R session:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="module-3.html#cb119-1" tabindex="-1"></a>R</span></code></pre></div>
<p>Now combine the ERCC expected concentration data with the observed RNA-seq expression values and produce x-y scatter plots that compare the expected and observed values for HTSEQ raw counts and StringTie TPM abundance estimates.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="module-3.html#cb120-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb120-2"><a href="module-3.html#cb120-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;data.table&quot;</span>)</span>
<span id="cb120-3"><a href="module-3.html#cb120-3" tabindex="-1"></a></span>
<span id="cb120-4"><a href="module-3.html#cb120-4" tabindex="-1"></a><span class="co">#load in the reference/expected concentration and fold change values for each ERCC transcript</span></span>
<span id="cb120-5"><a href="module-3.html#cb120-5" tabindex="-1"></a>ercc_ref <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;ERCC_Controls_Analysis.txt&quot;</span>, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb120-6"><a href="module-3.html#cb120-6" tabindex="-1"></a><span class="fu">names</span>(ercc_ref) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;ref_conc_mix_1&quot;</span>, <span class="st">&quot;ref_conc_mix_2&quot;</span>, <span class="st">&quot;ref_fc_mix1_vs_mix2&quot;</span>, <span class="st">&quot;ref_log2_mix1_vs_mix2&quot;</span>)</span>
<span id="cb120-7"><a href="module-3.html#cb120-7" tabindex="-1"></a><span class="fu">head</span>(ercc_ref)</span>
<span id="cb120-8"><a href="module-3.html#cb120-8" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref)</span>
<span id="cb120-9"><a href="module-3.html#cb120-9" tabindex="-1"></a></span>
<span id="cb120-10"><a href="module-3.html#cb120-10" tabindex="-1"></a><span class="co">#load the RNA-seq raw counts values for all samples and combined with the expected ERCC values</span></span>
<span id="cb120-11"><a href="module-3.html#cb120-11" tabindex="-1"></a>rna_counts_file <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/expression/htseq_counts/gene_read_counts_table_all_final.tsv&quot;</span>;</span>
<span id="cb120-12"><a href="module-3.html#cb120-12" tabindex="-1"></a>rna_counts <span class="ot">=</span> <span class="fu">read.table</span>(rna_counts_file, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb120-13"><a href="module-3.html#cb120-13" tabindex="-1"></a><span class="fu">dim</span>(rna_counts)</span>
<span id="cb120-14"><a href="module-3.html#cb120-14" tabindex="-1"></a></span>
<span id="cb120-15"><a href="module-3.html#cb120-15" tabindex="-1"></a><span class="co">#combine the ERCC expected concentration information with the observed RNA-seq counts</span></span>
<span id="cb120-16"><a href="module-3.html#cb120-16" tabindex="-1"></a>ercc_ref_counts <span class="ot">=</span> <span class="fu">merge</span>(<span class="at">x =</span> ercc_ref, <span class="at">y =</span> rna_counts, <span class="at">by.x =</span> <span class="st">&quot;ercc_id&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;GeneID&quot;</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb120-17"><a href="module-3.html#cb120-17" tabindex="-1"></a></span>
<span id="cb120-18"><a href="module-3.html#cb120-18" tabindex="-1"></a><span class="co">#convert UHR data to &quot;long&quot; format</span></span>
<span id="cb120-19"><a href="module-3.html#cb120-19" tabindex="-1"></a>uhr_data <span class="ot">=</span> ercc_ref_counts[,<span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_1&quot;</span>,<span class="st">&quot;UHR_Rep1&quot;</span>,<span class="st">&quot;UHR_Rep2&quot;</span>,<span class="st">&quot;UHR_Rep3&quot;</span>)]</span>
<span id="cb120-20"><a href="module-3.html#cb120-20" tabindex="-1"></a>uhr_data_long <span class="ot">=</span> <span class="fu">melt</span>(<span class="fu">setDT</span>(uhr_data), <span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_1&quot;</span>), <span class="at">variable.name =</span> <span class="st">&quot;sample&quot;</span>)</span>
<span id="cb120-21"><a href="module-3.html#cb120-21" tabindex="-1"></a>uhr_data_long<span class="sc">$</span>mix <span class="ot">=</span> <span class="st">&quot;mix 1&quot;</span></span>
<span id="cb120-22"><a href="module-3.html#cb120-22" tabindex="-1"></a><span class="fu">names</span>(uhr_data_long) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;concentration&quot;</span>, <span class="st">&quot;sample&quot;</span>, <span class="st">&quot;count&quot;</span>, <span class="st">&quot;mix&quot;</span>)</span>
<span id="cb120-23"><a href="module-3.html#cb120-23" tabindex="-1"></a></span>
<span id="cb120-24"><a href="module-3.html#cb120-24" tabindex="-1"></a><span class="co">#convert HBR data to &quot;long&quot; format</span></span>
<span id="cb120-25"><a href="module-3.html#cb120-25" tabindex="-1"></a>hbr_data <span class="ot">=</span> ercc_ref_counts[,<span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_2&quot;</span>,<span class="st">&quot;HBR_Rep1&quot;</span>,<span class="st">&quot;HBR_Rep2&quot;</span>,<span class="st">&quot;HBR_Rep3&quot;</span>)]</span>
<span id="cb120-26"><a href="module-3.html#cb120-26" tabindex="-1"></a>hbr_data_long <span class="ot">=</span> <span class="fu">melt</span>(<span class="fu">setDT</span>(hbr_data), <span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_2&quot;</span>), <span class="at">variable.name =</span> <span class="st">&quot;sample&quot;</span>)</span>
<span id="cb120-27"><a href="module-3.html#cb120-27" tabindex="-1"></a>hbr_data_long<span class="sc">$</span>mix <span class="ot">=</span> <span class="st">&quot;mix 2&quot;</span></span>
<span id="cb120-28"><a href="module-3.html#cb120-28" tabindex="-1"></a><span class="fu">names</span>(hbr_data_long) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;concentration&quot;</span>, <span class="st">&quot;sample&quot;</span>, <span class="st">&quot;count&quot;</span>, <span class="st">&quot;mix&quot;</span>)</span>
<span id="cb120-29"><a href="module-3.html#cb120-29" tabindex="-1"></a></span>
<span id="cb120-30"><a href="module-3.html#cb120-30" tabindex="-1"></a><span class="co">#rejoin the UHR and HBR tpm data</span></span>
<span id="cb120-31"><a href="module-3.html#cb120-31" tabindex="-1"></a>ercc_ref_counts_long <span class="ot">&lt;-</span> <span class="fu">rbind</span>(uhr_data_long, hbr_data_long)</span>
<span id="cb120-32"><a href="module-3.html#cb120-32" tabindex="-1"></a><span class="fu">head</span>(ercc_ref_counts_long)</span>
<span id="cb120-33"><a href="module-3.html#cb120-33" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref_counts_long)</span>
<span id="cb120-34"><a href="module-3.html#cb120-34" tabindex="-1"></a></span>
<span id="cb120-35"><a href="module-3.html#cb120-35" tabindex="-1"></a><span class="co">#fit a linear model and calculate correlation between expected concentations and observed TPM values</span></span>
<span id="cb120-36"><a href="module-3.html#cb120-36" tabindex="-1"></a></span>
<span id="cb120-37"><a href="module-3.html#cb120-37" tabindex="-1"></a>min_nonzero_count <span class="ot">=</span> <span class="fu">min</span>(ercc_ref_counts_long<span class="sc">$</span>count[ercc_ref_counts_long<span class="sc">$</span>count <span class="sc">&gt;</span> <span class="dv">0</span>])</span>
<span id="cb120-38"><a href="module-3.html#cb120-38" tabindex="-1"></a>ercc_ref_counts_long<span class="sc">$</span>log_count <span class="ot">=</span> <span class="fu">log2</span>(ercc_ref_counts_long<span class="sc">$</span>count <span class="sc">+</span> min_nonzero_count)</span>
<span id="cb120-39"><a href="module-3.html#cb120-39" tabindex="-1"></a></span>
<span id="cb120-40"><a href="module-3.html#cb120-40" tabindex="-1"></a>min_nonzero_conc <span class="ot">=</span> <span class="fu">min</span>(ercc_ref_counts_long<span class="sc">$</span>concentration[ercc_ref_counts_long<span class="sc">$</span>concentration <span class="sc">&gt;</span> <span class="dv">0</span>])</span>
<span id="cb120-41"><a href="module-3.html#cb120-41" tabindex="-1"></a>ercc_ref_counts_long<span class="sc">$</span>log_concentration<span class="ot">=</span> <span class="fu">log2</span>(ercc_ref_counts_long<span class="sc">$</span>concentration <span class="sc">+</span> min_nonzero_conc)</span>
<span id="cb120-42"><a href="module-3.html#cb120-42" tabindex="-1"></a></span>
<span id="cb120-43"><a href="module-3.html#cb120-43" tabindex="-1"></a>count_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(log_count <span class="sc">~</span> log_concentration, <span class="at">data=</span>ercc_ref_counts_long)</span>
<span id="cb120-44"><a href="module-3.html#cb120-44" tabindex="-1"></a>count_r_squared <span class="ot">=</span> <span class="fu">summary</span>(count_model)[[<span class="st">&quot;r.squared&quot;</span>]]</span>
<span id="cb120-45"><a href="module-3.html#cb120-45" tabindex="-1"></a>count_slope <span class="ot">=</span> <span class="fu">coef</span>(count_model)[<span class="st">&quot;log_concentration&quot;</span>]</span>
<span id="cb120-46"><a href="module-3.html#cb120-46" tabindex="-1"></a></span>
<span id="cb120-47"><a href="module-3.html#cb120-47" tabindex="-1"></a>p1 <span class="ot">=</span> <span class="fu">ggplot</span>(ercc_ref_counts_long, <span class="fu">aes</span>(<span class="at">x=</span>log_concentration, <span class="at">y=</span>log_count))</span>
<span id="cb120-48"><a href="module-3.html#cb120-48" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">shape=</span>mix, <span class="at">color=</span>sample))</span>
<span id="cb120-49"><a href="module-3.html#cb120-49" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm) </span>
<span id="cb120-50"><a href="module-3.html#cb120-50" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dv">10</span>, <span class="dv">5</span>, <span class="at">label=</span><span class="fu">paste</span>(<span class="st">&quot;R^2 =&quot;</span>, <span class="fu">round</span>(count_r_squared, <span class="at">digits=</span><span class="dv">2</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>)) </span>
<span id="cb120-51"><a href="module-3.html#cb120-51" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dv">10</span>, <span class="dv">4</span>, <span class="at">label=</span><span class="fu">paste</span>(<span class="st">&quot;Slope =&quot;</span>, <span class="fu">round</span>(count_slope, <span class="at">digits=</span><span class="dv">2</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>))</span>
<span id="cb120-52"><a href="module-3.html#cb120-52" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Log2 (expected concentration [amol/uL] + min non-zero value)&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Log2 (observed count + min non-zero value)&quot;</span>)</span>
<span id="cb120-53"><a href="module-3.html#cb120-53" tabindex="-1"></a></span>
<span id="cb120-54"><a href="module-3.html#cb120-54" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;ERCC_Count_Expression_vs_SpikeInConcentration.pdf&quot;</span>)</span>
<span id="cb120-55"><a href="module-3.html#cb120-55" tabindex="-1"></a><span class="fu">print</span>(p1)</span>
<span id="cb120-56"><a href="module-3.html#cb120-56" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb120-57"><a href="module-3.html#cb120-57" tabindex="-1"></a></span>
<span id="cb120-58"><a href="module-3.html#cb120-58" tabindex="-1"></a><span class="co">#load the RNA-seq TPM values for all samples and combine with expected ERCC values</span></span>
<span id="cb120-59"><a href="module-3.html#cb120-59" tabindex="-1"></a>rna_tpms_file <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/expression/stringtie/ref_only/gene_tpm_all_samples.tsv&quot;</span></span>
<span id="cb120-60"><a href="module-3.html#cb120-60" tabindex="-1"></a>rna_tpms <span class="ot">=</span> <span class="fu">read.table</span>(rna_tpms_file, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb120-61"><a href="module-3.html#cb120-61" tabindex="-1"></a><span class="fu">dim</span>(rna_tpms)</span>
<span id="cb120-62"><a href="module-3.html#cb120-62" tabindex="-1"></a></span>
<span id="cb120-63"><a href="module-3.html#cb120-63" tabindex="-1"></a><span class="co">#combine the ERCC expected concentration information with the observed RNA-seq TPM values</span></span>
<span id="cb120-64"><a href="module-3.html#cb120-64" tabindex="-1"></a>ercc_ref_tpms <span class="ot">=</span> <span class="fu">merge</span>(<span class="at">x =</span> ercc_ref, <span class="at">y =</span> rna_tpms, <span class="at">by.x =</span> <span class="st">&quot;ercc_id&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;Gene_ID&quot;</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb120-65"><a href="module-3.html#cb120-65" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref_tpms)</span>
<span id="cb120-66"><a href="module-3.html#cb120-66" tabindex="-1"></a></span>
<span id="cb120-67"><a href="module-3.html#cb120-67" tabindex="-1"></a><span class="co">#convert UHR data to &quot;long&quot; format</span></span>
<span id="cb120-68"><a href="module-3.html#cb120-68" tabindex="-1"></a>uhr_data <span class="ot">=</span> ercc_ref_tpms[,<span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_1&quot;</span>,<span class="st">&quot;UHR_Rep1&quot;</span>,<span class="st">&quot;UHR_Rep2&quot;</span>,<span class="st">&quot;UHR_Rep3&quot;</span>)]</span>
<span id="cb120-69"><a href="module-3.html#cb120-69" tabindex="-1"></a>uhr_data_long <span class="ot">=</span> <span class="fu">melt</span>(<span class="fu">setDT</span>(uhr_data), <span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_1&quot;</span>), <span class="at">variable.name =</span> <span class="st">&quot;sample&quot;</span>)</span>
<span id="cb120-70"><a href="module-3.html#cb120-70" tabindex="-1"></a>uhr_data_long<span class="sc">$</span>mix <span class="ot">=</span> <span class="st">&quot;mix 1&quot;</span></span>
<span id="cb120-71"><a href="module-3.html#cb120-71" tabindex="-1"></a><span class="fu">names</span>(uhr_data_long) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;concentration&quot;</span>, <span class="st">&quot;sample&quot;</span>, <span class="st">&quot;tpm&quot;</span>, <span class="st">&quot;mix&quot;</span>)</span>
<span id="cb120-72"><a href="module-3.html#cb120-72" tabindex="-1"></a></span>
<span id="cb120-73"><a href="module-3.html#cb120-73" tabindex="-1"></a><span class="co">#convert HBR data to &quot;long&quot; format</span></span>
<span id="cb120-74"><a href="module-3.html#cb120-74" tabindex="-1"></a>hbr_data <span class="ot">=</span> ercc_ref_tpms[,<span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_2&quot;</span>,<span class="st">&quot;HBR_Rep1&quot;</span>,<span class="st">&quot;HBR_Rep2&quot;</span>,<span class="st">&quot;HBR_Rep3&quot;</span>)]</span>
<span id="cb120-75"><a href="module-3.html#cb120-75" tabindex="-1"></a>hbr_data_long <span class="ot">=</span> <span class="fu">melt</span>(<span class="fu">setDT</span>(hbr_data), <span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>,<span class="st">&quot;subgroup&quot;</span>,<span class="st">&quot;ref_conc_mix_2&quot;</span>), <span class="at">variable.name =</span> <span class="st">&quot;sample&quot;</span>)</span>
<span id="cb120-76"><a href="module-3.html#cb120-76" tabindex="-1"></a>hbr_data_long<span class="sc">$</span>mix <span class="ot">=</span> <span class="st">&quot;mix 2&quot;</span></span>
<span id="cb120-77"><a href="module-3.html#cb120-77" tabindex="-1"></a><span class="fu">names</span>(hbr_data_long) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;concentration&quot;</span>, <span class="st">&quot;sample&quot;</span>, <span class="st">&quot;tpm&quot;</span>, <span class="st">&quot;mix&quot;</span>)</span>
<span id="cb120-78"><a href="module-3.html#cb120-78" tabindex="-1"></a></span>
<span id="cb120-79"><a href="module-3.html#cb120-79" tabindex="-1"></a><span class="co">#rejoin the UHR and HBR tpm data</span></span>
<span id="cb120-80"><a href="module-3.html#cb120-80" tabindex="-1"></a>ercc_ref_tpms_long <span class="ot">&lt;-</span> <span class="fu">rbind</span>(uhr_data_long, hbr_data_long)</span>
<span id="cb120-81"><a href="module-3.html#cb120-81" tabindex="-1"></a><span class="fu">head</span>(ercc_ref_tpms_long)</span>
<span id="cb120-82"><a href="module-3.html#cb120-82" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref_tpms_long)</span>
<span id="cb120-83"><a href="module-3.html#cb120-83" tabindex="-1"></a></span>
<span id="cb120-84"><a href="module-3.html#cb120-84" tabindex="-1"></a><span class="co">#fit a linear model and calculate correlation between expected concentations and observed TPM values</span></span>
<span id="cb120-85"><a href="module-3.html#cb120-85" tabindex="-1"></a>min_nonzero_tpm <span class="ot">=</span> <span class="fu">min</span>(ercc_ref_tpms_long<span class="sc">$</span>tpm[ercc_ref_tpms_long<span class="sc">$</span>tpm <span class="sc">&gt;</span> <span class="dv">0</span>])</span>
<span id="cb120-86"><a href="module-3.html#cb120-86" tabindex="-1"></a>ercc_ref_tpms_long<span class="sc">$</span>log_tpm <span class="ot">=</span> <span class="fu">log2</span>(ercc_ref_tpms_long<span class="sc">$</span>tpm <span class="sc">+</span> min_nonzero_tpm)</span>
<span id="cb120-87"><a href="module-3.html#cb120-87" tabindex="-1"></a></span>
<span id="cb120-88"><a href="module-3.html#cb120-88" tabindex="-1"></a>min_nonzero_conc <span class="ot">=</span> <span class="fu">min</span>(ercc_ref_tpms_long<span class="sc">$</span>concentration[ercc_ref_tpms_long<span class="sc">$</span>concentration <span class="sc">&gt;</span> <span class="dv">0</span>])</span>
<span id="cb120-89"><a href="module-3.html#cb120-89" tabindex="-1"></a>ercc_ref_tpms_long<span class="sc">$</span>log_concentration<span class="ot">=</span> <span class="fu">log2</span>(ercc_ref_tpms_long<span class="sc">$</span>concentration <span class="sc">+</span> min_nonzero_conc)</span>
<span id="cb120-90"><a href="module-3.html#cb120-90" tabindex="-1"></a></span>
<span id="cb120-91"><a href="module-3.html#cb120-91" tabindex="-1"></a>tpm_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(log_tpm <span class="sc">~</span> log_concentration, <span class="at">data=</span>ercc_ref_tpms_long)</span>
<span id="cb120-92"><a href="module-3.html#cb120-92" tabindex="-1"></a>tpm_r_squared <span class="ot">=</span> <span class="fu">summary</span>(tpm_model)[[<span class="st">&quot;r.squared&quot;</span>]]</span>
<span id="cb120-93"><a href="module-3.html#cb120-93" tabindex="-1"></a>tpm_slope <span class="ot">=</span> <span class="fu">coef</span>(tpm_model)[<span class="st">&quot;log_concentration&quot;</span>]</span>
<span id="cb120-94"><a href="module-3.html#cb120-94" tabindex="-1"></a></span>
<span id="cb120-95"><a href="module-3.html#cb120-95" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">ggplot</span>(ercc_ref_tpms_long, <span class="fu">aes</span>(<span class="at">x=</span>log_concentration, <span class="at">y=</span>log_tpm))</span>
<span id="cb120-96"><a href="module-3.html#cb120-96" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">shape=</span>mix, <span class="at">color=</span>sample))</span>
<span id="cb120-97"><a href="module-3.html#cb120-97" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm) </span>
<span id="cb120-98"><a href="module-3.html#cb120-98" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dv">10</span>, <span class="dv">5</span>, <span class="at">label=</span><span class="fu">paste</span>(<span class="st">&quot;R^2 =&quot;</span>, <span class="fu">round</span>(tpm_r_squared, <span class="at">digits=</span><span class="dv">2</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>)) </span>
<span id="cb120-99"><a href="module-3.html#cb120-99" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dv">10</span>, <span class="dv">4</span>, <span class="at">label=</span><span class="fu">paste</span>(<span class="st">&quot;Slope =&quot;</span>, <span class="fu">round</span>(tpm_slope, <span class="at">digits=</span><span class="dv">2</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>))</span>
<span id="cb120-100"><a href="module-3.html#cb120-100" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Log2 (expected concentration [amol/uL] + min non-zero value)&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Log2 (observed TPM estimate + min non-zero value)&quot;</span>)</span>
<span id="cb120-101"><a href="module-3.html#cb120-101" tabindex="-1"></a></span>
<span id="cb120-102"><a href="module-3.html#cb120-102" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;ERCC_TPM_Expression_vs_SpikeInConcentration.pdf&quot;</span>)</span>
<span id="cb120-103"><a href="module-3.html#cb120-103" tabindex="-1"></a><span class="fu">print</span>(p2)</span>
<span id="cb120-104"><a href="module-3.html#cb120-104" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb120-105"><a href="module-3.html#cb120-105" tabindex="-1"></a></span>
<span id="cb120-106"><a href="module-3.html#cb120-106" tabindex="-1"></a><span class="co"># Exit the R session</span></span>
<span id="cb120-107"><a href="module-3.html#cb120-107" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save=</span><span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>To view the resulting figures, navigate to the below URL replacing YOUR_IP_ADDRESS with your amazon instance IP address:
$RNA_HOME/ercc_spikein_analysis/</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/expression/ercc_spikein_analysis/ERCC_Count_Expression_vs_SpikeInConcentration.pdf</li>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/expression/ercc_spikein_analysis/ERCC_TPM_Expression_vs_SpikeInConcentration.pdf</li>
</ul>
<p>Which expression estimation (read counts or TPM values) are better representing the known/expected ERCC concentrations? Why?</p>
<p>Some notes of interpretation</p>
<ul>
<li>In general the expression estimates for ERCC transcripts we are getting from our data are highly correlated with the expected concentrations for the ERCC spike-in reagent</li>
<li>There are some ERRC transcripts that were not detected in our data at all (count and TPM of 0). These correspond to a range of expected concentrations in the spile-in reagent but they are all at the lower end. Essentially this indicates a sensitivity limitation. With our downsampled RNAseq data, we are failing to detect some of the less abundant spiked-in ERCC transcripts. This is probably hurting our R squared values slightly.</li>
<li>We observed a wide range of observed expression values for ERCC transcripts in both mixes. Remember that both mixes have ERCCs at low, medium, high levels (spread over 5-6 orders of magnitude). But between the two mixes the ERCCs at each expected concentration are different.</li>
<li>Details on the <a href="https://rnabio.org/assets/module_1/ERCC.pdf">ERCC spike-in reagent</a></li>
</ul>
</div>
</div>
<div id="differential-expression-mini-lecture" class="section level3 hasAnchor">
<h3>Differential Expression mini lecture<a href="module-3.html#differential-expression-mini-lecture" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a brief refresher on differential expression analysis, please refer to the <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_03_DifferentialExpression.pdf">mini lecture</a>.</p>
</div>
<div id="ballgown-de-analysis" class="section level3 hasAnchor">
<h3>Ballgown DE Analysis<a href="module-3.html#ballgown-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Use Ballgown to compare the UHR and HBR conditions. Refer to the Ballgown manual for a more detailed explanation:</p>
<ul>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/ballgown.html">https://www.bioconductor.org/packages/release/bioc/html/ballgown.html</a></li>
</ul>
<p>Create and change to ballgown ref-only results directory:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb121-1"><a href="module-3.html#cb121-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/</span>
<span id="cb121-2"><a href="module-3.html#cb121-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/</span></code></pre></div>
<p>Perform UHR vs. HBR comparison, using all replicates, for known (reference only mode) transcripts:</p>
<p>First, start an R session:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb122-1"><a href="module-3.html#cb122-1" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Run the following R commands in your R session.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="module-3.html#cb123-1" tabindex="-1"></a><span class="co"># load the required libraries</span></span>
<span id="cb123-2"><a href="module-3.html#cb123-2" tabindex="-1"></a><span class="fu">library</span>(ballgown)</span>
<span id="cb123-3"><a href="module-3.html#cb123-3" tabindex="-1"></a><span class="fu">library</span>(genefilter)</span>
<span id="cb123-4"><a href="module-3.html#cb123-4" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb123-5"><a href="module-3.html#cb123-5" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb123-6"><a href="module-3.html#cb123-6" tabindex="-1"></a></span>
<span id="cb123-7"><a href="module-3.html#cb123-7" tabindex="-1"></a><span class="co"># Create phenotype data needed for ballgown analysis</span></span>
<span id="cb123-8"><a href="module-3.html#cb123-8" tabindex="-1"></a>ids <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR_Rep1&quot;</span>, <span class="st">&quot;UHR_Rep2&quot;</span>, <span class="st">&quot;UHR_Rep3&quot;</span>, <span class="st">&quot;HBR_Rep1&quot;</span>, <span class="st">&quot;HBR_Rep2&quot;</span>, <span class="st">&quot;HBR_Rep3&quot;</span>)</span>
<span id="cb123-9"><a href="module-3.html#cb123-9" tabindex="-1"></a>type <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>)</span>
<span id="cb123-10"><a href="module-3.html#cb123-10" tabindex="-1"></a>inputs <span class="ot">=</span> <span class="st">&quot;/home/ubuntu/workspace/rnaseq/expression/stringtie/ref_only/&quot;</span></span>
<span id="cb123-11"><a href="module-3.html#cb123-11" tabindex="-1"></a>path <span class="ot">=</span> <span class="fu">paste</span>(inputs, ids, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb123-12"><a href="module-3.html#cb123-12" tabindex="-1"></a>pheno_data <span class="ot">=</span> <span class="fu">data.frame</span>(ids, type, path)</span>
<span id="cb123-13"><a href="module-3.html#cb123-13" tabindex="-1"></a></span>
<span id="cb123-14"><a href="module-3.html#cb123-14" tabindex="-1"></a><span class="co"># Load ballgown data structure and save it to a variable &quot;bg&quot;</span></span>
<span id="cb123-15"><a href="module-3.html#cb123-15" tabindex="-1"></a>bg <span class="ot">=</span> <span class="fu">ballgown</span>(<span class="at">samples =</span> <span class="fu">as.vector</span>(pheno_data<span class="sc">$</span>path), <span class="at">pData =</span> pheno_data)</span>
<span id="cb123-16"><a href="module-3.html#cb123-16" tabindex="-1"></a></span>
<span id="cb123-17"><a href="module-3.html#cb123-17" tabindex="-1"></a><span class="co"># Display a description of this object</span></span>
<span id="cb123-18"><a href="module-3.html#cb123-18" tabindex="-1"></a>bg</span>
<span id="cb123-19"><a href="module-3.html#cb123-19" tabindex="-1"></a></span>
<span id="cb123-20"><a href="module-3.html#cb123-20" tabindex="-1"></a><span class="co"># Load all attributes including gene name</span></span>
<span id="cb123-21"><a href="module-3.html#cb123-21" tabindex="-1"></a>bg_table <span class="ot">=</span> <span class="fu">texpr</span>(bg, <span class="st">&#39;all&#39;</span>)</span>
<span id="cb123-22"><a href="module-3.html#cb123-22" tabindex="-1"></a>bg_gene_names <span class="ot">=</span> <span class="fu">unique</span>(bg_table[, <span class="dv">9</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb123-23"><a href="module-3.html#cb123-23" tabindex="-1"></a>bg_transcript_names <span class="ot">=</span> <span class="fu">unique</span>(bg_table[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">6</span>)])</span>
<span id="cb123-24"><a href="module-3.html#cb123-24" tabindex="-1"></a></span>
<span id="cb123-25"><a href="module-3.html#cb123-25" tabindex="-1"></a><span class="co"># Save the ballgown object to a file for later use</span></span>
<span id="cb123-26"><a href="module-3.html#cb123-26" tabindex="-1"></a><span class="fu">save</span>(bg, <span class="at">file =</span> <span class="st">&#39;bg.rda&#39;</span>)</span>
<span id="cb123-27"><a href="module-3.html#cb123-27" tabindex="-1"></a></span>
<span id="cb123-28"><a href="module-3.html#cb123-28" tabindex="-1"></a><span class="co"># Pull the gene and transcript expression data frame from the ballgown object</span></span>
<span id="cb123-29"><a href="module-3.html#cb123-29" tabindex="-1"></a>gene_expression <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">gexpr</span>(bg))</span>
<span id="cb123-30"><a href="module-3.html#cb123-30" tabindex="-1"></a>transcript_expression <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">texpr</span>(bg))</span>
<span id="cb123-31"><a href="module-3.html#cb123-31" tabindex="-1"></a></span>
<span id="cb123-32"><a href="module-3.html#cb123-32" tabindex="-1"></a><span class="co"># Perform differential expression (DE) analysis with no filtering, at both gene and transcript level</span></span>
<span id="cb123-33"><a href="module-3.html#cb123-33" tabindex="-1"></a><span class="co"># Then add on transcript/gene names and sample level fpkm values for export</span></span>
<span id="cb123-34"><a href="module-3.html#cb123-34" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">stattest</span>(bg, <span class="at">feature =</span> <span class="st">&quot;transcript&quot;</span>, <span class="at">covariate =</span> <span class="st">&quot;type&quot;</span>, <span class="at">getFC =</span> <span class="cn">TRUE</span>, <span class="at">meas =</span> <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb123-35"><a href="module-3.html#cb123-35" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">merge</span>(results_transcripts, bg_transcript_names, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;t_id&quot;</span>))</span>
<span id="cb123-36"><a href="module-3.html#cb123-36" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">merge</span>(results_transcripts, transcript_expression, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;row.names&quot;</span>))</span>
<span id="cb123-37"><a href="module-3.html#cb123-37" tabindex="-1"></a></span>
<span id="cb123-38"><a href="module-3.html#cb123-38" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">stattest</span>(bg, <span class="at">feature =</span> <span class="st">&quot;gene&quot;</span>, <span class="at">covariate =</span> <span class="st">&quot;type&quot;</span>, <span class="at">getFC =</span> <span class="cn">TRUE</span>, <span class="at">meas =</span> <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb123-39"><a href="module-3.html#cb123-39" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">merge</span>(results_genes, bg_gene_names, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;gene_id&quot;</span>))</span>
<span id="cb123-40"><a href="module-3.html#cb123-40" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">merge</span>(results_genes, gene_expression, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;row.names&quot;</span>))</span>
<span id="cb123-41"><a href="module-3.html#cb123-41" tabindex="-1"></a></span>
<span id="cb123-42"><a href="module-3.html#cb123-42" tabindex="-1"></a><span class="co"># Save a tab delimited file for both the transcript and gene results</span></span>
<span id="cb123-43"><a href="module-3.html#cb123-43" tabindex="-1"></a><span class="fu">write.table</span>(results_transcripts, <span class="st">&quot;UHR_vs_HBR_transcript_results.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-44"><a href="module-3.html#cb123-44" tabindex="-1"></a><span class="fu">write.table</span>(results_genes, <span class="st">&quot;UHR_vs_HBR_gene_results.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-45"><a href="module-3.html#cb123-45" tabindex="-1"></a></span>
<span id="cb123-46"><a href="module-3.html#cb123-46" tabindex="-1"></a><span class="co"># Filter low-abundance genes. Here we remove all transcripts with a variance across the samples of less than one</span></span>
<span id="cb123-47"><a href="module-3.html#cb123-47" tabindex="-1"></a>bg_filt <span class="ot">=</span> <span class="fu">subset</span> (bg, <span class="st">&quot;SparseArray::rowVars(texpr(bg)) &gt; 1&quot;</span>, <span class="at">genomesubset =</span> <span class="cn">TRUE</span>)</span>
<span id="cb123-48"><a href="module-3.html#cb123-48" tabindex="-1"></a></span>
<span id="cb123-49"><a href="module-3.html#cb123-49" tabindex="-1"></a><span class="co"># Load all attributes including gene name</span></span>
<span id="cb123-50"><a href="module-3.html#cb123-50" tabindex="-1"></a>bg_filt_table <span class="ot">=</span> <span class="fu">texpr</span>(bg_filt , <span class="st">&#39;all&#39;</span>)</span>
<span id="cb123-51"><a href="module-3.html#cb123-51" tabindex="-1"></a>bg_filt_gene_names <span class="ot">=</span> <span class="fu">unique</span>(bg_filt_table[, <span class="dv">9</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb123-52"><a href="module-3.html#cb123-52" tabindex="-1"></a>bg_filt_transcript_names <span class="ot">=</span> <span class="fu">unique</span>(bg_filt_table[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">6</span>)])</span>
<span id="cb123-53"><a href="module-3.html#cb123-53" tabindex="-1"></a></span>
<span id="cb123-54"><a href="module-3.html#cb123-54" tabindex="-1"></a><span class="co"># Perform DE analysis now using the filtered data</span></span>
<span id="cb123-55"><a href="module-3.html#cb123-55" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">stattest</span>(bg_filt, <span class="at">feature =</span> <span class="st">&quot;transcript&quot;</span>, <span class="at">covariate =</span> <span class="st">&quot;type&quot;</span>, <span class="at">getFC =</span> <span class="cn">TRUE</span>, <span class="at">meas =</span> <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb123-56"><a href="module-3.html#cb123-56" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">merge</span>(results_transcripts, bg_filt_transcript_names, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;t_id&quot;</span>))</span>
<span id="cb123-57"><a href="module-3.html#cb123-57" tabindex="-1"></a>results_transcripts <span class="ot">=</span> <span class="fu">merge</span>(results_transcripts, transcript_expression, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;row.names&quot;</span>))</span>
<span id="cb123-58"><a href="module-3.html#cb123-58" tabindex="-1"></a></span>
<span id="cb123-59"><a href="module-3.html#cb123-59" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">stattest</span>(bg_filt, <span class="at">feature =</span> <span class="st">&quot;gene&quot;</span>, <span class="at">covariate =</span> <span class="st">&quot;type&quot;</span>, <span class="at">getFC =</span> <span class="cn">TRUE</span>, <span class="at">meas =</span> <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb123-60"><a href="module-3.html#cb123-60" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">merge</span>(results_genes, bg_filt_gene_names, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;gene_id&quot;</span>))</span>
<span id="cb123-61"><a href="module-3.html#cb123-61" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">merge</span>(results_genes, gene_expression, <span class="at">by.x =</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>), <span class="at">by.y =</span> <span class="fu">c</span>(<span class="st">&quot;row.names&quot;</span>))</span>
<span id="cb123-62"><a href="module-3.html#cb123-62" tabindex="-1"></a></span>
<span id="cb123-63"><a href="module-3.html#cb123-63" tabindex="-1"></a><span class="co"># Output the filtered list of genes and transcripts and save to tab delimited files</span></span>
<span id="cb123-64"><a href="module-3.html#cb123-64" tabindex="-1"></a><span class="fu">write.table</span>(results_transcripts, <span class="st">&quot;UHR_vs_HBR_transcript_results_filtered.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-65"><a href="module-3.html#cb123-65" tabindex="-1"></a><span class="fu">write.table</span>(results_genes, <span class="st">&quot;UHR_vs_HBR_gene_results_filtered.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-66"><a href="module-3.html#cb123-66" tabindex="-1"></a></span>
<span id="cb123-67"><a href="module-3.html#cb123-67" tabindex="-1"></a><span class="co"># Identify the significant genes with p-value &lt; 0.05</span></span>
<span id="cb123-68"><a href="module-3.html#cb123-68" tabindex="-1"></a>sig_transcripts <span class="ot">=</span> <span class="fu">subset</span>(results_transcripts, results_transcripts<span class="sc">$</span>pval<span class="sc">&lt;</span><span class="fl">0.05</span>)</span>
<span id="cb123-69"><a href="module-3.html#cb123-69" tabindex="-1"></a>sig_genes <span class="ot">=</span> <span class="fu">subset</span>(results_genes, results_genes<span class="sc">$</span>pval<span class="sc">&lt;</span><span class="fl">0.05</span>)</span>
<span id="cb123-70"><a href="module-3.html#cb123-70" tabindex="-1"></a></span>
<span id="cb123-71"><a href="module-3.html#cb123-71" tabindex="-1"></a><span class="co"># Output the significant gene results to a pair of tab delimited files</span></span>
<span id="cb123-72"><a href="module-3.html#cb123-72" tabindex="-1"></a><span class="fu">write.table</span>(sig_transcripts, <span class="st">&quot;UHR_vs_HBR_transcript_results_sig.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-73"><a href="module-3.html#cb123-73" tabindex="-1"></a><span class="fu">write.table</span>(sig_genes, <span class="st">&quot;UHR_vs_HBR_gene_results_sig.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-74"><a href="module-3.html#cb123-74" tabindex="-1"></a></span>
<span id="cb123-75"><a href="module-3.html#cb123-75" tabindex="-1"></a><span class="co"># Exit the R session</span></span>
<span id="cb123-76"><a href="module-3.html#cb123-76" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>Once you have completed the Ballgown analysis in R, exit the R session and continue with the steps below. A copy of the above R code is located <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/scripts/Tutorial_Part1_ballgown.R">here</a>.</p>
<p>What does the raw output from Ballgown look like?</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb124-1"><a href="module-3.html#cb124-1" tabindex="-1"></a><span class="fu">head</span> UHR_vs_HBR_gene_results.tsv</span></code></pre></div>
<p>How many genes are there on this chromosome?</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb125-1"><a href="module-3.html#cb125-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>How many passed filter in UHR or HBR?</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb126-1"><a href="module-3.html#cb126-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results_filtered.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>How many differentially expressed genes were found on this chromosome (p-value &lt; 0.05)?</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb127-1"><a href="module-3.html#cb127-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results_sig.tsv <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>Display the top 20 DE genes. Look at some of those genes in IGV - do they make sense?</p>
<p>In the following commands we use <code>grep -v feature</code> to remove lines that contain “feature”. Then we use <code>sort</code> to sort the data in various ways. The <code>k</code> option specifies that we want to sort on a particular column (<code>3</code> in this case which has the DE fold change values). The <code>n</code> option tells <code>sort</code> to sort numerically. The <code>r</code> option tells <code>sort</code> to reverse the sort.</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb128-1"><a href="module-3.html#cb128-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results_sig.tsv <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-rnk</span> 3 <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 20 <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="co">#Higher abundance in UHR</span></span>
<span id="cb128-2"><a href="module-3.html#cb128-2" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results_sig.tsv <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-nk</span> 3 <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 20 <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="co">#Higher abundance in HBR</span></span></code></pre></div>
<p>Save all genes with P&lt;0.05 to a new file.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb129-1"><a href="module-3.html#cb129-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> feature UHR_vs_HBR_gene_results_sig.tsv <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f</span> 6 <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/\&quot;//g&#39;</span> <span class="op">&gt;</span> DE_genes.txt</span>
<span id="cb129-2"><a href="module-3.html#cb129-2" tabindex="-1"></a><span class="fu">head</span> DE_genes.txt</span></code></pre></div>
<hr />
</div>
<div id="practical-exercise-9" class="section level3 hasAnchor">
<h3>PRACTICAL EXERCISE 9<a href="module-3.html#practical-exercise-9" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Assignment: Use Ballgown to identify differentially expressed genes from the StringTie expression estimates (i.e., Ballgown table files) which you created in Practical Exercise 8.</p>
<ul>
<li>Hint: Follow the example R code above.</li>
<li>Hint: You will need to change how the <code>pheno_data</code> object is created to point to the correct sample ids, type, and path to your inputs (the StringTie results files).</li>
<li>Hint: Make sure to save your ballgown data object to file (e.g., <code>bg.rda</code>) for use in subsequent practical exercises.</li>
<li>Hint: You may wish to save both a complete list of genes with differential expression results as well as a subset which are filtered and pass a significance test</li>
</ul>
<p>Solution: When you are ready you can check your approach against the <a href="/module-09-appendix/0009/05/01/Practical_Exercise_Solutions/#practical-exercise-9---differential-expression">Solutions</a></p>
<hr />
</div>
<div id="ercc-de-analysis" class="section level3 hasAnchor">
<h3>ERCC DE Analysis<a href="module-3.html#ercc-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This section will compare the differential expression estimates obtained by the RNAseq analysis against the expected differential expression results for the ERCC spike-in RNAs (mix1-UHR vs mix2-HBR):</p>
<p>First set up a directory to store the results of this analysis.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb130-1"><a href="module-3.html#cb130-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="va">$RNA_HOME</span>/de/ercc_spikein_analysis/</span>
<span id="cb130-2"><a href="module-3.html#cb130-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/ercc_spikein_analysis/</span>
<span id="cb130-3"><a href="module-3.html#cb130-3" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/ERCC_Controls_Analysis.txt</span>
<span id="cb130-4"><a href="module-3.html#cb130-4" tabindex="-1"></a><span class="fu">cat</span> ERCC_Controls_Analysis.txt</span></code></pre></div>
<p>Using R load the expected and observed ERCC DE results and produce a visualization.</p>
<p>First, start an R session:</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="module-3.html#cb131-1" tabindex="-1"></a>R</span></code></pre></div>
<p>Work through the following R commands</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="module-3.html#cb132-1" tabindex="-1"></a></span>
<span id="cb132-2"><a href="module-3.html#cb132-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb132-3"><a href="module-3.html#cb132-3" tabindex="-1"></a></span>
<span id="cb132-4"><a href="module-3.html#cb132-4" tabindex="-1"></a><span class="co">#load the ERCC expected fold change values for mix1 vs mix2</span></span>
<span id="cb132-5"><a href="module-3.html#cb132-5" tabindex="-1"></a>ercc_ref <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;ERCC_Controls_Analysis.txt&quot;</span>, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb132-6"><a href="module-3.html#cb132-6" tabindex="-1"></a><span class="fu">names</span>(ercc_ref) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;ercc_id&quot;</span>, <span class="st">&quot;subgroup&quot;</span>, <span class="st">&quot;ref_conc_mix_1&quot;</span>, <span class="st">&quot;ref_conc_mix_2&quot;</span>, <span class="st">&quot;ref_fc_mix1_vs_mix2&quot;</span>, <span class="st">&quot;ref_log2_mix1_vs_mix2&quot;</span>)</span>
<span id="cb132-7"><a href="module-3.html#cb132-7" tabindex="-1"></a><span class="fu">head</span>(ercc_ref)</span>
<span id="cb132-8"><a href="module-3.html#cb132-8" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref)</span>
<span id="cb132-9"><a href="module-3.html#cb132-9" tabindex="-1"></a></span>
<span id="cb132-10"><a href="module-3.html#cb132-10" tabindex="-1"></a><span class="co">#load the observed fold change values determined by our RNA-seq analysis</span></span>
<span id="cb132-11"><a href="module-3.html#cb132-11" tabindex="-1"></a>rna_de_file <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/de/ballgown/ref_only/UHR_vs_HBR_gene_results.tsv&quot;</span>;</span>
<span id="cb132-12"><a href="module-3.html#cb132-12" tabindex="-1"></a>rna_de <span class="ot">=</span> <span class="fu">read.table</span>(rna_de_file, <span class="at">header=</span><span class="cn">TRUE</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb132-13"><a href="module-3.html#cb132-13" tabindex="-1"></a><span class="fu">tail</span>(rna_de)</span>
<span id="cb132-14"><a href="module-3.html#cb132-14" tabindex="-1"></a><span class="fu">dim</span>(rna_de)</span>
<span id="cb132-15"><a href="module-3.html#cb132-15" tabindex="-1"></a></span>
<span id="cb132-16"><a href="module-3.html#cb132-16" tabindex="-1"></a><span class="co">#combine the expected and observed data into a single data table</span></span>
<span id="cb132-17"><a href="module-3.html#cb132-17" tabindex="-1"></a>ercc_ref_de <span class="ot">=</span> <span class="fu">merge</span>(<span class="at">x =</span> ercc_ref, <span class="at">y =</span> rna_de, <span class="at">by.x =</span> <span class="st">&quot;ercc_id&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;id&quot;</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb132-18"><a href="module-3.html#cb132-18" tabindex="-1"></a><span class="fu">head</span>(ercc_ref_de)</span>
<span id="cb132-19"><a href="module-3.html#cb132-19" tabindex="-1"></a><span class="fu">dim</span>(ercc_ref_de)</span>
<span id="cb132-20"><a href="module-3.html#cb132-20" tabindex="-1"></a></span>
<span id="cb132-21"><a href="module-3.html#cb132-21" tabindex="-1"></a><span class="co">#convert fold change values to log2 scale</span></span>
<span id="cb132-22"><a href="module-3.html#cb132-22" tabindex="-1"></a>ercc_ref_de<span class="sc">$</span>observed_log2_fc <span class="ot">=</span> <span class="fu">log2</span>(ercc_ref_de<span class="sc">$</span>fc)</span>
<span id="cb132-23"><a href="module-3.html#cb132-23" tabindex="-1"></a>ercc_ref_de<span class="sc">$</span>expected_log2_fc <span class="ot">=</span> ercc_ref_de<span class="sc">$</span>ref_log2_mix1_vs_mix2</span>
<span id="cb132-24"><a href="module-3.html#cb132-24" tabindex="-1"></a></span>
<span id="cb132-25"><a href="module-3.html#cb132-25" tabindex="-1"></a><span class="co">#fit a linear model and calculate R squared between the observed and expected fold change values</span></span>
<span id="cb132-26"><a href="module-3.html#cb132-26" tabindex="-1"></a>model <span class="ot">=</span> <span class="fu">lm</span>(observed_log2_fc <span class="sc">~</span> expected_log2_fc, <span class="at">data=</span>ercc_ref_de)</span>
<span id="cb132-27"><a href="module-3.html#cb132-27" tabindex="-1"></a>r_squared <span class="ot">=</span> <span class="fu">summary</span>(model)[[<span class="st">&quot;r.squared&quot;</span>]]</span>
<span id="cb132-28"><a href="module-3.html#cb132-28" tabindex="-1"></a></span>
<span id="cb132-29"><a href="module-3.html#cb132-29" tabindex="-1"></a><span class="co">#create a scatterplot to compare the observed and expected fold change values</span></span>
<span id="cb132-30"><a href="module-3.html#cb132-30" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ggplot</span>(ercc_ref_de, <span class="fu">aes</span>(<span class="at">x =</span> expected_log2_fc, <span class="at">y =</span> observed_log2_fc))</span>
<span id="cb132-31"><a href="module-3.html#cb132-31" tabindex="-1"></a>p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> subgroup)) </span>
<span id="cb132-32"><a href="module-3.html#cb132-32" tabindex="-1"></a>p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm) </span>
<span id="cb132-33"><a href="module-3.html#cb132-33" tabindex="-1"></a>p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">&quot;text&quot;</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="at">label=</span><span class="fu">paste</span>(<span class="st">&quot;R^2 =&quot;</span>, r_squared, <span class="at">sep=</span><span class="st">&quot; &quot;</span>))</span>
<span id="cb132-34"><a href="module-3.html#cb132-34" tabindex="-1"></a>p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Expected Fold Change (log2 scale)&quot;</span>) </span>
<span id="cb132-35"><a href="module-3.html#cb132-35" tabindex="-1"></a>p <span class="ot">=</span> p <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Observed Fold Change in RNA-seq data (log2 scale)&quot;</span>)</span>
<span id="cb132-36"><a href="module-3.html#cb132-36" tabindex="-1"></a></span>
<span id="cb132-37"><a href="module-3.html#cb132-37" tabindex="-1"></a><span class="co">#save the plot to a PDF</span></span>
<span id="cb132-38"><a href="module-3.html#cb132-38" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;ERCC_Ballgown-DE_vs_SpikeInDE.pdf&quot;</span>)</span>
<span id="cb132-39"><a href="module-3.html#cb132-39" tabindex="-1"></a><span class="fu">print</span>(p)</span>
<span id="cb132-40"><a href="module-3.html#cb132-40" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb132-41"><a href="module-3.html#cb132-41" tabindex="-1"></a></span>
<span id="cb132-42"><a href="module-3.html#cb132-42" tabindex="-1"></a><span class="co"># Exit the R session</span></span>
<span id="cb132-43"><a href="module-3.html#cb132-43" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save=</span><span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>View the results here:</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ercc_spikein_analysis/ERCC_Ballgown-DE_vs_SpikeInDE.pdf</li>
</ul>
</div>
<div id="differential-expression-mini-lecture-1" class="section level3 hasAnchor">
<h3>Differential Expression mini lecture<a href="module-3.html#differential-expression-mini-lecture-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a brief refresher on differential expression analysis, please refer to the <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_03_DifferentialExpression.pdf">mini lecture</a>.</p>
</div>
<div id="edger-de-analysis" class="section level3 hasAnchor">
<h3>edgeR DE Analysis<a href="module-3.html#edger-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial you will:</p>
<ul>
<li>Make use of the raw counts you generated previously using htseq-count</li>
<li>edgeR is a bioconductor package designed specifically for differential expression of count-based RNA-seq data</li>
<li>This is an alternative to using stringtie/ballgown to find differentially expressed genes</li>
</ul>
<p>First, create a directory for results:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb133-1"><a href="module-3.html#cb133-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/</span>
<span id="cb133-2"><a href="module-3.html#cb133-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> de/htseq_counts</span>
<span id="cb133-3"><a href="module-3.html#cb133-3" tabindex="-1"></a><span class="bu">cd</span> de/htseq_counts</span></code></pre></div>
<p>Note that the htseq-count results provide counts for each gene but uses only the Ensembl Gene ID (e.g. ENSG00000054611). This is not very convenient for biological interpretation. This next step creates a mapping file that will help us translate from ENSG IDs to Symbols. It does this by parsing the GTF transcriptome file we got from Ensembl. That file contains both gene names and IDs. Unfortunately, this file is a bit complex to parse. Furthermore, it contains the ERCC transcripts, and these have their own naming convention which also complicated the parsing.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb134-1"><a href="module-3.html#cb134-1" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;if ($_ =~ /gene_id\s\&quot;(ENSG\S+)\&quot;\;/) { $id = $1; $name = undef; if ($_ =~ /gene_name\s\&quot;(\S+)&quot;\;/) { $name = $1; }; }; if ($id &amp;&amp; $name) {print &quot;$id\t$name\n&quot;;} if ($_=~/gene_id\s\&quot;(ERCC\S+)\&quot;/){print &quot;$1\t$1\n&quot;;}&#39;</span> <span class="va">$RNA_REF_GTF</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> ENSG_ID2Name.txt</span>
<span id="cb134-2"><a href="module-3.html#cb134-2" tabindex="-1"></a><span class="fu">head</span> ENSG_ID2Name.txt</span></code></pre></div>
<p>Determine the number of unique Ensembl Gene IDs and symbols. What does this tell you?</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb135-1"><a href="module-3.html#cb135-1" tabindex="-1"></a><span class="co">#count unique gene ids</span></span>
<span id="cb135-2"><a href="module-3.html#cb135-2" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb135-3"><a href="module-3.html#cb135-3" tabindex="-1"></a><span class="co">#count unique gene names</span></span>
<span id="cb135-4"><a href="module-3.html#cb135-4" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb135-5"><a href="module-3.html#cb135-5" tabindex="-1"></a></span>
<span id="cb135-6"><a href="module-3.html#cb135-6" tabindex="-1"></a><span class="co">#show the most repeated gene names</span></span>
<span id="cb135-7"><a href="module-3.html#cb135-7" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 ENSG_ID2Name.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-r</span> <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<p>Launch R:</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb136-1"><a href="module-3.html#cb136-1" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>R code has been provided below. If you wish to have a script with all of the code, it can be found <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/scripts/Tutorial_edgeR.R">here</a>. Run the R commands below.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="module-3.html#cb137-1" tabindex="-1"></a></span>
<span id="cb137-2"><a href="module-3.html#cb137-2" tabindex="-1"></a><span class="co"># set working directory where output will go</span></span>
<span id="cb137-3"><a href="module-3.html#cb137-3" tabindex="-1"></a>working_dir <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/de/htseq_counts&quot;</span></span>
<span id="cb137-4"><a href="module-3.html#cb137-4" tabindex="-1"></a><span class="fu">setwd</span>(working_dir)</span>
<span id="cb137-5"><a href="module-3.html#cb137-5" tabindex="-1"></a></span>
<span id="cb137-6"><a href="module-3.html#cb137-6" tabindex="-1"></a><span class="co"># read in gene mapping</span></span>
<span id="cb137-7"><a href="module-3.html#cb137-7" tabindex="-1"></a>mapping <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;~/workspace/rnaseq/de/htseq_counts/ENSG_ID2Name.txt&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb137-8"><a href="module-3.html#cb137-8" tabindex="-1"></a></span>
<span id="cb137-9"><a href="module-3.html#cb137-9" tabindex="-1"></a><span class="co"># read in count matrix</span></span>
<span id="cb137-10"><a href="module-3.html#cb137-10" tabindex="-1"></a>rawdata <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;~/workspace/rnaseq/expression/htseq_counts/gene_read_counts_table_all_final.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb137-11"><a href="module-3.html#cb137-11" tabindex="-1"></a></span>
<span id="cb137-12"><a href="module-3.html#cb137-12" tabindex="-1"></a><span class="co"># Check dimensions</span></span>
<span id="cb137-13"><a href="module-3.html#cb137-13" tabindex="-1"></a><span class="fu">dim</span>(rawdata)</span>
<span id="cb137-14"><a href="module-3.html#cb137-14" tabindex="-1"></a></span>
<span id="cb137-15"><a href="module-3.html#cb137-15" tabindex="-1"></a><span class="co"># Require at least 1/6 of samples to have expressed count &gt;= 10</span></span>
<span id="cb137-16"><a href="module-3.html#cb137-16" tabindex="-1"></a>sample_cutoff <span class="ot">=</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">6</span>)</span>
<span id="cb137-17"><a href="module-3.html#cb137-17" tabindex="-1"></a>count_cutoff <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb137-18"><a href="module-3.html#cb137-18" tabindex="-1"></a></span>
<span id="cb137-19"><a href="module-3.html#cb137-19" tabindex="-1"></a><span class="co">#Define a function to calculate the fraction of values expressed above the count cutoff</span></span>
<span id="cb137-20"><a href="module-3.html#cb137-20" tabindex="-1"></a>getFE <span class="ot">=</span> <span class="cf">function</span>(data,count_cutoff){</span>
<span id="cb137-21"><a href="module-3.html#cb137-21" tabindex="-1"></a> FE <span class="ot">=</span> (<span class="fu">sum</span>(data <span class="sc">&gt;=</span> count_cutoff) <span class="sc">/</span> <span class="fu">length</span>(data))</span>
<span id="cb137-22"><a href="module-3.html#cb137-22" tabindex="-1"></a> <span class="fu">return</span>(FE)</span>
<span id="cb137-23"><a href="module-3.html#cb137-23" tabindex="-1"></a>}</span>
<span id="cb137-24"><a href="module-3.html#cb137-24" tabindex="-1"></a></span>
<span id="cb137-25"><a href="module-3.html#cb137-25" tabindex="-1"></a><span class="co">#Apply the function to all genes, and filter out genes not meeting the sample cutoff</span></span>
<span id="cb137-26"><a href="module-3.html#cb137-26" tabindex="-1"></a>fraction_expressed <span class="ot">=</span> <span class="fu">apply</span>(rawdata, <span class="dv">1</span>, getFE, count_cutoff)</span>
<span id="cb137-27"><a href="module-3.html#cb137-27" tabindex="-1"></a>keep <span class="ot">=</span> <span class="fu">which</span>(fraction_expressed <span class="sc">&gt;=</span> sample_cutoff)</span>
<span id="cb137-28"><a href="module-3.html#cb137-28" tabindex="-1"></a>rawdata <span class="ot">=</span> rawdata[keep, ]</span>
<span id="cb137-29"><a href="module-3.html#cb137-29" tabindex="-1"></a></span>
<span id="cb137-30"><a href="module-3.html#cb137-30" tabindex="-1"></a><span class="co"># Check dimensions again to see effect of filtering</span></span>
<span id="cb137-31"><a href="module-3.html#cb137-31" tabindex="-1"></a><span class="fu">dim</span>(rawdata)</span>
<span id="cb137-32"><a href="module-3.html#cb137-32" tabindex="-1"></a></span>
<span id="cb137-33"><a href="module-3.html#cb137-33" tabindex="-1"></a><span class="do">#################</span></span>
<span id="cb137-34"><a href="module-3.html#cb137-34" tabindex="-1"></a><span class="co"># Running edgeR #</span></span>
<span id="cb137-35"><a href="module-3.html#cb137-35" tabindex="-1"></a><span class="do">#################</span></span>
<span id="cb137-36"><a href="module-3.html#cb137-36" tabindex="-1"></a></span>
<span id="cb137-37"><a href="module-3.html#cb137-37" tabindex="-1"></a><span class="co"># load edgeR</span></span>
<span id="cb137-38"><a href="module-3.html#cb137-38" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;edgeR&quot;</span>)</span>
<span id="cb137-39"><a href="module-3.html#cb137-39" tabindex="-1"></a></span>
<span id="cb137-40"><a href="module-3.html#cb137-40" tabindex="-1"></a><span class="co"># make class labels</span></span>
<span id="cb137-41"><a href="module-3.html#cb137-41" tabindex="-1"></a>class <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;UHR&quot;</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="st">&quot;HBR&quot;</span>, <span class="dv">3</span>))</span>
<span id="cb137-42"><a href="module-3.html#cb137-42" tabindex="-1"></a></span>
<span id="cb137-43"><a href="module-3.html#cb137-43" tabindex="-1"></a><span class="co"># Get common gene names</span></span>
<span id="cb137-44"><a href="module-3.html#cb137-44" tabindex="-1"></a>Gene <span class="ot">=</span> <span class="fu">rownames</span>(rawdata)</span>
<span id="cb137-45"><a href="module-3.html#cb137-45" tabindex="-1"></a>Symbol <span class="ot">=</span> mapping[Gene, <span class="dv">1</span>]</span>
<span id="cb137-46"><a href="module-3.html#cb137-46" tabindex="-1"></a>gene_annotations <span class="ot">=</span> <span class="fu">cbind</span>(Gene, Symbol)</span>
<span id="cb137-47"><a href="module-3.html#cb137-47" tabindex="-1"></a></span>
<span id="cb137-48"><a href="module-3.html#cb137-48" tabindex="-1"></a><span class="co"># Make DGEList object</span></span>
<span id="cb137-49"><a href="module-3.html#cb137-49" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> rawdata, <span class="at">genes =</span> gene_annotations, <span class="at">group =</span> class)</span>
<span id="cb137-50"><a href="module-3.html#cb137-50" tabindex="-1"></a><span class="fu">nrow</span>(y)</span>
<span id="cb137-51"><a href="module-3.html#cb137-51" tabindex="-1"></a></span>
<span id="cb137-52"><a href="module-3.html#cb137-52" tabindex="-1"></a><span class="co"># TMM Normalization</span></span>
<span id="cb137-53"><a href="module-3.html#cb137-53" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">calcNormFactors</span>(y)</span>
<span id="cb137-54"><a href="module-3.html#cb137-54" tabindex="-1"></a></span>
<span id="cb137-55"><a href="module-3.html#cb137-55" tabindex="-1"></a><span class="co"># Estimate dispersion</span></span>
<span id="cb137-56"><a href="module-3.html#cb137-56" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">estimateCommonDisp</span>(y, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb137-57"><a href="module-3.html#cb137-57" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">estimateTagwiseDisp</span>(y)</span>
<span id="cb137-58"><a href="module-3.html#cb137-58" tabindex="-1"></a></span>
<span id="cb137-59"><a href="module-3.html#cb137-59" tabindex="-1"></a><span class="co"># Differential expression test</span></span>
<span id="cb137-60"><a href="module-3.html#cb137-60" tabindex="-1"></a>et <span class="ot">=</span> <span class="fu">exactTest</span>(y)</span>
<span id="cb137-61"><a href="module-3.html#cb137-61" tabindex="-1"></a></span>
<span id="cb137-62"><a href="module-3.html#cb137-62" tabindex="-1"></a><span class="co"># Extract raw counts to add back onto DE results</span></span>
<span id="cb137-63"><a href="module-3.html#cb137-63" tabindex="-1"></a>counts <span class="ot">=</span> <span class="fu">getCounts</span>(y)</span>
<span id="cb137-64"><a href="module-3.html#cb137-64" tabindex="-1"></a></span>
<span id="cb137-65"><a href="module-3.html#cb137-65" tabindex="-1"></a><span class="co"># Print top genes</span></span>
<span id="cb137-66"><a href="module-3.html#cb137-66" tabindex="-1"></a><span class="fu">topTags</span>(et)</span>
<span id="cb137-67"><a href="module-3.html#cb137-67" tabindex="-1"></a></span>
<span id="cb137-68"><a href="module-3.html#cb137-68" tabindex="-1"></a><span class="co"># Print number of up/down significant genes at FDR = 0.05  significance level</span></span>
<span id="cb137-69"><a href="module-3.html#cb137-69" tabindex="-1"></a><span class="fu">summary</span>(de <span class="ot">&lt;-</span> <span class="fu">decideTests</span>(et, <span class="at">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="at">p =</span> <span class="fl">0.05</span>))</span>
<span id="cb137-70"><a href="module-3.html#cb137-70" tabindex="-1"></a></span>
<span id="cb137-71"><a href="module-3.html#cb137-71" tabindex="-1"></a><span class="co">#Get output with BH-adjusted FDR values - all genes, any p-value, unsorted</span></span>
<span id="cb137-72"><a href="module-3.html#cb137-72" tabindex="-1"></a>out <span class="ot">=</span> <span class="fu">topTags</span>(et, <span class="at">n =</span> <span class="st">&quot;Inf&quot;</span>, <span class="at">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="at">sort.by =</span> <span class="st">&quot;none&quot;</span>, <span class="at">p.value =</span> <span class="dv">1</span>)<span class="sc">$</span>table</span>
<span id="cb137-73"><a href="module-3.html#cb137-73" tabindex="-1"></a></span>
<span id="cb137-74"><a href="module-3.html#cb137-74" tabindex="-1"></a><span class="co">#Add raw counts back onto results for convenience (make sure sort and total number of elements allows proper join)</span></span>
<span id="cb137-75"><a href="module-3.html#cb137-75" tabindex="-1"></a>out2 <span class="ot">=</span> <span class="fu">cbind</span>(out, counts)</span>
<span id="cb137-76"><a href="module-3.html#cb137-76" tabindex="-1"></a></span>
<span id="cb137-77"><a href="module-3.html#cb137-77" tabindex="-1"></a><span class="co">#Limit to significantly DE genes</span></span>
<span id="cb137-78"><a href="module-3.html#cb137-78" tabindex="-1"></a>out3 <span class="ot">=</span> out2[<span class="fu">as.logical</span>(de), ]</span>
<span id="cb137-79"><a href="module-3.html#cb137-79" tabindex="-1"></a></span>
<span id="cb137-80"><a href="module-3.html#cb137-80" tabindex="-1"></a><span class="co"># Order by p-value</span></span>
<span id="cb137-81"><a href="module-3.html#cb137-81" tabindex="-1"></a>o <span class="ot">=</span> <span class="fu">order</span>(et<span class="sc">$</span>table<span class="sc">$</span>PValue[<span class="fu">as.logical</span>(de)], <span class="at">decreasing=</span><span class="cn">FALSE</span>)</span>
<span id="cb137-82"><a href="module-3.html#cb137-82" tabindex="-1"></a>out4 <span class="ot">=</span> out3[o, ]</span>
<span id="cb137-83"><a href="module-3.html#cb137-83" tabindex="-1"></a></span>
<span id="cb137-84"><a href="module-3.html#cb137-84" tabindex="-1"></a><span class="co"># Save table</span></span>
<span id="cb137-85"><a href="module-3.html#cb137-85" tabindex="-1"></a><span class="fu">write.table</span>(out4, <span class="at">file =</span> <span class="st">&quot;DE_genes.txt&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb137-86"><a href="module-3.html#cb137-86" tabindex="-1"></a></span>
<span id="cb137-87"><a href="module-3.html#cb137-87" tabindex="-1"></a><span class="co">#To exit R type the following</span></span>
<span id="cb137-88"><a href="module-3.html#cb137-88" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>Once you have run the edgeR tutorial, compare the sigDE genes to those saved earlier from ballgown:</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb138-1"><a href="module-3.html#cb138-1" tabindex="-1"></a><span class="fu">head</span> <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/DE_genes.txt</span>
<span id="cb138-2"><a href="module-3.html#cb138-2" tabindex="-1"></a><span class="fu">head</span> <span class="va">$RNA_HOME</span>/de/htseq_counts/DE_genes.txt</span></code></pre></div>
<p>Pull out the gene IDs</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb139-1"><a href="module-3.html#cb139-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/</span>
<span id="cb139-2"><a href="module-3.html#cb139-2" tabindex="-1"></a></span>
<span id="cb139-3"><a href="module-3.html#cb139-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/DE_genes.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> ballgown_DE_gene_symbols.txt</span>
<span id="cb139-4"><a href="module-3.html#cb139-4" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 <span class="va">$RNA_HOME</span>/de/htseq_counts/DE_genes.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> Gene_Name <span class="op">&gt;</span> htseq_counts_edgeR_DE_gene_symbols.txt</span></code></pre></div>
<p>Visualize overlap with a venn diagram. This can be done with simple web tools like:</p>
<ul>
<li><a href="https://www.biovenn.nl/">https://www.biovenn.nl/</a></li>
<li><a href="http://bioinfogp.cnb.csic.es/tools/venny/">http://bioinfogp.cnb.csic.es/tools/venny/</a></li>
</ul>
<p>To get the two gene lists you could use <code>cat</code> to print out each list in your terminal and then copy/paste.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb140-1"><a href="module-3.html#cb140-1" tabindex="-1"></a><span class="fu">cat</span> ballgown_DE_gene_symbols.txt</span>
<span id="cb140-2"><a href="module-3.html#cb140-2" tabindex="-1"></a><span class="fu">cat</span> htseq_counts_edgeR_DE_gene_symbols.txt</span></code></pre></div>
<p>Alternatively you could view both lists in a web browser as you have done with other files. These two files should be here:</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ballgown_DE_gene_symbols.txt</li>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/htseq_counts_edgeR_DE_gene_symbols.txt</li>
</ul>
<div id="example-venn-diagram-de-genes-from-stringtieballgown-vs-htseqedger" class="section level5 hasAnchor">
<h5>Example Venn Diagram (DE genes from StringTie/Ballgown vs HTSeq/EdgeR)<a href="module-3.html#example-venn-diagram-de-genes-from-stringtieballgown-vs-htseqedger" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p><br>
{% include figure.html image=“/assets/module_3/venn-ballgown-vs-edger.png” width=“400” %}</p>
</div>
</div>
<div id="differential-expression-mini-lecture-2" class="section level3 hasAnchor">
<h3>Differential Expression mini lecture<a href="module-3.html#differential-expression-mini-lecture-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a brief refresher on differential expression analysis, please refer to the <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_03_03_DifferentialExpression.pdf">mini lecture</a>.</p>
</div>
<div id="deseq2-de-analysis" class="section level3 hasAnchor">
<h3>DESeq2 DE Analysis<a href="module-3.html#deseq2-de-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial you will:</p>
<ul>
<li>Make use of the raw counts you generated previously using htseq-count</li>
<li>DESeq2 is a bioconductor package designed specifically for differential expression of count-based RNA-seq data</li>
<li>This is an alternative to using stringtie/ballgown to find differentially expressed genes</li>
</ul>
</div>
<div id="setup" class="section level3 hasAnchor">
<h3>Setup<a href="module-3.html#setup" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, create a directory for results. Then start R:</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb141-1"><a href="module-3.html#cb141-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/</span>
<span id="cb141-2"><a href="module-3.html#cb141-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> de/deseq2</span>
<span id="cb141-3"><a href="module-3.html#cb141-3" tabindex="-1"></a><span class="bu">cd</span> de/deseq2</span>
<span id="cb141-4"><a href="module-3.html#cb141-4" tabindex="-1"></a></span>
<span id="cb141-5"><a href="module-3.html#cb141-5" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Once we start from R, the relevant packages should already be installed. We will load the libraries, set working directories, and read in the raw read counts data. Two pieces of information are required to perform analysis with DESeq2. A matrix of raw counts, such as was generated previously while running <a href="https://htseq.readthedocs.io/en/release_0.9.0/">HTseq</a> previously in this course. This is important as DESeq2 normalizes the data, correcting for differences in library size using using this type of data. DESeq2 also requires the experimental design which can be supplied as a data.frame, detailing the samples and conditions.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="module-3.html#cb142-1" tabindex="-1"></a><span class="co"># define working dir paths</span></span>
<span id="cb142-2"><a href="module-3.html#cb142-2" tabindex="-1"></a>datadir <span class="ot">=</span> <span class="st">&quot;/home/ubuntu/workspace/rnaseq/expression/htseq_counts&quot;</span></span>
<span id="cb142-3"><a href="module-3.html#cb142-3" tabindex="-1"></a>outdir <span class="ot">=</span> <span class="st">&quot;/home/ubuntu/workspace/rnaseq/de/deseq2&quot;</span></span>
<span id="cb142-4"><a href="module-3.html#cb142-4" tabindex="-1"></a></span>
<span id="cb142-5"><a href="module-3.html#cb142-5" tabindex="-1"></a><span class="co"># load R libraries we will use in this section</span></span>
<span id="cb142-6"><a href="module-3.html#cb142-6" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb142-7"><a href="module-3.html#cb142-7" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb142-8"><a href="module-3.html#cb142-8" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb142-9"><a href="module-3.html#cb142-9" tabindex="-1"></a></span>
<span id="cb142-10"><a href="module-3.html#cb142-10" tabindex="-1"></a><span class="co"># set working directory to data dir</span></span>
<span id="cb142-11"><a href="module-3.html#cb142-11" tabindex="-1"></a><span class="fu">setwd</span>(datadir)</span>
<span id="cb142-12"><a href="module-3.html#cb142-12" tabindex="-1"></a></span>
<span id="cb142-13"><a href="module-3.html#cb142-13" tabindex="-1"></a><span class="co"># read in the RNAseq read counts for each gene (produced by htseq-count)</span></span>
<span id="cb142-14"><a href="module-3.html#cb142-14" tabindex="-1"></a>htseqCounts <span class="ot">=</span> <span class="fu">fread</span>(<span class="st">&quot;gene_read_counts_table_all_final.tsv&quot;</span>)</span></code></pre></div>
</div>
<div id="format-htseq-counts-data-to-work-with-deseq2" class="section level3 hasAnchor">
<h3>Format htseq counts data to work with DESeq2<a href="module-3.html#format-htseq-counts-data-to-work-with-deseq2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>DESeq2 has a number of options for data import and has a function to read HTseq output files directly. Here the most universal option is used, reading in raw counts from a matrix in simple TSV format (one row per gene, one column per sample). The HTseq count data that was read in above is stored as an object of class “data.table”, this can be verified with the <code>class()</code> function. Before use in this exercise it is required to convert this object to an appropriate matrix format with gene names as rows and samples as columns.</p>
<p>It should be noted that while the replicate samples are technical replicates (i.e. the same library was sequenced), herein they are treated as biological replicates for illustrative purposes. DESeq2 does have a function to collapse technical replicates though.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="module-3.html#cb143-1" tabindex="-1"></a></span>
<span id="cb143-2"><a href="module-3.html#cb143-2" tabindex="-1"></a><span class="co"># view the class of the data</span></span>
<span id="cb143-3"><a href="module-3.html#cb143-3" tabindex="-1"></a><span class="fu">class</span>(htseqCounts)</span>
<span id="cb143-4"><a href="module-3.html#cb143-4" tabindex="-1"></a></span>
<span id="cb143-5"><a href="module-3.html#cb143-5" tabindex="-1"></a><span class="co"># convert the data.table to matrix format</span></span>
<span id="cb143-6"><a href="module-3.html#cb143-6" tabindex="-1"></a>htseqCounts <span class="ot">=</span> <span class="fu">as.matrix</span>(htseqCounts)</span>
<span id="cb143-7"><a href="module-3.html#cb143-7" tabindex="-1"></a><span class="fu">class</span>(htseqCounts)</span>
<span id="cb143-8"><a href="module-3.html#cb143-8" tabindex="-1"></a></span>
<span id="cb143-9"><a href="module-3.html#cb143-9" tabindex="-1"></a><span class="co"># set the gene ID values to be the row names for the matrix</span></span>
<span id="cb143-10"><a href="module-3.html#cb143-10" tabindex="-1"></a><span class="fu">rownames</span>(htseqCounts) <span class="ot">=</span> htseqCounts[, <span class="st">&quot;GeneID&quot;</span>]</span>
<span id="cb143-11"><a href="module-3.html#cb143-11" tabindex="-1"></a></span>
<span id="cb143-12"><a href="module-3.html#cb143-12" tabindex="-1"></a><span class="co"># now that the gene IDs are the row names, remove the redundant column that contains them</span></span>
<span id="cb143-13"><a href="module-3.html#cb143-13" tabindex="-1"></a>htseqCounts <span class="ot">=</span> htseqCounts[, <span class="fu">colnames</span>(htseqCounts) <span class="sc">!=</span> <span class="st">&quot;GeneID&quot;</span>]</span>
<span id="cb143-14"><a href="module-3.html#cb143-14" tabindex="-1"></a></span>
<span id="cb143-15"><a href="module-3.html#cb143-15" tabindex="-1"></a><span class="co"># convert the count values from strings (with spaces) to integers, because originally the gene column contained characters, the entire matrix was set to character</span></span>
<span id="cb143-16"><a href="module-3.html#cb143-16" tabindex="-1"></a><span class="fu">class</span>(htseqCounts) <span class="ot">=</span> <span class="st">&quot;integer&quot;</span></span>
<span id="cb143-17"><a href="module-3.html#cb143-17" tabindex="-1"></a></span>
<span id="cb143-18"><a href="module-3.html#cb143-18" tabindex="-1"></a><span class="co"># view the first few lines of the gene count matrix</span></span>
<span id="cb143-19"><a href="module-3.html#cb143-19" tabindex="-1"></a><span class="fu">head</span>(htseqCounts)</span></code></pre></div>
</div>
<div id="filter-raw-counts" class="section level3 hasAnchor">
<h3>Filter raw counts<a href="module-3.html#filter-raw-counts" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before running DESeq2 (or any differential expression analysis) it is useful to pre-filter data. There are computational benefits to doing this as the memory size of the objects within R will decrease and DESeq2 will have less data to work through and will be faster. By removing “low quality” data, it is also reduces the number of statistical tests that will be performed, which in turn reduces the impact of multiple test correction and can lead to more significant genes.</p>
<p>The amount of pre-filtering is up to the analyst however, it should be done in an unbiased way. DESeq2 recommends removing any gene with less than 10 reads across all samples. Below, we filter a gene if at least 1 sample does not have at least 10 reads. Either way, mostly what is being removed here are genes with very little evidence for expression in any sample (in many cases gene with 0 counts in all samples).</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="module-3.html#cb144-1" tabindex="-1"></a><span class="co"># run a filtering step</span></span>
<span id="cb144-2"><a href="module-3.html#cb144-2" tabindex="-1"></a><span class="co"># i.e. require that for every gene: at least 1 of 6 samples must have counts greater than 10</span></span>
<span id="cb144-3"><a href="module-3.html#cb144-3" tabindex="-1"></a><span class="co"># get index of rows that meet this criterion and use that to subset the matrix</span></span>
<span id="cb144-4"><a href="module-3.html#cb144-4" tabindex="-1"></a><span class="co"># note the dimensions of the matrix before and after filtering with dim()</span></span>
<span id="cb144-5"><a href="module-3.html#cb144-5" tabindex="-1"></a></span>
<span id="cb144-6"><a href="module-3.html#cb144-6" tabindex="-1"></a><span class="co"># breaking apart the command below to understand it&#39;s outcome</span></span>
<span id="cb144-7"><a href="module-3.html#cb144-7" tabindex="-1"></a><span class="fu">tail</span>(htseqCounts) <span class="co"># look at the raw counts</span></span>
<span id="cb144-8"><a href="module-3.html#cb144-8" tabindex="-1"></a><span class="fu">tail</span>(htseqCounts <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="co"># determine which cells have counts greater than 10</span></span>
<span id="cb144-9"><a href="module-3.html#cb144-9" tabindex="-1"></a><span class="fu">tail</span>(<span class="fu">rowSums</span>(htseqCounts <span class="sc">&gt;=</span> <span class="dv">10</span>)) <span class="co"># determine for which genes how many samples have counts greater than 10</span></span>
<span id="cb144-10"><a href="module-3.html#cb144-10" tabindex="-1"></a><span class="fu">tail</span>(<span class="fu">rowSums</span>(htseqCounts <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="sc">&gt;=</span> <span class="dv">1</span>) <span class="co"># filter to those entries/genes for which at least one sample has counts greater than 10</span></span>
<span id="cb144-11"><a href="module-3.html#cb144-11" tabindex="-1"></a><span class="fu">tail</span>(<span class="fu">which</span>(<span class="fu">rowSums</span>(htseqCounts <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="sc">&gt;=</span> <span class="dv">1</span>)) <span class="co">#obtain the index for the above filter criteria</span></span>
<span id="cb144-12"><a href="module-3.html#cb144-12" tabindex="-1"></a></span>
<span id="cb144-13"><a href="module-3.html#cb144-13" tabindex="-1"></a><span class="fu">dim</span>(htseqCounts)</span>
<span id="cb144-14"><a href="module-3.html#cb144-14" tabindex="-1"></a>htseqCounts <span class="ot">=</span> htseqCounts[<span class="fu">which</span>(<span class="fu">rowSums</span>(htseqCounts <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="sc">&gt;=</span> <span class="dv">1</span>), ]</span>
<span id="cb144-15"><a href="module-3.html#cb144-15" tabindex="-1"></a><span class="fu">dim</span>(htseqCounts)</span></code></pre></div>
</div>
<div id="specifying-the-experimental-design" class="section level3 hasAnchor">
<h3>Specifying the experimental design<a href="module-3.html#specifying-the-experimental-design" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As mentioned above DESeq2 also needs to know the experimental design, that is which samples belong to which condition to test. The experimental design for the example dataset herein is quite simple as there are 6 samples with two conditions to compare (UHR vs HBR), as such we can just create the experimental design right within R. There is one important thing to note, DESeq2 does not check sample names, it expects that the column names in the counts matrix we created correspond exactly to the row names we specify in the experimental design.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="module-3.html#cb145-1" tabindex="-1"></a><span class="co"># construct a mapping of the meta data for our experiment (comparing UHR cell lines to HBR brain tissues)</span></span>
<span id="cb145-2"><a href="module-3.html#cb145-2" tabindex="-1"></a><span class="co"># this is defining the biological condition/label for each experimental replicate</span></span>
<span id="cb145-3"><a href="module-3.html#cb145-3" tabindex="-1"></a><span class="co"># create a simple one column dataframe to start</span></span>
<span id="cb145-4"><a href="module-3.html#cb145-4" tabindex="-1"></a>metaData <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">&quot;Condition&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>))</span>
<span id="cb145-5"><a href="module-3.html#cb145-5" tabindex="-1"></a></span>
<span id="cb145-6"><a href="module-3.html#cb145-6" tabindex="-1"></a><span class="co"># convert the &quot;Condition&quot; column to a factor data type</span></span>
<span id="cb145-7"><a href="module-3.html#cb145-7" tabindex="-1"></a><span class="co"># the arbitrary order of these factors will determine the direction of log2 fold-changes for the genes (i.e. up or down regulated)</span></span>
<span id="cb145-8"><a href="module-3.html#cb145-8" tabindex="-1"></a>metaData<span class="sc">$</span>Condition <span class="ot">=</span> <span class="fu">factor</span>(metaData<span class="sc">$</span>Condition, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;UHR&quot;</span>))</span>
<span id="cb145-9"><a href="module-3.html#cb145-9" tabindex="-1"></a></span>
<span id="cb145-10"><a href="module-3.html#cb145-10" tabindex="-1"></a><span class="co"># set the row names of the metaData dataframe to be the names of our sample replicates from the read counts matrix</span></span>
<span id="cb145-11"><a href="module-3.html#cb145-11" tabindex="-1"></a><span class="fu">rownames</span>(metaData) <span class="ot">=</span> <span class="fu">colnames</span>(htseqCounts)</span>
<span id="cb145-12"><a href="module-3.html#cb145-12" tabindex="-1"></a></span>
<span id="cb145-13"><a href="module-3.html#cb145-13" tabindex="-1"></a><span class="co"># view the metadata dataframe</span></span>
<span id="cb145-14"><a href="module-3.html#cb145-14" tabindex="-1"></a><span class="fu">head</span>(metaData)</span>
<span id="cb145-15"><a href="module-3.html#cb145-15" tabindex="-1"></a></span>
<span id="cb145-16"><a href="module-3.html#cb145-16" tabindex="-1"></a><span class="co"># check that names of htseq count columns match the names of the meta data rows</span></span>
<span id="cb145-17"><a href="module-3.html#cb145-17" tabindex="-1"></a><span class="co"># use the &quot;all&quot; function which tests whether an entire logical vector is TRUE</span></span>
<span id="cb145-18"><a href="module-3.html#cb145-18" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(metaData) <span class="sc">==</span> <span class="fu">colnames</span>(htseqCounts))</span></code></pre></div>
</div>
<div id="construct-the-deseq2-object-piecing-all-the-data-together" class="section level3 hasAnchor">
<h3>Construct the DESeq2 object piecing all the data together<a href="module-3.html#construct-the-deseq2-object-piecing-all-the-data-together" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>With all the data properly formatted it is now possible to combine all the information required to run differental expression in one object. This object will hold the input data, and intermediary calculations. It is also where the condition to test is specified.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="module-3.html#cb146-1" tabindex="-1"></a><span class="co"># make deseq2 data sets</span></span>
<span id="cb146-2"><a href="module-3.html#cb146-2" tabindex="-1"></a><span class="co"># here we are setting up our experiment by supplying: (1) the gene counts matrix, (2) the sample/replicate for each column, and (3) the biological conditions we wish to compare.</span></span>
<span id="cb146-3"><a href="module-3.html#cb146-3" tabindex="-1"></a><span class="co"># this is a simple example that works for many experiments but these can also get more complex</span></span>
<span id="cb146-4"><a href="module-3.html#cb146-4" tabindex="-1"></a><span class="co"># for example, including designs with multiple variables such as &quot;~ group + condition&quot;,</span></span>
<span id="cb146-5"><a href="module-3.html#cb146-5" tabindex="-1"></a><span class="co"># and designs with interactions such as &quot;~ genotype + treatment + genotype:treatment&quot;.</span></span>
<span id="cb146-6"><a href="module-3.html#cb146-6" tabindex="-1"></a></span>
<span id="cb146-7"><a href="module-3.html#cb146-7" tabindex="-1"></a>dds <span class="ot">=</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> htseqCounts, <span class="at">colData =</span> metaData, <span class="at">design =</span> <span class="sc">~</span>Condition)</span>
<span id="cb146-8"><a href="module-3.html#cb146-8" tabindex="-1"></a></span>
<span id="cb146-9"><a href="module-3.html#cb146-9" tabindex="-1"></a><span class="co"># the design formula above is often a point of confussion, it is useful to put in words what is happening, when we specify &quot;design = ~Condition&quot; we are saying</span></span>
<span id="cb146-10"><a href="module-3.html#cb146-10" tabindex="-1"></a><span class="co"># regress gene expression on condition, or put another way model gene expression on condition</span></span>
<span id="cb146-11"><a href="module-3.html#cb146-11" tabindex="-1"></a><span class="co"># gene expression is the response variable and condition is the explanatory variable</span></span>
<span id="cb146-12"><a href="module-3.html#cb146-12" tabindex="-1"></a><span class="co"># you can put words to formulas using this [cheat sheet](https://www.econometrics.blog/post/the-r-formula-cheatsheet/)</span></span></code></pre></div>
</div>
<div id="running-deseq2" class="section level3 hasAnchor">
<h3>Running DESeq2<a href="module-3.html#running-deseq2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>With all the data now in place, DESeq2 can be run. Calling DESeq2 will perform the following actions:
- Estimation of size factors. i.e. accounting for differences in sequencing depth (or library size) across samples.
- Estimation of dispersion. i.e. estimate the biological variability (over and above the expected variation from sampling) in gene expression across biological replicates. This is needed to assess the significance of differences across conditions. Additional work is performed to correct for higher dispersion seen for genes with low expression.
- Perform “independent filtering” to reduce the number of statistical tests performed (see <code>?results</code> and <a href="https://doi.org/10.1073/pnas.0914005107">this paper</a> for details)
- Negative Binomial GLM fitting and performing the Wald statistical test
- Correct p values for multiple testing using the Benjamini and Hochberg method</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="module-3.html#cb147-1" tabindex="-1"></a><span class="co"># run the DESeq2 analysis on the &quot;dds&quot; object</span></span>
<span id="cb147-2"><a href="module-3.html#cb147-2" tabindex="-1"></a>dds <span class="ot">=</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb147-3"><a href="module-3.html#cb147-3" tabindex="-1"></a></span>
<span id="cb147-4"><a href="module-3.html#cb147-4" tabindex="-1"></a><span class="co"># view the first 5 lines of the DE results</span></span>
<span id="cb147-5"><a href="module-3.html#cb147-5" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">results</span>(dds)</span>
<span id="cb147-6"><a href="module-3.html#cb147-6" tabindex="-1"></a><span class="fu">head</span>(res, <span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="log-fold-change-shrinkage" class="section level3 hasAnchor">
<h3>Log-fold change shrinkage<a href="module-3.html#log-fold-change-shrinkage" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is good practice to shrink the log-fold change values, this does exactly what it sounds like, reducing the amount of log-fold change for genes where there are few counts which create huge variability that is not truly biological signal. Consider for example a gene for two samples, one sample has 1 read, and and one sample has 6 reads, that is a 6 fold change, that is likely not accurate. There are a number of algorithms that can be used to shrink log2 fold change, here we will use the apeglm algorithm, which does require the apeglm package to be installed.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="module-3.html#cb148-1" tabindex="-1"></a><span class="co"># shrink the log2 fold change estimates</span></span>
<span id="cb148-2"><a href="module-3.html#cb148-2" tabindex="-1"></a><span class="co">#   shrinkage of effect size (log fold change estimates) is useful for visualization and ranking of genes</span></span>
<span id="cb148-3"><a href="module-3.html#cb148-3" tabindex="-1"></a></span>
<span id="cb148-4"><a href="module-3.html#cb148-4" tabindex="-1"></a><span class="co">#   In simplistic terms, the goal of calculating &quot;dispersion estimates&quot; and &quot;shrinkage&quot; is also to account for the problem that</span></span>
<span id="cb148-5"><a href="module-3.html#cb148-5" tabindex="-1"></a><span class="co">#   genes with low mean counts across replicates tend to have higher variability than those with higher mean counts.</span></span>
<span id="cb148-6"><a href="module-3.html#cb148-6" tabindex="-1"></a><span class="co">#   Shrinkage attempts to correct for this. For a more detailed discussion of shrinkage refer to the DESeq2 vignette</span></span>
<span id="cb148-7"><a href="module-3.html#cb148-7" tabindex="-1"></a></span>
<span id="cb148-8"><a href="module-3.html#cb148-8" tabindex="-1"></a><span class="co"># first get the name of the coefficient (log fold change) to shrink</span></span>
<span id="cb148-9"><a href="module-3.html#cb148-9" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span>
<span id="cb148-10"><a href="module-3.html#cb148-10" tabindex="-1"></a></span>
<span id="cb148-11"><a href="module-3.html#cb148-11" tabindex="-1"></a><span class="co"># now apply the shrinkage approach</span></span>
<span id="cb148-12"><a href="module-3.html#cb148-12" tabindex="-1"></a>resLFC <span class="ot">=</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef =</span> <span class="st">&quot;Condition_UHR_vs_HBR&quot;</span>, <span class="at">type =</span> <span class="st">&quot;apeglm&quot;</span>)</span>
<span id="cb148-13"><a href="module-3.html#cb148-13" tabindex="-1"></a></span>
<span id="cb148-14"><a href="module-3.html#cb148-14" tabindex="-1"></a><span class="co"># make a copy of the shrinkage results to manipulate</span></span>
<span id="cb148-15"><a href="module-3.html#cb148-15" tabindex="-1"></a>deGeneResult <span class="ot">=</span> resLFC</span>
<span id="cb148-16"><a href="module-3.html#cb148-16" tabindex="-1"></a></span>
<span id="cb148-17"><a href="module-3.html#cb148-17" tabindex="-1"></a><span class="co"># contrast the values before and after shrinkage</span></span>
<span id="cb148-18"><a href="module-3.html#cb148-18" tabindex="-1"></a><span class="co"># create a temporary, simplified data structure with the values before/after shrinkage, and create a new column called &quot;group&quot; with this label</span></span>
<span id="cb148-19"><a href="module-3.html#cb148-19" tabindex="-1"></a>res_before <span class="ot">=</span> res</span>
<span id="cb148-20"><a href="module-3.html#cb148-20" tabindex="-1"></a>res_before<span class="sc">$</span>group <span class="ot">=</span> <span class="st">&quot;before shrinkage&quot;</span></span>
<span id="cb148-21"><a href="module-3.html#cb148-21" tabindex="-1"></a>res_after <span class="ot">=</span> deGeneResult</span>
<span id="cb148-22"><a href="module-3.html#cb148-22" tabindex="-1"></a>res_after<span class="sc">$</span>group <span class="ot">=</span> <span class="st">&quot;after shrinkage&quot;</span></span>
<span id="cb148-23"><a href="module-3.html#cb148-23" tabindex="-1"></a>res_combined <span class="ot">=</span> <span class="fu">rbind</span>(res_before[,<span class="fu">c</span>(<span class="st">&quot;log2FoldChange&quot;</span>,<span class="st">&quot;group&quot;</span>)], res_after[,<span class="fu">c</span>(<span class="st">&quot;log2FoldChange&quot;</span>,<span class="st">&quot;group&quot;</span>)])</span>
<span id="cb148-24"><a href="module-3.html#cb148-24" tabindex="-1"></a></span>
<span id="cb148-25"><a href="module-3.html#cb148-25" tabindex="-1"></a><span class="co"># adjust the order so that the legend has &quot;before shrinkage&quot; listed first</span></span>
<span id="cb148-26"><a href="module-3.html#cb148-26" tabindex="-1"></a>res_combined<span class="sc">$</span>group <span class="ot">=</span> <span class="fu">factor</span>(res_combined<span class="sc">$</span>group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;before shrinkage&quot;</span>, <span class="st">&quot;after shrinkage&quot;</span>))</span>
<span id="cb148-27"><a href="module-3.html#cb148-27" tabindex="-1"></a></span>
<span id="cb148-28"><a href="module-3.html#cb148-28" tabindex="-1"></a><span class="co"># look at the fold change values</span></span>
<span id="cb148-29"><a href="module-3.html#cb148-29" tabindex="-1"></a><span class="fu">head</span>(res_before)</span>
<span id="cb148-30"><a href="module-3.html#cb148-30" tabindex="-1"></a><span class="fu">head</span>(res_after)</span>
<span id="cb148-31"><a href="module-3.html#cb148-31" tabindex="-1"></a></span>
<span id="cb148-32"><a href="module-3.html#cb148-32" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_combined, <span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, <span class="at">color =</span> group)) <span class="sc">+</span> <span class="fu">geom_density</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb148-33"><a href="module-3.html#cb148-33" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;before shrinkage&quot;</span> <span class="ot">=</span> <span class="st">&quot;tomato4&quot;</span>, <span class="st">&quot;after shrinkage&quot;</span> <span class="ot">=</span> <span class="st">&quot;slategray&quot;</span>)) <span class="sc">+</span></span>
<span id="cb148-34"><a href="module-3.html#cb148-34" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">&quot;Shrinkage status&quot;</span>)</span>
<span id="cb148-35"><a href="module-3.html#cb148-35" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot=</span>p, <span class="at">filename =</span> <span class="st">&quot;before_after_shrinkage.pdf&quot;</span>, <span class="at">device=</span><span class="st">&quot;pdf&quot;</span>, <span class="at">width=</span><span class="dv">6</span>, <span class="at">height=</span><span class="dv">4</span>, <span class="at">units=</span><span class="st">&quot;in&quot;</span>)</span></code></pre></div>
<p>How did the results change before and after shinkage? What direction is each log2 fold change value moving?</p>
<p>Note that for these data, the impact of shrinkage is very subtle but the pattern is that fold change values move towards 0.</p>
</div>
<div id="annotate-gene-symbols-onto-the-de-results" class="section level3 hasAnchor">
<h3>Annotate gene symbols onto the DE results<a href="module-3.html#annotate-gene-symbols-onto-the-de-results" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>DESeq2 was run with ensembl gene IDs as identifiers, this is not the most human friendly way to interpret results. Here gene symbols are merged onto the differential expressed gene list to make the results a bit more interpretable.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="module-3.html#cb149-1" tabindex="-1"></a><span class="co"># read in gene ID to name mappings (using &quot;fread&quot; an alternative to &quot;read.table&quot;)</span></span>
<span id="cb149-2"><a href="module-3.html#cb149-2" tabindex="-1"></a>gene_id_mapping <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">&quot;/home/ubuntu/workspace/rnaseq/de/htseq_counts/ENSG_ID2Name.txt&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb149-3"><a href="module-3.html#cb149-3" tabindex="-1"></a></span>
<span id="cb149-4"><a href="module-3.html#cb149-4" tabindex="-1"></a><span class="co"># add names to the columns in the &quot;gene_id_mapping&quot; dataframe</span></span>
<span id="cb149-5"><a href="module-3.html#cb149-5" tabindex="-1"></a><span class="fu">setnames</span>(gene_id_mapping, <span class="fu">c</span>(<span class="st">&quot;ensemblID&quot;</span>, <span class="st">&quot;Symbol&quot;</span>))</span>
<span id="cb149-6"><a href="module-3.html#cb149-6" tabindex="-1"></a></span>
<span id="cb149-7"><a href="module-3.html#cb149-7" tabindex="-1"></a><span class="co"># view the first few lines of the gene ID to name mappings</span></span>
<span id="cb149-8"><a href="module-3.html#cb149-8" tabindex="-1"></a><span class="fu">head</span>(gene_id_mapping)</span>
<span id="cb149-9"><a href="module-3.html#cb149-9" tabindex="-1"></a></span>
<span id="cb149-10"><a href="module-3.html#cb149-10" tabindex="-1"></a><span class="co"># merge on gene names</span></span>
<span id="cb149-11"><a href="module-3.html#cb149-11" tabindex="-1"></a>deGeneResult<span class="sc">$</span>ensemblID <span class="ot">=</span> <span class="fu">rownames</span>(deGeneResult)</span>
<span id="cb149-12"><a href="module-3.html#cb149-12" tabindex="-1"></a>deGeneResult <span class="ot">=</span> <span class="fu">as.data.table</span>(deGeneResult)</span>
<span id="cb149-13"><a href="module-3.html#cb149-13" tabindex="-1"></a>deGeneResult <span class="ot">=</span> <span class="fu">merge</span>(deGeneResult, gene_id_mapping, <span class="at">by =</span> <span class="st">&quot;ensemblID&quot;</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb149-14"><a href="module-3.html#cb149-14" tabindex="-1"></a></span>
<span id="cb149-15"><a href="module-3.html#cb149-15" tabindex="-1"></a><span class="co"># merge the original raw count values onto this final dataframe to aid interpretation</span></span>
<span id="cb149-16"><a href="module-3.html#cb149-16" tabindex="-1"></a>original_counts <span class="ot">=</span> <span class="fu">as.data.frame</span>(htseqCounts)</span>
<span id="cb149-17"><a href="module-3.html#cb149-17" tabindex="-1"></a>original_counts[,<span class="st">&quot;ensemblID&quot;</span>] <span class="ot">=</span> <span class="fu">rownames</span>(htseqCounts)</span>
<span id="cb149-18"><a href="module-3.html#cb149-18" tabindex="-1"></a>deGeneResult <span class="ot">=</span> <span class="fu">merge</span>(deGeneResult, original_counts, <span class="at">by =</span> <span class="st">&#39;ensemblID&#39;</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb149-19"><a href="module-3.html#cb149-19" tabindex="-1"></a></span>
<span id="cb149-20"><a href="module-3.html#cb149-20" tabindex="-1"></a><span class="co"># view the final merged dataframe</span></span>
<span id="cb149-21"><a href="module-3.html#cb149-21" tabindex="-1"></a><span class="co"># based on the raw counts and fold change values what does a negative fold change mean with respect to our two conditions HBR and UHR?</span></span>
<span id="cb149-22"><a href="module-3.html#cb149-22" tabindex="-1"></a><span class="fu">head</span>(deGeneResult)</span></code></pre></div>
</div>
<div id="data-manipulation" class="section level3 hasAnchor">
<h3>Data manipulation<a href="module-3.html#data-manipulation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>With the DE analysis complete it is useful to view and filter the data frames to only the relevant genes. Here some basic data manipulation is performed, filtering to significant genes at specific thresholds.</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="module-3.html#cb150-1" tabindex="-1"></a><span class="co"># view the top genes according to adjusted p-value</span></span>
<span id="cb150-2"><a href="module-3.html#cb150-2" tabindex="-1"></a>deGeneResult[<span class="fu">order</span>(deGeneResult<span class="sc">$</span>padj), ]</span>
<span id="cb150-3"><a href="module-3.html#cb150-3" tabindex="-1"></a></span>
<span id="cb150-4"><a href="module-3.html#cb150-4" tabindex="-1"></a><span class="co"># view the top genes according to fold change</span></span>
<span id="cb150-5"><a href="module-3.html#cb150-5" tabindex="-1"></a>deGeneResult[<span class="fu">order</span>(deGeneResult<span class="sc">$</span>log2FoldChange), ]</span>
<span id="cb150-6"><a href="module-3.html#cb150-6" tabindex="-1"></a></span>
<span id="cb150-7"><a href="module-3.html#cb150-7" tabindex="-1"></a><span class="co"># determine the number of up/down significant genes at FDR &lt; 0.05 significance level</span></span>
<span id="cb150-8"><a href="module-3.html#cb150-8" tabindex="-1"></a><span class="fu">dim</span>(deGeneResult)[<span class="dv">1</span>] <span class="co"># number of genes tested</span></span>
<span id="cb150-9"><a href="module-3.html#cb150-9" tabindex="-1"></a><span class="fu">dim</span>(deGeneResult[deGeneResult<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>])[<span class="dv">1</span>] <span class="co">#number of significant genes</span></span>
<span id="cb150-10"><a href="module-3.html#cb150-10" tabindex="-1"></a></span>
<span id="cb150-11"><a href="module-3.html#cb150-11" tabindex="-1"></a><span class="co"># order the DE results by adjusted p-value</span></span>
<span id="cb150-12"><a href="module-3.html#cb150-12" tabindex="-1"></a>deGeneResultSorted <span class="ot">=</span> deGeneResult[<span class="fu">order</span>(deGeneResult<span class="sc">$</span>padj), ]</span>
<span id="cb150-13"><a href="module-3.html#cb150-13" tabindex="-1"></a></span>
<span id="cb150-14"><a href="module-3.html#cb150-14" tabindex="-1"></a><span class="co"># create a filtered data frame that limits to only the significant DE genes (adjusted p.value &lt; 0.05)</span></span>
<span id="cb150-15"><a href="module-3.html#cb150-15" tabindex="-1"></a>deGeneResultSignificant <span class="ot">=</span> deGeneResultSorted[deGeneResultSorted<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.05</span>]</span></code></pre></div>
</div>
<div id="save-results-to-files" class="section level3 hasAnchor">
<h3>Save results to files<a href="module-3.html#save-results-to-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The data generated is now written out as tab separated files. Some of the DESeq2 objects are also saved as serialized R (RDS) objects which can be read back into R later for visualization.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="module-3.html#cb151-1" tabindex="-1"></a><span class="co"># set the working directory to the output dir where we will store any results files</span></span>
<span id="cb151-2"><a href="module-3.html#cb151-2" tabindex="-1"></a><span class="fu">setwd</span>(outdir)</span>
<span id="cb151-3"><a href="module-3.html#cb151-3" tabindex="-1"></a></span>
<span id="cb151-4"><a href="module-3.html#cb151-4" tabindex="-1"></a><span class="co"># save the final DE result (all genes)  to an output file</span></span>
<span id="cb151-5"><a href="module-3.html#cb151-5" tabindex="-1"></a><span class="fu">fwrite</span>(deGeneResultSorted, <span class="at">file=</span><span class="st">&quot;DE_all_genes_DESeq2.tsv&quot;</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb151-6"><a href="module-3.html#cb151-6" tabindex="-1"></a></span>
<span id="cb151-7"><a href="module-3.html#cb151-7" tabindex="-1"></a><span class="co"># save the final DE result (significant genes only)  to an output file</span></span>
<span id="cb151-8"><a href="module-3.html#cb151-8" tabindex="-1"></a><span class="fu">fwrite</span>(deGeneResultSignificant, <span class="at">file=</span><span class="st">&quot;DE_sig_genes_DESeq2.tsv&quot;</span>, <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb151-9"><a href="module-3.html#cb151-9" tabindex="-1"></a></span>
<span id="cb151-10"><a href="module-3.html#cb151-10" tabindex="-1"></a><span class="co"># save the DESeq2 objects for the data visualization section</span></span>
<span id="cb151-11"><a href="module-3.html#cb151-11" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="st">&quot;dds.rds&quot;</span>)</span>
<span id="cb151-12"><a href="module-3.html#cb151-12" tabindex="-1"></a><span class="fu">saveRDS</span>(res, <span class="st">&quot;res.rds&quot;</span>)</span>
<span id="cb151-13"><a href="module-3.html#cb151-13" tabindex="-1"></a><span class="fu">saveRDS</span>(resLFC, <span class="st">&quot;resLFC.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="briefly-examine-the-top-over-expressed-genes" class="section level3 hasAnchor">
<h3>Briefly examine the top over-expressed genes<a href="module-3.html#briefly-examine-the-top-over-expressed-genes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For both conditions (HBR and UHR) lets take a look at the top n genes but this time according to fold-change instead of p-value.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="module-3.html#cb152-1" tabindex="-1"></a></span>
<span id="cb152-2"><a href="module-3.html#cb152-2" tabindex="-1"></a><span class="co"># use the dplyr library to manipulate the dataframe</span></span>
<span id="cb152-3"><a href="module-3.html#cb152-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb152-4"><a href="module-3.html#cb152-4" tabindex="-1"></a></span>
<span id="cb152-5"><a href="module-3.html#cb152-5" tabindex="-1"></a><span class="co"># create a new copy of the data frame, sorted by log2 fold change</span></span>
<span id="cb152-6"><a href="module-3.html#cb152-6" tabindex="-1"></a>deGeneResultSortedFoldchange <span class="ot">=</span> <span class="fu">arrange</span>(deGeneResultSignificant, log2FoldChange)</span>
<span id="cb152-7"><a href="module-3.html#cb152-7" tabindex="-1"></a></span>
<span id="cb152-8"><a href="module-3.html#cb152-8" tabindex="-1"></a><span class="co"># create a convenient data structure with just the top n genes from each condition</span></span>
<span id="cb152-9"><a href="module-3.html#cb152-9" tabindex="-1"></a>top_bottom <span class="ot">=</span> <span class="fu">bind_rows</span>(</span>
<span id="cb152-10"><a href="module-3.html#cb152-10" tabindex="-1"></a>  <span class="fu">head</span>(deGeneResultSortedFoldchange, <span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Set =</span> <span class="st">&quot;Bottom 10&quot;</span>),</span>
<span id="cb152-11"><a href="module-3.html#cb152-11" tabindex="-1"></a>  <span class="fu">tail</span>(deGeneResultSortedFoldchange, <span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Set =</span> <span class="st">&quot;Top 10&quot;</span>)</span>
<span id="cb152-12"><a href="module-3.html#cb152-12" tabindex="-1"></a>)</span>
<span id="cb152-13"><a href="module-3.html#cb152-13" tabindex="-1"></a></span>
<span id="cb152-14"><a href="module-3.html#cb152-14" tabindex="-1"></a><span class="co"># visualize data for the top n genes. Simplify the output a bit</span></span>
<span id="cb152-15"><a href="module-3.html#cb152-15" tabindex="-1"></a><span class="fu">print</span>(top_bottom[,<span class="fu">c</span>(<span class="st">&quot;log2FoldChange&quot;</span>,<span class="st">&quot;padj&quot;</span>,<span class="st">&quot;Symbol&quot;</span>,<span class="st">&quot;UHR_Rep1&quot;</span>,<span class="st">&quot;UHR_Rep2&quot;</span>,<span class="st">&quot;UHR_Rep3&quot;</span>,<span class="st">&quot;HBR_Rep1&quot;</span>,<span class="st">&quot;HBR_Rep2&quot;</span>,<span class="st">&quot;HBR_Rep3&quot;</span>,<span class="st">&quot;Set&quot;</span>)])</span>
<span id="cb152-16"><a href="module-3.html#cb152-16" tabindex="-1"></a></span>
<span id="cb152-17"><a href="module-3.html#cb152-17" tabindex="-1"></a><span class="co">#To exit R type the following</span></span>
<span id="cb152-18"><a href="module-3.html#cb152-18" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
</div>
<div id="perform-some-preliminary-exploration-of-de-genes-using-webtools" class="section level3 hasAnchor">
<h3>Perform some preliminary exploration of DE genes using webtools<a href="module-3.html#perform-some-preliminary-exploration-of-de-genes-using-webtools" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Download the file: <code>outdir/DE_sig_genes_DESeq2.tsv</code>. Open this spreadsheet, sort on “log2FoldChange” column and find the top 100 significant genes with higher expression in HBR (brain). Also download the file: <code>outdir/DE_all_genes_DESeq2.tsv</code> (to be used as a list of background genes or where we want the fold-change value for every gene).</p>
<ul>
<li>Try querying with the top 100 HBR over-expressed genes using: <a href="https://biit.cs.ut.ee/gprofiler/gost">g:Profiler</a></li>
<li>Try querying with the top 100 HBR over-expressed genes using: <a href="https://tissueenrich.gdcb.iastate.edu/">TissueEnrich</a>. Use the Tissue Enrichment tool. This tool also requires a Background Gene List. Use all genes in <code>DE_all_genes_DESeq2.tsv</code> for this purpose. You can also manually explore some individual genes over-expressed in UHR with the Tissue-Specific Genes tool. For example, try <em>PRAME</em> and <em>SERPIND1</em>, two of the top UHR genes.</li>
</ul>
<p><strong>g:Profiler example result</strong>
{% include figure.html image=“/assets/module_3/Gprofiler-example.png” width=“1200” %}</p>
<p><strong>TissueEnrich summary for top 100 HBR over-expressed genes</strong>
{% include figure.html image=“/assets/module_3/TissueEnrich-HBR-Genes-1.png” width=“1200” %}</p>
<p><strong>TissueEnrich result for Cerebral Cortex tissue</strong>
{% include figure.html image=“/assets/module_3/TissueEnrich-HBR-Genes-2.png” width=“1200” %}</p>
<p><strong>TissueEnrich example for UHR over-expressed gene: <em>PRAME</em></strong>
{% include figure.html image=“/assets/module_3/TissueEnrich-UHR-PRAME.png” width=“1200” %}</p>
<p><strong>TissueEnrich example for UHR over-expressed gene: <em>SERPIND1</em></strong>
{% include figure.html image=“/assets/module_3/TissueEnrich-UHR-SERPIND1.png” width=“1200” %}</p>
<p>Does all of this make sense when we think about the makeup of the HBR and UHR samples? Refer back to the <a href="module-01-inputs/0001/05/01/RNAseq_Data/">description of the samples</a>.</p>
<p>In this section we will compare the DE gene lists obtained from different DE methods (e.g. Ballgown, EdgeR, DESeq2)</p>
</div>
<div id="visualize-overlap-with-a-venn-diagram.-this-can-be-done-with-simple-web-tools-like" class="section level3 hasAnchor">
<h3>Visualize overlap with a venn diagram. This can be done with simple web tools like:<a href="module-3.html#visualize-overlap-with-a-venn-diagram.-this-can-be-done-with-simple-web-tools-like" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://www.biovenn.nl/">https://www.biovenn.nl/</a></li>
<li><a href="http://bioinfogp.cnb.csic.es/tools/venny/">http://bioinfogp.cnb.csic.es/tools/venny/</a></li>
</ul>
<p>Once you have run the DESeq2 tutorial, compare the sigDE genes to those saved earlier from ballgown and/or edgeR:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb153-1"><a href="module-3.html#cb153-1" tabindex="-1"></a><span class="fu">head</span> <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/DE_genes.txt</span>
<span id="cb153-2"><a href="module-3.html#cb153-2" tabindex="-1"></a><span class="fu">head</span> <span class="va">$RNA_HOME</span>/de/htseq_counts/DE_genes.txt</span>
<span id="cb153-3"><a href="module-3.html#cb153-3" tabindex="-1"></a><span class="fu">head</span> <span class="va">$RNA_HOME</span>/de/deseq2/DE_sig_genes_DESeq2.tsv</span></code></pre></div>
<p>Pull out the gene IDs</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb154-1"><a href="module-3.html#cb154-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/</span>
<span id="cb154-2"><a href="module-3.html#cb154-2" tabindex="-1"></a></span>
<span id="cb154-3"><a href="module-3.html#cb154-3" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 <span class="va">$RNA_HOME</span>/de/ballgown/ref_only/DE_genes.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> ballgown_DE_gene_symbols.txt</span>
<span id="cb154-4"><a href="module-3.html#cb154-4" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 2 <span class="va">$RNA_HOME</span>/de/htseq_counts/DE_genes.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> Gene_Name <span class="op">&gt;</span> htseq_counts_edgeR_DE_gene_symbols.txt</span>
<span id="cb154-5"><a href="module-3.html#cb154-5" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 7 <span class="va">$RNA_HOME</span>/de/deseq2/DE_sig_genes_DESeq2.tsv <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> Symbol <span class="op">&gt;</span> DESeq2_DE_gene_symbols.txt</span></code></pre></div>
<p>To get the two gene lists you could use <code>cat</code> to print out each list in your terminal and then copy/paste.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb155-1"><a href="module-3.html#cb155-1" tabindex="-1"></a><span class="fu">cat</span> ballgown_DE_gene_symbols.txt</span>
<span id="cb155-2"><a href="module-3.html#cb155-2" tabindex="-1"></a><span class="fu">cat</span> htseq_counts_edgeR_DE_gene_symbols.txt</span>
<span id="cb155-3"><a href="module-3.html#cb155-3" tabindex="-1"></a><span class="fu">cat</span> DESeq2_DE_gene_symbols.txt</span></code></pre></div>
<p>Alternatively you could view both lists in a web browser as you have done with other files. These two files should be here:</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ballgown_DE_gene_symbols.txt</li>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/DESeq2_DE_gene_symbols.txt</li>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/htseq_counts_edgeR_DE_gene_symbols.txt</li>
</ul>
<div id="example-venn-diagram-two-way-comparison-de-genes-from-stringtieballgown-vs-htseqdeseq2" class="section level5 hasAnchor">
<h5>Example Venn Diagram (Two-way comparison: DE genes from StringTie/Ballgown vs HTSeq/DESeq2)<a href="module-3.html#example-venn-diagram-two-way-comparison-de-genes-from-stringtieballgown-vs-htseqdeseq2" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p><br>
{% include figure.html image=“/assets/module_3/venn-ballgown-vs-deseq2.png” width=“400” %}</p>
</div>
<div id="example-venn-diagram-three-way-comparison-de-genes-from-stringtieballgown-vs-htseqdeseq2-vs-htseqedger" class="section level5 hasAnchor">
<h5>Example Venn Diagram (Three-way comparison: DE genes from StringTie/Ballgown vs HTSeq/DESeq2 vs HTSeq/EdgeR)<a href="module-3.html#example-venn-diagram-three-way-comparison-de-genes-from-stringtieballgown-vs-htseqdeseq2-vs-htseqedger" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p><br>
{% include figure.html image=“/assets/module_3/venn-ballgown-vs-deseq2-vs-edger.png” width=“500” %}</p>
</div>
</div>
<div id="ballgown-de-visualization" class="section level3 hasAnchor">
<h3>Ballgown DE Visualization<a href="module-3.html#ballgown-de-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Navigate to the correct directory and then launch R:</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb156-1"><a href="module-3.html#cb156-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/ballgown/ref_only</span>
<span id="cb156-2"><a href="module-3.html#cb156-2" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>A separate R tutorial file has been provided below. Run the R commands detailed in the R script. All results are directed to pdf file(s). The output pdf files can be viewed in your browser at the following urls. Note, you must replace <strong>YOUR_PUBLIC_IPv4_ADDRESS</strong> with your own amazon instance IP (e.g., 101.0.1.101)).</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ballgown/ref_only/Tutorial_Part2_ballgown_output.pdf</li>
</ul>
<p>First you’ll need to load the libraries needed for this analysis and define a path for the output PDF to be written.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="module-3.html#cb157-1" tabindex="-1"></a><span class="co">#load libraries</span></span>
<span id="cb157-2"><a href="module-3.html#cb157-2" tabindex="-1"></a><span class="fu">library</span>(ballgown)</span>
<span id="cb157-3"><a href="module-3.html#cb157-3" tabindex="-1"></a><span class="fu">library</span>(genefilter)</span>
<span id="cb157-4"><a href="module-3.html#cb157-4" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb157-5"><a href="module-3.html#cb157-5" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb157-6"><a href="module-3.html#cb157-6" tabindex="-1"></a></span>
<span id="cb157-7"><a href="module-3.html#cb157-7" tabindex="-1"></a><span class="co"># set the working directory</span></span>
<span id="cb157-8"><a href="module-3.html#cb157-8" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;~/workspace/rnaseq/de/ballgown/ref_only&quot;</span>)</span></code></pre></div>
<p>Next we’ll load our data into R.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="module-3.html#cb158-1" tabindex="-1"></a></span>
<span id="cb158-2"><a href="module-3.html#cb158-2" tabindex="-1"></a><span class="co"># Define the conditions being compared for use later</span></span>
<span id="cb158-3"><a href="module-3.html#cb158-3" tabindex="-1"></a>condition <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>)</span>
<span id="cb158-4"><a href="module-3.html#cb158-4" tabindex="-1"></a></span>
<span id="cb158-5"><a href="module-3.html#cb158-5" tabindex="-1"></a><span class="co"># Load the ballgown object from file</span></span>
<span id="cb158-6"><a href="module-3.html#cb158-6" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;bg.rda&quot;</span>)</span>
<span id="cb158-7"><a href="module-3.html#cb158-7" tabindex="-1"></a></span>
<span id="cb158-8"><a href="module-3.html#cb158-8" tabindex="-1"></a><span class="co"># The load command, loads an R object from a file into memory in our R session.</span></span>
<span id="cb158-9"><a href="module-3.html#cb158-9" tabindex="-1"></a><span class="co"># You can use ls() to view the names of variables that have been loaded</span></span>
<span id="cb158-10"><a href="module-3.html#cb158-10" tabindex="-1"></a><span class="fu">ls</span>()</span>
<span id="cb158-11"><a href="module-3.html#cb158-11" tabindex="-1"></a></span>
<span id="cb158-12"><a href="module-3.html#cb158-12" tabindex="-1"></a><span class="co"># Print a summary of the ballgown object</span></span>
<span id="cb158-13"><a href="module-3.html#cb158-13" tabindex="-1"></a>bg</span>
<span id="cb158-14"><a href="module-3.html#cb158-14" tabindex="-1"></a></span>
<span id="cb158-15"><a href="module-3.html#cb158-15" tabindex="-1"></a><span class="co"># Load gene names for lookup later in the tutorial</span></span>
<span id="cb158-16"><a href="module-3.html#cb158-16" tabindex="-1"></a>bg_table <span class="ot">=</span> <span class="fu">texpr</span>(bg, <span class="st">&quot;all&quot;</span>)</span>
<span id="cb158-17"><a href="module-3.html#cb158-17" tabindex="-1"></a>bg_gene_names <span class="ot">=</span> <span class="fu">unique</span>(bg_table[, <span class="dv">9</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb158-18"><a href="module-3.html#cb158-18" tabindex="-1"></a></span>
<span id="cb158-19"><a href="module-3.html#cb158-19" tabindex="-1"></a><span class="co"># Pull the gene and transcript expression data frame from the ballgown object</span></span>
<span id="cb158-20"><a href="module-3.html#cb158-20" tabindex="-1"></a>gene_expression <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">gexpr</span>(bg))</span>
<span id="cb158-21"><a href="module-3.html#cb158-21" tabindex="-1"></a>transcript_expression <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">texpr</span>(bg))</span>
<span id="cb158-22"><a href="module-3.html#cb158-22" tabindex="-1"></a></span>
<span id="cb158-23"><a href="module-3.html#cb158-23" tabindex="-1"></a><span class="co">#View expression values for the transcripts of a particular gene symbol of chromosome 22.  e.g. &#39;TST&#39;</span></span>
<span id="cb158-24"><a href="module-3.html#cb158-24" tabindex="-1"></a><span class="co">#First determine the transcript_ids in the data.frame that match &#39;TST&#39;, aka. ENSG00000128311, then display only those rows of the data.frame</span></span>
<span id="cb158-25"><a href="module-3.html#cb158-25" tabindex="-1"></a>i <span class="ot">=</span> bg_table[, <span class="st">&quot;gene_name&quot;</span>] <span class="sc">==</span> <span class="st">&quot;TST&quot;</span></span>
<span id="cb158-26"><a href="module-3.html#cb158-26" tabindex="-1"></a>bg_table[i,]</span>
<span id="cb158-27"><a href="module-3.html#cb158-27" tabindex="-1"></a></span>
<span id="cb158-28"><a href="module-3.html#cb158-28" tabindex="-1"></a><span class="co"># Display the transcript ID for a single row of data</span></span>
<span id="cb158-29"><a href="module-3.html#cb158-29" tabindex="-1"></a>ballgown<span class="sc">::</span><span class="fu">transcriptNames</span>(bg)[<span class="dv">2763</span>]</span>
<span id="cb158-30"><a href="module-3.html#cb158-30" tabindex="-1"></a></span>
<span id="cb158-31"><a href="module-3.html#cb158-31" tabindex="-1"></a><span class="co"># Display the gene name for a single row of data</span></span>
<span id="cb158-32"><a href="module-3.html#cb158-32" tabindex="-1"></a>ballgown<span class="sc">::</span><span class="fu">geneNames</span>(bg)[<span class="dv">2763</span>]</span>
<span id="cb158-33"><a href="module-3.html#cb158-33" tabindex="-1"></a></span>
<span id="cb158-34"><a href="module-3.html#cb158-34" tabindex="-1"></a><span class="co">#What if we want to view values for a list of genes of interest all at once?</span></span>
<span id="cb158-35"><a href="module-3.html#cb158-35" tabindex="-1"></a>genes_of_interest <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;TST&quot;</span>, <span class="st">&quot;MMP11&quot;</span>, <span class="st">&quot;LGALS2&quot;</span>, <span class="st">&quot;ISX&quot;</span>)</span>
<span id="cb158-36"><a href="module-3.html#cb158-36" tabindex="-1"></a>i <span class="ot">=</span> bg_table[, <span class="st">&quot;gene_name&quot;</span>] <span class="sc">%in%</span> genes_of_interest</span>
<span id="cb158-37"><a href="module-3.html#cb158-37" tabindex="-1"></a></span>
<span id="cb158-38"><a href="module-3.html#cb158-38" tabindex="-1"></a>bg_table[i,]</span>
<span id="cb158-39"><a href="module-3.html#cb158-39" tabindex="-1"></a></span>
<span id="cb158-40"><a href="module-3.html#cb158-40" tabindex="-1"></a><span class="co"># Load the transcript to gene index from the ballgown object</span></span>
<span id="cb158-41"><a href="module-3.html#cb158-41" tabindex="-1"></a>transcript_gene_table <span class="ot">=</span> <span class="fu">indexes</span>(bg)<span class="sc">$</span>t2g</span>
<span id="cb158-42"><a href="module-3.html#cb158-42" tabindex="-1"></a><span class="fu">head</span>(transcript_gene_table)</span>
<span id="cb158-43"><a href="module-3.html#cb158-43" tabindex="-1"></a></span>
<span id="cb158-44"><a href="module-3.html#cb158-44" tabindex="-1"></a><span class="co">#Each row of data represents a transcript. Many of these transcripts represent the same gene. Determine the numbers of transcripts and unique genes</span></span>
<span id="cb158-45"><a href="module-3.html#cb158-45" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(transcript_gene_table[, <span class="st">&quot;t_id&quot;</span>])) <span class="co">#Transcript count</span></span>
<span id="cb158-46"><a href="module-3.html#cb158-46" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(transcript_gene_table[, <span class="st">&quot;g_id&quot;</span>])) <span class="co">#Unique Gene count</span></span>
<span id="cb158-47"><a href="module-3.html#cb158-47" tabindex="-1"></a></span>
<span id="cb158-48"><a href="module-3.html#cb158-48" tabindex="-1"></a><span class="co"># Extract FPKM values from the &#39;bg&#39; object</span></span>
<span id="cb158-49"><a href="module-3.html#cb158-49" tabindex="-1"></a>fpkm <span class="ot">=</span> <span class="fu">texpr</span>(bg, <span class="at">meas =</span> <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb158-50"><a href="module-3.html#cb158-50" tabindex="-1"></a></span>
<span id="cb158-51"><a href="module-3.html#cb158-51" tabindex="-1"></a><span class="co"># View the last several rows of the FPKM table</span></span>
<span id="cb158-52"><a href="module-3.html#cb158-52" tabindex="-1"></a><span class="fu">tail</span>(fpkm)</span>
<span id="cb158-53"><a href="module-3.html#cb158-53" tabindex="-1"></a></span>
<span id="cb158-54"><a href="module-3.html#cb158-54" tabindex="-1"></a><span class="co"># Transform the FPKM values by adding 1 and convert to a log2 scale</span></span>
<span id="cb158-55"><a href="module-3.html#cb158-55" tabindex="-1"></a>fpkm <span class="ot">=</span> <span class="fu">log2</span>(fpkm <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb158-56"><a href="module-3.html#cb158-56" tabindex="-1"></a></span>
<span id="cb158-57"><a href="module-3.html#cb158-57" tabindex="-1"></a><span class="co"># View the last several rows of the transformed FPKM table</span></span>
<span id="cb158-58"><a href="module-3.html#cb158-58" tabindex="-1"></a><span class="fu">tail</span>(fpkm)</span></code></pre></div>
<p>Now we’ll start to generate figures with the following R code.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="module-3.html#cb159-1" tabindex="-1"></a></span>
<span id="cb159-2"><a href="module-3.html#cb159-2" tabindex="-1"></a><span class="do">#### Plot #1 - the number of transcripts per gene.</span></span>
<span id="cb159-3"><a href="module-3.html#cb159-3" tabindex="-1"></a><span class="co">#Many genes will have only 1 transcript, some genes will have several transcripts</span></span>
<span id="cb159-4"><a href="module-3.html#cb159-4" tabindex="-1"></a><span class="co">#Use the &#39;table()&#39; command to count the number of times each gene symbol occurs (i.e. the # of transcripts that have each gene symbol)</span></span>
<span id="cb159-5"><a href="module-3.html#cb159-5" tabindex="-1"></a><span class="co">#Then use the &#39;hist&#39; command to create a histogram of these counts</span></span>
<span id="cb159-6"><a href="module-3.html#cb159-6" tabindex="-1"></a><span class="co">#How many genes have 1 transcript?  More than one transcript?  What is the maximum number of transcripts for a single gene?</span></span>
<span id="cb159-7"><a href="module-3.html#cb159-7" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">&quot;TranscriptCountDistribution.pdf&quot;</span>)</span>
<span id="cb159-8"><a href="module-3.html#cb159-8" tabindex="-1"></a>counts<span class="ot">=</span><span class="fu">table</span>(transcript_gene_table[, <span class="st">&quot;g_id&quot;</span>])</span>
<span id="cb159-9"><a href="module-3.html#cb159-9" tabindex="-1"></a>c_one <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">which</span>(counts <span class="sc">==</span> <span class="dv">1</span>))</span>
<span id="cb159-10"><a href="module-3.html#cb159-10" tabindex="-1"></a>c_more_than_one <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">which</span>(counts <span class="sc">&gt;</span> <span class="dv">1</span>))</span>
<span id="cb159-11"><a href="module-3.html#cb159-11" tabindex="-1"></a>c_max <span class="ot">=</span> <span class="fu">max</span>(counts)</span>
<span id="cb159-12"><a href="module-3.html#cb159-12" tabindex="-1"></a><span class="fu">hist</span>(counts, <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">col =</span> <span class="st">&quot;bisque4&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;Transcripts per gene&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Distribution of transcript count per gene&quot;</span>)</span>
<span id="cb159-13"><a href="module-3.html#cb159-13" tabindex="-1"></a>legend_text <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">paste</span>(<span class="st">&quot;Genes with one transcript =&quot;</span>, c_one), <span class="fu">paste</span>(<span class="st">&quot;Genes with more than one transcript =&quot;</span>, c_more_than_one), <span class="fu">paste</span>(<span class="st">&quot;Max transcripts for single gene = &quot;</span>, c_max))</span>
<span id="cb159-14"><a href="module-3.html#cb159-14" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topright&quot;</span>, legend_text, <span class="at">lty =</span> <span class="cn">NULL</span>)</span>
<span id="cb159-15"><a href="module-3.html#cb159-15" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb159-16"><a href="module-3.html#cb159-16" tabindex="-1"></a></span>
<span id="cb159-17"><a href="module-3.html#cb159-17" tabindex="-1"></a><span class="do">#### Plot #2 - the distribution of transcript sizes as a histogram</span></span>
<span id="cb159-18"><a href="module-3.html#cb159-18" tabindex="-1"></a><span class="co">#In this analysis we supplied StringTie with transcript models so the lengths will be those of known transcripts</span></span>
<span id="cb159-19"><a href="module-3.html#cb159-19" tabindex="-1"></a><span class="co">#However, if we had used a de novo transcript discovery mode, this step would give us some idea of how well transcripts were being assembled</span></span>
<span id="cb159-20"><a href="module-3.html#cb159-20" tabindex="-1"></a><span class="co">#If we had a low coverage library, or other problems, we might get short &quot;transcripts&quot; that are actually only pieces of real transcripts</span></span>
<span id="cb159-21"><a href="module-3.html#cb159-21" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;TranscriptLengthDistribution.pdf&quot;</span>)</span>
<span id="cb159-22"><a href="module-3.html#cb159-22" tabindex="-1"></a><span class="fu">hist</span>(bg_table<span class="sc">$</span>length, <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">xlab =</span> <span class="st">&quot;Transcript length (bp)&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Distribution of transcript lengths&quot;</span>, <span class="at">col =</span> <span class="st">&quot;steelblue&quot;</span>)</span>
<span id="cb159-23"><a href="module-3.html#cb159-23" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb159-24"><a href="module-3.html#cb159-24" tabindex="-1"></a></span>
<span id="cb159-25"><a href="module-3.html#cb159-25" tabindex="-1"></a><span class="do">#### Plot #3 - distribution of gene expression levels for each sample</span></span>
<span id="cb159-26"><a href="module-3.html#cb159-26" tabindex="-1"></a><span class="co"># Create boxplots to display summary statistics for the FPKM values for each sample</span></span>
<span id="cb159-27"><a href="module-3.html#cb159-27" tabindex="-1"></a><span class="co"># set color based on condition which is UHR vs. HBR</span></span>
<span id="cb159-28"><a href="module-3.html#cb159-28" tabindex="-1"></a><span class="co"># set labels perpendicular to axis (las=2)</span></span>
<span id="cb159-29"><a href="module-3.html#cb159-29" tabindex="-1"></a><span class="co"># set ylab to indicate that values are log2 transformed</span></span>
<span id="cb159-30"><a href="module-3.html#cb159-30" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;All_samples_FPKM_boxplots.pdf&quot;</span>)</span>
<span id="cb159-31"><a href="module-3.html#cb159-31" tabindex="-1"></a><span class="fu">boxplot</span>(fpkm, <span class="at">col =</span> <span class="fu">as.numeric</span>(<span class="fu">as.factor</span>(condition)) <span class="sc">+</span> <span class="dv">1</span>,<span class="at">las =</span> <span class="dv">2</span>,<span class="at">ylab =</span> <span class="st">&quot;log2(FPKM + 1)&quot;</span>)</span>
<span id="cb159-32"><a href="module-3.html#cb159-32" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb159-33"><a href="module-3.html#cb159-33" tabindex="-1"></a></span>
<span id="cb159-34"><a href="module-3.html#cb159-34" tabindex="-1"></a><span class="do">#### Plot 4 - BoxPlot comparing the expression of a single gene for all replicates of both conditions</span></span>
<span id="cb159-35"><a href="module-3.html#cb159-35" tabindex="-1"></a><span class="co"># set border color for each of the boxplots</span></span>
<span id="cb159-36"><a href="module-3.html#cb159-36" tabindex="-1"></a><span class="co"># set title (main) to gene : transcript</span></span>
<span id="cb159-37"><a href="module-3.html#cb159-37" tabindex="-1"></a><span class="co"># set x label to Type</span></span>
<span id="cb159-38"><a href="module-3.html#cb159-38" tabindex="-1"></a><span class="co"># set ylab to indicate that values are log2 transformed</span></span>
<span id="cb159-39"><a href="module-3.html#cb159-39" tabindex="-1"></a></span>
<span id="cb159-40"><a href="module-3.html#cb159-40" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;TST_ENST00000249042_boxplot.pdf&quot;</span>)</span>
<span id="cb159-41"><a href="module-3.html#cb159-41" tabindex="-1"></a>transcript <span class="ot">=</span> <span class="fu">which</span>(ballgown<span class="sc">::</span><span class="fu">transcriptNames</span>(bg) <span class="sc">==</span> <span class="st">&quot;ENST00000249042&quot;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb159-42"><a href="module-3.html#cb159-42" tabindex="-1"></a><span class="fu">boxplot</span>(fpkm[transcript,] <span class="sc">~</span> condition, <span class="at">border =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>), <span class="at">main =</span> <span class="fu">paste</span>(ballgown<span class="sc">::</span><span class="fu">geneNames</span>(bg)[transcript],<span class="st">&quot;: &quot;</span>, ballgown<span class="sc">::</span><span class="fu">transcriptNames</span>(bg)[transcript]), <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;log2(FPKM+1)&quot;</span>)</span>
<span id="cb159-43"><a href="module-3.html#cb159-43" tabindex="-1"></a></span>
<span id="cb159-44"><a href="module-3.html#cb159-44" tabindex="-1"></a><span class="co"># Add the FPKM values for each sample onto the plot</span></span>
<span id="cb159-45"><a href="module-3.html#cb159-45" tabindex="-1"></a><span class="co"># set plot symbol to solid circle, default is empty circle</span></span>
<span id="cb159-46"><a href="module-3.html#cb159-46" tabindex="-1"></a><span class="fu">points</span>(fpkm[transcript,] <span class="sc">~</span> <span class="fu">jitter</span>(<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>)), <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>)<span class="sc">+</span><span class="dv">1</span>, <span class="at">pch =</span> <span class="dv">16</span>)</span>
<span id="cb159-47"><a href="module-3.html#cb159-47" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb159-48"><a href="module-3.html#cb159-48" tabindex="-1"></a></span>
<span id="cb159-49"><a href="module-3.html#cb159-49" tabindex="-1"></a></span>
<span id="cb159-50"><a href="module-3.html#cb159-50" tabindex="-1"></a><span class="do">#### Plot 5 - Plot of transcript structures observed and expression level for UHR vs HBR with representative replicate</span></span>
<span id="cb159-51"><a href="module-3.html#cb159-51" tabindex="-1"></a></span>
<span id="cb159-52"><a href="module-3.html#cb159-52" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;TST_transcript_structures_expression.pdf&quot;</span>)</span>
<span id="cb159-53"><a href="module-3.html#cb159-53" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb159-54"><a href="module-3.html#cb159-54" tabindex="-1"></a><span class="fu">plotTranscripts</span>(ballgown<span class="sc">::</span><span class="fu">geneIDs</span>(bg)[transcript], bg, <span class="at">main =</span> <span class="fu">c</span>(<span class="st">&quot;TST in HBR&quot;</span>), <span class="at">sample =</span> <span class="fu">c</span>(<span class="st">&quot;HBR_Rep1&quot;</span>), <span class="at">labelTranscripts =</span> <span class="cn">TRUE</span>)</span>
<span id="cb159-55"><a href="module-3.html#cb159-55" tabindex="-1"></a><span class="fu">plotTranscripts</span>(ballgown<span class="sc">::</span><span class="fu">geneIDs</span>(bg)[transcript], bg, <span class="at">main =</span> <span class="fu">c</span>(<span class="st">&quot;TST in UHR&quot;</span>), <span class="at">sample =</span> <span class="fu">c</span>(<span class="st">&quot;UHR_Rep1&quot;</span>), <span class="at">labelTranscripts =</span> <span class="cn">TRUE</span>)</span>
<span id="cb159-56"><a href="module-3.html#cb159-56" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb159-57"><a href="module-3.html#cb159-57" tabindex="-1"></a></span>
<span id="cb159-58"><a href="module-3.html#cb159-58" tabindex="-1"></a><span class="co"># Exit the R session</span></span>
<span id="cb159-59"><a href="module-3.html#cb159-59" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>Remember that you can view the output graphs of this step on your instance by navigating to this location in a web browser window:</p>
<p><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ballgown/ref_only/Tutorial_Part2_ballgown_output.pdf</p>
</div>
<div id="differential-expression-visualization" class="section level3 hasAnchor">
<h3>Differential Expression Visualization<a href="module-3.html#differential-expression-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this section we will be going over some basic visualizations of the DESeq2 results generated in the “Differential Expression with DESeq2” section of this course. Our goal is to quickly obtain some interpretable results using built-in visualization functions from DESeq2 or recommended packages. For a very extensive overview of DESeq2 and how to visualize and interpret the results, refer to the DESeq2 vignette.</p>
<div id="setup-1" class="section level4 hasAnchor">
<h4>Setup<a href="module-3.html#setup-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Navigate to the correct directory and then launch R:</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb160-1"><a href="module-3.html#cb160-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/deseq2</span>
<span id="cb160-2"><a href="module-3.html#cb160-2" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>If it is not already in your R environment, load the DESeqDataSet object and the results table into the R environment.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="module-3.html#cb161-1" tabindex="-1"></a><span class="co"># set the working directory</span></span>
<span id="cb161-2"><a href="module-3.html#cb161-2" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/home/ubuntu/workspace/rnaseq/de/deseq2&quot;</span>)</span>
<span id="cb161-3"><a href="module-3.html#cb161-3" tabindex="-1"></a></span>
<span id="cb161-4"><a href="module-3.html#cb161-4" tabindex="-1"></a><span class="co"># view the contents of this directory</span></span>
<span id="cb161-5"><a href="module-3.html#cb161-5" tabindex="-1"></a><span class="fu">dir</span>()</span>
<span id="cb161-6"><a href="module-3.html#cb161-6" tabindex="-1"></a></span>
<span id="cb161-7"><a href="module-3.html#cb161-7" tabindex="-1"></a><span class="co"># load libs</span></span>
<span id="cb161-8"><a href="module-3.html#cb161-8" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb161-9"><a href="module-3.html#cb161-9" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb161-10"><a href="module-3.html#cb161-10" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb161-11"><a href="module-3.html#cb161-11" tabindex="-1"></a></span>
<span id="cb161-12"><a href="module-3.html#cb161-12" tabindex="-1"></a><span class="co"># Load in the DESeqDataSet object (http://genomedata.org/cri-workshop/deseq2/dds.rds)</span></span>
<span id="cb161-13"><a href="module-3.html#cb161-13" tabindex="-1"></a>dds <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;dds.rds&quot;</span>)</span>
<span id="cb161-14"><a href="module-3.html#cb161-14" tabindex="-1"></a></span>
<span id="cb161-15"><a href="module-3.html#cb161-15" tabindex="-1"></a><span class="co"># Load in the results object before shrinkage (http://genomedata.org/cri-workshop/deseq2/res.rds)</span></span>
<span id="cb161-16"><a href="module-3.html#cb161-16" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;res.rds&quot;</span>)</span>
<span id="cb161-17"><a href="module-3.html#cb161-17" tabindex="-1"></a></span>
<span id="cb161-18"><a href="module-3.html#cb161-18" tabindex="-1"></a><span class="co"># Load in the results object after shrinkage (http://genomedata.org/cri-workshop/deseq2/resLFC.rds)</span></span>
<span id="cb161-19"><a href="module-3.html#cb161-19" tabindex="-1"></a>resLFC <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;resLFC.rds&quot;</span>)</span>
<span id="cb161-20"><a href="module-3.html#cb161-20" tabindex="-1"></a></span>
<span id="cb161-21"><a href="module-3.html#cb161-21" tabindex="-1"></a><span class="co"># Load in the final results file with all sorted DE results (http://genomedata.org/cri-workshop/deseq2/DE_all_genes_DESeq2.tsv)</span></span>
<span id="cb161-22"><a href="module-3.html#cb161-22" tabindex="-1"></a>deGeneResultSorted <span class="ot">=</span> <span class="fu">fread</span>(<span class="st">&quot;DE_all_genes_DESeq2.tsv&quot;</span>) </span></code></pre></div>
</div>
<div id="ma-plot-before-lfc-shrinkage" class="section level4 hasAnchor">
<h4>MA-plot before LFC shrinkage<a href="module-3.html#ma-plot-before-lfc-shrinkage" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>MA-plots were originally used to evaluate microarray expression data where M is the log ratio and A is the mean intensity for a gene (both based on scanned intensity measurements from the microarray).</p>
<p>These types of plots are still usefull in RNAseq DE experiments with two conditions, as they can immediately give us information on the number of signficantly differentially expressed genes, the ratio of up vs down regulated genes, and any outliers. To interpret these plots it is important to keep a couple of things in mind. The Y axis (M) is the log2 fold change between the two conditions tested, a higher fold-change indicates greater difference between condition A and condition B. The X axis (A) is a measure of read alignment to a gene, so as you go higher on on the X axis you are looking at regions which have higher totals of aligned reads, in other words the gene is “more” expressed overall (with the caveat that gene length is not being taken into account by raw read counts here).</p>
<p>Using the built-in <code>plotMA</code> function from DESeq2 we also see that the genes are color coded by a significance threshold (e.g., adjusted p-value &lt; 0.1). Genes with higher expression values and higher fold-changes are more often significant as one would expect.
})</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="module-3.html#cb162-1" tabindex="-1"></a><span class="co"># use DESeq2 built in MA-plot function</span></span>
<span id="cb162-2"><a href="module-3.html#cb162-2" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;maplot_preShrink.pdf&quot;</span>)</span>
<span id="cb162-3"><a href="module-3.html#cb162-3" tabindex="-1"></a><span class="fu">plotMA</span>(res, <span class="at">alpha =</span> <span class="fl">0.1</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">main =</span> <span class="st">&quot;MA-plot before LFC shrinkage&quot;</span>)</span>
<span id="cb162-4"><a href="module-3.html#cb162-4" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="ma-plot-after-lfc-shrinkage" class="section level4 hasAnchor">
<h4>MA-plot after LFC shrinkage<a href="module-3.html#ma-plot-after-lfc-shrinkage" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>When we ran DESeq2 we obtained two results, one with and one without log-fold change shrinkage. When you have genes with low hits you can get some very large fold changes. For example 1 hit on a gene vs 6 hits on a gene is a 6x fold change. This high level of variance though is probably quantifying noise instead of real biology. Running <code>plotMA</code> on our results where we applied an algorithm for log fold change shrinkage we can see that this “effect” is somewhat controlled for. I do want to note that while shrinking LFC is part of a typical DE workflow there are cases where you would not want to perform this, namely when there is already low variation amongst samples (i.e. from technical replicates) as most shrinkage algorithms rely on some variability to build a prior distribution.</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="module-3.html#cb163-1" tabindex="-1"></a><span class="co"># ma plot</span></span>
<span id="cb163-2"><a href="module-3.html#cb163-2" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;maplot_postShrink.pdf&quot;</span>)</span>
<span id="cb163-3"><a href="module-3.html#cb163-3" tabindex="-1"></a><span class="fu">plotMA</span>(resLFC, <span class="at">alpha =</span> <span class="fl">0.1</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">main =</span> <span class="st">&quot;MA-plot after LFC shrinkage&quot;</span>)</span>
<span id="cb163-4"><a href="module-3.html#cb163-4" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>The effect is very subtle here due to the focused nature of our dataset (chr22 genes only), but if you toggle between the two plots and look in the upper left and bottom left corners you can see some fold change values are moving closer to 0.</p>
</div>
<div id="viewing-individual-gene-counts-between-two-conditions" class="section level4 hasAnchor">
<h4>Viewing individual gene counts between two conditions<a href="module-3.html#viewing-individual-gene-counts-between-two-conditions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Often it may be useful to view the normalized counts for a gene amongst our samples. DESeq2 provides a built in function for that which works off of the dds object. Here we view SEPT3 which we can see in our DE output is significantly higher in the HBR cohort and PRAME which is significantly higher in the UHR cohort. This is useful as we can see the per-sample distribution of our corrected counts, we can immediately determine if there are any outliers within each group and investigate further if need be.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="module-3.html#cb164-1" tabindex="-1"></a></span>
<span id="cb164-2"><a href="module-3.html#cb164-2" tabindex="-1"></a><span class="co"># hint! you defined intgroup when creating the dds object, you can view the name by printing the dds objct in your R session</span></span>
<span id="cb164-3"><a href="module-3.html#cb164-3" tabindex="-1"></a><span class="co"># dds</span></span>
<span id="cb164-4"><a href="module-3.html#cb164-4" tabindex="-1"></a></span>
<span id="cb164-5"><a href="module-3.html#cb164-5" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;normalized_count_examples.pdf&quot;</span>)</span>
<span id="cb164-6"><a href="module-3.html#cb164-6" tabindex="-1"></a></span>
<span id="cb164-7"><a href="module-3.html#cb164-7" tabindex="-1"></a><span class="co"># view SEPT3 normalized counts</span></span>
<span id="cb164-8"><a href="module-3.html#cb164-8" tabindex="-1"></a><span class="fu">plotCounts</span>(dds, <span class="at">gene =</span> <span class="st">&quot;ENSG00000100167&quot;</span>, <span class="at">intgroup =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">main =</span> <span class="st">&quot;SEPT3&quot;</span>)</span>
<span id="cb164-9"><a href="module-3.html#cb164-9" tabindex="-1"></a></span>
<span id="cb164-10"><a href="module-3.html#cb164-10" tabindex="-1"></a><span class="co"># view PRAME normalized counts</span></span>
<span id="cb164-11"><a href="module-3.html#cb164-11" tabindex="-1"></a><span class="fu">plotCounts</span>(dds, <span class="at">gene =</span> <span class="st">&quot;ENSG00000185686&quot;</span>, <span class="at">intgroup =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">main =</span> <span class="st">&quot;PRAME&quot;</span>)</span>
<span id="cb164-12"><a href="module-3.html#cb164-12" tabindex="-1"></a></span>
<span id="cb164-13"><a href="module-3.html#cb164-13" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
</div>
<div id="viewing-pairwise-sample-clustering" class="section level3 hasAnchor">
<h3>Viewing pairwise sample clustering<a href="module-3.html#viewing-pairwise-sample-clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It may often be useful to view inter-sample relatedness. In other words, how similar or disimilar samples are to one another overall. While not part of the DESeq2 package, there is a convenient library that can easily construct a hierarchically clustered heatmap from our DESeq2 data. It should be noted that when doing a distance calculation using “raw count” data is not ideal, the count data should be transformed using <code>vst()</code> or <code>rlog()</code> which can be performed directly on the dds object. The reason for this is described in detail in the DESeq2 manuscript, suffice it to say that transforming gene variance to be more homoskedastic will make inferences of sample relatedness more interpretable.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="module-3.html#cb165-1" tabindex="-1"></a><span class="co"># note that we use rlog because we don&#39;t have a large number of genes, for a typical DE experiment with 1000&#39;s of genes use the vst() function</span></span>
<span id="cb165-2"><a href="module-3.html#cb165-2" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds, <span class="at">blind =</span> <span class="cn">FALSE</span>)</span>
<span id="cb165-3"><a href="module-3.html#cb165-3" tabindex="-1"></a></span>
<span id="cb165-4"><a href="module-3.html#cb165-4" tabindex="-1"></a><span class="co"># view the structure of this object</span></span>
<span id="cb165-5"><a href="module-3.html#cb165-5" tabindex="-1"></a>rld</span>
<span id="cb165-6"><a href="module-3.html#cb165-6" tabindex="-1"></a></span>
<span id="cb165-7"><a href="module-3.html#cb165-7" tabindex="-1"></a><span class="co"># compute sample distances (the dist function uses the euclidean distance metric by default)</span></span>
<span id="cb165-8"><a href="module-3.html#cb165-8" tabindex="-1"></a><span class="co"># in this command we will pull the rlog transformed data (&quot;regularized&quot; log2 transformed, see ?rlog for details) using &quot;assay&quot;</span></span>
<span id="cb165-9"><a href="module-3.html#cb165-9" tabindex="-1"></a><span class="co"># then we transpose that data using t()</span></span>
<span id="cb165-10"><a href="module-3.html#cb165-10" tabindex="-1"></a><span class="co"># then we calculate distance values using dist() </span></span>
<span id="cb165-11"><a href="module-3.html#cb165-11" tabindex="-1"></a><span class="co"># the distance is calculated for each vector of sample gene values, in a pairwise fashion comparing all samples</span></span>
<span id="cb165-12"><a href="module-3.html#cb165-12" tabindex="-1"></a></span>
<span id="cb165-13"><a href="module-3.html#cb165-13" tabindex="-1"></a><span class="co"># view the first few lines of raw data</span></span>
<span id="cb165-14"><a href="module-3.html#cb165-14" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(dds))</span>
<span id="cb165-15"><a href="module-3.html#cb165-15" tabindex="-1"></a></span>
<span id="cb165-16"><a href="module-3.html#cb165-16" tabindex="-1"></a><span class="co"># see the rlog transformed data</span></span>
<span id="cb165-17"><a href="module-3.html#cb165-17" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(rld))</span>
<span id="cb165-18"><a href="module-3.html#cb165-18" tabindex="-1"></a></span>
<span id="cb165-19"><a href="module-3.html#cb165-19" tabindex="-1"></a><span class="co"># see the impact of transposing the matrix</span></span>
<span id="cb165-20"><a href="module-3.html#cb165-20" tabindex="-1"></a><span class="fu">t</span>(<span class="fu">assay</span>(rld))[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb165-21"><a href="module-3.html#cb165-21" tabindex="-1"></a></span>
<span id="cb165-22"><a href="module-3.html#cb165-22" tabindex="-1"></a><span class="co"># see the distance values</span></span>
<span id="cb165-23"><a href="module-3.html#cb165-23" tabindex="-1"></a><span class="fu">dist</span>(<span class="fu">t</span>(<span class="fu">assay</span>(rld)))</span>
<span id="cb165-24"><a href="module-3.html#cb165-24" tabindex="-1"></a></span>
<span id="cb165-25"><a href="module-3.html#cb165-25" tabindex="-1"></a><span class="co"># put it all together and store the result</span></span>
<span id="cb165-26"><a href="module-3.html#cb165-26" tabindex="-1"></a>sampleDists <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(<span class="fu">assay</span>(rld)))</span>
<span id="cb165-27"><a href="module-3.html#cb165-27" tabindex="-1"></a></span>
<span id="cb165-28"><a href="module-3.html#cb165-28" tabindex="-1"></a><span class="co"># convert the distance result to a matrix</span></span>
<span id="cb165-29"><a href="module-3.html#cb165-29" tabindex="-1"></a>sampleDistMatrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(sampleDists)</span>
<span id="cb165-30"><a href="module-3.html#cb165-30" tabindex="-1"></a></span>
<span id="cb165-31"><a href="module-3.html#cb165-31" tabindex="-1"></a><span class="co"># view the distance numbers directly in the pairwise distance matrix</span></span>
<span id="cb165-32"><a href="module-3.html#cb165-32" tabindex="-1"></a><span class="fu">head</span>(sampleDistMatrix)</span>
<span id="cb165-33"><a href="module-3.html#cb165-33" tabindex="-1"></a></span>
<span id="cb165-34"><a href="module-3.html#cb165-34" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;distance_sample_heatmap.pdf&quot;</span>, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb165-35"><a href="module-3.html#cb165-35" tabindex="-1"></a></span>
<span id="cb165-36"><a href="module-3.html#cb165-36" tabindex="-1"></a><span class="co"># construct clustered heatmap, important to use the computed sample distances for clustering</span></span>
<span id="cb165-37"><a href="module-3.html#cb165-37" tabindex="-1"></a><span class="fu">pheatmap</span>(sampleDistMatrix, <span class="at">clustering_distance_rows =</span> sampleDists, <span class="at">clustering_distance_cols =</span> sampleDists)</span>
<span id="cb165-38"><a href="module-3.html#cb165-38" tabindex="-1"></a></span>
<span id="cb165-39"><a href="module-3.html#cb165-39" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>Instead of a distance metric we could also use a similarity metric such as a Peason correlation</p>
<p>There are many correlation and distance options:</p>
<p>Correlation: “pearson”, “kendall”, “spearman”
Distance: “euclidean”, “maximum”, “manhattan”, “canberra”, “binary” or “minkowski”</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="module-3.html#cb166-1" tabindex="-1"></a>sampleCorrs <span class="ot">=</span> <span class="fu">cor</span>(<span class="fu">assay</span>(rld), <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb166-2"><a href="module-3.html#cb166-2" tabindex="-1"></a>sampleCorrMatrix <span class="ot">=</span> <span class="fu">as.matrix</span>(sampleCorrs)</span>
<span id="cb166-3"><a href="module-3.html#cb166-3" tabindex="-1"></a><span class="fu">head</span>(sampleCorrMatrix)</span>
<span id="cb166-4"><a href="module-3.html#cb166-4" tabindex="-1"></a></span>
<span id="cb166-5"><a href="module-3.html#cb166-5" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;similarity_sample_heatmap.pdf&quot;</span>, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb166-6"><a href="module-3.html#cb166-6" tabindex="-1"></a></span>
<span id="cb166-7"><a href="module-3.html#cb166-7" tabindex="-1"></a><span class="fu">pheatmap</span>(sampleCorrMatrix)</span>
<span id="cb166-8"><a href="module-3.html#cb166-8" tabindex="-1"></a></span>
<span id="cb166-9"><a href="module-3.html#cb166-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>Instead of boiling all the gene count data for each sample down to a distance metric you can
get a similar sense of the pattern by just visualizing all the genes at once</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="module-3.html#cb167-1" tabindex="-1"></a></span>
<span id="cb167-2"><a href="module-3.html#cb167-2" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;all_gene_heatmap.pdf&quot;</span>, <span class="at">width =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">10</span>)</span>
<span id="cb167-3"><a href="module-3.html#cb167-3" tabindex="-1"></a></span>
<span id="cb167-4"><a href="module-3.html#cb167-4" tabindex="-1"></a><span class="co"># because there are so many gene we choose not to display them</span></span>
<span id="cb167-5"><a href="module-3.html#cb167-5" tabindex="-1"></a></span>
<span id="cb167-6"><a href="module-3.html#cb167-6" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="at">mat =</span> <span class="fu">t</span>(<span class="fu">assay</span>(rld)), <span class="at">show_colnames =</span> <span class="cn">FALSE</span>)</span>
<span id="cb167-7"><a href="module-3.html#cb167-7" tabindex="-1"></a></span>
<span id="cb167-8"><a href="module-3.html#cb167-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb167-9"><a href="module-3.html#cb167-9" tabindex="-1"></a></span>
<span id="cb167-10"><a href="module-3.html#cb167-10" tabindex="-1"></a></span>
<span id="cb167-11"><a href="module-3.html#cb167-11" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save=</span><span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<div id="supplementary-r-de-visualization" class="section level4 hasAnchor">
<h4>Supplementary R DE Visualization<a href="module-3.html#supplementary-r-de-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Occasionally you may wish to reformat and work with expression estimates in R in an ad hoc way. Here, we provide an optional/advanced tutorial on how to visualize your results for R and perform “old school” (non-ballgown, non-DESeq2) visualization of your data.</p>
<p>In this tutorial you will:</p>
<ul>
<li>Learn basic R usage and commands (common plots, and data manipulation tasks)</li>
<li>Examine the expression estimates</li>
<li>Create an MDS plot to visualize the differences between/among replicates, library prep methods and UHR versus HBR</li>
<li>Examine the differential expression estimates</li>
<li>Visualize the expression estimates and highlight those genes that appear to be differentially expressed</li>
<li>Ask how reproducible technical replicates are.</li>
</ul>
<p>Expression and differential expression files will be read into R. The R analysis will make use of the gene-level expression estimates from HISAT2/Stringtie (TPM values) and differential expression results from HISAT2/htseq-count/DESeq2 (fold-changes and p-values).</p>
<p>Start RStudio, or launch a posit Cloud session, or if you are using AWS, navigate to the correct directory and then launch R:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb168-1"><a href="module-3.html#cb168-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="va">$RNA_HOME</span>/de/visualization_advanced</span>
<span id="cb168-2"><a href="module-3.html#cb168-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/de/visualization_advanced</span>
<span id="cb168-3"><a href="module-3.html#cb168-3" tabindex="-1"></a></span>
<span id="cb168-4"><a href="module-3.html#cb168-4" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>First you’ll load your libraries and your data.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="module-3.html#cb169-1" tabindex="-1"></a><span class="co">#Load your libraries</span></span>
<span id="cb169-2"><a href="module-3.html#cb169-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb169-3"><a href="module-3.html#cb169-3" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb169-4"><a href="module-3.html#cb169-4" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb169-5"><a href="module-3.html#cb169-5" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb169-6"><a href="module-3.html#cb169-6" tabindex="-1"></a></span>
<span id="cb169-7"><a href="module-3.html#cb169-7" tabindex="-1"></a><span class="co">#Set a base working directory</span></span>
<span id="cb169-8"><a href="module-3.html#cb169-8" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/home/ubuntu/workspace/rnaseq/de/visualization_advanced/&quot;</span>)</span>
<span id="cb169-9"><a href="module-3.html#cb169-9" tabindex="-1"></a></span>
<span id="cb169-10"><a href="module-3.html#cb169-10" tabindex="-1"></a><span class="co">#Import expression results (TPM values) from the HISAT2/Stringtie pipeline (https://genomedata.org/cri-workshop/gene_tpm_all_samples.tsv)</span></span>
<span id="cb169-11"><a href="module-3.html#cb169-11" tabindex="-1"></a>gene_expression <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;~/workspace/rnaseq/expression/stringtie/ref_only/gene_tpm_all_samples.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb169-12"><a href="module-3.html#cb169-12" tabindex="-1"></a><span class="co">#gene_expression = read.table(&quot;data/bulk_rna/gene_tpm_all_samples.tsv&quot;, header = TRUE, stringsAsFactors = FALSE, row.names = 1)</span></span>
<span id="cb169-13"><a href="module-3.html#cb169-13" tabindex="-1"></a></span>
<span id="cb169-14"><a href="module-3.html#cb169-14" tabindex="-1"></a><span class="co">#Import gene name mapping file (https://genomedata.org/cri-workshop/ENSG_ID2Name.txt)</span></span>
<span id="cb169-15"><a href="module-3.html#cb169-15" tabindex="-1"></a>gene_names<span class="ot">=</span><span class="fu">read.table</span>(<span class="st">&quot;~/workspace/rnaseq/de/htseq_counts/ENSG_ID2Name.txt&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb169-16"><a href="module-3.html#cb169-16" tabindex="-1"></a><span class="co">#gene_names=read.table(&quot;data/bulk_rna/ENSG_ID2Name.txt&quot;, header = TRUE, stringsAsFactors = FALSE)</span></span>
<span id="cb169-17"><a href="module-3.html#cb169-17" tabindex="-1"></a></span>
<span id="cb169-18"><a href="module-3.html#cb169-18" tabindex="-1"></a><span class="fu">colnames</span>(gene_names) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;gene_id&quot;</span>, <span class="st">&quot;gene_name&quot;</span>)</span>
<span id="cb169-19"><a href="module-3.html#cb169-19" tabindex="-1"></a></span>
<span id="cb169-20"><a href="module-3.html#cb169-20" tabindex="-1"></a><span class="co">#Import DE results from the HISAT2/htseq-count/DESeq2 pipeline (http://genomedata.org/cri-workshop/deseq2/DE_all_genes_DESeq2.tsv)</span></span>
<span id="cb169-21"><a href="module-3.html#cb169-21" tabindex="-1"></a>results_genes <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;~/workspace/rnaseq/de/deseq2/DE_all_genes_DESeq2.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb169-22"><a href="module-3.html#cb169-22" tabindex="-1"></a><span class="co">#results_genes = read.table(&quot;outdir/DE_all_genes_DESeq2.tsv&quot;, sep = &quot;\t&quot;, header = TRUE, stringsAsFactors = FALSE)</span></span>
<span id="cb169-23"><a href="module-3.html#cb169-23" tabindex="-1"></a></span>
<span id="cb169-24"><a href="module-3.html#cb169-24" tabindex="-1"></a><span class="co">#Set a directory for the output to go to</span></span>
<span id="cb169-25"><a href="module-3.html#cb169-25" tabindex="-1"></a><span class="co"># output_dir = &quot;/cloud/project/outdir/visualization_advanced/&quot;</span></span>
<span id="cb169-26"><a href="module-3.html#cb169-26" tabindex="-1"></a><span class="co"># if (!dir.exists(output_dir)) {</span></span>
<span id="cb169-27"><a href="module-3.html#cb169-27" tabindex="-1"></a><span class="co">#   dir.create(output_dir, recursive = TRUE)</span></span>
<span id="cb169-28"><a href="module-3.html#cb169-28" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb169-29"><a href="module-3.html#cb169-29" tabindex="-1"></a><span class="co"># setwd(output_dir)</span></span></code></pre></div>
<p>Let’s briefly explore the imported data</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="module-3.html#cb170-1" tabindex="-1"></a></span>
<span id="cb170-2"><a href="module-3.html#cb170-2" tabindex="-1"></a><span class="do">#### Working with &#39;dataframes&#39;</span></span>
<span id="cb170-3"><a href="module-3.html#cb170-3" tabindex="-1"></a><span class="co">#View the first five rows of data (all columns) in the gene_expression (Stringtie TPM) dataframe</span></span>
<span id="cb170-4"><a href="module-3.html#cb170-4" tabindex="-1"></a><span class="fu">head</span>(gene_expression)</span>
<span id="cb170-5"><a href="module-3.html#cb170-5" tabindex="-1"></a></span>
<span id="cb170-6"><a href="module-3.html#cb170-6" tabindex="-1"></a><span class="co">#View the column names</span></span>
<span id="cb170-7"><a href="module-3.html#cb170-7" tabindex="-1"></a><span class="fu">colnames</span>(gene_expression)</span>
<span id="cb170-8"><a href="module-3.html#cb170-8" tabindex="-1"></a></span>
<span id="cb170-9"><a href="module-3.html#cb170-9" tabindex="-1"></a><span class="co">#View the row names</span></span>
<span id="cb170-10"><a href="module-3.html#cb170-10" tabindex="-1"></a><span class="fu">row.names</span>(gene_expression)</span>
<span id="cb170-11"><a href="module-3.html#cb170-11" tabindex="-1"></a></span>
<span id="cb170-12"><a href="module-3.html#cb170-12" tabindex="-1"></a><span class="co">#Determine the dimensions of the dataframe.  &#39;dim()&#39; will return the number of rows and columns</span></span>
<span id="cb170-13"><a href="module-3.html#cb170-13" tabindex="-1"></a><span class="fu">dim</span>(gene_expression)</span>
<span id="cb170-14"><a href="module-3.html#cb170-14" tabindex="-1"></a></span>
<span id="cb170-15"><a href="module-3.html#cb170-15" tabindex="-1"></a><span class="co">#Get the first 3 rows of data and a selection of columns</span></span>
<span id="cb170-16"><a href="module-3.html#cb170-16" tabindex="-1"></a>gene_expression[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">6</span>)]</span>
<span id="cb170-17"><a href="module-3.html#cb170-17" tabindex="-1"></a></span>
<span id="cb170-18"><a href="module-3.html#cb170-18" tabindex="-1"></a><span class="co">#Do the same thing, but using the column names instead of numbers</span></span>
<span id="cb170-19"><a href="module-3.html#cb170-19" tabindex="-1"></a>gene_expression[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="fu">c</span>(<span class="st">&quot;HBR_Rep1&quot;</span>, <span class="st">&quot;HBR_Rep2&quot;</span>, <span class="st">&quot;HBR_Rep3&quot;</span>, <span class="st">&quot;UHR_Rep3&quot;</span>)]</span>
<span id="cb170-20"><a href="module-3.html#cb170-20" tabindex="-1"></a></span>
<span id="cb170-21"><a href="module-3.html#cb170-21" tabindex="-1"></a><span class="co">#Now, exlore the differential expression (DESeq2 results) </span></span>
<span id="cb170-22"><a href="module-3.html#cb170-22" tabindex="-1"></a><span class="fu">head</span>(results_genes)</span>
<span id="cb170-23"><a href="module-3.html#cb170-23" tabindex="-1"></a><span class="fu">dim</span>(results_genes)</span>
<span id="cb170-24"><a href="module-3.html#cb170-24" tabindex="-1"></a></span>
<span id="cb170-25"><a href="module-3.html#cb170-25" tabindex="-1"></a><span class="co">#Assign some colors for use later.  You can specify color by RGB, Hex code, or name</span></span>
<span id="cb170-26"><a href="module-3.html#cb170-26" tabindex="-1"></a><span class="co">#To get a list of color names:</span></span>
<span id="cb170-27"><a href="module-3.html#cb170-27" tabindex="-1"></a><span class="fu">colours</span>()</span>
<span id="cb170-28"><a href="module-3.html#cb170-28" tabindex="-1"></a>data_colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;tomato1&quot;</span>, <span class="st">&quot;tomato2&quot;</span>, <span class="st">&quot;tomato3&quot;</span>, <span class="st">&quot;royalblue1&quot;</span>, <span class="st">&quot;royalblue2&quot;</span>, <span class="st">&quot;royalblue3&quot;</span>)</span></code></pre></div>
<p>The following code blocks are to generate various plots using the above data set.</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="module-3.html#cb171-1" tabindex="-1"></a></span>
<span id="cb171-2"><a href="module-3.html#cb171-2" tabindex="-1"></a><span class="do">#### Plot #1 - View the range of values and general distribution of TPM values for all 6 libraries</span></span>
<span id="cb171-3"><a href="module-3.html#cb171-3" tabindex="-1"></a><span class="co">#Create boxplots for this purpose</span></span>
<span id="cb171-4"><a href="module-3.html#cb171-4" tabindex="-1"></a><span class="co">#Display on a log2 scale and set a minimum non-zero value to avoid log2(0)</span></span>
<span id="cb171-5"><a href="module-3.html#cb171-5" tabindex="-1"></a>min_nonzero <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb171-6"><a href="module-3.html#cb171-6" tabindex="-1"></a></span>
<span id="cb171-7"><a href="module-3.html#cb171-7" tabindex="-1"></a><span class="co"># Set the columns for finding TPM and create shorter names for figures</span></span>
<span id="cb171-8"><a href="module-3.html#cb171-8" tabindex="-1"></a>data_columns <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb171-9"><a href="module-3.html#cb171-9" tabindex="-1"></a>short_names <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;HBR_1&quot;</span>, <span class="st">&quot;HBR_2&quot;</span>, <span class="st">&quot;HBR_3&quot;</span>, <span class="st">&quot;UHR_1&quot;</span>, <span class="st">&quot;UHR_2&quot;</span>, <span class="st">&quot;UHR_3&quot;</span>)</span>
<span id="cb171-10"><a href="module-3.html#cb171-10" tabindex="-1"></a></span>
<span id="cb171-11"><a href="module-3.html#cb171-11" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;All_samples_TPM_boxplots.pdf&quot;</span>)</span>
<span id="cb171-12"><a href="module-3.html#cb171-12" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">log2</span>(gene_expression[, data_columns] <span class="sc">+</span> min_nonzero), <span class="at">col =</span> data_colors, <span class="at">names =</span> short_names, <span class="at">las =</span> <span class="dv">2</span>, <span class="at">ylab =</span> <span class="st">&quot;log2(TPM)&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Distribution of TPMs for all 6 libraries&quot;</span>)</span>
<span id="cb171-13"><a href="module-3.html#cb171-13" tabindex="-1"></a><span class="co">#Note that the bold horizontal line on each boxplot is the median</span></span>
<span id="cb171-14"><a href="module-3.html#cb171-14" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-15"><a href="module-3.html#cb171-15" tabindex="-1"></a></span>
<span id="cb171-16"><a href="module-3.html#cb171-16" tabindex="-1"></a><span class="do">#### Plot #2 - plot a pair of replicates to assess reproducibility of technical replicates</span></span>
<span id="cb171-17"><a href="module-3.html#cb171-17" tabindex="-1"></a><span class="co">#Tranform the data by converting to log2 scale after adding an arbitrary small value to avoid log2(0)</span></span>
<span id="cb171-18"><a href="module-3.html#cb171-18" tabindex="-1"></a>x <span class="ot">=</span> gene_expression[, <span class="st">&quot;UHR_Rep1&quot;</span>]</span>
<span id="cb171-19"><a href="module-3.html#cb171-19" tabindex="-1"></a>y <span class="ot">=</span> gene_expression[, <span class="st">&quot;UHR_Rep2&quot;</span>]</span>
<span id="cb171-20"><a href="module-3.html#cb171-20" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_Rep1_vs_Rep2_scatter.pdf&quot;</span>)</span>
<span id="cb171-21"><a href="module-3.html#cb171-21" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">log2</span>(x <span class="sc">+</span> min_nonzero), <span class="at">y =</span> <span class="fu">log2</span>(y <span class="sc">+</span> min_nonzero), <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="st">&quot;blue&quot;</span>, <span class="at">cex =</span> <span class="fl">0.25</span>, <span class="at">xlab =</span> <span class="st">&quot;TPM (UHR, Replicate 1)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;TPM (UHR, Replicate 2)&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Comparison of expression values for a pair of replicates&quot;</span>)</span>
<span id="cb171-22"><a href="module-3.html#cb171-22" tabindex="-1"></a></span>
<span id="cb171-23"><a href="module-3.html#cb171-23" tabindex="-1"></a><span class="co">#Add a straight line of slope 1, and intercept 0</span></span>
<span id="cb171-24"><a href="module-3.html#cb171-24" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>)</span>
<span id="cb171-25"><a href="module-3.html#cb171-25" tabindex="-1"></a></span>
<span id="cb171-26"><a href="module-3.html#cb171-26" tabindex="-1"></a><span class="co">#Calculate the correlation coefficient and display in a legend</span></span>
<span id="cb171-27"><a href="module-3.html#cb171-27" tabindex="-1"></a>rs <span class="ot">=</span> <span class="fu">cor</span>(x, y)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb171-28"><a href="module-3.html#cb171-28" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="fu">paste</span>(<span class="st">&quot;R squared = &quot;</span>, <span class="fu">round</span>(rs, <span class="at">digits =</span> <span class="dv">3</span>), <span class="at">sep =</span> <span class="st">&quot;&quot;</span>), <span class="at">lwd =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb171-29"><a href="module-3.html#cb171-29" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-30"><a href="module-3.html#cb171-30" tabindex="-1"></a></span>
<span id="cb171-31"><a href="module-3.html#cb171-31" tabindex="-1"></a><span class="do">#### Plot #3 - Scatter plots with a large number of data points can be misleading ... regenerate this figure as a density scatter plot</span></span>
<span id="cb171-32"><a href="module-3.html#cb171-32" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_Rep1_vs_Rep2_SmoothScatter.pdf&quot;</span>)</span>
<span id="cb171-33"><a href="module-3.html#cb171-33" tabindex="-1"></a>colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;#007FFF&quot;</span>, <span class="st">&quot;cyan&quot;</span>,<span class="st">&quot;#7FFF7F&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;#FF7F00&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;#7F0000&quot;</span>))</span>
<span id="cb171-34"><a href="module-3.html#cb171-34" tabindex="-1"></a><span class="fu">smoothScatter</span>(<span class="at">x =</span> <span class="fu">log2</span>(x <span class="sc">+</span> min_nonzero), <span class="at">y =</span> <span class="fu">log2</span>(y <span class="sc">+</span> min_nonzero), <span class="at">xlab =</span> <span class="st">&quot;TPM (UHR, Replicate 1)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;TPM (UHR, Replicate 2)&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Comparison of expression values for a pair of replicates&quot;</span>, <span class="at">colramp =</span> colors, <span class="at">nbin =</span> <span class="dv">200</span>)</span>
<span id="cb171-35"><a href="module-3.html#cb171-35" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-36"><a href="module-3.html#cb171-36" tabindex="-1"></a></span>
<span id="cb171-37"><a href="module-3.html#cb171-37" tabindex="-1"></a><span class="do">#### Plot #4 - Scatter plots of all sets of replicates on a single plot</span></span>
<span id="cb171-38"><a href="module-3.html#cb171-38" tabindex="-1"></a><span class="co">#Create a function that generates an R plot.  This function will take as input the two libraries to be compared and a plot name</span></span>
<span id="cb171-39"><a href="module-3.html#cb171-39" tabindex="-1"></a>plotCor <span class="ot">=</span> <span class="cf">function</span>(lib1, lib2, name){</span>
<span id="cb171-40"><a href="module-3.html#cb171-40" tabindex="-1"></a>    x <span class="ot">=</span> gene_expression[, lib1]</span>
<span id="cb171-41"><a href="module-3.html#cb171-41" tabindex="-1"></a>    y <span class="ot">=</span> gene_expression[, lib2]</span>
<span id="cb171-42"><a href="module-3.html#cb171-42" tabindex="-1"></a>    colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;#007FFF&quot;</span>, <span class="st">&quot;cyan&quot;</span>, <span class="st">&quot;#7FFF7F&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;#FF7F00&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;#7F0000&quot;</span>))</span>
<span id="cb171-43"><a href="module-3.html#cb171-43" tabindex="-1"></a>    <span class="fu">smoothScatter</span>(<span class="at">x =</span> <span class="fu">log2</span>(x <span class="sc">+</span> min_nonzero), <span class="at">y =</span> <span class="fu">log2</span>(y <span class="sc">+</span> min_nonzero), <span class="at">xlab =</span> lib1, <span class="at">ylab =</span> lib2, <span class="at">main =</span> name, <span class="at">colramp =</span> colors, <span class="at">nbin =</span> <span class="dv">275</span>)</span>
<span id="cb171-44"><a href="module-3.html#cb171-44" tabindex="-1"></a>    <span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>)</span>
<span id="cb171-45"><a href="module-3.html#cb171-45" tabindex="-1"></a>        zero_count <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">which</span>(x <span class="sc">==</span> <span class="dv">0</span>)) <span class="sc">+</span> <span class="fu">length</span>(<span class="fu">which</span>(y <span class="sc">==</span> <span class="dv">0</span>))</span>
<span id="cb171-46"><a href="module-3.html#cb171-46" tabindex="-1"></a>    rs <span class="ot">=</span> <span class="fu">cor</span>(x, y, <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb171-47"><a href="module-3.html#cb171-47" tabindex="-1"></a>    legend_text <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">paste</span>(<span class="st">&quot;R squared = &quot;</span>, <span class="fu">round</span>(rs, <span class="at">digits =</span> <span class="dv">3</span>), <span class="at">sep=</span><span class="st">&quot;&quot;</span>), <span class="fu">paste</span>(<span class="st">&quot;Zero count = &quot;</span>, zero_count, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span>
<span id="cb171-48"><a href="module-3.html#cb171-48" tabindex="-1"></a>    <span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, legend_text, <span class="at">lwd =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="cn">NA</span>), <span class="at">col =</span> <span class="st">&quot;black&quot;</span>, <span class="at">bg =</span> <span class="st">&quot;white&quot;</span>, <span class="at">cex =</span> <span class="fl">0.8</span>)</span>
<span id="cb171-49"><a href="module-3.html#cb171-49" tabindex="-1"></a>}</span>
<span id="cb171-50"><a href="module-3.html#cb171-50" tabindex="-1"></a></span>
<span id="cb171-51"><a href="module-3.html#cb171-51" tabindex="-1"></a><span class="co">#Now make a call to our custom function created above, once for each library comparison</span></span>
<span id="cb171-52"><a href="module-3.html#cb171-52" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_All_Reps_SmoothScatter.pdf&quot;</span>)</span>
<span id="cb171-53"><a href="module-3.html#cb171-53" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb171-54"><a href="module-3.html#cb171-54" tabindex="-1"></a><span class="fu">plotCor</span>(<span class="st">&quot;UHR_Rep1&quot;</span>, <span class="st">&quot;UHR_Rep2&quot;</span>, <span class="st">&quot;UHR_1 vs UHR_2&quot;</span>)</span>
<span id="cb171-55"><a href="module-3.html#cb171-55" tabindex="-1"></a><span class="fu">plotCor</span>(<span class="st">&quot;UHR_Rep2&quot;</span>, <span class="st">&quot;UHR_Rep3&quot;</span>, <span class="st">&quot;UHR_2 vs UHR_3&quot;</span>)</span>
<span id="cb171-56"><a href="module-3.html#cb171-56" tabindex="-1"></a><span class="fu">plotCor</span>(<span class="st">&quot;UHR_Rep1&quot;</span>, <span class="st">&quot;UHR_Rep3&quot;</span>, <span class="st">&quot;UHR_1 vs UHR_3&quot;</span>)</span>
<span id="cb171-57"><a href="module-3.html#cb171-57" tabindex="-1"></a></span>
<span id="cb171-58"><a href="module-3.html#cb171-58" tabindex="-1"></a><span class="do">#### Compare the correlation between all replicates</span></span>
<span id="cb171-59"><a href="module-3.html#cb171-59" tabindex="-1"></a><span class="co">#Do we see the expected pattern for all eight libraries (i.e. replicates most similar, then tumor vs. normal)?</span></span>
<span id="cb171-60"><a href="module-3.html#cb171-60" tabindex="-1"></a></span>
<span id="cb171-61"><a href="module-3.html#cb171-61" tabindex="-1"></a><span class="co">#Calculate the TPM sum for all 6 libraries</span></span>
<span id="cb171-62"><a href="module-3.html#cb171-62" tabindex="-1"></a>gene_expression[,<span class="st">&quot;sum&quot;</span>] <span class="ot">=</span> <span class="fu">apply</span>(gene_expression[,data_columns], <span class="dv">1</span>, sum)</span>
<span id="cb171-63"><a href="module-3.html#cb171-63" tabindex="-1"></a></span>
<span id="cb171-64"><a href="module-3.html#cb171-64" tabindex="-1"></a><span class="co">#Identify the genes with a grand sum TPM of at least 5 - we will filter out the genes with very low expression across the board</span></span>
<span id="cb171-65"><a href="module-3.html#cb171-65" tabindex="-1"></a>i <span class="ot">=</span> <span class="fu">which</span>(gene_expression[,<span class="st">&quot;sum&quot;</span>] <span class="sc">&gt;</span> <span class="dv">5</span>)</span>
<span id="cb171-66"><a href="module-3.html#cb171-66" tabindex="-1"></a></span>
<span id="cb171-67"><a href="module-3.html#cb171-67" tabindex="-1"></a><span class="co">#Calculate the correlation between all pairs of data</span></span>
<span id="cb171-68"><a href="module-3.html#cb171-68" tabindex="-1"></a>r <span class="ot">=</span> <span class="fu">cor</span>(gene_expression[i,data_columns], <span class="at">use =</span> <span class="st">&quot;pairwise.complete.obs&quot;</span>, <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb171-69"><a href="module-3.html#cb171-69" tabindex="-1"></a></span>
<span id="cb171-70"><a href="module-3.html#cb171-70" tabindex="-1"></a><span class="co">#Print out these correlation values</span></span>
<span id="cb171-71"><a href="module-3.html#cb171-71" tabindex="-1"></a>r</span>
<span id="cb171-72"><a href="module-3.html#cb171-72" tabindex="-1"></a></span>
<span id="cb171-73"><a href="module-3.html#cb171-73" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-74"><a href="module-3.html#cb171-74" tabindex="-1"></a></span>
<span id="cb171-75"><a href="module-3.html#cb171-75" tabindex="-1"></a></span>
<span id="cb171-76"><a href="module-3.html#cb171-76" tabindex="-1"></a><span class="do">#### Plot #5 - Convert correlation to &#39;distance&#39;, and use &#39;multi-dimensional scaling&#39; to display the relative differences between libraries</span></span>
<span id="cb171-77"><a href="module-3.html#cb171-77" tabindex="-1"></a><span class="co">#This step calculates 2-dimensional coordinates to plot points for each library</span></span>
<span id="cb171-78"><a href="module-3.html#cb171-78" tabindex="-1"></a><span class="co">#Libraries with similar expression patterns (highly correlated to each other) should group together</span></span>
<span id="cb171-79"><a href="module-3.html#cb171-79" tabindex="-1"></a><span class="co">#What pattern do we expect to see, given the types of libraries we have (technical replicates, biologal replicates, tumor/normal)?</span></span>
<span id="cb171-80"><a href="module-3.html#cb171-80" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_vs_HBR_MDS.pdf&quot;</span>)</span>
<span id="cb171-81"><a href="module-3.html#cb171-81" tabindex="-1"></a>d <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> r</span>
<span id="cb171-82"><a href="module-3.html#cb171-82" tabindex="-1"></a>mds <span class="ot">=</span> <span class="fu">cmdscale</span>(d, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb171-83"><a href="module-3.html#cb171-83" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb171-84"><a href="module-3.html#cb171-84" tabindex="-1"></a><span class="fu">plot</span>(mds<span class="sc">$</span>points, <span class="at">type =</span> <span class="st">&quot;n&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="at">main =</span> <span class="st">&quot;MDS distance plot (all non-zero genes)&quot;</span>, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.12</span>, <span class="fl">0.12</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.12</span>, <span class="fl">0.12</span>))</span>
<span id="cb171-85"><a href="module-3.html#cb171-85" tabindex="-1"></a><span class="fu">points</span>(mds<span class="sc">$</span>points[, <span class="dv">1</span>], mds<span class="sc">$</span>points[, <span class="dv">2</span>], <span class="at">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="at">cex =</span> <span class="dv">2</span>, <span class="at">pch =</span> <span class="dv">16</span>)</span>
<span id="cb171-86"><a href="module-3.html#cb171-86" tabindex="-1"></a><span class="fu">text</span>(mds<span class="sc">$</span>points[, <span class="dv">1</span>], mds<span class="sc">$</span>points[, <span class="dv">2</span>], short_names, <span class="at">col =</span> data_colors)</span>
<span id="cb171-87"><a href="module-3.html#cb171-87" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-88"><a href="module-3.html#cb171-88" tabindex="-1"></a></span>
<span id="cb171-89"><a href="module-3.html#cb171-89" tabindex="-1"></a></span>
<span id="cb171-90"><a href="module-3.html#cb171-90" tabindex="-1"></a><span class="do">#### Plot #6 - View the distribution of differential expression values as a histogram</span></span>
<span id="cb171-91"><a href="module-3.html#cb171-91" tabindex="-1"></a><span class="co">#Display only those results that are significant according to DESeq2 (loaded above)</span></span>
<span id="cb171-92"><a href="module-3.html#cb171-92" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_vs_HBR_DE_FC_distribution.pdf&quot;</span>)</span>
<span id="cb171-93"><a href="module-3.html#cb171-93" tabindex="-1"></a>sig <span class="ot">=</span> <span class="fu">which</span>(results_genes<span class="sc">$</span>pvalue <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb171-94"><a href="module-3.html#cb171-94" tabindex="-1"></a><span class="fu">hist</span>(results_genes[sig, <span class="st">&quot;log2FoldChange&quot;</span>], <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">col =</span> <span class="st">&quot;seagreen&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;log2(Fold change) UHR vs HBR&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Distribution of differential expression values&quot;</span>)</span>
<span id="cb171-95"><a href="module-3.html#cb171-95" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb171-96"><a href="module-3.html#cb171-96" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;black&quot;</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb171-97"><a href="module-3.html#cb171-97" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="st">&quot;Fold-change &gt; 2&quot;</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb171-98"><a href="module-3.html#cb171-98" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-99"><a href="module-3.html#cb171-99" tabindex="-1"></a></span>
<span id="cb171-100"><a href="module-3.html#cb171-100" tabindex="-1"></a><span class="do">#### Plot #7 - Display the mean expression values from UHR and HBR and mark those that are significantly differentially expressed</span></span>
<span id="cb171-101"><a href="module-3.html#cb171-101" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">&quot;UHR_vs_HBR_mean_TPM_scatter.pdf&quot;</span>)</span>
<span id="cb171-102"><a href="module-3.html#cb171-102" tabindex="-1"></a></span>
<span id="cb171-103"><a href="module-3.html#cb171-103" tabindex="-1"></a>gene_expression[, <span class="st">&quot;HBR_mean&quot;</span>] <span class="ot">=</span> <span class="fu">apply</span>(gene_expression[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)], <span class="dv">1</span>, mean)</span>
<span id="cb171-104"><a href="module-3.html#cb171-104" tabindex="-1"></a>gene_expression[, <span class="st">&quot;UHR_mean&quot;</span>] <span class="ot">=</span> <span class="fu">apply</span>(gene_expression[,<span class="fu">c</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>)], <span class="dv">1</span>, mean)</span>
<span id="cb171-105"><a href="module-3.html#cb171-105" tabindex="-1"></a></span>
<span id="cb171-106"><a href="module-3.html#cb171-106" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">log2</span>(gene_expression[, <span class="st">&quot;UHR_mean&quot;</span>] <span class="sc">+</span> min_nonzero)</span>
<span id="cb171-107"><a href="module-3.html#cb171-107" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">log2</span>(gene_expression[, <span class="st">&quot;HBR_mean&quot;</span>] <span class="sc">+</span> min_nonzero)</span>
<span id="cb171-108"><a href="module-3.html#cb171-108" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.25</span>, <span class="at">xlab =</span> <span class="st">&quot;UHR TPM (log2)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;HBR TPM (log2)&quot;</span>, <span class="at">main =</span> <span class="st">&quot;UHR vs HBR TPMs&quot;</span>)</span>
<span id="cb171-109"><a href="module-3.html#cb171-109" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>)</span>
<span id="cb171-110"><a href="module-3.html#cb171-110" tabindex="-1"></a>xsig <span class="ot">=</span> x[sig]</span>
<span id="cb171-111"><a href="module-3.html#cb171-111" tabindex="-1"></a>ysig <span class="ot">=</span> y[sig]</span>
<span id="cb171-112"><a href="module-3.html#cb171-112" tabindex="-1"></a><span class="fu">points</span>(<span class="at">x =</span> xsig, <span class="at">y =</span> ysig, <span class="at">col =</span> <span class="st">&quot;magenta&quot;</span>, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb171-113"><a href="module-3.html#cb171-113" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="st">&quot;Significant&quot;</span>, <span class="at">col =</span> <span class="st">&quot;magenta&quot;</span>, <span class="at">pch =</span> <span class="dv">16</span>)</span>
<span id="cb171-114"><a href="module-3.html#cb171-114" tabindex="-1"></a></span>
<span id="cb171-115"><a href="module-3.html#cb171-115" tabindex="-1"></a><span class="co">#Get the gene symbols for the top N (according to corrected p-value) and display them on the plot</span></span>
<span id="cb171-116"><a href="module-3.html#cb171-116" tabindex="-1"></a>topn <span class="ot">=</span> <span class="fu">order</span>(results_genes[sig,<span class="st">&quot;padj&quot;</span>])[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>]</span>
<span id="cb171-117"><a href="module-3.html#cb171-117" tabindex="-1"></a><span class="fu">text</span>(x[topn], y[topn], results_genes[topn, <span class="st">&quot;Symbol&quot;</span>], <span class="at">col =</span> <span class="st">&quot;black&quot;</span>, <span class="at">cex =</span> <span class="fl">0.75</span>, <span class="at">srt =</span> <span class="dv">45</span>)</span>
<span id="cb171-118"><a href="module-3.html#cb171-118" tabindex="-1"></a></span>
<span id="cb171-119"><a href="module-3.html#cb171-119" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-120"><a href="module-3.html#cb171-120" tabindex="-1"></a></span>
<span id="cb171-121"><a href="module-3.html#cb171-121" tabindex="-1"></a><span class="do">#### Plot #8 - Create a heatmap to vizualize expression differences between the six samples</span></span>
<span id="cb171-122"><a href="module-3.html#cb171-122" tabindex="-1"></a><span class="co">#Define custom dist and hclust functions for use with heatmaps</span></span>
<span id="cb171-123"><a href="module-3.html#cb171-123" tabindex="-1"></a>mydist <span class="ot">=</span> <span class="cf">function</span>(c) {<span class="fu">dist</span>(c, <span class="at">method =</span> <span class="st">&quot;euclidian&quot;</span>)}</span>
<span id="cb171-124"><a href="module-3.html#cb171-124" tabindex="-1"></a>myclust <span class="ot">=</span> <span class="cf">function</span>(c) {<span class="fu">hclust</span>(c, <span class="at">method =</span> <span class="st">&quot;average&quot;</span>)}</span>
<span id="cb171-125"><a href="module-3.html#cb171-125" tabindex="-1"></a></span>
<span id="cb171-126"><a href="module-3.html#cb171-126" tabindex="-1"></a><span class="co">#Create a subset of significant genes with p-value&lt;0.05 and log2 fold-change &gt;= 2</span></span>
<span id="cb171-127"><a href="module-3.html#cb171-127" tabindex="-1"></a>sigpi <span class="ot">=</span> <span class="fu">which</span>(results_genes[, <span class="st">&quot;pvalue&quot;</span>] <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb171-128"><a href="module-3.html#cb171-128" tabindex="-1"></a>sigp <span class="ot">=</span> results_genes[sigpi, ]</span>
<span id="cb171-129"><a href="module-3.html#cb171-129" tabindex="-1"></a>sigfc <span class="ot">=</span> <span class="fu">which</span>(<span class="fu">abs</span>(sigp[, <span class="st">&quot;log2FoldChange&quot;</span>]) <span class="sc">&gt;=</span> <span class="dv">2</span>)</span>
<span id="cb171-130"><a href="module-3.html#cb171-130" tabindex="-1"></a>sigDE <span class="ot">=</span> sigp[sigfc, ]</span>
<span id="cb171-131"><a href="module-3.html#cb171-131" tabindex="-1"></a></span>
<span id="cb171-132"><a href="module-3.html#cb171-132" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_vs_HBR_heatmap.pdf&quot;</span>)</span>
<span id="cb171-133"><a href="module-3.html#cb171-133" tabindex="-1"></a>main_title <span class="ot">=</span> <span class="st">&quot;sig DE Genes&quot;</span></span>
<span id="cb171-134"><a href="module-3.html#cb171-134" tabindex="-1"></a><span class="fu">par</span>(<span class="at">cex.main =</span> <span class="fl">0.8</span>)</span>
<span id="cb171-135"><a href="module-3.html#cb171-135" tabindex="-1"></a>sigDE_genes <span class="ot">=</span> sigDE[, <span class="st">&quot;ensemblID&quot;</span>]</span>
<span id="cb171-136"><a href="module-3.html#cb171-136" tabindex="-1"></a>sigDE_genenames <span class="ot">=</span> sigDE[, <span class="st">&quot;Symbol&quot;</span>]</span>
<span id="cb171-137"><a href="module-3.html#cb171-137" tabindex="-1"></a></span>
<span id="cb171-138"><a href="module-3.html#cb171-138" tabindex="-1"></a>data <span class="ot">=</span> <span class="fu">log2</span>(<span class="fu">as.matrix</span>(gene_expression[<span class="fu">as.vector</span>(sigDE_genes), data_columns]) <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb171-139"><a href="module-3.html#cb171-139" tabindex="-1"></a><span class="fu">heatmap.2</span>(data, <span class="at">hclustfun =</span> myclust, <span class="at">distfun =</span> mydist, <span class="at">na.rm =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="st">&quot;none&quot;</span>, <span class="at">dendrogram =</span> <span class="st">&quot;both&quot;</span>, <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">10</span>,<span class="dv">4</span>), <span class="at">Rowv =</span> <span class="cn">TRUE</span>, <span class="at">Colv =</span> <span class="cn">TRUE</span>, <span class="at">symbreaks =</span> <span class="cn">FALSE</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">symkey =</span> <span class="cn">FALSE</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>, <span class="at">trace =</span> <span class="st">&quot;none&quot;</span>, <span class="at">main =</span> main_title, <span class="at">cexRow =</span> <span class="fl">0.3</span>, <span class="at">cexCol =</span> <span class="dv">1</span>, <span class="at">labRow =</span> sigDE_genenames, <span class="at">col =</span> <span class="fu">rev</span>(<span class="fu">heat.colors</span>(<span class="dv">75</span>)))</span>
<span id="cb171-140"><a href="module-3.html#cb171-140" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-141"><a href="module-3.html#cb171-141" tabindex="-1"></a></span>
<span id="cb171-142"><a href="module-3.html#cb171-142" tabindex="-1"></a><span class="do">#### Plot #9 - Volcano plot</span></span>
<span id="cb171-143"><a href="module-3.html#cb171-143" tabindex="-1"></a></span>
<span id="cb171-144"><a href="module-3.html#cb171-144" tabindex="-1"></a><span class="co"># Set differential expression status for each gene - default all genes to &quot;no change&quot;</span></span>
<span id="cb171-145"><a href="module-3.html#cb171-145" tabindex="-1"></a>results_genes<span class="sc">$</span>diffexpressed <span class="ot">=</span> <span class="st">&quot;No&quot;</span></span>
<span id="cb171-146"><a href="module-3.html#cb171-146" tabindex="-1"></a></span>
<span id="cb171-147"><a href="module-3.html#cb171-147" tabindex="-1"></a><span class="co"># if log2Foldchange &gt; 2 and pvalue &lt; 0.05, set as &quot;Up regulated&quot;</span></span>
<span id="cb171-148"><a href="module-3.html#cb171-148" tabindex="-1"></a>results_genes<span class="sc">$</span>diffexpressed[results_genes<span class="sc">$</span>log2FoldChange <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">&amp;</span> results_genes<span class="sc">$</span>pvalue <span class="sc">&lt;</span> <span class="fl">0.05</span>] <span class="ot">=</span> <span class="st">&quot;Higher in UHR&quot;</span></span>
<span id="cb171-149"><a href="module-3.html#cb171-149" tabindex="-1"></a></span>
<span id="cb171-150"><a href="module-3.html#cb171-150" tabindex="-1"></a><span class="co"># if log2Foldchange &lt; -2 and pvalue &lt; 0.05, set as &quot;Down regulated&quot;</span></span>
<span id="cb171-151"><a href="module-3.html#cb171-151" tabindex="-1"></a>results_genes<span class="sc">$</span>diffexpressed[results_genes<span class="sc">$</span>log2FoldChange <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">2</span> <span class="sc">&amp;</span> results_genes<span class="sc">$</span>pvalue <span class="sc">&lt;</span> <span class="fl">0.05</span>] <span class="ot">=</span> <span class="st">&quot;Higher in HBR&quot;</span></span>
<span id="cb171-152"><a href="module-3.html#cb171-152" tabindex="-1"></a></span>
<span id="cb171-153"><a href="module-3.html#cb171-153" tabindex="-1"></a><span class="co"># write the gene names of those significantly upregulated/downregulated to a new column</span></span>
<span id="cb171-154"><a href="module-3.html#cb171-154" tabindex="-1"></a>results_genes<span class="sc">$</span>gene_label <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb171-155"><a href="module-3.html#cb171-155" tabindex="-1"></a>results_genes<span class="sc">$</span>gene_label[results_genes<span class="sc">$</span>diffexpressed <span class="sc">!=</span> <span class="st">&quot;No&quot;</span>] <span class="ot">=</span> results_genes<span class="sc">$</span>Symbol[results_genes<span class="sc">$</span>diffexpressed <span class="sc">!=</span> <span class="st">&quot;No&quot;</span>]</span>
<span id="cb171-156"><a href="module-3.html#cb171-156" tabindex="-1"></a></span>
<span id="cb171-157"><a href="module-3.html#cb171-157" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;UHR_vs_HBR_volcano.pdf&quot;</span>)</span>
<span id="cb171-158"><a href="module-3.html#cb171-158" tabindex="-1"></a></span>
<span id="cb171-159"><a href="module-3.html#cb171-159" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> results_genes[results_genes<span class="sc">$</span>diffexpressed <span class="sc">!=</span> <span class="st">&quot;No&quot;</span>,], <span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue), <span class="at">label =</span> gene_label, <span class="at">color =</span> diffexpressed)) <span class="sc">+</span></span>
<span id="cb171-160"><a href="module-3.html#cb171-160" tabindex="-1"></a>             <span class="fu">xlab</span>(<span class="st">&quot;log2Foldchange&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-161"><a href="module-3.html#cb171-161" tabindex="-1"></a>             <span class="fu">scale_color_manual</span>(<span class="at">name =</span> <span class="st">&quot;Differentially expressed&quot;</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span></span>
<span id="cb171-162"><a href="module-3.html#cb171-162" tabindex="-1"></a>             <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb171-163"><a href="module-3.html#cb171-163" tabindex="-1"></a>             <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb171-164"><a href="module-3.html#cb171-164" tabindex="-1"></a>             <span class="fu">geom_text_repel</span>() <span class="sc">+</span></span>
<span id="cb171-165"><a href="module-3.html#cb171-165" tabindex="-1"></a>             <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.6</span>, <span class="fl">0.6</span>), <span class="at">col =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-166"><a href="module-3.html#cb171-166" tabindex="-1"></a>             <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), <span class="at">col =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb171-167"><a href="module-3.html#cb171-167" tabindex="-1"></a>             <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="dv">5</span>))) <span class="sc">+</span></span>
<span id="cb171-168"><a href="module-3.html#cb171-168" tabindex="-1"></a>             <span class="fu">geom_point</span>(<span class="at">data =</span> results_genes[results_genes<span class="sc">$</span>diffexpressed <span class="sc">==</span> <span class="st">&quot;No&quot;</span>,], <span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(pvalue)), <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb171-169"><a href="module-3.html#cb171-169" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb171-170"><a href="module-3.html#cb171-170" tabindex="-1"></a></span>
<span id="cb171-171"><a href="module-3.html#cb171-171" tabindex="-1"></a><span class="co">#To exit R type:</span></span>
<span id="cb171-172"><a href="module-3.html#cb171-172" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>The output file can be viewed in your browser at the following url. Note, you must replace <strong>YOUR_PUBLIC_IPv4_ADDRESS</strong> with your own amazon instance IP (e.g., 101.0.1.101)).</p>
<ul>
<li><a href="http://" class="uri">http://</a><strong>YOUR_PUBLIC_IPv4_ADDRESS</strong>/rnaseq/de/ballgown/ref_only/Tutorial_Part3_Supplementary_R_output.pdf</li>
</ul>
</div>
<div id="visual-comparison-of-example-genes-from-the-volcano-plot" class="section level4 hasAnchor">
<h4>Visual comparison of example genes from the volcano plot<a href="module-3.html#visual-comparison-of-example-genes-from-the-volcano-plot" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>One can manually explore interesting looking genes from the volcano plot. In this case our analysis involves comparison of RNA isolated from tissues of different types (HBR -&gt; brain tissue, UHR -&gt; a collection of cancer cell lines). So, in this analysis it might make sense to explore candidates in a tissue expression atlas such as <a href="https://www.gtexportal.org/">GTEX</a>.</p>
<p>Looking at our gene plot, two example genes we could look at are: SEPT3 (significantly higher in HBR) and PRAME (significantly higher in UHR).</p>
<p><em>Expression of SEPT3 across tissues according to GTEX</em>
<img src="assets/module_3/SEPT3.png" alt="SEPT3" />
Note that this gene appears to be highly expressed in brain tissues.</p>
<p><em>Expression of PRAME across tissues according to GTEX</em>
<img src="assets/module_3/PRAME.png" alt="PRAME" />
Note that this gene appears to be almost uniquely expressed in testis tissue. Since one of the cell lines in the UHR sample is a testicular cancer cell line, this makes sense.</p>
<hr />
</div>
<div id="practical-exercise-10-advanced" class="section level4 hasAnchor">
<h4>PRACTICAL EXERCISE 10 (ADVANCED)<a href="module-3.html#practical-exercise-10-advanced" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Assignment: Use R to create a volcano plot for the differentially expressed genes you identified with Ballgown in Practical Exercise 9.</p>
<ul>
<li>Hint: Follow the example R code above.</li>
<li>Hint: You could import the ballgown data object (e.g., <code>bg.rda</code>) that you should have saved in Practical Exercise 9 as a source of DE results.</li>
</ul>
<p>Solution: When you are ready you can check your approach against the <a href="/module-09-appendix/0009/05/01/Practical_Exercise_Solutions/#practical-exercise-10---volcano-plot">Solutions</a></p>
<hr />
<p>In this section we will use the GAGE tool in R to test for significantly enriched sets of genes within those genes found to be significantly “up” and “down” in our UHR vs HBR differential gene expression analysis. Do we see enrichment for genes associated with brain related cell types and processes in the list of DE genes that have significant differential expression beween the UHR samples compared to the HBR samples?</p>
</div>
<div id="what-is-gage" class="section level4 hasAnchor">
<h4>What is gage?<a href="module-3.html#what-is-gage" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The Generally Applicable Gene-set Enrichment tool (<a href="https://bioconductor.org/packages/release/bioc/html/gage.html">GAGE</a>) is a popular bioconductor package used to perform gene-set enrichment and pathway analysis. The package works independent of sample sizes, experimental designs, assay platforms, and is applicable to both microarray and RNAseq data sets. In this section we will use <a href="https://bioconductor.org/packages/release/bioc/html/gage.html">GAGE</a> and gene sets from the “Gene Ontology” (<a href="http://www.geneontology.org/">GO</a>) and the <a href="https://www.gsea-msigdb.org/gsea/msigdb">MSigDB</a> databases to perform pathway analysis.</p>
<p>Now, we will start a docker session to run the analysis in this section</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb172-1"><a href="module-3.html#cb172-1" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">-v</span> /home/ubuntu/workspace:/workspace cnithin7/bioc-custom:1.0 /bin/bash</span></code></pre></div>
<p>Launch R at the commandline, start RStudio, or launch a posit Cloud session:</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb173-1"><a href="module-3.html#cb173-1" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
</div>
<div id="importing-de-results-for-gage" class="section level4 hasAnchor">
<h4>Importing DE results for gage<a href="module-3.html#importing-de-results-for-gage" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Before we perform the pathway analysis we need to read in our differential expression results from the previous DE analysis.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="module-3.html#cb174-1" tabindex="-1"></a></span>
<span id="cb174-2"><a href="module-3.html#cb174-2" tabindex="-1"></a><span class="co"># Define working dir paths and load in data</span></span>
<span id="cb174-3"><a href="module-3.html#cb174-3" tabindex="-1"></a>datadir <span class="ot">=</span> <span class="st">&quot;/workspace/rnaseq/de/deseq2&quot;</span></span>
<span id="cb174-4"><a href="module-3.html#cb174-4" tabindex="-1"></a><span class="co">#datadir = &quot;/cloud/project/outdir/&quot;</span></span>
<span id="cb174-5"><a href="module-3.html#cb174-5" tabindex="-1"></a></span>
<span id="cb174-6"><a href="module-3.html#cb174-6" tabindex="-1"></a><span class="fu">setwd</span>(datadir)</span>
<span id="cb174-7"><a href="module-3.html#cb174-7" tabindex="-1"></a></span>
<span id="cb174-8"><a href="module-3.html#cb174-8" tabindex="-1"></a><span class="co"># Load in the DE results file with only significant genes (e.g., http://genomedata.org/cri-workshop/deseq2/DE_sig_genes_DESeq2.tsv)</span></span>
<span id="cb174-9"><a href="module-3.html#cb174-9" tabindex="-1"></a>DE_genes <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;DE_sig_genes_DESeq2.tsv&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb174-10"><a href="module-3.html#cb174-10" tabindex="-1"></a></span>
<span id="cb174-11"><a href="module-3.html#cb174-11" tabindex="-1"></a>output_dir <span class="ot">=</span> <span class="st">&quot;/workspace/rnaseq/de/deseq2/pathway/&quot;</span></span>
<span id="cb174-12"><a href="module-3.html#cb174-12" tabindex="-1"></a><span class="co">#output_dir = &quot;/cloud/project/outdir/pathway/&quot;</span></span>
<span id="cb174-13"><a href="module-3.html#cb174-13" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(output_dir)) {</span>
<span id="cb174-14"><a href="module-3.html#cb174-14" tabindex="-1"></a>  <span class="fu">dir.create</span>(output_dir, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb174-15"><a href="module-3.html#cb174-15" tabindex="-1"></a>}</span>
<span id="cb174-16"><a href="module-3.html#cb174-16" tabindex="-1"></a></span>
<span id="cb174-17"><a href="module-3.html#cb174-17" tabindex="-1"></a><span class="fu">setwd</span>(output_dir)</span></code></pre></div>
<p>Now let’s go ahead and load <a href="https://bioconductor.org/packages/release/bioc/html/gage.html">GAGE</a> and some other useful packages.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="module-3.html#cb175-1" tabindex="-1"></a><span class="fu">library</span>(AnnotationDbi)</span>
<span id="cb175-2"><a href="module-3.html#cb175-2" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb175-3"><a href="module-3.html#cb175-3" tabindex="-1"></a><span class="fu">library</span>(GO.db)</span>
<span id="cb175-4"><a href="module-3.html#cb175-4" tabindex="-1"></a><span class="fu">library</span>(gage)</span></code></pre></div>
</div>
<div id="setting-up-gene-set-databases" class="section level4 hasAnchor">
<h4>Setting up gene set databases<a href="module-3.html#setting-up-gene-set-databases" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In order to perform our pathway analysis we need a list of pathways and their respective genes. There are many databases that contain collections of genes (or gene sets) that can be used to understand whether a set of mutated or differentially expressed genes are functionally related. Some of these resources include: <a href="http://www.geneontology.org/">GO</a>, <a href="https://www.kegg.jp">KEGG</a>, <a href="https://www.gsea-msigdb.org/gsea/msigdb">MSigDB</a>, and <a href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>. For this exercise we are going to investigate <a href="http://www.geneontology.org/">GO</a> and <a href="https://www.gsea-msigdb.org/gsea/msigdb">MSigDB</a>. The <a href="https://bioconductor.org/packages/release/bioc/html/gage.html">GAGE</a> package has a function for querying <a href="http://www.geneontology.org/">GO</a> in real time: <a href="https://www.rdocumentation.org/packages/gage/versions/2.22.0/topics/go.gsets">go.gsets()</a>. This function takes a species as an argument and will return a list of gene sets and some helpful meta information for subsetting these lists. If you are unfamiliar with <a href="http://www.geneontology.org/">GO</a>, it is helpful to know that GO terms are categorized into three gene ontologies: “Biological Process”, “Molecular Function”, and “Cellular Component”. This information will come in handy later in our exercise. GAGE does not provide a similar tool to investigate the gene sets available in MSigDB. Fortunately, MSigDB provides a download-able <code>.gmt</code> file for all gene sets. This format is easily read into GAGE using a function called <a href="https://www.rdocumentation.org/packages/gage/versions/2.22.0/topics/readList">readList()</a>. If you check out <a href="https://www.gsea-msigdb.org/gsea/msigdb">MSigDB</a> you will see that there are 8 unique gene set collections, each with slightly different features. For this exercise we will use the <a href="https://www.gsea-msigdb.org/gsea/msigdb/collection_details.jsp#C8">C8 - cell type signature gene sets collection</a>, which is a collection of gene sets that contain cluster markers for cell types identified from single-cell sequencing studies of human tissue.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="module-3.html#cb176-1" tabindex="-1"></a><span class="co"># Set up go database</span></span>
<span id="cb176-2"><a href="module-3.html#cb176-2" tabindex="-1"></a>go.hs <span class="ot">=</span> <span class="fu">go.gsets</span>(<span class="at">species =</span> <span class="st">&quot;human&quot;</span>)</span>
<span id="cb176-3"><a href="module-3.html#cb176-3" tabindex="-1"></a>go.bp.gs <span class="ot">=</span> go.hs<span class="sc">$</span>go.sets[go.hs<span class="sc">$</span>go.subs<span class="sc">$</span>BP]</span>
<span id="cb176-4"><a href="module-3.html#cb176-4" tabindex="-1"></a>go.mf.gs <span class="ot">=</span> go.hs<span class="sc">$</span>go.sets[go.hs<span class="sc">$</span>go.subs<span class="sc">$</span>MF]</span>
<span id="cb176-5"><a href="module-3.html#cb176-5" tabindex="-1"></a>go.cc.gs <span class="ot">=</span> go.hs<span class="sc">$</span>go.sets[go.hs<span class="sc">$</span>go.subs<span class="sc">$</span>CC]</span>
<span id="cb176-6"><a href="module-3.html#cb176-6" tabindex="-1"></a></span>
<span id="cb176-7"><a href="module-3.html#cb176-7" tabindex="-1"></a><span class="co"># Here we will read in an MSigDB gene set that was selected for this exercise and saved to the course website. </span></span>
<span id="cb176-8"><a href="module-3.html#cb176-8" tabindex="-1"></a>c8 <span class="ot">=</span> <span class="st">&quot;http://genomedata.org/rnaseq-tutorial/c8.all.v7.2.entrez.gmt&quot;</span></span>
<span id="cb176-9"><a href="module-3.html#cb176-9" tabindex="-1"></a>all_cell_types <span class="ot">=</span> <span class="fu">readList</span>(c8)</span></code></pre></div>
</div>
<div id="annotating-genes" class="section level4 hasAnchor">
<h4>Annotating genes<a href="module-3.html#annotating-genes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>OK, so we have our differentially expressed genes and we have our gene sets. However, if you look at one of the objects containing the gene sets you’ll notice that each gene set contains a series of integers. These integers are <a href="https://www.ncbi.nlm.nih.gov/gquery/">Entrez</a> gene identifiers. But do we have comparable information in our DE gene list? Right now, no. Our previous results use Ensembl IDs as gene identifiers. We will need to convert our gene identifiers to the format used in the GO and MSigDB gene sets before we can perform the pathway analysis. Fortunately, Bioconductor maintains genome wide annotation data for many species, you can view these species with the <a href="https://bioconductor.org/packages/release/BiocViews.html#___OrgDb">OrgDb bioc view</a>. This makes converting the gene identifiers relatively straightforward, below we use the <a href="https://www.rdocumentation.org/packages/OrganismDbi/versions/1.14.1/topics/MultiDb-class">mapIds()</a> function to query the OrganismDb object for the Entrez id based on the Ensembl id. Because there might not be a one-to-one relationship with these identifiers we also use <code>multiVals="first"</code> to specify that only the first identifier should be returned. Another option would be to use <code>multiVals="asNA"</code> to ignore one-to-many mappings.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="module-3.html#cb177-1" tabindex="-1"></a>DE_genes<span class="sc">$</span>entrez <span class="ot">=</span> <span class="fu">mapIds</span>(org.Hs.eg.db, <span class="at">column =</span> <span class="st">&quot;ENTREZID&quot;</span>, <span class="at">keys =</span> DE_genes<span class="sc">$</span>ensemblID, <span class="at">keytype =</span> <span class="st">&quot;ENSEMBL&quot;</span>, <span class="at">multiVals =</span> <span class="st">&quot;first&quot;</span>)</span></code></pre></div>
</div>
<div id="some-clean-up-and-identifier-mapping" class="section level4 hasAnchor">
<h4>Some clean-up and identifier mapping<a href="module-3.html#some-clean-up-and-identifier-mapping" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>After completing the annotation above you will notice that some of our Ensembl gene IDs were not mapped to an Entrez gene ID. Why did this happen? Well, this is actually a complicated point and gets at some nuanced concepts of how to define and annotate a gene. The short answer is that we are using two different resources that have annotated the human genome and there are some differences in how these resources have completed this task. Therefore, it is expected that there are some discrepencies. In the next few steps we will clean up what we can by first removing the ERCC spike-in genes and then will use a different identifier for futher mapping.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="module-3.html#cb178-1" tabindex="-1"></a><span class="co">#Remove spike-in</span></span>
<span id="cb178-2"><a href="module-3.html#cb178-2" tabindex="-1"></a>DE_genes_clean <span class="ot">=</span> DE_genes[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;ERCC&quot;</span>, DE_genes<span class="sc">$</span>ensemblID), ]</span>
<span id="cb178-3"><a href="module-3.html#cb178-3" tabindex="-1"></a></span>
<span id="cb178-4"><a href="module-3.html#cb178-4" tabindex="-1"></a><span class="do">##Just so we know what we have removed </span></span>
<span id="cb178-5"><a href="module-3.html#cb178-5" tabindex="-1"></a>ERCC_gene_count <span class="ot">=</span> <span class="fu">nrow</span>(DE_genes[<span class="fu">grepl</span>(<span class="st">&quot;ERCC&quot;</span>, DE_genes<span class="sc">$</span>ensemblID), ])</span>
<span id="cb178-6"><a href="module-3.html#cb178-6" tabindex="-1"></a>ERCC_gene_count</span>
<span id="cb178-7"><a href="module-3.html#cb178-7" tabindex="-1"></a></span>
<span id="cb178-8"><a href="module-3.html#cb178-8" tabindex="-1"></a><span class="do">###Deal with genes that we do not have an Entrez ID for </span></span>
<span id="cb178-9"><a href="module-3.html#cb178-9" tabindex="-1"></a>missing_ensembl_key <span class="ot">=</span> DE_genes_clean[<span class="fu">is.na</span>(DE_genes_clean<span class="sc">$</span>entrez), ]</span>
<span id="cb178-10"><a href="module-3.html#cb178-10" tabindex="-1"></a>DE_genes_clean <span class="ot">=</span> DE_genes_clean[<span class="sc">!</span>DE_genes_clean<span class="sc">$</span>ensemblID <span class="sc">%in%</span> missing_ensembl_key<span class="sc">$</span>ensemblID, ]</span>
<span id="cb178-11"><a href="module-3.html#cb178-11" tabindex="-1"></a></span>
<span id="cb178-12"><a href="module-3.html#cb178-12" tabindex="-1"></a><span class="do">###Try mapping using a different key</span></span>
<span id="cb178-13"><a href="module-3.html#cb178-13" tabindex="-1"></a>missing_ensembl_key<span class="sc">$</span>entrez <span class="ot">=</span> <span class="fu">mapIds</span>(org.Hs.eg.db, <span class="at">column =</span> <span class="st">&quot;ENTREZID&quot;</span>, <span class="at">keys =</span> missing_ensembl_key<span class="sc">$</span>Symbol, <span class="at">keytype =</span> <span class="st">&quot;SYMBOL&quot;</span>, <span class="at">multiVal =</span> <span class="st">&quot;first&quot;</span>)</span>
<span id="cb178-14"><a href="module-3.html#cb178-14" tabindex="-1"></a></span>
<span id="cb178-15"><a href="module-3.html#cb178-15" tabindex="-1"></a><span class="co">#Remove remaining genes </span></span>
<span id="cb178-16"><a href="module-3.html#cb178-16" tabindex="-1"></a>missing_ensembl_key_update <span class="ot">=</span> missing_ensembl_key[<span class="sc">!</span><span class="fu">is.na</span>(missing_ensembl_key<span class="sc">$</span>entrez),]</span>
<span id="cb178-17"><a href="module-3.html#cb178-17" tabindex="-1"></a></span>
<span id="cb178-18"><a href="module-3.html#cb178-18" tabindex="-1"></a><span class="co">#Create a Final Gene list of all genes where we were able to find an Entrez ID (using two approaches)</span></span>
<span id="cb178-19"><a href="module-3.html#cb178-19" tabindex="-1"></a>DE_genes_clean <span class="ot">=</span> <span class="fu">rbind</span>(DE_genes_clean, missing_ensembl_key_update)</span>
<span id="cb178-20"><a href="module-3.html#cb178-20" tabindex="-1"></a></span>
<span id="cb178-21"><a href="module-3.html#cb178-21" tabindex="-1"></a><span class="co">#Reduce to only the unique set of entrez genes in case different genes were mapped to the same entrez ID using the two approaches</span></span>
<span id="cb178-22"><a href="module-3.html#cb178-22" tabindex="-1"></a>DE_genes_clean<span class="ot">=</span>DE_genes_clean[<span class="sc">!</span><span class="fu">duplicated</span>(DE_genes_clean<span class="sc">$</span>entrez),]</span></code></pre></div>
</div>
<div id="final-preparation-of-deseq2-results-for-gage" class="section level4 hasAnchor">
<h4>Final preparation of DESeq2 results for gage<a href="module-3.html#final-preparation-of-deseq2-results-for-gage" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>OK, last step. Let’s format the differential expression results into a format suitable for the <a href="">GAGE</a> package. Basically this means obtaining the log2 fold change values and assigning entrez gene identifiers to these values.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="module-3.html#cb179-1" tabindex="-1"></a><span class="co"># grab the log fold changes for everything</span></span>
<span id="cb179-2"><a href="module-3.html#cb179-2" tabindex="-1"></a>De_gene.fc <span class="ot">=</span> DE_genes_clean<span class="sc">$</span>log2FoldChange</span>
<span id="cb179-3"><a href="module-3.html#cb179-3" tabindex="-1"></a></span>
<span id="cb179-4"><a href="module-3.html#cb179-4" tabindex="-1"></a><span class="co"># set the name for each row to be the Entrez Gene ID</span></span>
<span id="cb179-5"><a href="module-3.html#cb179-5" tabindex="-1"></a><span class="fu">names</span>(De_gene.fc) <span class="ot">=</span> DE_genes_clean<span class="sc">$</span>entrez</span></code></pre></div>
</div>
<div id="running-pathway-analysis" class="section level4 hasAnchor">
<h4>Running pathway analysis<a href="module-3.html#running-pathway-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We can now use the <a href="https://www.rdocumentation.org/packages/gage/versions/2.22.0/topics/gage">gage()</a> function to obtain the significantly perturbed pathways from our differential expression experiment.</p>
<p>Note on the abbreviations below: “bp” refers to biological process, “mf” refers to molecular function, and “cc” refers to cellular process. These are the three main categories of gene ontology terms/annotations that were mentioned above.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="module-3.html#cb180-1" tabindex="-1"></a><span class="co">#Run GAGE</span></span>
<span id="cb180-2"><a href="module-3.html#cb180-2" tabindex="-1"></a><span class="co">#go </span></span>
<span id="cb180-3"><a href="module-3.html#cb180-3" tabindex="-1"></a>fc.go.bp.p <span class="ot">=</span> <span class="fu">gage</span>(De_gene.fc, <span class="at">gsets =</span> go.bp.gs)</span>
<span id="cb180-4"><a href="module-3.html#cb180-4" tabindex="-1"></a>fc.go.mf.p <span class="ot">=</span> <span class="fu">gage</span>(De_gene.fc, <span class="at">gsets =</span> go.mf.gs)</span>
<span id="cb180-5"><a href="module-3.html#cb180-5" tabindex="-1"></a>fc.go.cc.p <span class="ot">=</span> <span class="fu">gage</span>(De_gene.fc, <span class="at">gsets =</span> go.cc.gs)</span>
<span id="cb180-6"><a href="module-3.html#cb180-6" tabindex="-1"></a></span>
<span id="cb180-7"><a href="module-3.html#cb180-7" tabindex="-1"></a><span class="co">#msigdb</span></span>
<span id="cb180-8"><a href="module-3.html#cb180-8" tabindex="-1"></a>fc.c8.p <span class="ot">=</span> <span class="fu">gage</span>(De_gene.fc, <span class="at">gsets =</span> all_cell_types)</span>
<span id="cb180-9"><a href="module-3.html#cb180-9" tabindex="-1"></a></span>
<span id="cb180-10"><a href="module-3.html#cb180-10" tabindex="-1"></a><span class="do">###Convert to dataframes </span></span>
<span id="cb180-11"><a href="module-3.html#cb180-11" tabindex="-1"></a><span class="co">#Results for testing for GO terms which are up-regulated</span></span>
<span id="cb180-12"><a href="module-3.html#cb180-12" tabindex="-1"></a>fc.go.bp.p.up <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.bp.p<span class="sc">$</span>greater)</span>
<span id="cb180-13"><a href="module-3.html#cb180-13" tabindex="-1"></a>fc.go.mf.p.up <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.mf.p<span class="sc">$</span>greater)</span>
<span id="cb180-14"><a href="module-3.html#cb180-14" tabindex="-1"></a>fc.go.cc.p.up <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.cc.p<span class="sc">$</span>greater)</span>
<span id="cb180-15"><a href="module-3.html#cb180-15" tabindex="-1"></a></span>
<span id="cb180-16"><a href="module-3.html#cb180-16" tabindex="-1"></a><span class="co">#Results for testing for GO terms which are down-regulated</span></span>
<span id="cb180-17"><a href="module-3.html#cb180-17" tabindex="-1"></a>fc.go.bp.p.down <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.bp.p<span class="sc">$</span>less)</span>
<span id="cb180-18"><a href="module-3.html#cb180-18" tabindex="-1"></a>fc.go.mf.p.down <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.mf.p<span class="sc">$</span>less)</span>
<span id="cb180-19"><a href="module-3.html#cb180-19" tabindex="-1"></a>fc.go.cc.p.down <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.go.cc.p<span class="sc">$</span>less)</span>
<span id="cb180-20"><a href="module-3.html#cb180-20" tabindex="-1"></a></span>
<span id="cb180-21"><a href="module-3.html#cb180-21" tabindex="-1"></a><span class="co">#Results for testing for MSigDB C8 gene sets which are up-regulated</span></span>
<span id="cb180-22"><a href="module-3.html#cb180-22" tabindex="-1"></a>fc.c8.p.up <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.c8.p<span class="sc">$</span>greater)</span>
<span id="cb180-23"><a href="module-3.html#cb180-23" tabindex="-1"></a></span>
<span id="cb180-24"><a href="module-3.html#cb180-24" tabindex="-1"></a><span class="co">#Results for testing for MSigDB C8 gene sets which are down-regulated</span></span>
<span id="cb180-25"><a href="module-3.html#cb180-25" tabindex="-1"></a>fc.c8.p.down <span class="ot">=</span> <span class="fu">as.data.frame</span>(fc.c8.p<span class="sc">$</span>less)</span></code></pre></div>
</div>
<div id="explore-significant-results" class="section level4 hasAnchor">
<h4>Explore significant results<a href="module-3.html#explore-significant-results" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Alright, now we have results with accompanying p-values (yay!).</p>
<p>What does “up-” or “down-regulated” mean here, in the context of our UHR vs HBR comparison? It may help to open and review the data in your DE_genes_DESeq2.tsv file.</p>
<p>Look at the cellular process results from our GO analysis. Do the results match your expectation?</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="module-3.html#cb181-1" tabindex="-1"></a></span>
<span id="cb181-2"><a href="module-3.html#cb181-2" tabindex="-1"></a><span class="co">#Try doing something like this to find some significant results:</span></span>
<span id="cb181-3"><a href="module-3.html#cb181-3" tabindex="-1"></a><span class="co">#View the top 20 significantly up- or down-regulated GO terms from the Cellular Component Ontology</span></span>
<span id="cb181-4"><a href="module-3.html#cb181-4" tabindex="-1"></a><span class="fu">head</span>(fc.go.cc.p.up[<span class="fu">order</span>(fc.go.cc.p.up<span class="sc">$</span>p.val),], <span class="at">n =</span> <span class="dv">20</span>)</span>
<span id="cb181-5"><a href="module-3.html#cb181-5" tabindex="-1"></a><span class="fu">head</span>(fc.go.cc.p.down[<span class="fu">order</span>(fc.go.cc.p.down<span class="sc">$</span>p.val),], <span class="at">n =</span> <span class="dv">20</span>)</span>
<span id="cb181-6"><a href="module-3.html#cb181-6" tabindex="-1"></a></span>
<span id="cb181-7"><a href="module-3.html#cb181-7" tabindex="-1"></a><span class="co">#You can do the same thing with your results from MSigDB</span></span>
<span id="cb181-8"><a href="module-3.html#cb181-8" tabindex="-1"></a><span class="fu">head</span>(fc.c8.p.up)</span>
<span id="cb181-9"><a href="module-3.html#cb181-9" tabindex="-1"></a><span class="fu">head</span>(fc.c8.p.down)</span>
<span id="cb181-10"><a href="module-3.html#cb181-10" tabindex="-1"></a></span>
<span id="cb181-11"><a href="module-3.html#cb181-11" tabindex="-1"></a><span class="co">#What if we want to know which specific genes from our DE gene result were found in a specific significant pathway?</span></span>
<span id="cb181-12"><a href="module-3.html#cb181-12" tabindex="-1"></a><span class="co">#For example, one significant pathway from fc.go.cc.p.down was &quot;GO:0045202 synapse&quot; with set.size = 22 genes.</span></span>
<span id="cb181-13"><a href="module-3.html#cb181-13" tabindex="-1"></a><span class="co">#Let&#39;s extract the postsynapse DE gene results</span></span>
<span id="cb181-14"><a href="module-3.html#cb181-14" tabindex="-1"></a>synapse <span class="ot">=</span> DE_genes_clean[<span class="fu">which</span>(DE_genes_clean<span class="sc">$</span>entrez <span class="sc">%in%</span> go.cc.gs<span class="sc">$</span><span class="st">`</span><span class="at">GO:0045202 synapse</span><span class="st">`</span>),]</span>
<span id="cb181-15"><a href="module-3.html#cb181-15" tabindex="-1"></a></span>
<span id="cb181-16"><a href="module-3.html#cb181-16" tabindex="-1"></a><span class="co">#How many total synapse genes are there in GO? How many total DE genes? How many overlap?</span></span>
<span id="cb181-17"><a href="module-3.html#cb181-17" tabindex="-1"></a><span class="fu">length</span>(go.cc.gs<span class="sc">$</span><span class="st">`</span><span class="at">GO:0045202 synapse</span><span class="st">`</span>)</span>
<span id="cb181-18"><a href="module-3.html#cb181-18" tabindex="-1"></a><span class="fu">length</span>(DE_genes_clean<span class="sc">$</span>entrez)</span>
<span id="cb181-19"><a href="module-3.html#cb181-19" tabindex="-1"></a><span class="fu">length</span>(synapse<span class="sc">$</span>entrez)</span>
<span id="cb181-20"><a href="module-3.html#cb181-20" tabindex="-1"></a></span>
<span id="cb181-21"><a href="module-3.html#cb181-21" tabindex="-1"></a><span class="co">#Are the synapse DE genes consistently down-regulated? Let&#39;s print out a subset of columns from the DE result for synapse genes</span></span>
<span id="cb181-22"><a href="module-3.html#cb181-22" tabindex="-1"></a>synapse[,<span class="fu">c</span>(<span class="st">&quot;Symbol&quot;</span>, <span class="st">&quot;entrez&quot;</span>, <span class="st">&quot;log2FoldChange&quot;</span>, <span class="st">&quot;padj&quot;</span>, <span class="st">&quot;UHR_Rep1&quot;</span>, <span class="st">&quot;UHR_Rep2&quot;</span>, <span class="st">&quot;UHR_Rep3&quot;</span>, <span class="st">&quot;HBR_Rep1&quot;</span>, <span class="st">&quot;HBR_Rep2&quot;</span>, <span class="st">&quot;HBR_Rep3&quot;</span>)]</span></code></pre></div>
</div>
<div id="more-exploration" class="section level4 hasAnchor">
<h4>More exploration<a href="module-3.html#more-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>At this point, it will be helpful to move out of R and further explore our results locally. We will use an online tool to visualize how the GO terms we uncovered are related to each other.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="module-3.html#cb182-1" tabindex="-1"></a><span class="fu">write.table</span>(fc.go.cc.p.up, <span class="st">&quot;fc.go.cc.p.up.tsv&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb182-2"><a href="module-3.html#cb182-2" tabindex="-1"></a><span class="fu">write.table</span>(fc.go.cc.p.down, <span class="st">&quot;fc.go.cc.p.down.tsv&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="visualize" class="section level4 hasAnchor">
<h4>Visualize<a href="module-3.html#visualize" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For this next step we will do a very brief introduction to visualizing our results. We will use a tool called <a href="https://2019.webgestalt.org/2017/GOView/">GOView</a>, which is part of the <a href="http://www.webgestalt.org/">WEB-based Gene Set Ananlysis ToolKit (WebGestalt)</a> suite of tools.</p>
<p><strong>Step One</strong>
* Use a web browser to download your results</p>
<ul>
<li><p>For AWS: Navigate to the URL below replacing YOUR_IP_ADDRESS with your amazon instance IP address:
<a href="http://" class="uri">http://</a><strong>YOUR_IP_ADDRESS</strong>/rnaseq/de/deseq2</p></li>
<li><p>Download the linked files by right clicking on the two saved result files: <code>fc.go.cc.p.up.tsv</code> and <code>fc.go.cc.p.down.tsv</code>.</p></li>
<li><p>For posit Cloud: Navigate to the <code>outdir</code> folder in the ‘Files’ pane. Select <code>fc.go.cc.p.up.tsv</code> and <code>fc.go.cc.p.down.tsv</code> then ‘More’ -&gt; ‘Export…’. You may need to unzip the downloaded files.</p></li>
<li><p>Open the result file in your text editor of choice. We like <a href="https://www.barebones.com/products/textwrangler/">text wrangler</a>.
You should also be able to open the file in excel, google sheets, or another spreadsheet tool. This might help you visualize the data in rows and columns (NB: There might be a small amount of formatting necessary to get the header to line up properly).</p></li>
<li><p>You can either create an input file using <a href="http://genomedata.org/rnaseq-tutorial/fc.go.cc.p.down_web_ges.tsv">this file</a> as a guide, or you can simply use your downloaded data to cut and paste your GO terms of interest directly into GOView.</p></li>
</ul>
<p><strong>Step Two</strong></p>
<ul>
<li>Navigate to the <a href="http://www.webgestalt.org/">WEB-based Gene Set Analysis ToolKit (WebGestalt)</a></li>
</ul>
<p><strong>Step Three</strong></p>
<ul>
<li><p>Navigate to the <a href="http://www.webgestalt.org/2017/GOView/">GOView</a> tool</p></li>
<li><p>Then, input the GO terms you would like to explore into the <a href="http://www.webgestalt.org/2017/GOView/">GOView</a> interface by following the steps described in the “Beginning an analysis” section of the webpage. We will walk through a sample analysis.</p></li>
<li><p>Explore the outputs!</p></li>
</ul>
</div>
<div id="alternative-pathway-analysis-tools-and-strategies" class="section level4 hasAnchor">
<h4>Alternative Pathway analysis tools and strategies!<a href="module-3.html#alternative-pathway-analysis-tools-and-strategies" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>At this point, to generate additional visualizations and gain more insight from our enrichment analysis, we can turn to two powerful functions in the clusterProfiler package: gseGO and enrichKEGG. These tools allow us to go beyond simply identifying enriched pathways and give us the ability to visualize and explore these pathways in a much more detailed way.</p>
<p>First, to use gseGO and enrichKEGG effectively, we need a ranked list of DE genes based on their log2 fold change values. This ranked list allows us to analyze and visualize enrichment based on the degree of differential expression, rather than just a binary presence or absence in a pathway.</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="module-3.html#cb183-1" tabindex="-1"></a><span class="co"># Create a named vector of log2 fold changes with Entrez IDs as names</span></span>
<span id="cb183-2"><a href="module-3.html#cb183-2" tabindex="-1"></a>ranked_genes <span class="ot">&lt;-</span> <span class="fu">setNames</span>(DE_genes_clean<span class="sc">$</span>log2FoldChange, DE_genes_clean<span class="sc">$</span>entrez)</span>
<span id="cb183-3"><a href="module-3.html#cb183-3" tabindex="-1"></a></span>
<span id="cb183-4"><a href="module-3.html#cb183-4" tabindex="-1"></a><span class="co"># Sort the genes by log2 fold change</span></span>
<span id="cb183-5"><a href="module-3.html#cb183-5" tabindex="-1"></a>ranked_genes <span class="ot">&lt;-</span> <span class="fu">sort</span>(ranked_genes, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Using gseGO, we can analyze the ranked DE genes list to create classic GSEA enrichment plots. These plots help us see how gene expression levels are distributed across the pathways that were identified as significant in our initial analysis. By examining the ranking of gene expression within these pathways, we can get a clearer picture of how specific pathways are activated or suppressed in our data set.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="module-3.html#cb184-1" tabindex="-1"></a><span class="co"># Install a package missing from the template</span></span>
<span id="cb184-2"><a href="module-3.html#cb184-2" tabindex="-1"></a><span class="co">#if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))</span></span>
<span id="cb184-3"><a href="module-3.html#cb184-3" tabindex="-1"></a><span class="co">#    install.packages(&quot;BiocManager&quot;, quiet = TRUE)</span></span>
<span id="cb184-4"><a href="module-3.html#cb184-4" tabindex="-1"></a><span class="co">#BiocManager::install(&quot;pathview&quot;, update = FALSE, ask = FALSE, quiet = TRUE)</span></span>
<span id="cb184-5"><a href="module-3.html#cb184-5" tabindex="-1"></a></span>
<span id="cb184-6"><a href="module-3.html#cb184-6" tabindex="-1"></a><span class="co"># Load relevant packages</span></span>
<span id="cb184-7"><a href="module-3.html#cb184-7" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span>
<span id="cb184-8"><a href="module-3.html#cb184-8" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb184-9"><a href="module-3.html#cb184-9" tabindex="-1"></a><span class="fu">library</span>(pathview)</span>
<span id="cb184-10"><a href="module-3.html#cb184-10" tabindex="-1"></a><span class="fu">library</span>(ggnewscale)</span>
<span id="cb184-11"><a href="module-3.html#cb184-11" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb184-12"><a href="module-3.html#cb184-12" tabindex="-1"></a></span>
<span id="cb184-13"><a href="module-3.html#cb184-13" tabindex="-1"></a>gsea_res <span class="ot">&lt;-</span> <span class="fu">gseGO</span>(</span>
<span id="cb184-14"><a href="module-3.html#cb184-14" tabindex="-1"></a>  <span class="at">geneList =</span> ranked_genes,       <span class="co"># Ranked list of genes</span></span>
<span id="cb184-15"><a href="module-3.html#cb184-15" tabindex="-1"></a>  <span class="at">OrgDb =</span> org.Hs.eg.db,          <span class="co"># Specify organism database</span></span>
<span id="cb184-16"><a href="module-3.html#cb184-16" tabindex="-1"></a>  <span class="at">ont =</span> <span class="st">&quot;CC&quot;</span>,                    <span class="co"># Use &quot;BP&quot; for Biological Process, &quot;MF&quot; for Molecular Function, &quot;CC&quot; for Cellular Component</span></span>
<span id="cb184-17"><a href="module-3.html#cb184-17" tabindex="-1"></a>  <span class="at">keyType =</span> <span class="st">&quot;ENTREZID&quot;</span>,          <span class="co"># Ensure your gene IDs match the key type in OrgDb</span></span>
<span id="cb184-18"><a href="module-3.html#cb184-18" tabindex="-1"></a>  <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>,           <span class="co"># Set a p-value cutoff for significant pathways</span></span>
<span id="cb184-19"><a href="module-3.html#cb184-19" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb184-20"><a href="module-3.html#cb184-20" tabindex="-1"></a>)</span></code></pre></div>
<p>Generate the classic GSEA enrichment plot</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="module-3.html#cb185-1" tabindex="-1"></a><span class="co"># Plot the enrichment plot for a specific GO term or pathway - Synapse</span></span>
<span id="cb185-2"><a href="module-3.html#cb185-2" tabindex="-1"></a>gsea_plot <span class="ot">&lt;-</span> <span class="fu">gseaplot2</span>(gsea_res, <span class="at">geneSetID =</span> <span class="st">&quot;GO:0045202&quot;</span>, <span class="at">title =</span> <span class="st">&quot;Enrichment Plot for Synapse&quot;</span>)</span>
<span id="cb185-3"><a href="module-3.html#cb185-3" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;gsea_go_synapse_plot.pdf&quot;</span>, <span class="at">plot=</span>gsea_plot, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">8</span>)</span></code></pre></div>
<p>We can use additional visualizations, such as dot plots, ridge plots, and concept network plots, to gain further insights into the enriched pathways.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="module-3.html#cb186-1" tabindex="-1"></a><span class="co"># Dotplot for top GO pathways enriched with DE genes</span></span>
<span id="cb186-2"><a href="module-3.html#cb186-2" tabindex="-1"></a>gsea_dot_plot <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(gsea_res, <span class="at">showCategory =</span> <span class="dv">30</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;GSEA Dotplot - Top 30 GO Categories&quot;</span>)</span>
<span id="cb186-3"><a href="module-3.html#cb186-3" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;gsea_dot_plot.pdf&quot;</span>, <span class="at">plot=</span>gsea_dot_plot, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb186-4"><a href="module-3.html#cb186-4" tabindex="-1"></a></span>
<span id="cb186-5"><a href="module-3.html#cb186-5" tabindex="-1"></a><span class="co">#Ridgeplot for top GO pathways enriched with DE genes</span></span>
<span id="cb186-6"><a href="module-3.html#cb186-6" tabindex="-1"></a>gsea_ridge_plot <span class="ot">&lt;-</span><span class="fu">ridgeplot</span>(gsea_res)</span>
<span id="cb186-7"><a href="module-3.html#cb186-7" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;gsea_ridge_plot.pdf&quot;</span>, <span class="at">plot=</span>gsea_ridge_plot, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">8</span>)</span>
<span id="cb186-8"><a href="module-3.html#cb186-8" tabindex="-1"></a></span>
<span id="cb186-9"><a href="module-3.html#cb186-9" tabindex="-1"></a><span class="co"># Concept network plot to illustrate relationships between the top enriched GO terms and DE genes</span></span>
<span id="cb186-10"><a href="module-3.html#cb186-10" tabindex="-1"></a>gsea_cnetplot <span class="ot">&lt;-</span> <span class="fu">cnetplot</span>(gsea_res, <span class="at">foldChange =</span> ranked_genes, <span class="at">showCategory =</span> <span class="dv">10</span>)</span>
<span id="cb186-11"><a href="module-3.html#cb186-11" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;gsea_cnetplot.pdf&quot;</span>, <span class="at">plot=</span>gsea_cnetplot, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>)</span></code></pre></div>
<p>The enrichKEGG function can be used to visualize KEGG pathways, showing detailed diagrams with our DE genes highlighted. This approach is especially useful for understanding the biological roles of up- and down-regulated genes within specific metabolic or signaling pathways. By using the pathview package, we can generate pathway diagrams where each DE gene is displayed in its functional context and color-coded by expression level. This makes it easy to see which parts of a pathway are impacted and highlights any potential regulatory or metabolic shifts in a clear, intuitive format. We will start by downloading and installing the KEGG database and then run the <code>enrichKEGG</code> function.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="module-3.html#cb187-1" tabindex="-1"></a><span class="co"># Download KEGG DB file and install</span></span>
<span id="cb187-2"><a href="module-3.html#cb187-2" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">&#39;https://www.bioconductor.org/packages/3.11/data/annotation/src/contrib/KEGG.db_3.2.4.tar.gz&#39;</span>, <span class="at">destfile=</span><span class="st">&#39;KEGG.db_3.2.4.tar.gz&#39;</span>)</span>
<span id="cb187-3"><a href="module-3.html#cb187-3" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;KEGG.db_3.2.4.tar.gz&quot;</span>, <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span>
<span id="cb187-4"><a href="module-3.html#cb187-4" tabindex="-1"></a></span>
<span id="cb187-5"><a href="module-3.html#cb187-5" tabindex="-1"></a><span class="co"># Run enrichKEGG with local database option</span></span>
<span id="cb187-6"><a href="module-3.html#cb187-6" tabindex="-1"></a>pathways <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene =</span> <span class="fu">names</span>(ranked_genes), <span class="at">organism =</span> <span class="st">&quot;hsa&quot;</span>, <span class="at">keyType =</span> <span class="st">&quot;kegg&quot;</span>, <span class="at">use_internal_data=</span><span class="cn">TRUE</span>)</span>
<span id="cb187-7"><a href="module-3.html#cb187-7" tabindex="-1"></a><span class="fu">head</span>(pathways<span class="sc">@</span>result)</span></code></pre></div>
<p>Let’s choose one of the pathways above, for example <code>hsa04010 - MAPK signaling pathway</code> to visualize.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="module-3.html#cb188-1" tabindex="-1"></a><span class="co"># Define the KEGG pathway ID based on above, and run pathview (note this automatically generates and saves plots to your current directory)</span></span>
<span id="cb188-2"><a href="module-3.html#cb188-2" tabindex="-1"></a>pathway_id <span class="ot">&lt;-</span> <span class="st">&quot;hsa04010&quot;</span>  <span class="co"># Replace with the KEGG pathway ID of interest</span></span>
<span id="cb188-3"><a href="module-3.html#cb188-3" tabindex="-1"></a><span class="fu">pathview</span>(</span>
<span id="cb188-4"><a href="module-3.html#cb188-4" tabindex="-1"></a>  <span class="at">gene.data =</span> ranked_genes,    <span class="co"># DE gene data with Entrez IDs</span></span>
<span id="cb188-5"><a href="module-3.html#cb188-5" tabindex="-1"></a>  <span class="at">pathway.id =</span> pathway_id,  <span class="co"># KEGG pathway ID</span></span>
<span id="cb188-6"><a href="module-3.html#cb188-6" tabindex="-1"></a>  <span class="at">species =</span> <span class="st">&quot;hsa&quot;</span>,          <span class="co"># Species code for human</span></span>
<span id="cb188-7"><a href="module-3.html#cb188-7" tabindex="-1"></a>  <span class="at">limit =</span> <span class="fu">list</span>(<span class="at">gene =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>)),  <span class="co"># Set color scale limits for log2 fold changes</span></span>
<span id="cb188-8"><a href="module-3.html#cb188-8" tabindex="-1"></a>  <span class="at">low =</span> <span class="st">&quot;blue&quot;</span>,             <span class="co"># Color for down-regulated genes</span></span>
<span id="cb188-9"><a href="module-3.html#cb188-9" tabindex="-1"></a>  <span class="at">mid =</span> <span class="st">&quot;white&quot;</span>,            <span class="co"># Color for neutral genes</span></span>
<span id="cb188-10"><a href="module-3.html#cb188-10" tabindex="-1"></a>  <span class="at">high =</span> <span class="st">&quot;red&quot;</span>              <span class="co"># Color for up-regulated genes</span></span>
<span id="cb188-11"><a href="module-3.html#cb188-11" tabindex="-1"></a>)</span>
<span id="cb188-12"><a href="module-3.html#cb188-12" tabindex="-1"></a></span>
<span id="cb188-13"><a href="module-3.html#cb188-13" tabindex="-1"></a><span class="co">#To exit R type:</span></span>
<span id="cb188-14"><a href="module-3.html#cb188-14" tabindex="-1"></a><span class="fu">quit</span>(<span class="at">save =</span> <span class="st">&quot;no&quot;</span>)</span></code></pre></div>
<p>Leave the docker session</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb189-1"><a href="module-3.html#cb189-1" tabindex="-1"></a><span class="bu">exit</span></span></code></pre></div>
<p>In this section we will obtain a dataset to allow demonstration of batch correction using the ComBat-Seq tool in R (Bioconductor).</p>
<div id="download-and-prepare-some-test-data-where-some-batch-effects-are-expected" class="section level5 hasAnchor">
<h5>Download and prepare some test data where some batch effects are expected<a href="module-3.html#download-and-prepare-some-test-data-where-some-batch-effects-are-expected" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>For this exercise we will obtain public RNA-seq data from an extensive multi-platform comparison of sequencing platforms that also examined the impact of generating data at multiple sites, using polyA vs ribo-reduction for enrichment, and the impact of RNA degradation (<a href="https://pubmed.ncbi.nlm.nih.gov/25150835/">PMID: 25150835</a>): “Multi-platform and cross-methodological reproducibility of transcriptome profiling by RNA-seq in the ABRF Next-Generation Sequencing Study”.</p>
<p>This publication used the same UHR (cancer cell lines) and HBR (brain tissue) samples we have been using throughout this course. To examine a strong batch effect, we will consider a DE analysis of UHR vs HBR where we compare Ribo-depleted (“Ribo”) and polyA-enriched (“Poly”) samples.</p>
<p>The entire RNA-seq dataset <a href="https://pubmed.ncbi.nlm.nih.gov/25150835/">PMID: 25150835</a> used for this module has been deposited in GEO. In GEO, these data are organized as a <code>superseries</code>: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE46876">GSE46876</a> which has data for several sequencing platforms. The data from the Illumina Platform are part of this <code>subseries</code>: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48035">GSE48035</a>.</p>
<p>To do this analysis quickly, we will download pre-computed raw read counts for this dataset: <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE48035&amp;format=file&amp;file=GSE48035%5FILMN%2Ecounts%2Etxt%2Egz">GSE48035_ILMN.counts.txt.gz</a></p>
<p>Set up a working directory and download the RNA-seq counts file needed for the following exercise as follows:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb190-1"><a href="module-3.html#cb190-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span></span>
<span id="cb190-2"><a href="module-3.html#cb190-2" tabindex="-1"></a><span class="fu">mkdir</span> batch_correction</span>
<span id="cb190-3"><a href="module-3.html#cb190-3" tabindex="-1"></a><span class="bu">cd</span> batch_correction</span>
<span id="cb190-4"><a href="module-3.html#cb190-4" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/batch_correction/GSE48035_ILMN.counts.txt.gz</span></code></pre></div>
<p>Create a simplified version of this file that has only the counts for the samples we wish to use for this analysis as follows:</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb191-1"><a href="module-3.html#cb191-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/batch_correction</span>
<span id="cb191-2"><a href="module-3.html#cb191-2" tabindex="-1"></a></span>
<span id="cb191-3"><a href="module-3.html#cb191-3" tabindex="-1"></a><span class="co">#remove all quotes from file</span></span>
<span id="cb191-4"><a href="module-3.html#cb191-4" tabindex="-1"></a><span class="fu">zcat</span> GSE48035_ILMN.counts.txt.gz <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="st">&#39;&quot;&#39;</span> <span class="op">&gt;</span> GSE48035_ILMN.counts.tmp.txt</span>
<span id="cb191-5"><a href="module-3.html#cb191-5" tabindex="-1"></a></span>
<span id="cb191-6"><a href="module-3.html#cb191-6" tabindex="-1"></a><span class="co">#create a fixed version of the header and store for later</span></span>
<span id="cb191-7"><a href="module-3.html#cb191-7" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 1 GSE48035_ILMN.counts.tmp.txt <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;print &quot;Gene\tChr\t$_&quot;&#39;</span> <span class="op">&gt;</span> header.txt</span>
<span id="cb191-8"><a href="module-3.html#cb191-8" tabindex="-1"></a></span>
<span id="cb191-9"><a href="module-3.html#cb191-9" tabindex="-1"></a><span class="co">#split the chromosome and gene names on each line, sort the file by gene name</span></span>
<span id="cb191-10"><a href="module-3.html#cb191-10" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp; if ($_ =~ /^(chr\w+)\!(\S+)(.*)/){print &quot;$2\t$1$3\n&quot;}else{print &quot;$_\n&quot;}&#39;</span> GSE48035_ILMN.counts.tmp.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="op">&gt;</span> GSE48035_ILMN.counts.tmp2.txt</span>
<span id="cb191-11"><a href="module-3.html#cb191-11" tabindex="-1"></a></span>
<span id="cb191-12"><a href="module-3.html#cb191-12" tabindex="-1"></a><span class="co">#remove the old header line</span></span>
<span id="cb191-13"><a href="module-3.html#cb191-13" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="at">--color</span><span class="op">=</span>never ABRF GSE48035_ILMN.counts.tmp2.txt <span class="op">&gt;</span> GSE48035_ILMN.counts.clean.txt</span>
<span id="cb191-14"><a href="module-3.html#cb191-14" tabindex="-1"></a></span>
<span id="cb191-15"><a href="module-3.html#cb191-15" tabindex="-1"></a><span class="co">#cut out only the columns for the UHR (A) and HBR (B) samples, replicates 1-4, and PolyA vs Enrichment </span></span>
<span id="cb191-16"><a href="module-3.html#cb191-16" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1-2,3-6,7-10,19-22,23-26 GSE48035_ILMN.counts.clean.txt <span class="op">&gt;</span> GSE48035_ILMN.Counts.SampleSubset.txt</span>
<span id="cb191-17"><a href="module-3.html#cb191-17" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1-2,3-6,7-10,19-22,23-26 header.txt <span class="op">&gt;</span> header.SampleSubset.txt</span>
<span id="cb191-18"><a href="module-3.html#cb191-18" tabindex="-1"></a></span>
<span id="cb191-19"><a href="module-3.html#cb191-19" tabindex="-1"></a><span class="co">#how many gene lines are we starting with?</span></span>
<span id="cb191-20"><a href="module-3.html#cb191-20" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> GSE48035_ILMN.Counts.SampleSubset.txt</span>
<span id="cb191-21"><a href="module-3.html#cb191-21" tabindex="-1"></a></span>
<span id="cb191-22"><a href="module-3.html#cb191-22" tabindex="-1"></a><span class="co">#cleanup intermediate files created above</span></span>
<span id="cb191-23"><a href="module-3.html#cb191-23" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-f</span> GSE48035_ILMN.counts.txt.gz GSE48035_ILMN.counts.tmp.txt GSE48035_ILMN.counts.tmp2.txt GSE48035_ILMN.counts.clean.txt header.txt</span></code></pre></div>
<p>Further limit these counts to those that correspond to <strong>known protein coding genes</strong>:</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb192-1"><a href="module-3.html#cb192-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/batch_correction</span>
<span id="cb192-2"><a href="module-3.html#cb192-2" tabindex="-1"></a></span>
<span id="cb192-3"><a href="module-3.html#cb192-3" tabindex="-1"></a><span class="co">#download complete Ensembl GTF file</span></span>
<span id="cb192-4"><a href="module-3.html#cb192-4" tabindex="-1"></a><span class="fu">wget</span> ftp://ftp.ensembl.org/pub/release-101/gtf/homo_sapiens/Homo_sapiens.GRCh38.101.gtf.gz</span>
<span id="cb192-5"><a href="module-3.html#cb192-5" tabindex="-1"></a></span>
<span id="cb192-6"><a href="module-3.html#cb192-6" tabindex="-1"></a><span class="co">#grab all the gene records, limit to gene with &quot;protein_coding&quot; biotype, create unique gene name list</span></span>
<span id="cb192-7"><a href="module-3.html#cb192-7" tabindex="-1"></a><span class="fu">zcat</span> Homo_sapiens.GRCh38.101.gtf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-w</span> gene <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;gene_biotype </span><span class="dt">\&quot;</span><span class="st">protein_coding</span><span class="dt">\&quot;</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f</span> 9 <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">&quot;;&quot;</span> <span class="at">-f</span> 3 <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="st">&quot; gene_name &quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="st">&#39;&quot;&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> Ensembl101_ProteinCodingGeneNames.txt</span>
<span id="cb192-8"><a href="module-3.html#cb192-8" tabindex="-1"></a></span>
<span id="cb192-9"><a href="module-3.html#cb192-9" tabindex="-1"></a><span class="co">#how many unique protein coding genes names does Ensembl have?</span></span>
<span id="cb192-10"><a href="module-3.html#cb192-10" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> Ensembl101_ProteinCodingGeneNames.txt</span>
<span id="cb192-11"><a href="module-3.html#cb192-11" tabindex="-1"></a></span>
<span id="cb192-12"><a href="module-3.html#cb192-12" tabindex="-1"></a><span class="co">#filter our gene count matrix down to only the protein coding genes</span></span>
<span id="cb192-13"><a href="module-3.html#cb192-13" tabindex="-1"></a><span class="fu">join</span> <span class="at">-j</span> 1 <span class="at">-t</span> <span class="st">$&#39;</span><span class="dt">\t</span><span class="st">&#39;</span> Ensembl101_ProteinCodingGeneNames.txt GSE48035_ILMN.Counts.SampleSubset.txt <span class="kw">|</span> <span class="fu">cat</span> header.SampleSubset.txt <span class="at">-</span> <span class="op">&gt;</span> GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv</span>
<span id="cb192-14"><a href="module-3.html#cb192-14" tabindex="-1"></a></span>
<span id="cb192-15"><a href="module-3.html#cb192-15" tabindex="-1"></a><span class="co">#how many lines of RNA-seq counts do we still have?</span></span>
<span id="cb192-16"><a href="module-3.html#cb192-16" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv</span>
<span id="cb192-17"><a href="module-3.html#cb192-17" tabindex="-1"></a></span>
<span id="cb192-18"><a href="module-3.html#cb192-18" tabindex="-1"></a><span class="co">#clean up </span></span>
<span id="cb192-19"><a href="module-3.html#cb192-19" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-f</span> header.SampleSubset.txt GSE48035_ILMN.Counts.SampleSubset.txt</span>
<span id="cb192-20"><a href="module-3.html#cb192-20" tabindex="-1"></a></span>
<span id="cb192-21"><a href="module-3.html#cb192-21" tabindex="-1"></a><span class="co">#take a look at the final filtered read count matrix to be used for the following analysis</span></span>
<span id="cb192-22"><a href="module-3.html#cb192-22" tabindex="-1"></a><span class="ex">column</span> <span class="at">-t</span> GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code></pre></div>
<p>Note that filtering gene lists by gene name as we have done above is generally not advised as we usually can’t guarantee that gene names from two different lists are compatible. Mapping between unique identifiers would be preferable. But for demonstrating the batch analysis below this should be fine…</p>
<p>In this section we will use the ComBat-Seq tool in R (Bioconductor) to demonstrate the principles and application of batch correction. Due to the way our test data was generated (at a single center, at one time, with consistent methodology) we do NOT expect batch effects in these data. Therefore we will use a different (but highly related) dataset to demonstrate the impact of Batch correction in this module.</p>
</div>
</div>
<div id="introduction-to-batch-correction" class="section level4 hasAnchor">
<h4>Introduction to batch correction<a href="module-3.html#introduction-to-batch-correction" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We highly recommend reading the entire <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7518324/">ComBat-Seq manuscript</a> by Yuqing Zhang, Giovanni Parmigiani, and W Evan Johnson. This manuscript does a beautiful job of briefly introducing the concept of batch correction and the differences between <code>normalization</code> and <code>batch correction</code>. If you find this exercise helpful in your research, please cite the ComBat-Seq paper (<a href="https://pubmed.ncbi.nlm.nih.gov/33015620/">PMID: 33015620</a>).</p>
<p>In particular, this excerpt covers the basics:
&gt; Genomic data are often produced in batches due to logistical or practical restrictions, but technical variation and differences across batches, often called batch effects, can cause significant heterogeneity across batches of data. Batch effects often result in discrepancies in the statistical distributions across data from different technical processing batches, and can have unfavorable impact on downstream biological analysis … Batch effects often cannot be fully addressed by normalization methods and procedures. The differences in the overall expression distribution of each sample across batch may be corrected by normalization methods, such as transforming the raw counts to (logarithms of) CPM, TPM or RPKM/FPKM, the trimmed mean of M values (TMM), or relative log expression (RLE). However, batch effects in composition, i.e. the level of expression of genes scaled by the total expression (coverage) in each sample, cannot be fully corrected with normalization. … While the overall distribution of samples may be normalized to the same level across batches, individual genes may still be affected by batch-level bias.</p>
</div>
<div id="introduction-to-this-demonstration-of-a-batch-correction-approach" class="section level4 hasAnchor">
<h4>Introduction to this demonstration of a batch correction approach<a href="module-3.html#introduction-to-this-demonstration-of-a-batch-correction-approach" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For this exercise we have obtained public RNA-seq data from an extensive multi-platform comparison of sequencing platforms that also examined the impact of: (1) generating data at multiple sites, (2) using polyA vs ribo-reduction for enrichment, and (3) different levels of RNA degradation (<a href="https://pubmed.ncbi.nlm.nih.gov/25150835/">PMID: 25150835</a>): “Multi-platform and cross-methodological reproducibility of transcriptome profiling by RNA-seq in the ABRF Next-Generation Sequencing Study”.</p>
<p>This publication used the same UHR (cancer cell lines) and HBR (brain tissue) samples we have been using throughout this course. To examine a strong batch effect, we will consider a DE analysis of UHR vs HBR where we compare Ribo-depleted (“Ribo”) and polyA-enriched (“Poly”) samples.</p>
<p><strong>Questions</strong>
If we do DE analysis of UHR vs. HBR for replicates that are consistent with respect to Ribo-depletion or PolyA-enrichment, how does the result compare to when we mix the Ribo-depleted data and PolyA-enriched data together?</p>
<p>If you do batch correction and redo these comparisons does it make the results more comparable? i.e. can we correct for the technical differences introduced by the library construction approach and see the same biological differences? Can we improve our statistical power by then benefitting from more samples?</p>
<p>This is a bit contrived because there really are true biological differences expected for polyA vs. Ribo-depleted data. To counter this, we will limit the analysis to only know protein coding genes. For those genes we expect essentially the same biological answer when comparing UHR vs HBR expression patterns.</p>
<p>This exercise is also a bit simplistic in the sense that we have perfectly balanced conditions and batches. Our conditions of interest are: HBR (brain) vs. UHR (cancer cell line) expression patterns. Our batches are the two methods of processing: Riboreduction and PolyA enrichment. And we have 4 replicates of both conditions in both batches. To perform this kind of batch correction you need at least some representation of each of your conditions of interest in each batch. So, for example, if we processed all the HBR samples with Riboreduction and all the UHR samples with PolyA enrichment, we would be unable to model the batch effect vs the condition effect.</p>
<p>There are also other experiments from this published dataset we could use instead. For example, different levels of degradation?, data generated by different labs? <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4167418/">Figure 1</a> and <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4167418/">Figure 2</a> give a nice high level summary of all the data generated. We chose the Ribo-reduction and PolyA enrichment data for this exercise because we expect this would introduce a very strong batch effect. It is also the kind of thing that one could imagine really coming up in a meta-analysis where one you are pooling data from multiple studies. For example, imagine we find three published studies from three labs that all assayed some number of normal breast tissue and breast tumor. Each used a different approach to generate their data but they all involved this same biological comparison and by combining the three datasets we hope to substantially increase our power. If we can correct for batch effects arising from the three labs, this may be successful.</p>
</div>
<div id="samples-abbreviations-used-in-the-following-analysis" class="section level4 hasAnchor">
<h4>Samples abbreviations used in the following analysis<a href="module-3.html#samples-abbreviations-used-in-the-following-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>HBR -&gt; Human Brain Reference, Biological condition (pool of adult brain tissues)</li>
<li>UHR -&gt; Universal Human Reference, Biological condition (pool of cancer cell lines)</li>
<li>Ribo -&gt; Library preparation method using ribosomal reduction, Batch group</li>
<li>Poly -&gt; Library preparation method using polyA enrichment, Batch group</li>
<li>1-4 -&gt; Replicate number: 1, 2, 3, 4.</li>
</ul>
</div>
<div id="perform-principal-component-analysis-pca-on-the-uncorrected-counts" class="section level4 hasAnchor">
<h4>Perform principal component analysis (PCA) on the uncorrected counts<a href="module-3.html#perform-principal-component-analysis-pca-on-the-uncorrected-counts" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>PCA analysis can be used to identify potential batch effects in your data. The general strategy is to use PCA to identify patterns of similarity/difference in the expression signatures of your samples and to ask whether it appears to be driven by the expected biological conditions of interest. The PCA plot can be labeled with the biological conditions and also with potential sources of batch effects such as: sequencing source, date of data generation, lab technician, library construction kit batches, matrigel batches, mouse litters, software or instrumentation versions, etc.</p>
<p><a href="https://en.wikipedia.org/wiki/Principal_component_analysis">Principal component analysis</a> is a <a href="https://en.wikipedia.org/wiki/Dimensionality_reduction">dimensionality-reduction</a> method that can be applied to large datasets (e.g. thousands of gene expression values for many samples). PCA tries to represent a large set of variables as a smaller set of variables that maximally capture the information content of the larger set. PCA is a general exploratory data analysis approach with many applications and nuances, the details of which are beyond the scope of this demonstration of batch effect correction. However, in the context of this module, PCA provides a way to visualize samples as “clusters” based on their overall pattern of gene expression values. The composition and arrangement of these clusters (usually visualized in 2D or interactive 3D plots) can be helpful in interpreting high level differences between samples and testing prior expectations about the similarity between conditions, replicates, etc.</p>
<p>We will perform PCA analysis before AND after batch correction. Samples will be labelled according to biological condition (UHR vs HBR) and library preparation type (Ribo vs PolyA).</p>
<p>Does the analysis below suggest that sample grouping according to PCA is being influenced by batch?</p>
<p>Perform the following analyses in <code>R</code>:</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="module-3.html#cb193-1" tabindex="-1"></a></span>
<span id="cb193-2"><a href="module-3.html#cb193-2" tabindex="-1"></a><span class="co">#Define working dir paths</span></span>
<span id="cb193-3"><a href="module-3.html#cb193-3" tabindex="-1"></a><span class="co"># datadir = &quot;/cloud/project/data/bulk_rna&quot;</span></span>
<span id="cb193-4"><a href="module-3.html#cb193-4" tabindex="-1"></a><span class="co"># outdir = &quot;/cloud/project/outdir&quot;</span></span>
<span id="cb193-5"><a href="module-3.html#cb193-5" tabindex="-1"></a></span>
<span id="cb193-6"><a href="module-3.html#cb193-6" tabindex="-1"></a>datadir <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/batch_correction&quot;</span></span>
<span id="cb193-7"><a href="module-3.html#cb193-7" tabindex="-1"></a>outdir <span class="ot">=</span> <span class="st">&quot;~/workspace/rnaseq/batch_correction/outputs&quot;</span></span>
<span id="cb193-8"><a href="module-3.html#cb193-8" tabindex="-1"></a></span>
<span id="cb193-9"><a href="module-3.html#cb193-9" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(outdir)) <span class="fu">dir.create</span>(outdir)</span>
<span id="cb193-10"><a href="module-3.html#cb193-10" tabindex="-1"></a></span>
<span id="cb193-11"><a href="module-3.html#cb193-11" tabindex="-1"></a></span>
<span id="cb193-12"><a href="module-3.html#cb193-12" tabindex="-1"></a></span>
<span id="cb193-13"><a href="module-3.html#cb193-13" tabindex="-1"></a><span class="co">#load neccessary libraries</span></span>
<span id="cb193-14"><a href="module-3.html#cb193-14" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;sva&quot;</span>) <span class="co">#Note this exercise requires sva (&gt;= v3.36.0) which is only available for R (&gt;= v4.x)</span></span>
<span id="cb193-15"><a href="module-3.html#cb193-15" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb193-16"><a href="module-3.html#cb193-16" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;gridExtra&quot;</span>)</span>
<span id="cb193-17"><a href="module-3.html#cb193-17" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;edgeR&quot;</span>)</span>
<span id="cb193-18"><a href="module-3.html#cb193-18" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;UpSetR&quot;</span>)</span>
<span id="cb193-19"><a href="module-3.html#cb193-19" tabindex="-1"></a><span class="fu">library</span>(grid)</span>
<span id="cb193-20"><a href="module-3.html#cb193-20" tabindex="-1"></a></span>
<span id="cb193-21"><a href="module-3.html#cb193-21" tabindex="-1"></a><span class="co">#load in the uncorrected data as raw counts</span></span>
<span id="cb193-22"><a href="module-3.html#cb193-22" tabindex="-1"></a><span class="fu">setwd</span>(datadir)</span>
<span id="cb193-23"><a href="module-3.html#cb193-23" tabindex="-1"></a>uncorrected_data <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">as.is =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb193-24"><a href="module-3.html#cb193-24" tabindex="-1"></a><span class="fu">setwd</span>(outdir)</span>
<span id="cb193-25"><a href="module-3.html#cb193-25" tabindex="-1"></a></span>
<span id="cb193-26"><a href="module-3.html#cb193-26" tabindex="-1"></a><span class="co">#simplify the names of the data columns</span></span>
<span id="cb193-27"><a href="module-3.html#cb193-27" tabindex="-1"></a><span class="co"># (A = Universal Human Reference RNA and B = Human Brain Reference RNA)</span></span>
<span id="cb193-28"><a href="module-3.html#cb193-28" tabindex="-1"></a><span class="co"># RNA = polyA enrichment and RIBO = ribosomal RNA depletion</span></span>
<span id="cb193-29"><a href="module-3.html#cb193-29" tabindex="-1"></a><span class="co"># 1, 2, 3, 4 are replicates</span></span>
<span id="cb193-30"><a href="module-3.html#cb193-30" tabindex="-1"></a><span class="fu">names</span>(uncorrected_data) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>, <span class="st">&quot;Chr&quot;</span>, <span class="st">&quot;UHR_Ribo_1&quot;</span>, <span class="st">&quot;UHR_Ribo_2&quot;</span>, <span class="st">&quot;UHR_Ribo_3&quot;</span>, <span class="st">&quot;UHR_Ribo_4&quot;</span>, <span class="st">&quot;HBR_Ribo_1&quot;</span>, <span class="st">&quot;HBR_Ribo_2&quot;</span>, <span class="st">&quot;HBR_Ribo_3&quot;</span>, <span class="st">&quot;HBR_Ribo_4&quot;</span>, </span>
<span id="cb193-31"><a href="module-3.html#cb193-31" tabindex="-1"></a>                            <span class="st">&quot;UHR_Poly_1&quot;</span>, <span class="st">&quot;UHR_Poly_2&quot;</span>, <span class="st">&quot;UHR_Poly_3&quot;</span>, <span class="st">&quot;UHR_Poly_4&quot;</span>, <span class="st">&quot;HBR_Poly_1&quot;</span>, <span class="st">&quot;HBR_Poly_2&quot;</span>, <span class="st">&quot;HBR_Poly_3&quot;</span>, <span class="st">&quot;HBR_Poly_4&quot;</span>)</span>
<span id="cb193-32"><a href="module-3.html#cb193-32" tabindex="-1"></a>sample_names <span class="ot">=</span> <span class="fu">names</span>(uncorrected_data)[<span class="dv">3</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">names</span>(uncorrected_data))]</span>
<span id="cb193-33"><a href="module-3.html#cb193-33" tabindex="-1"></a></span>
<span id="cb193-34"><a href="module-3.html#cb193-34" tabindex="-1"></a><span class="co">#review data structure</span></span>
<span id="cb193-35"><a href="module-3.html#cb193-35" tabindex="-1"></a><span class="fu">head</span>(uncorrected_data)</span>
<span id="cb193-36"><a href="module-3.html#cb193-36" tabindex="-1"></a><span class="fu">dim</span>(uncorrected_data)</span>
<span id="cb193-37"><a href="module-3.html#cb193-37" tabindex="-1"></a></span>
<span id="cb193-38"><a href="module-3.html#cb193-38" tabindex="-1"></a><span class="co">#define conditions, library methods, and replicates</span></span>
<span id="cb193-39"><a href="module-3.html#cb193-39" tabindex="-1"></a>conditions <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;UHR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>, <span class="st">&quot;HBR&quot;</span>)</span>
<span id="cb193-40"><a href="module-3.html#cb193-40" tabindex="-1"></a>library_methods <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Ribo&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>, <span class="st">&quot;Poly&quot;</span>)</span>
<span id="cb193-41"><a href="module-3.html#cb193-41" tabindex="-1"></a>replicates <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)</span>
<span id="cb193-42"><a href="module-3.html#cb193-42" tabindex="-1"></a></span>
<span id="cb193-43"><a href="module-3.html#cb193-43" tabindex="-1"></a><span class="co">#calculate principal components for the uncorrected data</span></span>
<span id="cb193-44"><a href="module-3.html#cb193-44" tabindex="-1"></a>pca_uncorrected_obj <span class="ot">=</span> <span class="fu">prcomp</span>(uncorrected_data[,sample_names])</span>
<span id="cb193-45"><a href="module-3.html#cb193-45" tabindex="-1"></a></span>
<span id="cb193-46"><a href="module-3.html#cb193-46" tabindex="-1"></a><span class="co">#pull PCA values out of the PCA object</span></span>
<span id="cb193-47"><a href="module-3.html#cb193-47" tabindex="-1"></a>pca_uncorrected <span class="ot">=</span> <span class="fu">as.data.frame</span>(pca_uncorrected_obj[<span class="dv">2</span>]<span class="sc">$</span>rotation)</span>
<span id="cb193-48"><a href="module-3.html#cb193-48" tabindex="-1"></a></span>
<span id="cb193-49"><a href="module-3.html#cb193-49" tabindex="-1"></a><span class="co">#assign labels to the data frame</span></span>
<span id="cb193-50"><a href="module-3.html#cb193-50" tabindex="-1"></a>pca_uncorrected[,<span class="st">&quot;condition&quot;</span>] <span class="ot">=</span> conditions</span>
<span id="cb193-51"><a href="module-3.html#cb193-51" tabindex="-1"></a>pca_uncorrected[,<span class="st">&quot;library_method&quot;</span>] <span class="ot">=</span> library_methods</span>
<span id="cb193-52"><a href="module-3.html#cb193-52" tabindex="-1"></a>pca_uncorrected[,<span class="st">&quot;replicate&quot;</span>] <span class="ot">=</span> replicates</span>
<span id="cb193-53"><a href="module-3.html#cb193-53" tabindex="-1"></a></span>
<span id="cb193-54"><a href="module-3.html#cb193-54" tabindex="-1"></a><span class="co">#plot the PCA</span></span>
<span id="cb193-55"><a href="module-3.html#cb193-55" tabindex="-1"></a><span class="co">#create a classic 2-dimension PCA plot (first two principal components) with conditions and library methods indicated</span></span>
<span id="cb193-56"><a href="module-3.html#cb193-56" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span> <span class="ot">=</span> <span class="st">&quot;#481567FF&quot;</span>, <span class="st">&quot;HBR&quot;</span> <span class="ot">=</span> <span class="st">&quot;#1F968BFF&quot;</span>)</span>
<span id="cb193-57"><a href="module-3.html#cb193-57" tabindex="-1"></a>p1 <span class="ot">=</span> <span class="fu">ggplot</span>(<span class="at">data =</span> pca_uncorrected, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> condition, <span class="at">shape =</span> library_method))</span>
<span id="cb193-58"><a href="module-3.html#cb193-58" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>)</span>
<span id="cb193-59"><a href="module-3.html#cb193-59" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">&quot;norm&quot;</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span>
<span id="cb193-60"><a href="module-3.html#cb193-60" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;PCA, RNA-seq counts for 16 HBR/UHR and Ribo/PolyA samples (uncorrected data)&quot;</span>, <span class="at">color =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">shape=</span><span class="st">&quot;Library Method&quot;</span>)</span>
<span id="cb193-61"><a href="module-3.html#cb193-61" tabindex="-1"></a>p1 <span class="ot">=</span> p1 <span class="sc">+</span> <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> cols)</span></code></pre></div>
</div>
<div id="introduction-to-bioconductor-sva-and-combat-seq-in-r" class="section level4 hasAnchor">
<h4>Introduction to Bioconductor SVA and ComBat-Seq in R<a href="module-3.html#introduction-to-bioconductor-sva-and-combat-seq-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The ComBat-Seq package is made available as part of the <a href="https://www.bioconductor.org/packages/release/bioc/html/sva.html">SVA package</a> for Surrogate Variable Analysis. This package is a collection of methods for removing batch effects and other unwanted variation in large datasets. It includes the ComBat method that has been widely used for batch correction of gene expression datasets, especially those generated on microarray platforms. ComBat-Seq is a modification of the ComBat approach. ComBat-Seq has been tailored to the count based data of bulk RNA-seq datasets. Particular advantages of the ComBat-Seq approach are that it: (1) uses a negative binomial regression model (the negative binomial distribution is thought to model the characteristics of bulk RNA-seq count data), and (2) allows the output of corrected data that retain the count nature of the data and can be safely fed into many existing methods for DE analysis (such as EdgeR and DESeq2).</p>
<p>ComBat-Seq has a relatively short list of arguments, and for several of these we will use the default setting. Very basic documentation of these arguments can be found <a href="https://rdrr.io/bioc/sva/man/ComBat_seq.html">here</a> and <a href="https://github.com/zhangyuqing/ComBat-seq">here</a>.</p>
<p>Our attempt to explain each of the ComBat-Seq arguments (for optional arguments, default is shown):</p>
<ul>
<li><code>counts</code>. This is your matrix of gene expression read counts (raw counts). Each row is a gene, each column is a sample, and each cell has an integer count for the number of RNA-seq counts observed for that gene/sample combination. In R we want this data to be passed into ComBat-Seq in matrix format (use <code>as.matrix()</code> if neccessary).</li>
<li><code>batch</code>. This is a vector describing the batches you are concerned about. For example, if you have eight samples that you created RNA-seq data for, but for the first four you used library kit (A) and for the last four samples you used library kit (B). In this situation you would define your <code>batch</code> vector as: <code>c(1,1,1,1,2,2,2,2)</code>.</li>
<li><code>group = NULL</code>. This is a vector describing your biological condition of interest. For example, if your experiment involved <em>pairs</em> of drug treated and untreated cells, and you did 4 biological replicates. You would define your <code>group</code> vector as: c(1,2,1,2,1,2,1,2).</li>
<li><code>covar_mod = NULL</code>. If you have multiple biological conditions of interest, you can define these with <code>covar_mod</code> (covariates) instead of <code>group</code>. For example, lets assume we have the same experiment as described above, except that we did four replicates (treated vs untreated pairs), but we alternated use of male and female cells for each of the replicates. You then would define a covariate matrix to supply to <code>covar_mod</code> as follows:</li>
</ul>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="module-3.html#cb194-1" tabindex="-1"></a>  <span class="co">#treatment_group = c(1,2,1,2,1,2,1,2)</span></span>
<span id="cb194-2"><a href="module-3.html#cb194-2" tabindex="-1"></a>  <span class="co">#sex_group = c(1,1,2,2,1,1,2,2)</span></span>
<span id="cb194-3"><a href="module-3.html#cb194-3" tabindex="-1"></a>  <span class="co">#covariate_matrix = cbind(treatment_group, sex_group)</span></span></code></pre></div>
<ul>
<li><code>full_mod = TRUE</code>. If TRUE include the condition of interest in model. Generally we believe this should be set to the default TRUE. We have yet to find a cohesive explanation for a situation where one would want this to be FALSE.</li>
<li><code>shrink = FALSE</code>. Whether to apply shrinkage on parameter estimation.<br />
</li>
<li><code>shrink.disp = FALSE</code>. Whether to apply shrinkage on dispersion.<br />
</li>
<li><code>gene.subset.n = NULL</code>. Number of genes to use in empirical Bayes estimation, only useful when shrink = TRUE.</li>
</ul>
<p>A detailed discussion of <em>shrinkage</em> (related to the <code>shrink</code>, <code>shrink.disp</code>, and <code>gene_subset.n</code> arguments is beyond the scope of this tutorial. Briefly, shrinkage refers to a set of methods that attempt to correct for gene-specific variability in the counts observed in RNA-seq datasets. More specifically, it relates to the <em>dispersion parameter</em> of the <a href="https://en.wikipedia.org/wiki/Negative_binomial_distribution">negative binomial distribution</a> used to model RNA-seq count data that can suffer from <a href="https://en.wikipedia.org/wiki/Overdispersion">overdispersion</a>. The dispersion parameter describes how much variance deviates from the mean. In simple terms, shrinkage methods are an attempt to correct for problematic dispersion. A more detailed discussion of these statistical concepts can be found in the <a href="https://pubmed.ncbi.nlm.nih.gov/25516281/">DESeq2 paper</a>. However, for our purposes here, the bottom line is that the ComBat-Seq authors state that “We have shown that applying empirical Bayes shrinkage is not necessary for ComBat-seq because the approach is already sufficiently robust due to the distributional assumption.” So we will leave these arguments at their default <code>FALSE</code> settings.</p>
</div>
<div id="demonstration-of-combat-seq-on-the-uhrhbr-data-with-two-library-types-ribopoly" class="section level4 hasAnchor">
<h4>Demonstration of ComBat-Seq on the UHR/HBR data with two library types (Ribo/Poly)<a href="module-3.html#demonstration-of-combat-seq-on-the-uhrhbr-data-with-two-library-types-ribopoly" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Continuing the R session started above, use ComBat-Seq to perform batch correction as follows:</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="module-3.html#cb195-1" tabindex="-1"></a></span>
<span id="cb195-2"><a href="module-3.html#cb195-2" tabindex="-1"></a><span class="co">#perform the batch correction</span></span>
<span id="cb195-3"><a href="module-3.html#cb195-3" tabindex="-1"></a></span>
<span id="cb195-4"><a href="module-3.html#cb195-4" tabindex="-1"></a><span class="co">#first we need to transform the format of our groups and batches from names (e.g. &quot;UHR&quot;, &quot;HBR&quot;, etc.) to numbers (e.g. 1, 2, etc.)</span></span>
<span id="cb195-5"><a href="module-3.html#cb195-5" tabindex="-1"></a><span class="co">#in the command below &quot;sapply&quot; is used to apply the &quot;switch&quot; command to each element and convert names to numbers as we define</span></span>
<span id="cb195-6"><a href="module-3.html#cb195-6" tabindex="-1"></a>groups <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">as.character</span>(conditions), <span class="cf">switch</span>, <span class="st">&quot;UHR&quot;</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">&quot;HBR&quot;</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="at">USE.NAMES =</span> <span class="cn">FALSE</span>)</span>
<span id="cb195-7"><a href="module-3.html#cb195-7" tabindex="-1"></a>batches <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">as.character</span>(library_methods), <span class="cf">switch</span>, <span class="st">&quot;Ribo&quot;</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">&quot;Poly&quot;</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="at">USE.NAMES =</span> <span class="cn">FALSE</span>)</span>
<span id="cb195-8"><a href="module-3.html#cb195-8" tabindex="-1"></a></span>
<span id="cb195-9"><a href="module-3.html#cb195-9" tabindex="-1"></a><span class="co">#now run ComBat_seq</span></span>
<span id="cb195-10"><a href="module-3.html#cb195-10" tabindex="-1"></a>corrected_data <span class="ot">=</span> <span class="fu">ComBat_seq</span>(<span class="at">counts =</span> <span class="fu">as.matrix</span>(uncorrected_data[,sample_names]), <span class="at">batch =</span> batches, <span class="at">group =</span> groups)</span>
<span id="cb195-11"><a href="module-3.html#cb195-11" tabindex="-1"></a></span>
<span id="cb195-12"><a href="module-3.html#cb195-12" tabindex="-1"></a><span class="co">#join the gene and chromosome names onto the now corrected counts from ComBat_seq</span></span>
<span id="cb195-13"><a href="module-3.html#cb195-13" tabindex="-1"></a>corrected_data <span class="ot">=</span> <span class="fu">cbind</span>(uncorrected_data[, <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>, <span class="st">&quot;Chr&quot;</span>)], corrected_data)</span>
<span id="cb195-14"><a href="module-3.html#cb195-14" tabindex="-1"></a></span>
<span id="cb195-15"><a href="module-3.html#cb195-15" tabindex="-1"></a><span class="co">#compare dimensions of corrected and uncorrected data sets</span></span>
<span id="cb195-16"><a href="module-3.html#cb195-16" tabindex="-1"></a><span class="fu">dim</span>(uncorrected_data)</span>
<span id="cb195-17"><a href="module-3.html#cb195-17" tabindex="-1"></a><span class="fu">dim</span>(corrected_data)</span>
<span id="cb195-18"><a href="module-3.html#cb195-18" tabindex="-1"></a></span>
<span id="cb195-19"><a href="module-3.html#cb195-19" tabindex="-1"></a><span class="co">#visually compare values of corrected and uncorrected data sets</span></span>
<span id="cb195-20"><a href="module-3.html#cb195-20" tabindex="-1"></a><span class="fu">head</span>(uncorrected_data)</span>
<span id="cb195-21"><a href="module-3.html#cb195-21" tabindex="-1"></a><span class="fu">head</span>(corrected_data)</span></code></pre></div>
</div>
<div id="perform-pca-analysis-on-the-batch-corrected-data-and-contrast-with-the-uncorrected-data" class="section level4 hasAnchor">
<h4>Perform PCA analysis on the batch corrected data and contrast with the uncorrected data<a href="module-3.html#perform-pca-analysis-on-the-batch-corrected-data-and-contrast-with-the-uncorrected-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>As performed above, use PCA to examine whether batch correction changes the grouping of samples by the expression patterns. Does the corrected data cluster according to biological condition (UHR vs HBR) better now regardless of library preparation type (Ribo vs PolyA)?</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="module-3.html#cb196-1" tabindex="-1"></a></span>
<span id="cb196-2"><a href="module-3.html#cb196-2" tabindex="-1"></a><span class="co">#calculate principal components for the uncorrected data</span></span>
<span id="cb196-3"><a href="module-3.html#cb196-3" tabindex="-1"></a>pca_corrected_obj <span class="ot">=</span> <span class="fu">prcomp</span>(corrected_data[, sample_names])</span>
<span id="cb196-4"><a href="module-3.html#cb196-4" tabindex="-1"></a></span>
<span id="cb196-5"><a href="module-3.html#cb196-5" tabindex="-1"></a><span class="co">#pull PCA values out of the PCA object</span></span>
<span id="cb196-6"><a href="module-3.html#cb196-6" tabindex="-1"></a>pca_corrected <span class="ot">=</span> <span class="fu">as.data.frame</span>(pca_corrected_obj[<span class="dv">2</span>]<span class="sc">$</span>rotation)</span>
<span id="cb196-7"><a href="module-3.html#cb196-7" tabindex="-1"></a></span>
<span id="cb196-8"><a href="module-3.html#cb196-8" tabindex="-1"></a><span class="co">#assign labels to the data frame</span></span>
<span id="cb196-9"><a href="module-3.html#cb196-9" tabindex="-1"></a>pca_corrected[,<span class="st">&quot;condition&quot;</span>] <span class="ot">=</span> conditions</span>
<span id="cb196-10"><a href="module-3.html#cb196-10" tabindex="-1"></a>pca_corrected[,<span class="st">&quot;library_method&quot;</span>] <span class="ot">=</span> library_methods</span>
<span id="cb196-11"><a href="module-3.html#cb196-11" tabindex="-1"></a>pca_corrected[,<span class="st">&quot;replicate&quot;</span>] <span class="ot">=</span> replicates</span>
<span id="cb196-12"><a href="module-3.html#cb196-12" tabindex="-1"></a></span>
<span id="cb196-13"><a href="module-3.html#cb196-13" tabindex="-1"></a><span class="co">#as above, create a PCA plot for comparison to the uncorrected data</span></span>
<span id="cb196-14"><a href="module-3.html#cb196-14" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;UHR&quot;</span> <span class="ot">=</span> <span class="st">&quot;#481567FF&quot;</span>, <span class="st">&quot;HBR&quot;</span> <span class="ot">=</span> <span class="st">&quot;#1F968BFF&quot;</span>)</span>
<span id="cb196-15"><a href="module-3.html#cb196-15" tabindex="-1"></a>p2 <span class="ot">=</span> <span class="fu">ggplot</span>(<span class="at">data =</span> pca_corrected, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">color =</span> condition, <span class="at">shape =</span> library_method))</span>
<span id="cb196-16"><a href="module-3.html#cb196-16" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>)</span>
<span id="cb196-17"><a href="module-3.html#cb196-17" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">stat_ellipse</span>(<span class="at">type =</span> <span class="st">&quot;norm&quot;</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span>
<span id="cb196-18"><a href="module-3.html#cb196-18" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;PCA, RNA-seq counts for 16 HBR/UHR and Ribo/PolyA samples (batch corrected data)&quot;</span>, <span class="at">color =</span> <span class="st">&quot;Condition&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;Library Method&quot;</span>)</span>
<span id="cb196-19"><a href="module-3.html#cb196-19" tabindex="-1"></a>p2 <span class="ot">=</span> p2 <span class="sc">+</span> <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> cols)</span>
<span id="cb196-20"><a href="module-3.html#cb196-20" tabindex="-1"></a></span>
<span id="cb196-21"><a href="module-3.html#cb196-21" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;Uncorrected-vs-BatchCorrected-PCA.pdf&quot;</span>)</span>
<span id="cb196-22"><a href="module-3.html#cb196-22" tabindex="-1"></a><span class="fu">grid.arrange</span>(p1, p2, <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb196-23"><a href="module-3.html#cb196-23" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>If the above analysis worked you should have an image that looks like this:
<img src="assets/module_3/Uncorrected-vs-BatchCorrected-PCA.png" alt="PCA-Plot" /></p>
<p>Note that prior to correction the UHR samples are separate from the HBR samples. Note that the PolyA and Ribo-reduction samples are also separated. The 16 samples group into 4 fairly distinct clusters. In other words, the overall expression signatures of these samples seem to reflect both the biological condition (UHR vs HBR) and library construction approach (Ribo vs PolyA).</p>
<p>However, after correcting for the batch effect of library construction approach, we see a marked improvement. The library construction approach still seems to be noticeable, but the 16 samples now essentially group into two distinct clusters: HBR and UHR.</p>
</div>
<div id="perform-differential-expression-analysis-of-the-corrected-and-uncorrected-data" class="section level4 hasAnchor">
<h4>Perform differential expression analysis of the corrected and uncorrected data<a href="module-3.html#perform-differential-expression-analysis-of-the-corrected-and-uncorrected-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>How does batch correction influence differential gene expression results? Use UpSet plots to examine the overlap of significant DE genes found for the following comparisons:</p>
<ul>
<li>UHR-Ribo vs HBR-Ribo (same library type, 4 vs 4 replicates)</li>
<li>UHR-Poly vs HBR-Poly (same library type, 4 vs 4 replicates)</li>
<li>UHR-Ribo vs HBR-Poly (different library types, 4 vs 4 replicates)</li>
<li>UHR-Poly vs HBR-Ribo (different library types, 4 vs 4 replicates)</li>
<li>UHR-Comb vs HBR-Comb (combined library types, 8 vs 8 replicates)</li>
</ul>
<p>These five differential expression analysis comparisons will be performed with both the uncorrected and corrected data.</p>
<ul>
<li>Does correction increase agreement between the five comparisons?</li>
<li>Does it appear to increase statistical power when combining all 8 replicates of UHR and HBR?</li>
<li>What do we expect to see for comparisons like UHR-Ribo vs HBR-Poly before and after batch correction? Do we expect correction to increase or decrease the number of significant results?</li>
</ul>
<p>Explore these questions by continuing on with the R session started above and doing the following:</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="module-3.html#cb197-1" tabindex="-1"></a><span class="co">#perform differential expression analysis on the uncorrected data and batch corrected data sets</span></span>
<span id="cb197-2"><a href="module-3.html#cb197-2" tabindex="-1"></a></span>
<span id="cb197-3"><a href="module-3.html#cb197-3" tabindex="-1"></a><span class="co">#first define the sets of samples to be compared to each other</span></span>
<span id="cb197-4"><a href="module-3.html#cb197-4" tabindex="-1"></a>uhr_ribo_samples <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR_Ribo_1&quot;</span>, <span class="st">&quot;UHR_Ribo_2&quot;</span>, <span class="st">&quot;UHR_Ribo_3&quot;</span>, <span class="st">&quot;UHR_Ribo_4&quot;</span>)</span>
<span id="cb197-5"><a href="module-3.html#cb197-5" tabindex="-1"></a>uhr_poly_samples <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;UHR_Poly_1&quot;</span>, <span class="st">&quot;UHR_Poly_2&quot;</span>, <span class="st">&quot;UHR_Poly_3&quot;</span>, <span class="st">&quot;UHR_Poly_4&quot;</span>)</span>
<span id="cb197-6"><a href="module-3.html#cb197-6" tabindex="-1"></a>hbr_ribo_samples <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;HBR_Ribo_1&quot;</span>, <span class="st">&quot;HBR_Ribo_2&quot;</span>, <span class="st">&quot;HBR_Ribo_3&quot;</span>, <span class="st">&quot;HBR_Ribo_4&quot;</span>)</span>
<span id="cb197-7"><a href="module-3.html#cb197-7" tabindex="-1"></a>hbr_poly_samples <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;HBR_Poly_1&quot;</span>, <span class="st">&quot;HBR_Poly_2&quot;</span>, <span class="st">&quot;HBR_Poly_3&quot;</span>, <span class="st">&quot;HBR_Poly_4&quot;</span>)</span>
<span id="cb197-8"><a href="module-3.html#cb197-8" tabindex="-1"></a>uhr_samples <span class="ot">=</span> <span class="fu">c</span>(uhr_ribo_samples, uhr_poly_samples)</span>
<span id="cb197-9"><a href="module-3.html#cb197-9" tabindex="-1"></a>hbr_samples <span class="ot">=</span> <span class="fu">c</span>(hbr_ribo_samples, hbr_poly_samples)</span>
<span id="cb197-10"><a href="module-3.html#cb197-10" tabindex="-1"></a></span>
<span id="cb197-11"><a href="module-3.html#cb197-11" tabindex="-1"></a><span class="co">#create a function that will run edgeR (DE analysis) for a particular pair of sample sets</span></span>
<span id="cb197-12"><a href="module-3.html#cb197-12" tabindex="-1"></a>run_edgeR <span class="ot">=</span> <span class="cf">function</span>(data, group_a_name, group_a_samples, group_b_samples, group_b_name){</span>
<span id="cb197-13"><a href="module-3.html#cb197-13" tabindex="-1"></a>  <span class="co">#create a list of all samples for this current comparison</span></span>
<span id="cb197-14"><a href="module-3.html#cb197-14" tabindex="-1"></a>  samples_for_comparison <span class="ot">=</span> <span class="fu">c</span>(group_a_samples, group_b_samples)</span>
<span id="cb197-15"><a href="module-3.html#cb197-15" tabindex="-1"></a>  </span>
<span id="cb197-16"><a href="module-3.html#cb197-16" tabindex="-1"></a>  <span class="co">#define the class factor for this pair of sample sets</span></span>
<span id="cb197-17"><a href="module-3.html#cb197-17" tabindex="-1"></a>  class <span class="ot">=</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(group_a_name, <span class="fu">length</span>(group_a_samples)), <span class="fu">rep</span>(group_b_name, <span class="fu">length</span>(group_b_samples))))</span>
<span id="cb197-18"><a href="module-3.html#cb197-18" tabindex="-1"></a>  </span>
<span id="cb197-19"><a href="module-3.html#cb197-19" tabindex="-1"></a>  <span class="co">#create a simplified data matrix for only these samples</span></span>
<span id="cb197-20"><a href="module-3.html#cb197-20" tabindex="-1"></a>  rawdata <span class="ot">=</span> data[, samples_for_comparison]</span>
<span id="cb197-21"><a href="module-3.html#cb197-21" tabindex="-1"></a>  </span>
<span id="cb197-22"><a href="module-3.html#cb197-22" tabindex="-1"></a>  <span class="co">#store gene names for later</span></span>
<span id="cb197-23"><a href="module-3.html#cb197-23" tabindex="-1"></a>  genes <span class="ot">=</span> <span class="fu">rownames</span>(data)</span>
<span id="cb197-24"><a href="module-3.html#cb197-24" tabindex="-1"></a>  gene_names <span class="ot">=</span> data[,<span class="st">&quot;Gene&quot;</span>]</span>
<span id="cb197-25"><a href="module-3.html#cb197-25" tabindex="-1"></a>  </span>
<span id="cb197-26"><a href="module-3.html#cb197-26" tabindex="-1"></a>  <span class="co">#make DGElist object</span></span>
<span id="cb197-27"><a href="module-3.html#cb197-27" tabindex="-1"></a>  y <span class="ot">=</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> rawdata, <span class="at">genes =</span> genes, <span class="at">group =</span> class)</span>
<span id="cb197-28"><a href="module-3.html#cb197-28" tabindex="-1"></a>  </span>
<span id="cb197-29"><a href="module-3.html#cb197-29" tabindex="-1"></a>  <span class="co">#perform TMM normalization</span></span>
<span id="cb197-30"><a href="module-3.html#cb197-30" tabindex="-1"></a>  y <span class="ot">=</span> <span class="fu">calcNormFactors</span>(y)</span>
<span id="cb197-31"><a href="module-3.html#cb197-31" tabindex="-1"></a>  </span>
<span id="cb197-32"><a href="module-3.html#cb197-32" tabindex="-1"></a>  <span class="co">#estimate dispersion</span></span>
<span id="cb197-33"><a href="module-3.html#cb197-33" tabindex="-1"></a>  y <span class="ot">=</span> <span class="fu">estimateCommonDisp</span>(y, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb197-34"><a href="module-3.html#cb197-34" tabindex="-1"></a>  y <span class="ot">=</span> <span class="fu">estimateTagwiseDisp</span>(y)</span>
<span id="cb197-35"><a href="module-3.html#cb197-35" tabindex="-1"></a>  </span>
<span id="cb197-36"><a href="module-3.html#cb197-36" tabindex="-1"></a>  <span class="co">#perform the differential expression test</span></span>
<span id="cb197-37"><a href="module-3.html#cb197-37" tabindex="-1"></a>  et <span class="ot">=</span> <span class="fu">exactTest</span>(y)</span>
<span id="cb197-38"><a href="module-3.html#cb197-38" tabindex="-1"></a>  </span>
<span id="cb197-39"><a href="module-3.html#cb197-39" tabindex="-1"></a>  <span class="co">#print number of up/down significant genes at FDR = 0.05 significance level and store the DE status in a new variable (de)</span></span>
<span id="cb197-40"><a href="module-3.html#cb197-40" tabindex="-1"></a>  de <span class="ot">=</span> <span class="fu">decideTests</span>(et, <span class="at">adjust.method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">p =</span> <span class="fl">0.05</span>)</span>
<span id="cb197-41"><a href="module-3.html#cb197-41" tabindex="-1"></a>  <span class="fu">summary</span>(de)</span>
<span id="cb197-42"><a href="module-3.html#cb197-42" tabindex="-1"></a>  </span>
<span id="cb197-43"><a href="module-3.html#cb197-43" tabindex="-1"></a>  <span class="co">#create a matrix of the DE results</span></span>
<span id="cb197-44"><a href="module-3.html#cb197-44" tabindex="-1"></a>  mat <span class="ot">=</span> <span class="fu">cbind</span>(</span>
<span id="cb197-45"><a href="module-3.html#cb197-45" tabindex="-1"></a>    genes, </span>
<span id="cb197-46"><a href="module-3.html#cb197-46" tabindex="-1"></a>    gene_names,</span>
<span id="cb197-47"><a href="module-3.html#cb197-47" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;%0.3f&quot;</span>, <span class="fu">log10</span>(et<span class="sc">$</span>table<span class="sc">$</span>PValue)),</span>
<span id="cb197-48"><a href="module-3.html#cb197-48" tabindex="-1"></a>    <span class="fu">sprintf</span>(<span class="st">&quot;%0.3f&quot;</span>, et<span class="sc">$</span>table<span class="sc">$</span>logFC)</span>
<span id="cb197-49"><a href="module-3.html#cb197-49" tabindex="-1"></a>  )</span>
<span id="cb197-50"><a href="module-3.html#cb197-50" tabindex="-1"></a></span>
<span id="cb197-51"><a href="module-3.html#cb197-51" tabindex="-1"></a>  <span class="co">#create a version of this matrix that is limited to only the *significant* results</span></span>
<span id="cb197-52"><a href="module-3.html#cb197-52" tabindex="-1"></a>  mat <span class="ot">=</span> mat[<span class="fu">as.logical</span>(de),]</span>
<span id="cb197-53"><a href="module-3.html#cb197-53" tabindex="-1"></a>  </span>
<span id="cb197-54"><a href="module-3.html#cb197-54" tabindex="-1"></a>  <span class="co">#add name to the columns of the final matrix</span></span>
<span id="cb197-55"><a href="module-3.html#cb197-55" tabindex="-1"></a>  <span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>, <span class="st">&quot;Gene_Name&quot;</span>, <span class="st">&quot;Log10_Pvalue&quot;</span>, <span class="st">&quot;Log_fold_change&quot;</span>)</span>
<span id="cb197-56"><a href="module-3.html#cb197-56" tabindex="-1"></a>  </span>
<span id="cb197-57"><a href="module-3.html#cb197-57" tabindex="-1"></a>  <span class="fu">return</span>(mat)</span>
<span id="cb197-58"><a href="module-3.html#cb197-58" tabindex="-1"></a>}</span>
<span id="cb197-59"><a href="module-3.html#cb197-59" tabindex="-1"></a></span>
<span id="cb197-60"><a href="module-3.html#cb197-60" tabindex="-1"></a><span class="co">#run the five comparisons through edgeR using the *uncorrected data*</span></span>
<span id="cb197-61"><a href="module-3.html#cb197-61" tabindex="-1"></a>uhr_ribo_vs_hbr_ribo_uncorrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> uncorrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_ribo_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_ribo_samples)</span>
<span id="cb197-62"><a href="module-3.html#cb197-62" tabindex="-1"></a>uhr_poly_vs_hbr_poly_uncorrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> uncorrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_poly_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_poly_samples)</span>
<span id="cb197-63"><a href="module-3.html#cb197-63" tabindex="-1"></a>uhr_ribo_vs_hbr_poly_uncorrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> uncorrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_ribo_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_poly_samples)</span>
<span id="cb197-64"><a href="module-3.html#cb197-64" tabindex="-1"></a>uhr_poly_vs_hbr_ribo_uncorrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> uncorrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_poly_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_ribo_samples)</span>
<span id="cb197-65"><a href="module-3.html#cb197-65" tabindex="-1"></a>uhr_vs_hbr_uncorrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> uncorrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_samples)</span>
<span id="cb197-66"><a href="module-3.html#cb197-66" tabindex="-1"></a></span>
<span id="cb197-67"><a href="module-3.html#cb197-67" tabindex="-1"></a><span class="co">#run the same five comparisons through edgeR using the *batch corrected data*</span></span>
<span id="cb197-68"><a href="module-3.html#cb197-68" tabindex="-1"></a>uhr_ribo_vs_hbr_ribo_corrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> corrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_ribo_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_ribo_samples)</span>
<span id="cb197-69"><a href="module-3.html#cb197-69" tabindex="-1"></a>uhr_poly_vs_hbr_poly_corrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> corrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_poly_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_poly_samples)</span>
<span id="cb197-70"><a href="module-3.html#cb197-70" tabindex="-1"></a>uhr_ribo_vs_hbr_poly_corrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> corrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_ribo_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_poly_samples)</span>
<span id="cb197-71"><a href="module-3.html#cb197-71" tabindex="-1"></a>uhr_poly_vs_hbr_ribo_corrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> corrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_poly_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_ribo_samples)</span>
<span id="cb197-72"><a href="module-3.html#cb197-72" tabindex="-1"></a>uhr_vs_hbr_corrected <span class="ot">=</span> <span class="fu">run_edgeR</span>(<span class="at">data =</span> corrected_data, <span class="at">group_a_name =</span> <span class="st">&quot;UHR&quot;</span>, <span class="at">group_a_samples =</span> uhr_samples, <span class="at">group_b_name =</span> <span class="st">&quot;HBR&quot;</span>, <span class="at">group_b_samples =</span> hbr_samples)</span>
<span id="cb197-73"><a href="module-3.html#cb197-73" tabindex="-1"></a></span>
<span id="cb197-74"><a href="module-3.html#cb197-74" tabindex="-1"></a><span class="co">#how much of a difference does batch correction make when doing the comparison of all UHR vs all HBR samples?</span></span>
<span id="cb197-75"><a href="module-3.html#cb197-75" tabindex="-1"></a><span class="fu">dim</span>(uhr_vs_hbr_uncorrected)</span>
<span id="cb197-76"><a href="module-3.html#cb197-76" tabindex="-1"></a><span class="fu">dim</span>(uhr_vs_hbr_corrected)</span>
<span id="cb197-77"><a href="module-3.html#cb197-77" tabindex="-1"></a></span>
<span id="cb197-78"><a href="module-3.html#cb197-78" tabindex="-1"></a><span class="co">#create upset plots to summarize the overlap between the comparisons performed above</span></span>
<span id="cb197-79"><a href="module-3.html#cb197-79" tabindex="-1"></a></span>
<span id="cb197-80"><a href="module-3.html#cb197-80" tabindex="-1"></a><span class="co">#first create upset plot from the *uncorrected* data</span></span>
<span id="cb197-81"><a href="module-3.html#cb197-81" tabindex="-1"></a>listInput1 <span class="ot">=</span> <span class="fu">list</span>(<span class="st">&quot;4 UHR Ribo vs 4 HBR Ribo&quot;</span> <span class="ot">=</span> uhr_ribo_vs_hbr_ribo_uncorrected[, <span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-82"><a href="module-3.html#cb197-82" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Poly vs 4HBR Poly&quot;</span> <span class="ot">=</span> uhr_poly_vs_hbr_poly_uncorrected[, <span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-83"><a href="module-3.html#cb197-83" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Ribo vs 4 HBR Poly&quot;</span> <span class="ot">=</span> uhr_ribo_vs_hbr_poly_uncorrected[, <span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-84"><a href="module-3.html#cb197-84" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Poly vs 4 HBR Ribo&quot;</span> <span class="ot">=</span> uhr_poly_vs_hbr_ribo_uncorrected[, <span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-85"><a href="module-3.html#cb197-85" tabindex="-1"></a>                  <span class="st">&quot;8 UHR vs 8 HBR&quot;</span> <span class="ot">=</span> uhr_vs_hbr_uncorrected[, <span class="st">&quot;Gene&quot;</span>])</span>
<span id="cb197-86"><a href="module-3.html#cb197-86" tabindex="-1"></a></span>
<span id="cb197-87"><a href="module-3.html#cb197-87" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;Uncorrected-UpSet.pdf&quot;</span>, <span class="at">onefile =</span> <span class="cn">FALSE</span>)</span>
<span id="cb197-88"><a href="module-3.html#cb197-88" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(listInput1), <span class="at">order.by =</span> <span class="st">&quot;freq&quot;</span>, <span class="at">number.angles =</span> <span class="dv">45</span>, <span class="at">point.size =</span> <span class="dv">3</span>)</span>
<span id="cb197-89"><a href="module-3.html#cb197-89" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb197-90"><a href="module-3.html#cb197-90" tabindex="-1"></a></span>
<span id="cb197-91"><a href="module-3.html#cb197-91" tabindex="-1"></a><span class="co">#now create an upset plot from the *batch corrected* data</span></span>
<span id="cb197-92"><a href="module-3.html#cb197-92" tabindex="-1"></a>listInput2 <span class="ot">=</span> <span class="fu">list</span>(<span class="st">&quot;4 UHR Ribo vs 4 HBR Ribo&quot;</span> <span class="ot">=</span> uhr_ribo_vs_hbr_ribo_corrected[,<span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-93"><a href="module-3.html#cb197-93" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Poly vs 4 HBR Poly&quot;</span> <span class="ot">=</span> uhr_poly_vs_hbr_poly_corrected[,<span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-94"><a href="module-3.html#cb197-94" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Ribo vs 4 HBR Poly&quot;</span> <span class="ot">=</span> uhr_ribo_vs_hbr_poly_corrected[,<span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-95"><a href="module-3.html#cb197-95" tabindex="-1"></a>                  <span class="st">&quot;4 UHR Poly vs 4 HBR Ribo&quot;</span> <span class="ot">=</span> uhr_poly_vs_hbr_ribo_corrected[,<span class="st">&quot;Gene&quot;</span>],</span>
<span id="cb197-96"><a href="module-3.html#cb197-96" tabindex="-1"></a>                  <span class="st">&quot;8 UHR vs 8 HBR&quot;</span> <span class="ot">=</span> uhr_vs_hbr_corrected[,<span class="st">&quot;Gene&quot;</span>])</span>
<span id="cb197-97"><a href="module-3.html#cb197-97" tabindex="-1"></a></span>
<span id="cb197-98"><a href="module-3.html#cb197-98" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">&quot;BatchCorrected-UpSet.pdf&quot;</span>, <span class="at">onefile =</span> <span class="cn">FALSE</span>)</span>
<span id="cb197-99"><a href="module-3.html#cb197-99" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(listInput2), <span class="at">order.by =</span> <span class="st">&quot;freq&quot;</span>, <span class="at">number.angles=</span><span class="dv">45</span>, <span class="at">point.size=</span><span class="dv">3</span>)</span>
<span id="cb197-100"><a href="module-3.html#cb197-100" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb197-101"><a href="module-3.html#cb197-101" tabindex="-1"></a></span>
<span id="cb197-102"><a href="module-3.html#cb197-102" tabindex="-1"></a><span class="co">#write out the final set of DE genes where all UHR and HBR samples were compared using the corrected data</span></span>
<span id="cb197-103"><a href="module-3.html#cb197-103" tabindex="-1"></a><span class="fu">write.table</span>(uhr_vs_hbr_corrected, <span class="at">file =</span> <span class="st">&quot;DE_genes_uhr_vs_hbr_corrected.tsv&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb197-104"><a href="module-3.html#cb197-104" tabindex="-1"></a></span>
<span id="cb197-105"><a href="module-3.html#cb197-105" tabindex="-1"></a><span class="co">#To exit R type the following</span></span>
<span id="cb197-106"><a href="module-3.html#cb197-106" tabindex="-1"></a><span class="co">#quit(save = &quot;no&quot;)</span></span></code></pre></div>
<p>Note that an UpSet plot is an alternative to a Venn Diagram. It shows the overlap (intersection) between an arbitrary number of sets of values. In this case we are comparing the list of genes identified as significantly DE by five different comparisons. The black circles connected by a line indicate each combination of sets being considered. The bar graph above each column shows how many genes are shared across those sets. For example, the first column has all five black circles. The bar above that column indicates how many genes were found in all five DE comparisons performed.</p>
<p>What differs in each comparison is:</p>
<ul>
<li>whether we are comparing replicates prepared with the same library construction approach (Ribo reduction or PolyA) or whether we are comparing data prepared with different approaches</li>
<li>whether we are comparing 4 UHR vs 4 HBR replicates according to these library construction approaches, or pooling all 8 UHR and all 8 HBR samples (ignoring their different library types)</li>
<li>whether we are doing these comparisons before or after batch correction</li>
</ul>
<div id="before-batch-correction" class="section level5 hasAnchor">
<h5>Before batch correction<a href="module-3.html#before-batch-correction" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="float">
<img src="assets/module_3/Uncorrected-UpSet.png" alt="Uncorrected-UpSet" />
<div class="figcaption">Uncorrected-UpSet</div>
</div>
</div>
<div id="after-batch-correction" class="section level5 hasAnchor">
<h5>After batch correction<a href="module-3.html#after-batch-correction" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="float">
<img src="assets/module_3/BatchCorrected-UpSet.png" alt="BatchCorrected-UpSet" />
<div class="figcaption">BatchCorrected-UpSet</div>
</div>
<p>There are several notable observations from the analysis above and the two UpSet plots.</p>
<ul>
<li>In the uncorrected data, we actually see more DE genes when comparing a mix of library contruction approaches (e.g. UHR-Ribo vs UHR-Poly). There are likely false positives in these results. i.e. Genes that appear to be different between UHR and HBR, but where the difference is actually caused by differences in the library preparation not the biology.</li>
<li>If we combine all 8 samples together for each biological condition we can see that we actually get considerably <em>fewer</em> significant genes with the uncorrected data. Presumably this is because we are now introducing noise caused by a mix of different library construction approaches and this impacts the statistical analysis (more variability).</li>
<li>When we apply batch correction, we see that now all five comparisons tend to agree with each other for the most part on what genes are differentially expressed. Overall agreement across comparisons is improved.</li>
<li>With the batch corrected data we now see that combining all 8 samples actually improves statistical power and results in a <em>larger</em> number of significant DE genes relative to the 4 vs 4 comparisons. This is presumably the most accurate result of all the comparisons we did.</li>
</ul>
<p>In the <a href="/module-02-alignment/0002/07/01/Team_Assignment_Alignment/">previous Team Assignment</a>, teams have trimmed, aligned, visualized and performed QC evaluations of their RNAseq data. Using this aligned data, students will now apply the concepts they have learned regarding expression estimation and differential expression analysis. To complete this assignment, students will need to review commands we performed in earlier sections.</p>
<p>Before starting this team exercise, first find the folder containing your <strong>6</strong> aligned bam files (along with the index files). Note: In the previous exercise, you merged bams files for easy visualization in IGV, we will not be using that for expression and de analysis.</p>
</div>
</div>
</div>
<div id="part-i---expression-estimation" class="section level3 hasAnchor">
<h3>Part I - Expression Estimation<a href="module-3.html#part-i---expression-estimation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Familiarize yourself with Stringtie options</li>
<li>Run Stringtie to obtain expression values</li>
<li>Optionally, run htseq-count to obtain raw count values</li>
</ul>
<div id="remember-to-do-this-in-a-new-directory-under-team_exercises" class="section level4 hasAnchor">
<h4>Remember to do this in a new directory under team_exercises<a href="module-3.html#remember-to-do-this-in-a-new-directory-under-team_exercises" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb198"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb198-1"><a href="module-3.html#cb198-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$RNA_HOME</span>/team_exercise/expression/</span>
<span id="cb198-2"><a href="module-3.html#cb198-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/team_exercise/expression/</span></code></pre></div>
<p>Teams can now use <code>Stringtie</code> to estimate the gene expression levels in their sample and answer the following questions:</p>
<p><strong>Q1.</strong> Based on your stringtie results, what are the top 5 genes with highest average expression levels across all knockout samples? What about in your rescue samples? (Hint: You can use R, command-line tools, or download files to your desktop for this analysis)</p>
</div>
</div>
<div id="part-ii---differential-expression" class="section level3 hasAnchor">
<h3>Part II - Differential Expression<a href="module-3.html#part-ii---differential-expression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Perform differential analysis between the knockout and rescued samples</li>
<li>Check which genes are differentially expressed with statistical significance</li>
<li>Visualize DE results. For example you could create an MDS plot, x-y scatter plot of mean KO vs Rescue FPKM values, or a volcano plot.</li>
</ul>
<p>Teams will now use ballgown to perform differential expression analysis followed by visualization of their results. Alternatively, if raw counts were generated teams can use edgeR or DESeq2 for differential expression analysis.</p>
<p>Hint: You will should create a separate directory under your team_exercises folder for your ballgown (or edgeR or DESeq2) outputs.</p>
<p><strong>Q2.</strong> How many significant differentially expressed genes do you observe?</p>
<p><strong>Q3.</strong> By referring back to the supplementary tutorial in the DE Visualization Module, can you construct a volcano plot showcasing the significantly de genes?</p>
<p>Additionally, students should feel free to explore other visualization methods, including those they may have used in past research experiences and share with the class.</p>
</div>
<div id="presenting-your-results-1" class="section level3 hasAnchor">
<h3>Presenting Your Results<a href="module-3.html#presenting-your-results-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>At the end of this team exercise, students will show how they visualized their differential expression results to the class.</p>
<p><strong>Preamble:</strong> Note that the following integrated assignment asks you to work on new RNA-seq data and apply the concepts you have learned up to this point. To complete this assignment you will need to review commands we performed in many of the earlier sections. Try to construct these commands on your own and get all the way to the end of the assignment. If you get very stuck or would like to compare your solutions to those suggested by the instructors, refer to the answers page. The integrated assignment answers page is an expanded version of this page with all of the questions plus detailed code solutions to all problems. Avoid the temptation to look at it without trying on your own.</p>
<p><strong>Background:</strong> Cell lines are often used to study different experimental conditions and to study the function of specific genes by various perturbation approaches. One such type of study involves knocking down expression of a target of interest by shRNA and then using RNA-seq to measure the impact on gene expression. These eperiments often include use of a control shRNA to account for any expression changes that may occur from just the introduction of these molecules. Differential expression is performed by comparing biological replicates of shRNA knockdown vs shRNA control.</p>
<p><strong>Objectives:</strong> In this assignment, we will be using a subset of the <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA471072">GSE114360 dataset</a>, which consists of 6 RNA-seq datasets generated from a cell line (3 transfected with shRNA, and 3 controls). Our goal will be to determine differentially expressed genes.</p>
<p>Experimental information and other things to keep in mind:</p>
<ul>
<li>The libraries are prepared as paired end.</li>
<li>The samples are sequenced on an Illumina 4000.</li>
<li>Each read is 150 bp long</li>
<li>The dataset is located here: <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA471072">GSE114360</a></li>
<li>3 samples transfected with target shRNA and 3 samples with control shRNA</li>
<li>Libraries were prepared using standard Illumina protocols</li>
<li>For this exercise we will be using a subset of the reads (first 1,000,000 reads from each pair).</li>
<li>The files are named based on their SRR id’s, and obey the following key:
<ul>
<li>SRR7155055 = CBSLR knockdown sample 1 (T1 - aka transfected 1)</li>
<li>SRR7155056 = CBSLR knockdown sample 2 (T2 - aka transfected 2)</li>
<li>SRR7155057 = CBSLR knockdown sample 3 (T3 - aka transfected 3)</li>
<li>SRR7155058 = control sample 1 (C1 - aka control 1)</li>
<li>SRR7155059 = control sample 2 (C2 - aka control 2)</li>
<li>SRR7155060 = control sample 3 (C3 - aka control 3)</li>
</ul></li>
</ul>
<p>Experimental descriptions from the study authors:</p>
<p>Experimental details from the <a href="https://pubmed.ncbi.nlm.nih.gov/35499052/">paper</a>:
“An RNA transcriptome-sequencing analysis was performed in shRNA-NC or shRNA-CBSLR-1 MKN45 cells cultured under hypoxic conditions for 24 h (Fig. 2A).”</p>
<p>Experimental details from the GEO submission:
“An RNA transcriptome sequencing analysis was performed in MKN45 cells that were transfected with tcons_00001221 shRNA or control shRNA.”</p>
<p>Note that according to <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=CBSLR">GeneCards</a> and <a href="https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/55459">HGNC</a>, <em>CBSLR</em> and <em>tcons_00001221</em> refer to the same gene.</p>
</div>
<div id="part-0-obtaining-data-and-references" class="section level3 hasAnchor">
<h3>PART 0 : Obtaining Data and References<a href="module-3.html#part-0-obtaining-data-and-references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Obtain the files necessary for data processing</li>
<li>Familiarize yourself with reference and annotation file format</li>
<li>Familiarize yourself with sequence FASTQ format</li>
</ul>
<p>Create a working directory ~/workspace/rnaseq/integrated_assignment/ to store this exercise. Then create a unix environment variable named <code>RNA_INT_ASSIGNMENT</code> that stores this path for convenience in later commands.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb199-1"><a href="module-3.html#cb199-1" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_HOME</span><span class="op">=</span>~/workspace/rnaseq</span>
<span id="cb199-2"><a href="module-3.html#cb199-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span></span>
<span id="cb199-3"><a href="module-3.html#cb199-3" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/workspace/rnaseq/integrated_assignment/</span>
<span id="cb199-4"><a href="module-3.html#cb199-4" tabindex="-1"></a><span class="bu">export</span> <span class="va">RNA_INT_DIR</span><span class="op">=</span>~/workspace/rnaseq/integrated_assignment</span></code></pre></div>
<p>Obtain reference, annotation, adapter and data files and place them in the integrated assignment directory</p>
<p>Remember: when initiating an environment variable, we do NOT need the $; however, everytime we call the variable, it needs to be preceeded by a $.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb200-1"><a href="module-3.html#cb200-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_INT_DIR</span></span>
<span id="cb200-2"><a href="module-3.html#cb200-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_INT_DIR</span></span>
<span id="cb200-3"><a href="module-3.html#cb200-3" tabindex="-1"></a></span>
<span id="cb200-4"><a href="module-3.html#cb200-4" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/Integrated_Assignment_RNA_Data.tar.gz</span></code></pre></div>
<p><strong>Q1.)</strong> How many items are there under the “reference” directory (counting all files in all sub-directories)? What if this reference file was not provided for you - how would you obtain/create a reference genome fasta file. How about the GTF transcripts file from Ensembl?</p>
<p><strong>Q2.)</strong> How many exons does the gene SOX4 have? How about the longest isoform of PCA3?</p>
<p><strong>Q3.)</strong> How many samples do you see under the data directory?</p>
<p>NOTE: The fastq files you have copied above contain only the first 1,000,000 reads. Keep this in mind when you are combing through the results of the differential expression analysis.</p>
</div>
<div id="part-1-data-preprocessing" class="section level3 hasAnchor">
<h3>Part 1 : Data preprocessing<a href="module-3.html#part-1-data-preprocessing" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Run a quality check with <code>fastqc</code> before and after trimming</li>
<li>Familiarize yourself with the options for <code>fastqc</code> to be able to redirect your output</li>
<li>Perform adapter trimming and data cleanup on your data using <code>fastp</code></li>
<li>Familiarize yourself with the output metrics from adapter trimming</li>
<li>Examine <code>fastqc</code> and/or <code>multiqc</code> reports for the pre- and post-trimmed data</li>
</ul>
<p><strong>Q4.)</strong> What metrics, if any, have the samples failed? Are the errors related?</p>
<p><strong>Q5.)</strong> What average percentage of reads remain after adapter trimming? Why do reads get tossed out?</p>
<p><strong>Q6.)</strong> What sample has the largest number of reads after trimming/cleanup?</p>
</div>
<div id="part-2-data-alignment" class="section level3 hasAnchor">
<h3>Part 2: Data alignment<a href="module-3.html#part-2-data-alignment" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong>
- Familiarize yourself with HISAT2 alignment options
- Perform alignments
- Obtain alignment summary
- Convert your alignment into compressed bam format</p>
<p><strong>Q7.)</strong> How would you obtain summary statistics for each aligned file?</p>
<p><strong>Q8.)</strong> Approximately how much space is saved by converting the sam to a bam format?</p>
<p>In order to make visualization easier, you should now merge each of your replicate sample bams into one combined BAM for each condition. Make sure to index these bams afterwards to be able to view them on IGV.</p>
<p>Try viewing genes such as TP53 to get a sense of how the data is aligned. To do this:
- Load up IGV
- Change the reference genome to “Human hg38” in the top-left category
- Click on File &gt; Load from URL, and in the File URL enter: “<a href="http://" class="uri">http://</a><your IP>/rnaseq/integrated_assignment/alignments/transfected.bam”. Repeat this step and enter “<a href="http://" class="uri">http://</a><your IP>/rnaseq/integrated_assignment/alignments/control.bam” to load the other bam.
- Right-click on the alignments track in the middle, and Group alignments by “Library”
- Jump to TP53 by typing it into the search bar above</p>
<p><strong>Q9.)</strong> What portion of the gene do the reads seem to be piling up on? What would be different if we were viewing whole-genome sequencing data?</p>
<p><strong>Q10.)</strong> What are the lines connecting the reads trying to convey?</p>
</div>
<div id="part-3-expression-estimation" class="section level3 hasAnchor">
<h3>Part 3: Expression Estimation<a href="module-3.html#part-3-expression-estimation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Familiarize yourself with Stringtie options</li>
<li>Run Stringtie to obtain expression values</li>
<li>Obtain expression values for the gene SOX4</li>
<li>Create an expression results directory, run Stringtie on all samples, and store the results in appropriately named subdirectories in this results dir</li>
</ul>
<p><strong>Q11.)</strong> How can you obtain the expression of the gene SOX4 across the transfected and control samples?</p>
</div>
<div id="part-4-differential-expression-analysis" class="section level3 hasAnchor">
<h3>Part 4: Differential Expression Analysis<a href="module-3.html#part-4-differential-expression-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Goals:</strong></p>
<ul>
<li>Perform differential analysis between the transfected and control samples</li>
</ul>
<p>Adapt the R tutorial code that was used in <a href="https://rnabio.org/module-03-expression/0003/03/01/Differential_Expression/">Differential Expression</a> section. Modify it to work on these data (which are also a 3x3 replicate comparison of two conditions).</p>
<p><strong>Q12.)</strong> Are there any significant differentially expressed genes? How many in total do you see? If we expected SOX4 to be differentially expressed, why don’t we see it in this case?</p>
</div>
<div id="part-5-differential-expression-analysis-visualization" class="section level3 hasAnchor">
<h3>Part 5: Differential Expression Analysis Visualization<a href="module-3.html#part-5-differential-expression-analysis-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Q13.)</strong> What plots can you generate to help you visualize this gene expression profile?</p>
<p>Hint, a volcano plot is popular approach to provide a high level summary of a differential expression analysis. Refer to the <a href="https://rnabio.org/module-03-expression/0003/04/01/DE_Visualization/">DE Visualization</a> section for example R code.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/RNA_2025/blob/main/030-module-3.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
