<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 1 | RNA-seq Analysis 2025</title>
  <meta name="description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 1 | RNA-seq Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/RNA_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 1 | RNA-seq Analysis 2025" />
  
  <meta name="twitter:description" content="Workshop website for the 2025 RNA-seq Analysis Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/RNA_2025/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Malachi Griffith and Obi Griffith" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-0-introductions-and-environment-setup.html"/>
<link rel="next" href="module-2.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>RNA-Seq Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_RNA_Icon.png" width="70%" alt="WORKSHOP NAME logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html"><i class="fa fa-check"></i>Module 0: Introductions and Environment Setup</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#environment-setup"><i class="fa fa-check"></i>Environment Setup</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tool-installation"><i class="fa fa-check"></i>Tool Installation</a>
<ul>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#note"><i class="fa fa-check"></i>Note</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#samtools"><i class="fa fa-check"></i>SAMtools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bam-readcount"><i class="fa fa-check"></i>bam-readcount</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#hisat2"><i class="fa fa-check"></i>HISAT2</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#stringtie"><i class="fa fa-check"></i>StringTie</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gffcompare"><i class="fa fa-check"></i>gffcompare</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#htseq-count"><i class="fa fa-check"></i>htseq-count</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#tophat"><i class="fa fa-check"></i>TopHat</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#kallisto"><i class="fa fa-check"></i>kallisto</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastqc"><i class="fa fa-check"></i>FastQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#fastp"><i class="fa fa-check"></i>Fastp</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#multiqc"><i class="fa fa-check"></i>MultiQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#picard"><i class="fa fa-check"></i>Picard</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#regtools"><i class="fa fa-check"></i>RegTools</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#rseqc"><i class="fa fa-check"></i>RSeQC</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bedops"><i class="fa fa-check"></i>bedops</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#gtftogenepred"><i class="fa fa-check"></i>gtfToGenePred</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#genepredtobed"><i class="fa fa-check"></i>genePredToBed</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#how_are_we_stranded_here"><i class="fa fa-check"></i>how_are_we_stranded_here</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-cell-ranger"><i class="fa fa-check"></i>Install Cell Ranger</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#install-r"><i class="fa fa-check"></i>Install R</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#r-libraries"><i class="fa fa-check"></i>R Libraries</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#bioconductor"><i class="fa fa-check"></i>Bioconductor</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#sleuth"><i class="fa fa-check"></i>Sleuth</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#practical-exercise-1---software-installation"><i class="fa fa-check"></i>PRACTICAL EXERCISE 1 - Software Installation</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#add-locally-installed-tools-to-your-path-optional"><i class="fa fa-check"></i>Add locally installed tools to your PATH [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-from-official-ubuntu-packages-optional"><i class="fa fa-check"></i>Installing tools from official ubuntu packages [OPTIONAL]</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image"><i class="fa fa-check"></i>Installing tools by Docker image</a></li>
<li class="chapter" data-level="" data-path="module-0-introductions-and-environment-setup.html"><a href="module-0-introductions-and-environment-setup.html#installing-tools-by-docker-image-using-singularity"><i class="fa fa-check"></i>Installing tools by Docker image (using Singularity)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html"><i class="fa fa-check"></i>Module 1</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#labs"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---key-concepts"><i class="fa fa-check"></i>Module 1 - Key concepts</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#module-1---learning-objectives"><i class="fa fa-check"></i>Module 1 - Learning objectives</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-a-reference-genome-from-ensembl-igenomes-ncbi-or-ucsc."><i class="fa fa-check"></i>Obtain a reference genome from Ensembl, iGenomes, NCBI or UCSC.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#note-on-complex-commands-and-scripting-in-unix"><i class="fa fa-check"></i>Note on complex commands and scripting in Unix</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-2-advanced"><i class="fa fa-check"></i>PRACTICAL EXERCISE 2 (ADVANCED)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#fastafastqgtf-mini-lecture-1"><i class="fa fa-check"></i>FASTA/FASTQ/GTF mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-known-genetranscript-annotations"><i class="fa fa-check"></i>Obtain Known Gene/Transcript Annotations</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#definitions"><i class="fa fa-check"></i>Definitions:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#the-purpose-of-gene-annotations-.gtf-file"><i class="fa fa-check"></i>The Purpose of Gene Annotations (.gtf file)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#sources-for-obtaining-gene-annotation-files-formatted-for-hisat2stringtieballgown"><i class="fa fa-check"></i>Sources for obtaining gene annotation files formatted for HISAT2/StringTie/Ballgown</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#important-notes"><i class="fa fa-check"></i>Important notes:</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#indexing-mini-lecture"><i class="fa fa-check"></i>Indexing mini lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#create-a-hisat2-index"><i class="fa fa-check"></i>Create a HISAT2 index</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#obtain-rna-seq-test-data."><i class="fa fa-check"></i>Obtain RNA-seq test data.</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#determining-the-strandedness-of-rna-seq-data-optional"><i class="fa fa-check"></i>Determining the strandedness of RNA-seq data (Optional)</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#practical-exercise-3"><i class="fa fa-check"></i>PRACTICAL EXERCISE 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html"><i class="fa fa-check"></i>Module 2</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#labs-1"><i class="fa fa-check"></i>Labs</a></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">RNA-seq Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-1" class="section level1 hasAnchor">
<h1>Module 1<a href="module-1.html#module-1" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-1" class="section level2 hasAnchor">
<h2>Lecture<a href="module-1.html#lecture-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe src="https://drive.google.com/file/d/1zjmIjjvpgiJTzuFFiWF8_8h1bn7xKL8e/preview" width="640" height="480" allow="autoplay">
</iframe>
</div>
<div id="labs" class="section level2 hasAnchor">
<h2>Labs<a href="module-1.html#labs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://rnabio.org/module-01-inputs/0001/01/01/Intro_to_Inputs/">Module 1 Labs</a></p>
<div id="module-1---key-concepts" class="section level3 hasAnchor">
<h3>Module 1 - Key concepts<a href="module-1.html#module-1---key-concepts" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Review central dogma, RNA sequencing, RNAseq study design, library construction strategies, biological vs technical replicates, alignment strategies, etc.</li>
</ul>
</div>
<div id="module-1---learning-objectives" class="section level3 hasAnchor">
<h3>Module 1 - Learning objectives<a href="module-1.html#module-1---learning-objectives" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>Introduction to the theory and practice of RNA sequencing (RNA-seq) analysis</li>
<li>Rationale for sequencing RNA</li>
<li>Challenges specific to RNA-seq</li>
<li>General goals and themes of RNA-seq analysis work flows</li>
<li>Common technical questions related to RNA-seq analysis</li>
<li>Getting help outside of this course</li>
<li>Introduction to the RNA-seq hands on tutorial</li>
</ul>
</div>
<div id="lecture-2" class="section level3 hasAnchor">
<h3>Lecture<a href="module-1.html#lecture-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/full/RNASeq_Module1_IntrotoRNA.pdf">Introduction to RNA sequencing lecture</a></li>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_01_01_FASTA_FASTQ_GTF.pdf">FASTA/FASTQ/GTF formats mini lecture</a></li>
<li><a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_01_02_Indexing.pdf">Indexing mini lecture</a></li>
</ul>
<hr />
<p><img src="img/module_1/RNA-seq_Flowchart2.png" alt="RNA-seq_Flowchart" /><br />
</p>
<hr />
</div>
<div id="fastafastqgtf-mini-lecture" class="section level3 hasAnchor">
<h3>FASTA/FASTQ/GTF mini lecture<a href="module-1.html#fastafastqgtf-mini-lecture" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a refresher on common file formats such as FASTA, FASTQ, and GTF files, we have made a <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_01_01_FASTA_FASTQ_GTF.pdf">mini lecture</a> briefly covering these.</p>
</div>
<div id="obtain-a-reference-genome-from-ensembl-igenomes-ncbi-or-ucsc." class="section level3 hasAnchor">
<h3>Obtain a reference genome from Ensembl, iGenomes, NCBI or UCSC.<a href="module-1.html#obtain-a-reference-genome-from-ensembl-igenomes-ncbi-or-ucsc." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this example analysis we will use the human GRCh38 version of the genome from Ensembl. Furthermore, we are actually going to perform the analysis using only a single chromosome (chr22) and the ERCC spike-in to make it run faster.</p>
<p>First we will create the necessary working directory.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb45-1"><a href="module-1.html#cb45-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span></span>
<span id="cb45-2"><a href="module-1.html#cb45-2" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_REFS_DIR</span></span>
<span id="cb45-3"><a href="module-1.html#cb45-3" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$RNA_REFS_DIR</span></span></code></pre></div>
<p>The complete data from which these files were obtained can be found at: <a href="ftp://ftp.ensembl.org/pub/release-86/fasta/homo_sapiens/dna/">ftp://ftp.ensembl.org/pub/release-86/fasta/homo_sapiens/dna/</a>. You could use wget to download the Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz file, then unzip/untar.</p>
<p>We have prepared this simplified reference for you. It contains chr22 (and ERCC transcript) fasta files in both a single combined file and individual files. Download the reference genome file to the rnaseq working directory</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb46-1"><a href="module-1.html#cb46-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_REFS_DIR</span></span>
<span id="cb46-2"><a href="module-1.html#cb46-2" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/fasta/GRCh38/chr22_with_ERCC92.fa</span>
<span id="cb46-3"><a href="module-1.html#cb46-3" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>View the first 10 lines of this file. Why does it look like this?</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb47-1"><a href="module-1.html#cb47-1" tabindex="-1"></a><span class="fu">head</span> chr22_with_ERCC92.fa</span></code></pre></div>
<p>How many lines and characters are in this file? How long is this chromosome (in bases and Mbp)?</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb48-1"><a href="module-1.html#cb48-1" tabindex="-1"></a><span class="fu">wc</span> chr22_with_ERCC92.fa</span></code></pre></div>
<p>View 10 lines from approximately the middle of this file. What is the significance of the upper and lower case characters?</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb49-1"><a href="module-1.html#cb49-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 425000 chr22_with_ERCC92.fa <span class="kw">|</span> <span class="fu">tail</span></span></code></pre></div>
<p>What is the count of each base in the entire reference genome file (skipping the header lines for each sequence)?</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb50-1"><a href="module-1.html#cb50-1" tabindex="-1"></a><span class="fu">cat</span> chr22_with_ERCC92.fa <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;&gt;&quot;</span> <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp $_; $bases{$_}++ for split //; if (eof){print &quot;$_ $bases{$_}\n&quot; for sort keys %bases}&#39;</span></span></code></pre></div>
<p>Note: Instead of the above, you might consider getting reference genomes and associated annotations from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/">UCSC. e.g., UCSC GRCh38 download</a>.</p>
<p>Wherever you get them from, remember that the names of your reference sequences (chromosomes) must those matched in your annotation gtf files (described in the next section).</p>
<p>View a list of all sequences in our reference genome fasta file.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb51-1"><a href="module-1.html#cb51-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">&quot;&gt;&quot;</span> chr22_with_ERCC92.fa</span></code></pre></div>
<hr />
</div>
<div id="note-on-complex-commands-and-scripting-in-unix" class="section level3 hasAnchor">
<h3>Note on complex commands and scripting in Unix<a href="module-1.html#note-on-complex-commands-and-scripting-in-unix" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Take a closer look at the command above that counts the occurrence of each nucleotide base in our chr22 reference sequence. Note that for even a seemingly simple question, commands can become quite complex. In that approach, a combination of Unix commands, pipes, and the scripting language Perl are used to answer the question. In bioinformatics, generally this kind of scripting comes up before too long, because you have an analysis question that is so specific there is no out of the box tool available. Or an existing tool will give perform a much more complex and involved analysis than needed to answer a very focused question.</p>
<p>In Unix there are usually many ways to solve the same problem. Perl as a language has mostly fallen out of favor. This kind of simple text parsing problem is one area it perhaps still remains relevant. Let’s benchmark the run time of the previous approach and constrast with several alternatives that do not rely on Perl.</p>
<p>Each of the following gives exactly the same answer. <code>time</code> is used to measure the run time of each alternative. Each starts by using <code>cat</code> to dump the file to standard out and then using <code>grep</code> to remove the header lines starting with “&gt;”. Each ends with <code>column -t</code> to make the output line up consistently.</p>
<pre><code>#1. The Perl approach. This command removes the end of line character with chomp, then it splits each line into an array of individual characters, amd it creates a data structure called a hash to store counts of each letter on each line. Once the end of the file is reached it prints out the contents of this data structure in order.  
time cat chr22_with_ERCC92.fa | grep -v &quot;&gt;&quot; | perl -ne &#39;chomp $_; $bases{$_}++ for split //; if (eof){print &quot;$bases{$_} $_\n&quot; for sort keys %bases}&#39; | column -t

#2. The Awk approach. Awk is an alternative scripting language include in most linux distributions. This command is conceptually very similar to the Perl approach but with a different syntax. A for loop is used to iterate over each character until the end (&quot;NF&quot;) is reached. Again the counts for each letter are stored in a simple data structure and once the end of the file is reach the results are printed.  
time cat chr22_with_ERCC92.fa | grep -v &quot;&gt;&quot; | awk &#39;{for (i=1; i&lt;=NF; i++){a[$i]++}}END{for (i in a){print a[i], i}}&#39; FS= - | sort -k 2 | column -t

#3. The Sed approach. Sed is an alternative scripting language. &quot;tr&quot; is used to remove newline characters. Then sed is used simply to split each character onto its own line, effectively creating a file with millions of lines. Then unix sort and uniq are used to produce counts of each unique character, and sort is used to order the results consistently with the previous approaches.
time cat chr22_with_ERCC92.fa | grep -v &quot;&gt;&quot; | tr -d &#39;\n&#39; | sed &#39;s/\(.\)/\1\n/g&#39;  - | sort | uniq -c | sort -k 2 | column -t

#4. The grep appoach. The &quot;-o&quot; option of grep splits each match onto a line which we then use to get a count. The &quot;-i&quot; option makes the matching work for upper/lower case. The &quot;-P&quot; option allows us to use Perl style regular expressions with Greg.
time cat chr22_with_ERCC92.fa | grep -v &quot;&gt;&quot; | grep -i -o -P &quot;a|c|g|t|y|n&quot; | sort | uniq -c

#5. Finally, the simplest/shortest approach that leverages the unix fold command to split each character onto its own line as in the Sed example.
time cat chr22_with_ERCC92.fa | grep -v &quot;&gt;&quot; | fold -w1 | sort | uniq -c | column -t

</code></pre>
<p>Which method is fastest? Why are the first two approaches so much faster than the others?</p>
<hr />
</div>
<div id="practical-exercise-2-advanced" class="section level3 hasAnchor">
<h3>PRACTICAL EXERCISE 2 (ADVANCED)<a href="module-1.html#practical-exercise-2-advanced" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Assignment: Use a commandline scripting approach of your choice to further examine our chr22 reference genome file and answer the following questions.</p>
<p>Questions:
- How many bases on chromosome 22 correspond to repetitive elements?
- What is the percentage of the whole length?
- How many occurences of the EcoRI (GAATTC) restriction site are present in the chromosome 22 sequence?</p>
<p>Hint: Each question can be tackled using approaches similar to those above, using the file ‘chr22_with_ERCC92.fa’ as a starting point.
Hint: To make things simpler, first produce a file with only the chr22 sequence.
Hint: Remember that repetitive elements in the sequence are represented in lower case</p>
<p>Solution: When you are ready you can check your approach against the <a href="">Solutions</a>.</p>
</div>
<div id="fastafastqgtf-mini-lecture-1" class="section level3 hasAnchor">
<h3>FASTA/FASTQ/GTF mini lecture<a href="module-1.html#fastafastqgtf-mini-lecture-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you would like a refresher on common file formats such as FASTA, FASTQ, and GTF files, we have made a <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_01_01_FASTA_FASTQ_GTF.pdf">mini lecture</a> briefly covering these.</p>
</div>
<div id="obtain-known-genetranscript-annotations" class="section level3 hasAnchor">
<h3>Obtain Known Gene/Transcript Annotations<a href="module-1.html#obtain-known-genetranscript-annotations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial we will use annotations obtained from Ensembl (<a href="ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.gtf.gz">Homo_sapiens.GRCh38.86.gtf.gz</a>) for chromosome 22 only. For time reasons, these are prepared for you and made available on your AWS instance. But you should get familiar with sources of gene annotations for RNA-seq analysis.</p>
<p>Copy the gene annotation files to the working directory.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb53-1"><a href="module-1.html#cb53-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_REFS_DIR</span></span>
<span id="cb53-2"><a href="module-1.html#cb53-2" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_REFS_DIR</span></span>
<span id="cb53-3"><a href="module-1.html#cb53-3" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/annotations/GRCh38/chr22_with_ERCC92.gtf</span></code></pre></div>
<p>Take a look at the contents of the .gtf file. Press <code>q</code> to exit the <code>less</code> display.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb54-1"><a href="module-1.html#cb54-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_REF_GTF</span></span>
<span id="cb54-2"><a href="module-1.html#cb54-2" tabindex="-1"></a><span class="fu">less</span> <span class="at">-p</span> start_codon <span class="at">-S</span> <span class="va">$RNA_REF_GTF</span></span></code></pre></div>
<p>Note how the <code>-S</code> option makes it easier to veiw this file with <code>less</code>. Make the formatting a bit nicer still:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb55-1"><a href="module-1.html#cb55-1" tabindex="-1"></a><span class="fu">cat</span> chr22_with_ERCC92.gtf <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="kw">|</span> <span class="fu">less</span> <span class="at">-p</span> exon <span class="at">-S</span></span></code></pre></div>
<p>How many unique gene IDs are in the .gtf file?</p>
<p>We can use a perl command-line command to find out:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb56-1"><a href="module-1.html#cb56-1" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;if ($_ =~ /(gene_id\s\&quot;ENSG\w+\&quot;)/){print &quot;$1\n&quot;}&#39;</span> <span class="va">$RNA_REF_GTF</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<ul>
<li><p>Using <code>perl -ne ''</code> will execute the code between single quotes, on the .gtf file, line-by-line.</p></li>
<li><p>The <code>$_</code> variable holds the contents of each line.</p></li>
<li><p>The <code>'if ($_ =~//)'</code> is a pattern-matching command which will look for the pattern “gene_id” followed by a space followed by “ENSG” and one or more word characters (indicated by <code>\w+</code>) surrounded by double quotes.</p></li>
<li><p>The pattern to be matched is enclosed in parentheses. This allows us to print it out from the special variable <code>$1</code>.</p></li>
<li><p>The output of this perl command will be a long list of ENSG Ids.</p></li>
<li><p>By piping to <code>sort</code>, then <code>uniq</code>, then word count we can count the unique number of genes in the file.</p></li>
</ul>
<p>We can also use <code>grep</code> to find this same information.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb57-1"><a href="module-1.html#cb57-1" tabindex="-1"></a><span class="fu">cat</span> chr22_with_ERCC92.gtf <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-w</span> gene <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<ul>
<li><code>grep -w gene</code> is telling grep to do an exact match for the string ‘gene’. This means that it will return lines that are of the feature type <code>gene</code>.</li>
</ul>
<p>Now view the structure of a single transcript in GTF format. Press <code>q</code> to exit the <code>less</code> display when you are done.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb58-1"><a href="module-1.html#cb58-1" tabindex="-1"></a><span class="fu">grep</span> ENST00000342247 <span class="va">$RNA_REF_GTF</span> <span class="kw">|</span> <span class="fu">less</span> <span class="at">-p</span> <span class="st">&quot;exon\s&quot;</span> <span class="at">-S</span></span></code></pre></div>
<p>To learn more, see:</p>
<ul>
<li><a href="http://perldoc.perl.org/perlre.html#Regular-Expressions">http://perldoc.perl.org/perlre.html#Regular-Expressions</a></li>
<li><a href="http://www.perl.com/pub/2004/08/09/commandline.html">http://www.perl.com/pub/2004/08/09/commandline.html</a></li>
</ul>
</div>
<div id="definitions" class="section level3 hasAnchor">
<h3>Definitions:<a href="module-1.html#definitions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Reference genome</strong> - The nucleotide sequence of the chromosomes of a species. Genes are the functional units of a reference genome and gene annotations describe the structure of transcripts expressed from those gene loci.</p>
<p><strong>Gene annotations</strong> - Descriptions of gene/transcript models for a genome. A transcript model consists of the coordinates of the exons of a transcript on a reference genome. Additional information such as the strand the transcript is generated from, gene name, coding portion of the transcript, alternate transcript start sites, and other information may be provided.</p>
<p><strong>GTF (.gtf) file</strong> - A common file format referred to as Gene Transfer Format used to store gene and transcript annotation information. You can learn more about this format here: <a href="http://genome.ucsc.edu/FAQ/FAQformat#format4">http://genome.ucsc.edu/FAQ/FAQformat#format4</a></p>
</div>
<div id="the-purpose-of-gene-annotations-.gtf-file" class="section level3 hasAnchor">
<h3>The Purpose of Gene Annotations (.gtf file)<a href="module-1.html#the-purpose-of-gene-annotations-.gtf-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When running the HISAT2/StringTie/Ballgown pipeline, known gene/transcript annotations are used for several purposes:</p>
<ul>
<li><p>During the HISAT2 index creation step, annotations may be provided to create local indexes to represent transcripts as well as a global index for the entire reference genome. This allows for faster mapping and better mapping across exon boundaries and splice sites. If an alignment still can not be found it will attempt to determine if the read corresponds to a novel exon-exon junction. See the <a href="/module-01-inputs/0001/04/01/Indexing/">Indexing section</a> and the HISAT2 publication for more details.</p></li>
<li><p>During the StringTie step, a .gtf file can be used to specify transcript models to guide the assembly process and limit expression estimates to predefined transcripts using the <code>-G</code> and <code>-e</code> options together. The <code>-e</code> option will give you one expression estimate for each of the transcripts in your .gtf file, giving you a ‘microarray like’ expression result.</p></li>
<li><p>During the StringTie step, if the <code>-G</code> option is specified without the <code>-e</code> option the .gtf file is used only to ‘guide’ the assembly of transcripts. Instead of assuming only the known transcript models are correct, the resulting expression estimates will correspond to both known and novel/predicted transcripts.</p></li>
<li><p>During the StringTie and gffcompare steps, a .gtf file is used to determine the transcripts that will be examined for differential expression using Ballgown. These may be known transcripts that you download from a public source, or a .gtf of transcripts predicted by StringTie from the read data in an earlier step.</p></li>
</ul>
</div>
<div id="sources-for-obtaining-gene-annotation-files-formatted-for-hisat2stringtieballgown" class="section level3 hasAnchor">
<h3>Sources for obtaining gene annotation files formatted for HISAT2/StringTie/Ballgown<a href="module-1.html#sources-for-obtaining-gene-annotation-files-formatted-for-hisat2stringtieballgown" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are many possible sources of .gtf gene/transcript annotation files. For example, from Ensembl, UCSC, RefSeq, etc. Several options and related instructions for obtaining the gene annotation files are provided below.</p>
<div id="i.-ensembl-ftp-site" class="section level4 hasAnchor">
<h4>I. ENSEMBL FTP SITE<a href="module-1.html#i.-ensembl-ftp-site" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Based on Ensembl annotations only. Available for many species. <a href="http://useast.ensembl.org/info/data/ftp/index.html">http://useast.ensembl.org/info/data/ftp/index.html</a></p>
</div>
<div id="ii.-ucsc-table-browser" class="section level4 hasAnchor">
<h4>II. UCSC TABLE BROWSER<a href="module-1.html#ii.-ucsc-table-browser" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Based on UCSC annotations or several other possible annotation sources collected by UCSC. You might chose this option if you want to have a lot of flexibility in the annotations you obtain. e.g. to grab only the transcripts from chromosome 22 as in the following example:</p>
<ul>
<li>Open the following in your browser: <a href="http://genome.ucsc.edu/">http://genome.ucsc.edu/</a></li>
<li>Select ‘Tools’ and then ‘Table Browser’ at the top of the page.</li>
<li>Select ‘Mammal’, ‘Human’, and ‘Dec. 2013 (GRCh38/hg38)’ from the first row of drop down menus.</li>
<li>Select ‘Genes and Gene Predictions’ and ‘GENCODE v29’ from the second row of drop down menus. To limit your selection to only chromosome 22, select the ‘position’ option beside ‘region’, enter ‘chr22’ in the ‘position’ box.</li>
<li>Select ‘GTF - gene transfer format’ for output format and enter ‘UCSC_Genes.gtf’ for output file.</li>
<li>Hit the ‘get output’ button and save the file. Make note of its location</li>
</ul>
<p>In addition to the .gtf file you may find uses for some extra files providing alternatively formatted or additional information on the same transcripts. For example:</p>
<div id="how-to-get-a-gene-bed-file" class="section level5 hasAnchor">
<h5>How to get a Gene bed file:<a href="module-1.html#how-to-get-a-gene-bed-file" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<ul>
<li>Change the output format to ‘BED - browser extensible data’.</li>
<li>Change the output file to ‘UCSC_Genes.bed’, and hit the ‘get output’ button.</li>
<li>Make sure ‘Whole Gene’ is selected, hit the ‘get BED’ button, and save the file.</li>
</ul>
</div>
<div id="how-to-get-an-exon-bed-file" class="section level5 hasAnchor">
<h5>How to get an Exon bed file:<a href="module-1.html#how-to-get-an-exon-bed-file" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<ul>
<li>Go back one page in your browser and change the output file to ‘UCSC_Exons.bed’, then hit the ‘get output’ button again.</li>
<li>Select ‘Exons plus’, enter 0 in the adjacent box, hit the ‘get BED’ button, and save the file.</li>
</ul>
</div>
<div id="how-to-get-gene-symbols-and-descriptions-for-all-ucsc-genes" class="section level5 hasAnchor">
<h5>How to get gene symbols and descriptions for all UCSC genes:<a href="module-1.html#how-to-get-gene-symbols-and-descriptions-for-all-ucsc-genes" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<ul>
<li>Again go back one page in your browser and change the ‘output format’ to ‘selected fields from primary and related tables’.</li>
<li>Change the output file to ‘UCSC_Names.txt’, and hit the ‘get output’ button.</li>
<li>Make sure ‘chrom’ is selected near the top of the page.</li>
<li>Under ‘Linked Tables’ make sure ‘kgXref’ is selected, and then hit ‘Allow Selection From Checked Tables’. This will link the table and give you access to its fields.</li>
<li>Under ‘hg38.kgXref fields’ select: ‘kgID’, ‘geneSymbol’, ‘description’.</li>
<li>Hit the ‘get output’ button and save the file.</li>
<li>To get annotations for the whole genome, make sure ‘genome’ is selected beside ‘region’. By default, the files downloaded above will be compressed. To decompress, use ‘gunzip filename’ in linux.</li>
</ul>
</div>
</div>
<div id="iii.-hisat2-precomputed-genome-index" class="section level4 hasAnchor">
<h4>III. HISAT2 Precomputed Genome Index<a href="module-1.html#iii.-hisat2-precomputed-genome-index" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>HISAT2 has prebuilt reference genome index files for both DNA and RNA alignment. Various versions of the index files include SNPs and/or transcript splice sites. Versions of both the Ensembl and UCSC genomes for human build 38 are linked from the main HISAT2 page: <a href="https://ccb.jhu.edu/software/hisat2/index.shtml">https://ccb.jhu.edu/software/hisat2/index.shtml</a></p>
<p>Or those same files are directly available from their FTP site: <a href="ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/">ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/</a></p>
</div>
</div>
<div id="important-notes" class="section level3 hasAnchor">
<h3>Important notes:<a href="module-1.html#important-notes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>On chromosome naming conventions:</strong>
In order for your RNA-seq analysis to work, the chromosome names in your .gtf file must match those in your reference genome (i.e. your reference genome fasta file). If you get a StringTie result where all transcripts have an expression value of 0, you may have overlooked this. Unfortunately, Ensembl, NCBI, and UCSC can not agree on how to name the chromosomes in many species, so this problem may come up often. You can avoid this by getting a complete reference genome and gene annotation package from the same source (e.g., Ensembl) to maintain consistency.</p>
<p><strong>On reference genome builds:</strong>
Your annotations must correspond to the same reference genome build as your reference genome fasta file. e.g., both correspond to UCSC human build ‘hg38’, NCBI human build ‘GRCh38’, etc. Even if both your reference genome and annotations are from UCSC or Ensembl they could still correspond to different versions of that genome. This would cause problems in any RNA-seq pipeline.</p>
<p>A more detailed discussion of commonly used version of the human reference genome can be found in a companion workshop <a href="https://pmbio.org/module-02-inputs/0002/02/01/Reference_Genome/">PMBIO Reference Genomes</a>.</p>
</div>
<div id="indexing-mini-lecture" class="section level3 hasAnchor">
<h3>Indexing mini lecture<a href="module-1.html#indexing-mini-lecture" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you want a refresher on indexing, we have made an <a href="https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cbw/2025/mini/RNASeq_MiniLecture_01_02_Indexing.pdf">indexing mini lecture</a> available.</p>
</div>
<div id="create-a-hisat2-index" class="section level3 hasAnchor">
<h3>Create a HISAT2 index<a href="module-1.html#create-a-hisat2-index" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Create a HISAT2 index for chr22 and the ERCC spike-in sequences. HISAT2 can incorporate exons and splice sites into the index file for alignment. First create a splice site file, then an exon file. Finally make the aligner FM index.</p>
<p>To learn more about how the HISAT2 indexing strategy is distinct from other next gen aligners refer to the <a href="https://www.ncbi.nlm.nih.gov/pubmed/25751142">HISAT publication</a>.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb59-1"><a href="module-1.html#cb59-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_REFS_DIR</span></span>
<span id="cb59-2"><a href="module-1.html#cb59-2" tabindex="-1"></a><span class="ex">hisat2_extract_splice_sites.py</span> <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> <span class="va">$RNA_REFS_DIR</span>/splicesites.tsv</span>
<span id="cb59-3"><a href="module-1.html#cb59-3" tabindex="-1"></a><span class="ex">hisat2_extract_exons.py</span> <span class="va">$RNA_REF_GTF</span> <span class="op">&gt;</span> <span class="va">$RNA_REFS_DIR</span>/exons.tsv</span>
<span id="cb59-4"><a href="module-1.html#cb59-4" tabindex="-1"></a><span class="ex">hisat2-build</span> <span class="at">-p</span> 4 <span class="at">--ss</span> <span class="va">$RNA_REFS_DIR</span>/splicesites.tsv <span class="at">--exon</span> <span class="va">$RNA_REFS_DIR</span>/exons.tsv <span class="va">$RNA_REF_FASTA</span> <span class="va">$RNA_REF_INDEX</span></span>
<span id="cb59-5"><a href="module-1.html#cb59-5" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>Perform a visual survey on the contents of your refs directory. What is the source/purpose of each file.</p>
<p><strong>[OPTIONAL]</strong> To create an index for all chromosomes instead of just chr22 you would do something like the following:</p>
<p><strong>WARNING:</strong> In order to index the entire human genome, HISAT2 requires 160GB of RAM. Your AWS instance size will run out of RAM.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb60-1"><a href="module-1.html#cb60-1" tabindex="-1"></a><span class="co">#hisat2_extract_splice_sites.py Homo_sapiens.GRCh38.86.gtf &gt; splicesites.tsv</span></span>
<span id="cb60-2"><a href="module-1.html#cb60-2" tabindex="-1"></a><span class="co">#hisat2_extract_exons.py Homo_sapiens.GRCh38.86.gtf &gt; exons.tsv</span></span>
<span id="cb60-3"><a href="module-1.html#cb60-3" tabindex="-1"></a><span class="co">#hisat2-build -p 4 --ss splicesites.tsv --exon exons.tsv Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa Homo_sapiens.GRCh38.dna_sm.primary_assembly</span></span></code></pre></div>
</div>
<div id="obtain-rna-seq-test-data." class="section level3 hasAnchor">
<h3>Obtain RNA-seq test data.<a href="module-1.html#obtain-rna-seq-test-data." class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The test data consists of two commercially available RNA samples: <a href="/assets/module_1/UHR.pdf">Universal Human Reference (UHR)</a> and <a href="/assets/module_1/HBR.pdf">Human Brain Reference (HBR)</a>. The UHR is total RNA isolated from a diverse set of 10 cancer cell lines (breast, liver, cervix, testis, brain, skin, fatty tissue, histocyte, macrophage, T cell, B cell). The HBR is total RNA isolated from the brains of 23 Caucasians, male and female, of varying age but mostly 60-80 years old.</p>
<p>In addition, a spike-in control was used. Specifically we added an aliquot of the <a href="/assets/module_1/ERCC.pdf">ERCC ExFold RNA Spike-In Control Mixes</a> to each sample. The spike-in consists of 92 transcripts that are present in known concentrations across a wide abundance range (from very few copies to many copies). This range allows us to test the degree to which the RNA-seq assay (including all laboratory and analysis steps) accurately reflects the relative abundance of transcript species within a sample. There are two ‘mixes’ of these transcripts to allow an assessment of differential expression output between samples if you put one mix in each of your two comparisons. In our case, Mix1 was added to the UHR sample, and Mix2 was added to the HBR sample. We also have 3 complete experimental replicates for each sample. This allows us to assess the technical variability of our overall process of producing RNA-seq data in the lab.</p>
<p>For all libraries we prepared low-throughput (Set A) TruSeq Stranded Total RNA Sample Prep Kit libraries with Ribo-Zero Gold to remove both cytoplasmic and mitochondrial rRNA. Triplicate, indexed libraries were made starting with 100ng Agilent/Strategene Universal Human Reference total RNA and 100ng Ambion Human Brain Reference total RNA. The Universal Human Reference replicates received 2 ul of 1:1000 ERCC Mix 1. The Human Brain Reference replicates received 1:1000 ERCC Mix 2. The libraries were quantified with KAPA Library Quantification qPCR and adjusted to the appropriate concentration for sequencing. The triplicate, indexed libraries were then pooled prior to sequencing. Each pool of three replicate libraries were sequenced across 2 lanes of a HiSeq 2000 using paired-end sequence chemistry with 100bp read lengths.</p>
<p>So to summarize we have:</p>
<ul>
<li>UHR + ERCC Spike-In Mix1, Replicate 1</li>
<li>UHR + ERCC Spike-In Mix1, Replicate 2</li>
<li>UHR + ERCC Spike-In Mix1, Replicate 3</li>
<li>HBR + ERCC Spike-In Mix2, Replicate 1</li>
<li>HBR + ERCC Spike-In Mix2, Replicate 2</li>
<li>HBR + ERCC Spike-In Mix2, Replicate 3</li>
</ul>
<p>Each data set has a corresponding pair of FASTQ files (read 1 and read 2 of paired end reads).</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb61-1"><a href="module-1.html#cb61-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$RNA_DATA_DIR</span></span>
<span id="cb61-2"><a href="module-1.html#cb61-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$RNA_DATA_DIR</span></span>
<span id="cb61-3"><a href="module-1.html#cb61-3" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_DATA_DIR</span></span>
<span id="cb61-4"><a href="module-1.html#cb61-4" tabindex="-1"></a><span class="fu">wget</span> http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar</span></code></pre></div>
<p>Unpack the test data using tar. You should see 6 sets of paired end fastq files. One for each of our sample replicates above. We have 6 pairs (12 files) because in fastq format, read 1 and read 2 of a each read pair (fragment) are stored in separate files.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb62-1"><a href="module-1.html#cb62-1" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> HBR_UHR_ERCC_ds_5pc.tar</span>
<span id="cb62-2"><a href="module-1.html#cb62-2" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>Enter the data directory and view the first two read records of a file (in fastq format each read corresponds to 4 lines of data)</p>
<p>The reads are paired-end 101-mers generated on an Illumina HiSeq instrument. The test data has been pre-filtered for reads that appear to map to chromosome 22. Lets copy the raw input data to our tutorial working directory.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb63-1"><a href="module-1.html#cb63-1" tabindex="-1"></a><span class="fu">zcat</span> UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 8</span></code></pre></div>
<p>Identify the following components of each read: read name, read sequence, and quality string</p>
<p>How many reads are there in the first library? Decompress file on the fly with ‘zcat’, pipe into ‘grep’, search for the read name prefix and pipe into ‘wc’ to do a word count (‘-l’ gives lines)</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb64-1"><a href="module-1.html#cb64-1" tabindex="-1"></a><span class="fu">zcat</span> UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-P</span> <span class="st">&quot;^\@HWI&quot;</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
</div>
<div id="determining-the-strandedness-of-rna-seq-data-optional" class="section level3 hasAnchor">
<h3>Determining the strandedness of RNA-seq data (Optional)<a href="module-1.html#determining-the-strandedness-of-rna-seq-data-optional" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In order to determine strandedness, we will be using <a href="https://github.com/betsig/how_are_we_stranded_here">check_strandedness</a>(<a href="https://hub.docker.com/r/mgibio/checkstrandedness">docker image</a>). In order use this tool, there are a few steps we need to get our inputs ready, specifically creating a fasta of our GTF file.</p>
<p>To create a transcripts FASTA file from our reference transcriptome in GTF format we need to get our transcript information into BED12 format. To get to that format we will convert GTF -&gt; GenePred -&gt; BED12. The GenePred format is just an intermediate that we create because we did not find a tool that would convert straight from GTF to BED12.</p>
<p>Once we have our transcriptome information (the position of exons for each transcript on each chromosome) we can use <code>bedtools getfasta</code> to splice the exon sequences from the chromosome sequence together into a full length transcript sequence.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb65-1"><a href="module-1.html#cb65-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/refs/</span>
<span id="cb65-2"><a href="module-1.html#cb65-2" tabindex="-1"></a></span>
<span id="cb65-3"><a href="module-1.html#cb65-3" tabindex="-1"></a><span class="co"># Convert our reference transcriptome GTF to genePred format</span></span>
<span id="cb65-4"><a href="module-1.html#cb65-4" tabindex="-1"></a><span class="ex">gtfToGenePred</span> chr22_with_ERCC92.gtf chr22_with_ERCC92.genePred</span>
<span id="cb65-5"><a href="module-1.html#cb65-5" tabindex="-1"></a></span>
<span id="cb65-6"><a href="module-1.html#cb65-6" tabindex="-1"></a><span class="co"># Convert the genPred format to bed12 format</span></span>
<span id="cb65-7"><a href="module-1.html#cb65-7" tabindex="-1"></a><span class="ex">genePredToBed</span> chr22_with_ERCC92.genePred chr22_with_ERCC92.bed12</span>
<span id="cb65-8"><a href="module-1.html#cb65-8" tabindex="-1"></a></span>
<span id="cb65-9"><a href="module-1.html#cb65-9" tabindex="-1"></a><span class="co"># Use bedtools to create fasta from GTF</span></span>
<span id="cb65-10"><a href="module-1.html#cb65-10" tabindex="-1"></a><span class="ex">bedtools</span> getfasta <span class="at">-fi</span> chr22_with_ERCC92.fa <span class="at">-bed</span> chr22_with_ERCC92.bed12 <span class="at">-s</span> <span class="at">-split</span> <span class="at">-name</span> <span class="at">-fo</span> chr22_ERCC92_transcripts.fa</span></code></pre></div>
<p>Use less to view the file chr22_ERCC92_transcripts.fa. Note that this file has messy transcript names. Use the following hairball perl one-liner to tidy up the header line for each fasta sequence.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb66-1"><a href="module-1.html#cb66-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/refs</span>
<span id="cb66-2"><a href="module-1.html#cb66-2" tabindex="-1"></a><span class="fu">cat</span> chr22_ERCC92_transcripts.fa <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;if($_ =~/^\&gt;\S+\:\:(ERCC\-\d+)\:.*/){print &quot;&gt;$1\n&quot;}elsif ($_ =~/^\&gt;(\S+)\:\:.*/){print &quot;&gt;$1\n&quot;}else{print $_}&#39;</span> <span class="op">&gt;</span> chr22_ERCC92_transcripts.clean.fa</span></code></pre></div>
<p>View the resulting ‘clean’ file using <code>less chr22_ERCC92_transcripts.clean.fa</code> (use ‘q’ to exit). View the end of this file using <code>tail chr22_ERCC92_transcripts.clean.fa</code>. Note that we have one fasta record for each Ensembl transcript on chromosome 22 and we have an additional fasta record for each ERCC spike-in sequence.</p>
<p>We also need to reformat our GTF file slightly. Rows that correspond to genes are missing the “transcript_id” field. We are going to add in this field but leave it blank for these rows using the following command.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb67-1"><a href="module-1.html#cb67-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$RNA_HOME</span>/refs</span>
<span id="cb67-2"><a href="module-1.html#cb67-2" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&#39;{ if ($0 ~ &quot;transcript_id&quot;) print $0; else print $0&quot; transcript_id \&quot;\&quot;;&quot;; }&#39;</span> chr22_with_ERCC92.gtf <span class="op">&gt;</span> chr22_with_ERCC92_tidy.gtf</span></code></pre></div>
<p>Now that we have created our input files, we can now run the check_strandedness tool on some of our instrument data. Note: we are using a docker image for this tool.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb68-1"><a href="module-1.html#cb68-1" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-v</span> /home/ubuntu/workspace/rnaseq:/docker_workspace mgibio/checkstrandedness:latest check_strandedness <span class="at">--gtf</span> /docker_workspace/refs/chr22_with_ERCC92_tidy.gtf <span class="at">--transcripts</span> /docker_workspace/refs/chr22_ERCC92_transcripts.clean.fa <span class="at">--reads_1</span> /docker_workspace/data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz <span class="at">--reads_2</span> /docker_workspace/data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz</span></code></pre></div>
<p><code>docker run</code> is how you initialize a docker container to run a command</p>
<p><code>-v</code> is the parameter used to mount your workspace so that the docker container can see the files that you’re working with. In the example above, <code>/home/ubuntu/workspace/rnaseq</code> from the EC2 instance has been mounted as <code>/docker_workspace</code> within the docker container.</p>
<p><code>mgibio/checkstrandedness</code> is the docker container name. The <code>:latest</code> refers to the specific tag and release of the docker container.</p>
<p>The output of this command should look like so:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb69-1"><a href="module-1.html#cb69-1" tabindex="-1"></a><span class="ex">Fraction</span> of reads failed to determine: 0.1123</span>
<span id="cb69-2"><a href="module-1.html#cb69-2" tabindex="-1"></a><span class="ex">Fraction</span> of reads explained by <span class="st">&quot;1++,1--,2+-,2-+&quot;</span>: 0.0155</span>
<span id="cb69-3"><a href="module-1.html#cb69-3" tabindex="-1"></a><span class="ex">Fraction</span> of reads explained by <span class="st">&quot;1+-,1-+,2++,2--&quot;</span>: 0.8722</span>
<span id="cb69-4"><a href="module-1.html#cb69-4" tabindex="-1"></a><span class="ex">Over</span> 75% of reads explained by <span class="st">&quot;1+-,1-+,2++,2--&quot;</span></span>
<span id="cb69-5"><a href="module-1.html#cb69-5" tabindex="-1"></a><span class="ex">Data</span> is likely RF/fr-firststrand</span></code></pre></div>
<p>Using this <a href="https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/">table</a>, we can see if this is what we expect. Note that since the UHR and HBR data were generated with the TruSeq Stranded Kit, as mentioned above, the correct strand setting for kallisto is <code>--rf-stranded</code>, which is what check_strandedness confirms. Similarly when we run HISAT we will use <code>--rna-strandness RF</code>, when we run StringTie we will use <code>--rf</code>, and when we run htseq-count we will use <code>--stranded reverse</code>.</p>
<hr />
</div>
<div id="practical-exercise-3" class="section level3 hasAnchor">
<h3>PRACTICAL EXERCISE 3<a href="module-1.html#practical-exercise-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Assignment: Download an additional dataset and unpack it. This data will be used in future practical exercises.</p>
<ul>
<li>Hint: Do this in a separate working directory called ‘practice’ and create sub-directories for organization (data, alignments, etc).</li>
<li>In this exercise you will download an archive of publicly available read data from here: <a href="http://genomedata.org/rnaseq-tutorial/practical.tar">http://genomedata.org/rnaseq-tutorial/practical.tar</a></li>
</ul>
<p>The practice dataset includes 3 replicates of data from the HCC1395 breast cancer cell line and 3 replicates of data from HCC1395BL matched lymphoblastoid line. So, this will be a tumor vs normal (cell line) comparison. The reads are paired-end 151-mers generated on an Illumina HiSeq instrument from TruSeq Stranded Total RNA Libraries. The test data has been pre-filtered for reads that appear to map to chromosome 22.</p>
<p><strong>Questions</strong></p>
<ul>
<li>How many data files were contained in the ‘practical.tar’ archive? What commonly used sequence data file format are they?</li>
<li>In the first read of the hcc1395, normal, replicate 1, read 1 file, what was the physical location of the read on the flow cell (i.e. lane, tile, x, y)?</li>
<li>In the first read of this same file, how many ‘T’ bases are there?</li>
</ul>
<p>Solution: When you are ready you can check your approach against the <a href="/module-09-appendix/0009/05/01/Practical_Exercise_Solutions/#practical-exercise-3---data">Solutions</a>.</p>
<hr />
<p><strong>NOTE</strong>: The complete RAW HCC1395 data sets can be found here: <a href="http://genomedata.org/pmbio-workshop/fastqs/all/">http://genomedata.org/pmbio-workshop/fastqs/all/</a></p>
<p>If you use this data, please cite our paper: <a href="https://github.com/griffithlab/rnaseq_tutorial/wiki/Citation">Citation</a></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-0-introductions-and-environment-setup.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/RNA_2025/blob/main/010-module-1.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
